{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Added C:\\Users\\heine\\github\\expression_broad_data to path: \n",
      "['', 'C:\\\\Users\\\\heine\\\\AppData\\\\Local\\\\Enthought\\\\Canopy\\\\edm\\\\envs\\\\bmh_python35_ts\\\\python35.zip', 'C:\\\\Users\\\\heine\\\\AppData\\\\Local\\\\Enthought\\\\Canopy\\\\edm\\\\envs\\\\bmh_python35_ts\\\\DLLs', 'C:\\\\Users\\\\heine\\\\AppData\\\\Local\\\\Enthought\\\\Canopy\\\\edm\\\\envs\\\\bmh_python35_ts\\\\lib', 'C:\\\\Users\\\\heine\\\\AppData\\\\Local\\\\Enthought\\\\Canopy\\\\edm\\\\envs\\\\bmh_python35_ts', 'C:\\\\Users\\\\heine\\\\AppData\\\\Local\\\\Enthought\\\\Canopy\\\\edm\\\\envs\\\\bmh_python35_ts\\\\lib\\\\site-packages', 'C:\\\\Users\\\\heine\\\\AppData\\\\Local\\\\Enthought\\\\Canopy\\\\edm\\\\envs\\\\bmh_python35_ts\\\\lib\\\\site-packages\\\\IPython\\\\extensions', 'C:\\\\Users\\\\heine\\\\.ipython', 'C:\\\\Users\\\\heine\\\\github\\\\expression_broad_data']\n",
      "I am about to import a library\n",
      "I am importing io_library\n",
      "C:\\Users\\heine\\github\\expression_broad_data\\expression_data\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "\n",
    "#only do this once in a kernel\n",
    "#sys.path.append(\"/home/heineike/github/expression_broad_data\")\n",
    "base_dir = \"C:\\\\Users\\\\heine\\\\github\\\\expression_broad_data\"\n",
    "if sys.path[-1] != base_dir:\n",
    "    sys.path.append(base_dir)\n",
    "    print(\"Added \" + base_dir + \" to path: \" )\n",
    "    print(sys.path)\n",
    "\n",
    "#Indicate operating environment and import core modules\n",
    "#import os\n",
    "# location_input = input(\"what computer are you on? a = Bens, b = gpucluster, c = other   \")\n",
    "# location_dict = {'a': \"C:\\\\Users\\Ben\\Documents\\GitHub\\expression_broad_data\", 'b': \"/home/heineike/github/expression_broad_data/\",'c':'you need to add your location to the location_dict'}\n",
    "# data_processing_dir = location_dict[location_input]\n",
    "# print(\"base directory is \" + base_dir)\n",
    "# os.chdir(base_dir) \n",
    "print(\"I am about to import a library\")\n",
    "from core import expression_plots \n",
    "from core import io_library "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "#from intermine.webservice import Service\n",
    "#data_processing_dir = '/home/heineike/github/expression_broad_data/expression_data/'\n",
    "data_processing_dir = base_dir + os.path.normpath(\"\\expression_data\") + os.sep\n",
    "#service = Service(\"https://yeastmine.yeastgenome.org:443/yeastmine/service\", token = \"L155qee34elf4eY1a9H5\")\n",
    "import scipy.stats as stats\n",
    "import plotly.plotly as py\n",
    "py.sign_in('heineike02', 'APPjKrtARaN2ZgUYIkqr')\n",
    "#import plotly.figure_factory as ff\n",
    "#import pandas as pd\n",
    "from plotly.graph_objs import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "kl_sc_PKA_data = pd.read_pickle(data_processing_dir + \"20170817_klscpka\" + os.sep + \"kl_sc_PKA_data_SDC_t40_AsMinvAsPlus.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genes\n",
    "kl_sc_PKA_data[[kl_sc_PKA_data['SC_common_name'].isin([])]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N SCer Orthologs</th>\n",
       "      <th>kl_genename</th>\n",
       "      <th>sc_genename</th>\n",
       "      <th>PKA(AS)+1NMPP1_KL</th>\n",
       "      <th>padj_KL</th>\n",
       "      <th>PKA(AS)+1NMPP1_SC</th>\n",
       "      <th>padj_SC</th>\n",
       "      <th>SC_common_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>332</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0D05643g</td>\n",
       "      <td>YLR075W</td>\n",
       "      <td>-3.305108</td>\n",
       "      <td>6.044718e-14</td>\n",
       "      <td>-1.655790</td>\n",
       "      <td>2.180809e-03</td>\n",
       "      <td>RPL10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4741</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0D12540g</td>\n",
       "      <td>YPL249C-A</td>\n",
       "      <td>-4.099702</td>\n",
       "      <td>1.917858e-18</td>\n",
       "      <td>-3.259622</td>\n",
       "      <td>1.015249e-29</td>\n",
       "      <td>RPL36B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      N SCer Orthologs   kl_genename sc_genename  PKA(AS)+1NMPP1_KL  \\\n",
       "332                  1  KLLA0D05643g     YLR075W          -3.305108   \n",
       "4741                 2  KLLA0D12540g   YPL249C-A          -4.099702   \n",
       "\n",
       "           padj_KL  PKA(AS)+1NMPP1_SC       padj_SC SC_common_name  \n",
       "332   6.044718e-14          -1.655790  2.180809e-03          RPL10  \n",
       "4741  1.917858e-18          -3.259622  1.015249e-29         RPL36B  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_of_interest = ['RPL36B','RPL10']\n",
    "#['YAP1','HAP2','HAP4','HAP5','SIP2','CAT8','SIP4']\n",
    "\n",
    "kl_sc_PKA_data[kl_sc_PKA_data['SC_common_name'].isin(genes_of_interest)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[462, 486, 319, 3709]\n"
     ]
    }
   ],
   "source": [
    "#Highlight hits that are statistically significant for K.Lactis \n",
    "\n",
    "#Add column which highlights only pvalues below a threshold\n",
    "\n",
    "#should eventually have a principled way of picking these\n",
    "pthreshold_KL = 10e-5\n",
    "pthreshold_SC = 10e-9\n",
    "\n",
    "# stre_eq1 = [kl_gene for kl_gene, n in jso_prom_hits_kl_hitcount.items() if n==1]\n",
    "# stre_gt1 = [kl_gene for kl_gene, n in jso_prom_hits_kl_hitcount.items() if n>1]\n",
    "\n",
    "kl_sc_PKA_data_klsig = kl_sc_PKA_data[kl_sc_PKA_data['padj_KL'] < pthreshold_KL]\n",
    "kl_sc_PKA_data_scsig = kl_sc_PKA_data[kl_sc_PKA_data['padj_SC'] < pthreshold_SC]\n",
    "\n",
    "klscsig_ind = set(kl_sc_PKA_data_klsig.index)&set(kl_sc_PKA_data_scsig.index)\n",
    "klsig_scunsig_ind = set(kl_sc_PKA_data_klsig.index)-klscsig_ind\n",
    "scsig_klunsig_ind = set(kl_sc_PKA_data_scsig.index)-klscsig_ind\n",
    "unsig_ind = set(kl_sc_PKA_data.index)-(set(kl_sc_PKA_data_klsig.index)|set(kl_sc_PKA_data_scsig.index))\n",
    "\n",
    "print([len(klscsig_ind),len(klsig_scunsig_ind),len(scsig_klunsig_ind),len(unsig_ind)])\n",
    "\n",
    "kl_sc_PKA_data_klsig_scunsig = kl_sc_PKA_data.loc[list(klsig_scunsig_ind)]\n",
    "kl_sc_PKA_data_scsig_klunsig = kl_sc_PKA_data.loc[list(scsig_klunsig_ind)]\n",
    "kl_sc_PKA_data_klscsig = kl_sc_PKA_data.loc[list(klscsig_ind)]\n",
    "kl_sc_PKA_data_unsig = kl_sc_PKA_data.loc[list(unsig_ind)]\n",
    "\n",
    "# print([len(kl_sc_PKA_data_klscsig),len(kl_sc_PKA_data_klsig_scunsig),len(kl_sc_PKA_data_scsig_klunsig),len(kl_sc_PKA_data_unsig)])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'YKR076W YOR163W YNL015W YNL305C YOR176W YPR160W YPR098C YDR357C YPR155C YJL161W YOL071W YKR049C YLR178C YJL057C YBR280C YDL149W YOR223W YHR087W YML128C YPL222W YOL032W YJL068C YAL060W YMR169C YMR087W YIL155C YHL021C YML004C YOR161C YBR285W YHL036W YJR036C YPR026W YKL107W YLR149C YJL153C YGR255C YCL039W YOL096C YER079W YKL086W YOL048C YDR262W YFR045W YHR176W YJL020C YKL065W-A YOR367W YCR061W YBR126C YIL160C YOR374W YHR075C YMR196W YKL067W YMR107W YPL004C YBR056W YOL084W YPL196W YDR100W YDR231C YPL154C YFL014W YJL016W YBR269C YLR345W YIL136W YCL038C YDL020C YJL042W YPL003W YBR149W YCR004C YOR292C YGL227W YBL064C YDL078C YIL124W YML070W YKL085W YLL023C YBR062C YKL151C YPR151C YHR140W YKR011C YKL103C YHR104W YNL115C YDL126C YMR110C YDL124W YKL142W YBR026C YIL017C YOR228C YNL274C YBR230C YGL037C YLR438W YNL242W YBL078C YHR008C YNL200C YOR289W YBR053C YNL134C YIL065C YBL086C YGR088W YOL065C YEL012W YLL039C YER142C YDR171W YEL011W YPL123C YPL109C YKL193C YGL141W YDL169C YBR072W YBR286W YBR006W YIL033C YML054C YAL061W YGR236C YMR170C YGR174C YCR011C YOR285W YMR280C YCR045C YPL206C YLR070C YNL237W YFR003C YDR379C-A YDR070C YMR250W YMR135C YJL132W YIL108W YGL006W YJR008W YIL107C YKL162C YGL205W YJL163C YER182W YPL113C YDL110C YLR152C YIL055C YIL101C YOL013C YML042W YGR194C YOR152C YJL185C YIL087C YHR138C YMR297W YPL006W YJR096W YNL202W YKL010C YOR036W YDR074W YLL026W YAL008W YLR241W YGR130C YPL186C YBL029C-A YGR281W YPL247C YNL321W YOR215C YLR284C YNL100W YIR014W YNL045W YGR052W YKL121W YMR031C YBR001C YCL040W YDR516C YDR018C YER096W YGR256W YGR254W YHR106W YAL028W YLR177W YDR055W YBL075C YER103W YHR160C YFR015C YLR258W YPR149W YHR016C YDL204W YAL034C YKL141W YMR118C YEL060C YER119C YBL101C YPR030W YDL085W YHR096C YMR271C YNR002C YBR037C YDR272W YAL030W YER054C YER067W YNL055C YBR147W YLR270W YOR173W YJL079C YNL270C YHR097C YIL111W YNL014W YGR248W YNR001C YNL194C YFR053C YDR034W-B YNR014W YFR017C YLR248W YAL017W YLR080W YJL164C YPR172W YPL230W YBR052C YDR032C YJR059W YEL041W YOR185C YMR175W YPL223C YGR008C YLR327C YDL130W-A YGR043C YIL113W YBR241C YGL104C YGR209C YFL042C YCL035C YDR513W YCL048W-A YGL229C YNL195C YDL223C YDR436W YML100W YMR261C YIL066C YJL045W YDR085C YDR475C YOR019W YDR059C YBR214W YKR058W YER101C YKL091C YLR356W YKL035W YGR243W YLR164W YDL022W YNR034W-A YDL019C YER150W YDR453C YBR117C YLR174W YMR105C YGR070W YER039C YAL062W YBR169C YDL021W YBR183W YEL039C'"
      ]
     },
     "execution_count": 284,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\" \".join(kl_sc_PKA_data_scsig[kl_sc_PKA_data_scsig['PKA(AS)+1NMPP1_SC']>0]['sc_genename'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes that are up in S.Cer and Down in K.Lac\n",
      "['YNL014W']\n",
      "Genes that are up in K.Lac and Down in S.Cer\n",
      "['YJL208C']\n"
     ]
    }
   ],
   "source": [
    "#Separate out activated from repressed genes\n",
    "\n",
    "#Identifying genes that switch from activated to repressed\n",
    "\n",
    "#Significant for both, but up in SC and down in kl. \n",
    "upsc_downkl = list(kl_sc_PKA_data_klscsig[(kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_SC']>0)&(kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_KL']<0)]['sc_genename'])\n",
    "print(\"Genes that are up in S.Cer and Down in K.Lac\")\n",
    "print(upsc_downkl)\n",
    "#HEF3 {'YNL014W'}\n",
    "\n",
    "upkl_downsc = list(kl_sc_PKA_data_klscsig[(kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_SC']<0)&(kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_KL']>0)]['sc_genename'])\n",
    "print(\"Genes that are up in K.Lac and Down in S.Cer\")\n",
    "print(upkl_downsc)\n",
    "\n",
    "#YJL208C\n",
    "#NUC1 up in KL and down in SC\n",
    "#mitochondrial nuclease.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Set up go term list and database: \n",
    "GO_aspect = 'P'\n",
    "#The three GO_aspect are: \n",
    "#C = cellular_component\n",
    "#F = molecular_function\n",
    "#P = biological_process\n",
    "\n",
    "\n",
    "go_slims = pd.read_table(data_processing_dir + os.path.normpath('go_terms/go_slim_mapping.tab'),header = None)\n",
    "go_slims.columns = ['sc_genename','sc_common_name','sgd_ID','GO_aspect','GO_term','GO_term_ID','feature_type']\n",
    "\n",
    "go_slims_aspect = go_slims[go_slims['GO_aspect']==GO_aspect]\n",
    "go_term_list = list(set(go_slims_aspect['GO_term']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#See which genes correspond to JSO targets. \n",
    "jso_msn2_targets = ['SSA1', 'SSA4', 'HSP104', 'CTT1', 'TPK1', 'TFS1', 'DDR2', 'TSA1', 'HSP12', 'TMA10', 'TPS2', 'RTC3', 'GPH1', 'HXK1', 'TSA2', 'CYC7', 'PNC1', 'GSY1', 'FMP48', 'SS13', 'ALD4', 'YNR014W', 'HOR2', 'YOR173W', 'RAS2', 'GLK1', 'GPD1', 'HSP26', 'SIP18', 'TKL2', 'FMP16', 'ALD3', 'PGM2']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes common name</th>\n",
       "      <th>N genes in goterm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>transmembrane transport</th>\n",
       "      <td>[GPD1, SSA4]</td>\n",
       "      <td>233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein targeting</th>\n",
       "      <td>[GPD1, RAS2, SSA4]</td>\n",
       "      <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sporulation</th>\n",
       "      <td>[RAS2]</td>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein folding</th>\n",
       "      <td>[HSP26, TSA2, SSA4, TSA1, HSP104]</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>response to chemical</th>\n",
       "      <td>[RAS2, SIP18, CTT1, HSP12, TSA1, HSP104, TSA2]</td>\n",
       "      <td>446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regulation of cell cycle</th>\n",
       "      <td>[TSA1]</td>\n",
       "      <td>233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>meiotic cell cycle</th>\n",
       "      <td>[RAS2]</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>response to oxidative stress</th>\n",
       "      <td>[CTT1, TSA2, HSP12, TSA1, HSP104]</td>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>signaling</th>\n",
       "      <td>[TPK1, TFS1]</td>\n",
       "      <td>245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellular amino acid metabolic process</th>\n",
       "      <td>[ALD3]</td>\n",
       "      <td>199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regulation of DNA metabolic process</th>\n",
       "      <td>[PNC1]</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>response to heat</th>\n",
       "      <td>[HSP12, HSP26, SSA4, TPS2, HSP104]</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chromatin organization</th>\n",
       "      <td>[PNC1]</td>\n",
       "      <td>318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>carbohydrate metabolic process</th>\n",
       "      <td>[GLK1, HXK1, TPS2, PGM2, GSY1, TSA1, HSP104, G...</td>\n",
       "      <td>198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cofactor metabolic process</th>\n",
       "      <td>[GLK1, HXK1, ALD4, GPD1, TKL2, PNC1]</td>\n",
       "      <td>179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein phosphorylation</th>\n",
       "      <td>[TPK1, FMP48]</td>\n",
       "      <td>197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellular respiration</th>\n",
       "      <td>[CYC7]</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biological_process</th>\n",
       "      <td>[FMP16, YNR014W, TMA10]</td>\n",
       "      <td>1801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oligosaccharide metabolic process</th>\n",
       "      <td>[PGM2, TPS2, HSP104]</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>monocarboxylic acid metabolic process</th>\n",
       "      <td>[ALD4, GLK1, HXK1]</td>\n",
       "      <td>140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>other</th>\n",
       "      <td>[RTC3]</td>\n",
       "      <td>249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nucleobase-containing small molecule metabolic process</th>\n",
       "      <td>[RAS2, GLK1, HXK1, ALD4, GPD1, CYC7, TKL2, PNC1]</td>\n",
       "      <td>198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudohyphal growth</th>\n",
       "      <td>[RAS2]</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellular response to DNA damage stimulus</th>\n",
       "      <td>[TSA1]</td>\n",
       "      <td>299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>response to osmotic stress</th>\n",
       "      <td>[CTT1, GPD1, HSP12]</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellular ion homeostasis</th>\n",
       "      <td>[PGM2]</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>peroxisome organization</th>\n",
       "      <td>[GPD1]</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>carbohydrate transport</th>\n",
       "      <td>[GLK1, HXK1]</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>generation of precursor metabolites and energy</th>\n",
       "      <td>[GLK1, HXK1, PGM2, GSY1, CYC7, GPH1]</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                    genes common name  \\\n",
       "transmembrane transport                                                                  [GPD1, SSA4]   \n",
       "protein targeting                                                                  [GPD1, RAS2, SSA4]   \n",
       "sporulation                                                                                    [RAS2]   \n",
       "protein folding                                                     [HSP26, TSA2, SSA4, TSA1, HSP104]   \n",
       "response to chemical                                   [RAS2, SIP18, CTT1, HSP12, TSA1, HSP104, TSA2]   \n",
       "regulation of cell cycle                                                                       [TSA1]   \n",
       "meiotic cell cycle                                                                             [RAS2]   \n",
       "response to oxidative stress                                        [CTT1, TSA2, HSP12, TSA1, HSP104]   \n",
       "signaling                                                                                [TPK1, TFS1]   \n",
       "cellular amino acid metabolic process                                                          [ALD3]   \n",
       "regulation of DNA metabolic process                                                            [PNC1]   \n",
       "response to heat                                                   [HSP12, HSP26, SSA4, TPS2, HSP104]   \n",
       "chromatin organization                                                                         [PNC1]   \n",
       "carbohydrate metabolic process                      [GLK1, HXK1, TPS2, PGM2, GSY1, TSA1, HSP104, G...   \n",
       "cofactor metabolic process                                       [GLK1, HXK1, ALD4, GPD1, TKL2, PNC1]   \n",
       "protein phosphorylation                                                                 [TPK1, FMP48]   \n",
       "cellular respiration                                                                           [CYC7]   \n",
       "biological_process                                                            [FMP16, YNR014W, TMA10]   \n",
       "oligosaccharide metabolic process                                                [PGM2, TPS2, HSP104]   \n",
       "monocarboxylic acid metabolic process                                              [ALD4, GLK1, HXK1]   \n",
       "other                                                                                          [RTC3]   \n",
       "nucleobase-containing small molecule metabolic ...   [RAS2, GLK1, HXK1, ALD4, GPD1, CYC7, TKL2, PNC1]   \n",
       "pseudohyphal growth                                                                            [RAS2]   \n",
       "cellular response to DNA damage stimulus                                                       [TSA1]   \n",
       "response to osmotic stress                                                        [CTT1, GPD1, HSP12]   \n",
       "cellular ion homeostasis                                                                       [PGM2]   \n",
       "peroxisome organization                                                                        [GPD1]   \n",
       "carbohydrate transport                                                                   [GLK1, HXK1]   \n",
       "generation of precursor metabolites and energy                   [GLK1, HXK1, PGM2, GSY1, CYC7, GPH1]   \n",
       "\n",
       "                                                    N genes in goterm  \n",
       "transmembrane transport                                           233  \n",
       "protein targeting                                                 310  \n",
       "sporulation                                                       138  \n",
       "protein folding                                                    95  \n",
       "response to chemical                                              446  \n",
       "regulation of cell cycle                                          233  \n",
       "meiotic cell cycle                                                287  \n",
       "response to oxidative stress                                      105  \n",
       "signaling                                                         245  \n",
       "cellular amino acid metabolic process                             199  \n",
       "regulation of DNA metabolic process                               107  \n",
       "response to heat                                                   73  \n",
       "chromatin organization                                            318  \n",
       "carbohydrate metabolic process                                    198  \n",
       "cofactor metabolic process                                        179  \n",
       "protein phosphorylation                                           197  \n",
       "cellular respiration                                               84  \n",
       "biological_process                                               1801  \n",
       "oligosaccharide metabolic process                                  27  \n",
       "monocarboxylic acid metabolic process                             140  \n",
       "other                                                             249  \n",
       "nucleobase-containing small molecule metabolic ...                198  \n",
       "pseudohyphal growth                                                74  \n",
       "cellular response to DNA damage stimulus                          299  \n",
       "response to osmotic stress                                         92  \n",
       "cellular ion homeostasis                                          128  \n",
       "peroxisome organization                                            50  \n",
       "carbohydrate transport                                             33  \n",
       "generation of precursor metabolites and energy                    155  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_set_data = kl_sc_PKA_data[kl_sc_PKA_data['SC_common_name'].isin(jso_msn2_targets)]\n",
    "gene_set_list = gene_set_data['sc_genename']\n",
    "\n",
    "\n",
    "go_term_data = []\n",
    "\n",
    "go_term_index = []\n",
    "\n",
    "for term in go_term_list: \n",
    "    term_genes = list(go_slims_aspect[go_slims_aspect['GO_term']==term]['sc_genename'])\n",
    "    if len(term_genes)> len(set(term_genes)):\n",
    "        print(\"Duplicate Term: \" + term)\n",
    "    subset_genes_in_goterm =  set(gene_set_list) & set(term_genes)\n",
    "    N_subset_genes_in_goterm = len(subset_genes_in_goterm)\n",
    "    N_genes_in_goterm = len(term_genes)\n",
    "    if N_subset_genes_in_goterm >0:\n",
    "        subset_genes_in_goterm_commonname = io_library.SC_common_name_lookup(subset_genes_in_goterm)\n",
    "        go_term_data.append((N_subset_genes_in_goterm,\n",
    "                             subset_genes_in_goterm,\n",
    "                             subset_genes_in_goterm_commonname,\n",
    "                             N_genes_in_goterm))\n",
    "        go_term_index.append(term)\n",
    "\n",
    "go_term_df = pd.DataFrame(go_term_data, index = go_term_index,columns = ['N subset genes in goterm',\n",
    "                                                                            'genes',\n",
    "                                                                            'genes common name',\n",
    "                                                                            'N genes in goterm'])\n",
    "go_term_df[['genes common name','N genes in goterm']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The following missing genes are filtered out of the background set:\n",
      "{'Scer_YGOB_ADL119W', 'Scer_YGOB_Anc_7.495', 'Scer_YGOB_YDR134C', 'Scer_RDN25-1', 'Scer_YGOB_SDC25', 'Scer_RDN5-1', 'Scer_RDN58-1', 'Scer_RDN18-1'}\n",
      "klsig_act\n",
      "The following missing genes are filtered out:\n",
      "{'Scer_YGOB_SDC25'}\n",
      "klsig_act not_yet_annotated\n",
      "klsig_rep\n",
      "The following missing genes are filtered out:\n",
      "set()\n",
      "klsig_rep not_yet_annotated\n",
      "scsig_act\n",
      "The following missing genes are filtered out:\n",
      "set()\n",
      "scsig_act not_yet_annotated\n",
      "scsig_rep\n",
      "The following missing genes are filtered out:\n",
      "set()\n",
      "scsig_rep not_yet_annotated\n",
      "klscsig_act\n",
      "The following missing genes are filtered out:\n",
      "set()\n",
      "klscsig_act not_yet_annotated\n",
      "klscsig_rep\n",
      "The following missing genes are filtered out:\n",
      "set()\n",
      "klscsig_rep not_yet_annotated\n"
     ]
    }
   ],
   "source": [
    "#Add in the one gene that is significant for both but repressed in SC and activated in KL. \n",
    "klsig_act_genes = list(kl_sc_PKA_data_klsig_scunsig[kl_sc_PKA_data_klsig_scunsig['PKA(AS)+1NMPP1_KL']>0]['sc_genename'])\n",
    "for gene in upkl_downsc: \n",
    "    klsig_act_genes.append(gene)\n",
    "\n",
    "#Add in the one gene that is significant for both but activated in SC and repressed in KL.\n",
    "klsig_rep_genes = list(kl_sc_PKA_data_klsig_scunsig[kl_sc_PKA_data_klsig_scunsig['PKA(AS)+1NMPP1_KL']<0]['sc_genename'])\n",
    "for gene in upsc_downkl: \n",
    "    klsig_rep_genes.append(gene)\n",
    "\n",
    "#Add in the one gene that is significant for both but activated in SC and repressed in KL.\n",
    "scsig_act_genes = list(kl_sc_PKA_data_scsig_klunsig[kl_sc_PKA_data_scsig_klunsig['PKA(AS)+1NMPP1_SC']>0]['sc_genename'])\n",
    "for gene in upsc_downkl: \n",
    "    scsig_act_genes.append(gene)\n",
    "\n",
    "#Add in the one gene that is significant for both but repressed in SC and activated in KL. \n",
    "scsig_rep_genes = list(kl_sc_PKA_data_scsig_klunsig[kl_sc_PKA_data_scsig_klunsig['PKA(AS)+1NMPP1_SC']<0]['sc_genename'])\n",
    "for gene in upkl_downsc: \n",
    "    scsig_rep_genes.append(gene)\n",
    "\n",
    "gene_sets = ['klsig_act', 'klsig_rep','scsig_act','scsig_rep','klscsig_act','klscsig_rep']\n",
    "\n",
    "gene_set_dict = {'klsig_act': klsig_act_genes,\n",
    "                 'klsig_rep': klsig_rep_genes,\n",
    "                 'scsig_act': scsig_act_genes,\n",
    "                 'scsig_rep': scsig_rep_genes,\n",
    "                 'klscsig_act': list(kl_sc_PKA_data_klscsig[(kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_SC']>0) & (kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_KL']>0)]['sc_genename']),\n",
    "                 'klscsig_rep': list(kl_sc_PKA_data_klscsig[(kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_SC']<0) & (kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_KL']<0)]['sc_genename'])\n",
    "                }\n",
    "\n",
    "missing_genes = ['Scer_YGOB_YDR134C', 'Scer_RDN18-1', \n",
    "                 'Scer_YGOB_SDC25', 'Scer_RDN25-1', \n",
    "                 'Scer_RDN58-1', 'Scer_YGOB_Anc_7.495', \n",
    "                 'Scer_YGOB_ADL119W', 'Scer_RDN5-1']\n",
    "\n",
    "pval_cutoff = 0.01\n",
    "\n",
    "background_genes = list(kl_sc_PKA_data['sc_genename'])\n",
    "print(\"The following missing genes are filtered out of the background set:\")\n",
    "print(set(background_genes) & set(missing_genes))\n",
    "background_genes_filtered = list(set(background_genes)-set(missing_genes))\n",
    "\n",
    "\n",
    "# #Set up go term list and database: \n",
    "# GO_aspect = 'P'\n",
    "# #The three GO_aspect are: \n",
    "# #C = cellular_component\n",
    "# #F = molecular_function\n",
    "# #P = biological_process\n",
    "\n",
    "\n",
    "# go_slims = pd.read_table(data_processing_dir + os.path.normpath('go_terms/go_slim_mapping.tab'),header = None)\n",
    "# go_slims.columns = ['sc_genename','sc_common_name','sgd_ID','GO_aspect','GO_term','GO_term_ID','feature_type']\n",
    "\n",
    "# go_slims_aspect = go_slims[go_slims['GO_aspect']==GO_aspect]\n",
    "# go_term_list = list(set(go_slims_aspect['GO_term']))\n",
    "\n",
    "gene_set_enrichment = {}\n",
    "\n",
    "for gene_set in gene_sets: \n",
    "    print(gene_set)\n",
    "    gene_set_list = gene_set_dict[gene_set]\n",
    "\n",
    "    #filter out missing genes \n",
    "    print(\"The following missing genes are filtered out:\")\n",
    "    print(set(gene_set_list) & set(missing_genes))\n",
    "    gene_set_list_filtered = list(set(gene_set_list)-set(missing_genes))\n",
    "    go_term_data = []\n",
    "    go_term_list_filtered = go_term_list.copy()\n",
    "    for term in go_term_list: \n",
    "        term_genes = list(go_slims_aspect[go_slims_aspect['GO_term']==term]['sc_genename'])\n",
    "        if len(term_genes)> len(set(term_genes)):\n",
    "            print(\"Duplicate Term for term \" + term)\n",
    "        subset_genes_in_goterm =  set(gene_set_list) & set(term_genes)\n",
    "        N_subset_genes_in_goterm = len(subset_genes_in_goterm)\n",
    "        N_subset_genes_notin_goterm = len(gene_set_list)-N_subset_genes_in_goterm\n",
    "        N_bg_genes_in_goterm = len(set(background_genes_filtered) & set(term_genes))\n",
    "        if N_bg_genes_in_goterm >0:\n",
    "            N_bg_genes_notin_goterm = len(background_genes_filtered)-N_bg_genes_in_goterm\n",
    "            oddsratio, pvalue = stats.fisher_exact([[N_subset_genes_in_goterm, N_bg_genes_in_goterm], [N_subset_genes_notin_goterm, N_bg_genes_notin_goterm]],alternative = 'greater')\n",
    "            subset_genes_in_goterm_commonname = io_library.SC_common_name_lookup(subset_genes_in_goterm)\n",
    "            go_term_data.append((N_subset_genes_in_goterm,\n",
    "                                 len(gene_set_list),\n",
    "                                 float(N_subset_genes_in_goterm)/float(len(gene_set_list)),\n",
    "                                 float(N_bg_genes_in_goterm)/float(len(background_genes_filtered)),\n",
    "                                 float(N_subset_genes_in_goterm)/float(N_bg_genes_in_goterm),\n",
    "                                 pvalue,\n",
    "                                 subset_genes_in_goterm,\n",
    "                                 subset_genes_in_goterm_commonname,\n",
    "                                 oddsratio))\n",
    "        else:\n",
    "            print(gene_set + \" \" + term)\n",
    "            go_term_list_filtered.remove(term) \n",
    "            \n",
    "    go_term_df = pd.DataFrame(go_term_data, index = go_term_list_filtered,columns = ['N subset genes in goterm',\n",
    "                                                                            'N genes in subset',\n",
    "                                                                            'pct goterm in subset',\n",
    "                                                                            'pct go term in background',\n",
    "                                                                            'pct of go terms genes in subset',\n",
    "                                                                            'pvalue',\n",
    "                                                                            'genes',\n",
    "                                                                            'genes common name',\n",
    "                                                                            'oddsratio'])\n",
    "    gene_set_enrichment[gene_set] = go_term_df\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "#     new_query = query_all.clone()\n",
    "    \n",
    "#     #subset query based on desired gene set\n",
    "#     new_query.add_constraint(\"Gene.secondaryIdentifier\", \"ONE OF\", gene_set_list_filtered)\n",
    "    \n",
    "#     #Generate a list\n",
    "#     new_list = service.create_list(new_query.select('Gene.id'), \"Gene\", name = \"20170905\" + gene_set)\n",
    "    \n",
    "#     #Run Go enrichment on the list\n",
    "#     results = new_list.calculate_enrichment('go_enrichment_for_gene')\n",
    "#     results_list = list(results)    \n",
    "#     for result in results_list: \n",
    "#         print(result)\n",
    "#         if result['p-value']<pval_cutoff:\n",
    "#             print(result)\n",
    "#             print(result)\n",
    "            \n",
    "#     #Given a P-value cutoff\n",
    "#     # Save the go term \n",
    "#     # Save the contingency table\n",
    "#     # S"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['generation of precursor metabolites and energy',\n",
       " 'regulation of cell cycle',\n",
       " 'biological_process',\n",
       " 'carbohydrate metabolic process',\n",
       " 'monocarboxylic acid metabolic process',\n",
       " 'transposition',\n",
       " 'response to oxidative stress',\n",
       " 'sporulation',\n",
       " 'not_yet_annotated',\n",
       " 'endosomal transport',\n",
       " 'protein modification by small protein conjugation or removal',\n",
       " 'Golgi vesicle transport',\n",
       " 'translational elongation',\n",
       " 'ion transport',\n",
       " 'cytoskeleton organization',\n",
       " 'mitochondrion organization',\n",
       " 'protein lipidation',\n",
       " 'ribosomal large subunit biogenesis',\n",
       " 'histone modification',\n",
       " 'organelle fusion',\n",
       " 'protein dephosphorylation',\n",
       " 'oligosaccharide metabolic process',\n",
       " 'RNA splicing',\n",
       " 'pseudohyphal growth',\n",
       " 'RNA modification',\n",
       " 'DNA replication',\n",
       " 'cytokinesis',\n",
       " 'telomere organization',\n",
       " 'other',\n",
       " 'nuclear transport',\n",
       " 'translational initiation',\n",
       " 'conjugation',\n",
       " 'transmembrane transport',\n",
       " 'protein phosphorylation',\n",
       " 'invasive growth in response to glucose limitation',\n",
       " 'organelle inheritance',\n",
       " 'response to chemical',\n",
       " 'regulation of organelle organization',\n",
       " 'cell wall organization or biogenesis',\n",
       " 'amino acid transport',\n",
       " 'snoRNA processing',\n",
       " 'DNA-templated transcription, termination',\n",
       " 'cellular ion homeostasis',\n",
       " 'vitamin metabolic process',\n",
       " 'protein folding',\n",
       " 'vesicle organization',\n",
       " 'nucleobase-containing compound transport',\n",
       " 'cytoplasmic translation',\n",
       " 'protein acylation',\n",
       " 'protein glycosylation',\n",
       " 'protein targeting',\n",
       " 'regulation of DNA metabolic process',\n",
       " 'cofactor metabolic process',\n",
       " 'transcription from RNA polymerase II promoter',\n",
       " 'lipid transport',\n",
       " 'endocytosis',\n",
       " 'mRNA processing',\n",
       " 'ribosome assembly',\n",
       " 'rRNA processing',\n",
       " 'regulation of translation',\n",
       " 'peroxisome organization',\n",
       " 'protein alkylation',\n",
       " 'DNA repair',\n",
       " 'protein maturation',\n",
       " 'RNA catabolic process',\n",
       " 'response to heat',\n",
       " 'meiotic cell cycle',\n",
       " 'regulation of transport',\n",
       " 'lipid metabolic process',\n",
       " 'transcription from RNA polymerase I promoter',\n",
       " 'DNA-templated transcription, elongation',\n",
       " 'tRNA processing',\n",
       " 'nucleobase-containing small molecule metabolic process',\n",
       " 'cell budding',\n",
       " 'carbohydrate transport',\n",
       " 'cellular respiration',\n",
       " 'organelle fission',\n",
       " 'mitotic cell cycle',\n",
       " 'chromosome segregation',\n",
       " 'proteolysis involved in cellular protein catabolic process',\n",
       " 'organelle assembly',\n",
       " 'mitochondrial translation',\n",
       " 'signaling',\n",
       " 'nucleus organization',\n",
       " 'vacuole organization',\n",
       " 'cellular response to DNA damage stimulus',\n",
       " 'ribosomal small subunit biogenesis',\n",
       " 'DNA-templated transcription, initiation',\n",
       " 'membrane fusion',\n",
       " 'DNA recombination',\n",
       " 'chromatin organization',\n",
       " 'cellular amino acid metabolic process',\n",
       " 'transcription from RNA polymerase III promoter',\n",
       " 'ribosomal subunit export from nucleus',\n",
       " 'response to starvation',\n",
       " 'tRNA aminoacylation for protein translation',\n",
       " 'regulation of protein modification process',\n",
       " 'peptidyl-amino acid modification',\n",
       " 'protein complex biogenesis',\n",
       " 'cell morphogenesis',\n",
       " 'exocytosis',\n",
       " 'response to osmotic stress']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_term_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['not_yet_annotated']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_term_list_filtered = go_term_list.copy()\n",
    "go_term_list_filtered.remove('not_yet_annotated')\n",
    "[term for term in go_term_list if term =='not_yet_annotated']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pd.set_option('display.max_colwidth', -1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pvalue</th>\n",
       "      <th>N subset genes in goterm</th>\n",
       "      <th>pct goterm in subset</th>\n",
       "      <th>pct go term in background</th>\n",
       "      <th>pct of go terms genes in subset</th>\n",
       "      <th>genes common name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>lipid metabolic process</th>\n",
       "      <td>0.000327</td>\n",
       "      <td>28</td>\n",
       "      <td>0.120172</td>\n",
       "      <td>0.057770</td>\n",
       "      <td>0.097561</td>\n",
       "      <td>[TGL2, PIP2, CHO1, VPS34, TGL4, FAA2, DPL1, MPO1, TSC13, ATG26, ICL2, TGL3, PLB1, FOX2, CIT3, UME6, ATG15, NEM1, PDH1, LPX1, HTD2, SCS3, INP51, OAF1, GPT2, TGL5, PLB3, CRC1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>biological_process</th>\n",
       "      <td>0.000896</td>\n",
       "      <td>45</td>\n",
       "      <td>0.193133</td>\n",
       "      <td>0.118357</td>\n",
       "      <td>0.076531</td>\n",
       "      <td>[NIT2, YLR173W, YLR257W, YGR015C, YDR415C, EHD3, YIL024C, YLR456W, DOS2, YML020W, YDR222W, OSW5, YMR226C, GPM3, YML119W, YNL165W, SCM4, MRX3, YIR035C, YET3, YMR102C, YGL081W, YOL024W, YNL011C, YBR138C, YKL050C, MOH1, SNA2, YMR160W, YGR127W, YMR114C, YPL071C, YDL114W, FUB1, YLR361C-A, YIL152W, LEE1, JID1, TMS1, YDR109C, YJR111C, YLR225C, UBP11, YNL010W, YDR461C-A]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>monocarboxylic acid metabolic process</th>\n",
       "      <td>0.003994</td>\n",
       "      <td>14</td>\n",
       "      <td>0.060086</td>\n",
       "      <td>0.025564</td>\n",
       "      <td>0.110236</td>\n",
       "      <td>[TSC13, UGA1, PIP2, FOX2, CIT3, HTD2, ACS1, FAA2, OAF1, ICL2, GLO4, PDH1, CRC1, PGI1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>conjugation</th>\n",
       "      <td>0.027618</td>\n",
       "      <td>11</td>\n",
       "      <td>0.047210</td>\n",
       "      <td>0.023551</td>\n",
       "      <td>0.094017</td>\n",
       "      <td>[RCE1, FAR1, GPA1, FUS2, STE5, AXL1, AFB1, STE2, FUS3, PRM2, STE18]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>other</th>\n",
       "      <td>0.033653</td>\n",
       "      <td>15</td>\n",
       "      <td>0.064378</td>\n",
       "      <td>0.037238</td>\n",
       "      <td>0.081081</td>\n",
       "      <td>[DAL2, SYM1, YAT1, YBR139W, ATG38, OXP1, GDI1, YHR112C, YOL075C, YSP3, SRL3, LPL1, YGL114W, STE6, YUH1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>response to chemical</th>\n",
       "      <td>0.034977</td>\n",
       "      <td>27</td>\n",
       "      <td>0.115880</td>\n",
       "      <td>0.079308</td>\n",
       "      <td>0.068528</td>\n",
       "      <td>[DIG2, ASI1, YKL069W, PIP2, ECM38, STE5, AHP1, AFB1, STE2, TRX3, TMC1, ATG19, FAR1, IZH2, FUS3, ASI3, RDS2, DIG1, UME6, GPA1, CIN5, FUS2, FRD1, OAF1, MXR2, CTA1, STE18]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein maturation</th>\n",
       "      <td>0.035121</td>\n",
       "      <td>6</td>\n",
       "      <td>0.025751</td>\n",
       "      <td>0.009863</td>\n",
       "      <td>0.122449</td>\n",
       "      <td>[ATG19, ATG18, RCE1, DAP2, AXL1, ATG15]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>carbohydrate metabolic process</th>\n",
       "      <td>0.039895</td>\n",
       "      <td>14</td>\n",
       "      <td>0.060086</td>\n",
       "      <td>0.034823</td>\n",
       "      <td>0.080925</td>\n",
       "      <td>[RBK1, PGM1, PYC2, RDS2, FYV10, GAL80, UME6, RMD5, PIG2, SGA1, PYC1, DOG1, FBP1, PGI1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sporulation</th>\n",
       "      <td>0.060024</td>\n",
       "      <td>10</td>\n",
       "      <td>0.042918</td>\n",
       "      <td>0.023752</td>\n",
       "      <td>0.084746</td>\n",
       "      <td>[MSO1, VPS13, UBX6, SPO71, PMD1, MDS3, FMP45, ATG26, UBX7, PTP2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>endosomal transport</th>\n",
       "      <td>0.071305</td>\n",
       "      <td>8</td>\n",
       "      <td>0.034335</td>\n",
       "      <td>0.018116</td>\n",
       "      <td>0.088889</td>\n",
       "      <td>[ATG20, ATG18, SNX4, BTN2, SNX41, VPS68, MUK1, VPS13]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lipid transport</th>\n",
       "      <td>0.082644</td>\n",
       "      <td>6</td>\n",
       "      <td>0.025751</td>\n",
       "      <td>0.012480</td>\n",
       "      <td>0.096774</td>\n",
       "      <td>[DET1, YSP2, ACB1, LAM4, PXA1, PXA2]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         pvalue  N subset genes in goterm  \\\n",
       "lipid metabolic process                0.000327  28                         \n",
       "biological_process                     0.000896  45                         \n",
       "monocarboxylic acid metabolic process  0.003994  14                         \n",
       "conjugation                            0.027618  11                         \n",
       "other                                  0.033653  15                         \n",
       "response to chemical                   0.034977  27                         \n",
       "protein maturation                     0.035121  6                          \n",
       "carbohydrate metabolic process         0.039895  14                         \n",
       "sporulation                            0.060024  10                         \n",
       "endosomal transport                    0.071305  8                          \n",
       "lipid transport                        0.082644  6                          \n",
       "\n",
       "                                       pct goterm in subset  \\\n",
       "lipid metabolic process                0.120172               \n",
       "biological_process                     0.193133               \n",
       "monocarboxylic acid metabolic process  0.060086               \n",
       "conjugation                            0.047210               \n",
       "other                                  0.064378               \n",
       "response to chemical                   0.115880               \n",
       "protein maturation                     0.025751               \n",
       "carbohydrate metabolic process         0.060086               \n",
       "sporulation                            0.042918               \n",
       "endosomal transport                    0.034335               \n",
       "lipid transport                        0.025751               \n",
       "\n",
       "                                       pct go term in background  \\\n",
       "lipid metabolic process                0.057770                    \n",
       "biological_process                     0.118357                    \n",
       "monocarboxylic acid metabolic process  0.025564                    \n",
       "conjugation                            0.023551                    \n",
       "other                                  0.037238                    \n",
       "response to chemical                   0.079308                    \n",
       "protein maturation                     0.009863                    \n",
       "carbohydrate metabolic process         0.034823                    \n",
       "sporulation                            0.023752                    \n",
       "endosomal transport                    0.018116                    \n",
       "lipid transport                        0.012480                    \n",
       "\n",
       "                                       pct of go terms genes in subset  \\\n",
       "lipid metabolic process                0.097561                          \n",
       "biological_process                     0.076531                          \n",
       "monocarboxylic acid metabolic process  0.110236                          \n",
       "conjugation                            0.094017                          \n",
       "other                                  0.081081                          \n",
       "response to chemical                   0.068528                          \n",
       "protein maturation                     0.122449                          \n",
       "carbohydrate metabolic process         0.080925                          \n",
       "sporulation                            0.084746                          \n",
       "endosomal transport                    0.088889                          \n",
       "lipid transport                        0.096774                          \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                   genes common name  \n",
       "lipid metabolic process                [TGL2, PIP2, CHO1, VPS34, TGL4, FAA2, DPL1, MPO1, TSC13, ATG26, ICL2, TGL3, PLB1, FOX2, CIT3, UME6, ATG15, NEM1, PDH1, LPX1, HTD2, SCS3, INP51, OAF1, GPT2, TGL5, PLB3, CRC1]                                                                                                                                                                                                  \n",
       "biological_process                     [NIT2, YLR173W, YLR257W, YGR015C, YDR415C, EHD3, YIL024C, YLR456W, DOS2, YML020W, YDR222W, OSW5, YMR226C, GPM3, YML119W, YNL165W, SCM4, MRX3, YIR035C, YET3, YMR102C, YGL081W, YOL024W, YNL011C, YBR138C, YKL050C, MOH1, SNA2, YMR160W, YGR127W, YMR114C, YPL071C, YDL114W, FUB1, YLR361C-A, YIL152W, LEE1, JID1, TMS1, YDR109C, YJR111C, YLR225C, UBP11, YNL010W, YDR461C-A]  \n",
       "monocarboxylic acid metabolic process  [TSC13, UGA1, PIP2, FOX2, CIT3, HTD2, ACS1, FAA2, OAF1, ICL2, GLO4, PDH1, CRC1, PGI1]                                                                                                                                                                                                                                                                                          \n",
       "conjugation                            [RCE1, FAR1, GPA1, FUS2, STE5, AXL1, AFB1, STE2, FUS3, PRM2, STE18]                                                                                                                                                                                                                                                                                                            \n",
       "other                                  [DAL2, SYM1, YAT1, YBR139W, ATG38, OXP1, GDI1, YHR112C, YOL075C, YSP3, SRL3, LPL1, YGL114W, STE6, YUH1]                                                                                                                                                                                                                                                                        \n",
       "response to chemical                   [DIG2, ASI1, YKL069W, PIP2, ECM38, STE5, AHP1, AFB1, STE2, TRX3, TMC1, ATG19, FAR1, IZH2, FUS3, ASI3, RDS2, DIG1, UME6, GPA1, CIN5, FUS2, FRD1, OAF1, MXR2, CTA1, STE18]                                                                                                                                                                                                       \n",
       "protein maturation                     [ATG19, ATG18, RCE1, DAP2, AXL1, ATG15]                                                                                                                                                                                                                                                                                                                                        \n",
       "carbohydrate metabolic process         [RBK1, PGM1, PYC2, RDS2, FYV10, GAL80, UME6, RMD5, PIG2, SGA1, PYC1, DOG1, FBP1, PGI1]                                                                                                                                                                                                                                                                                         \n",
       "sporulation                            [MSO1, VPS13, UBX6, SPO71, PMD1, MDS3, FMP45, ATG26, UBX7, PTP2]                                                                                                                                                                                                                                                                                                               \n",
       "endosomal transport                    [ATG20, ATG18, SNX4, BTN2, SNX41, VPS68, MUK1, VPS13]                                                                                                                                                                                                                                                                                                                          \n",
       "lipid transport                        [DET1, YSP2, ACB1, LAM4, PXA1, PXA2]                                                                                                                                                                                                                                                                                                                                           "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_term_df = gene_set_enrichment['klsig_act']\n",
    "go_term_df_cutoff = go_term_df[go_term_df['pvalue']<0.1].sort_values('pvalue')\n",
    "#\",\".join(list(go_term_df_cutoff.loc['lipid metabolic process']['genes']))\n",
    "go_term_df_cutoff[['pvalue','N subset genes in goterm','pct goterm in subset',\n",
    "                   'pct go term in background','pct of go terms genes in subset','genes common name']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TGL2',\n",
       " 'PIP2',\n",
       " 'CHO1',\n",
       " 'VPS34',\n",
       " 'TGL4',\n",
       " 'FAA2',\n",
       " 'DPL1',\n",
       " 'MPO1',\n",
       " 'TSC13',\n",
       " 'ATG26',\n",
       " 'ICL2',\n",
       " 'TGL3',\n",
       " 'PLB1',\n",
       " 'FOX2',\n",
       " 'CIT3',\n",
       " 'UME6',\n",
       " 'ATG15',\n",
       " 'NEM1',\n",
       " 'PDH1',\n",
       " 'LPX1',\n",
       " 'HTD2',\n",
       " 'SCS3',\n",
       " 'INP51',\n",
       " 'OAF1',\n",
       " 'GPT2',\n",
       " 'TGL5',\n",
       " 'PLB3',\n",
       " 'CRC1']"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_term_df_cutoff.loc[\"lipid metabolic process\"]['genes common name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N SCer Orthologs</th>\n",
       "      <th>kl_genename</th>\n",
       "      <th>sc_genename</th>\n",
       "      <th>PKA(AS)+1NMPP1_KL</th>\n",
       "      <th>padj_KL</th>\n",
       "      <th>PKA(AS)+1NMPP1_SC</th>\n",
       "      <th>padj_SC</th>\n",
       "      <th>SC_common_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0E12651g</td>\n",
       "      <td>YHR067W</td>\n",
       "      <td>2.001996</td>\n",
       "      <td>1.328496e-05</td>\n",
       "      <td>0.891624</td>\n",
       "      <td>4.304409e-01</td>\n",
       "      <td>HTD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0E13905g</td>\n",
       "      <td>YPR006C</td>\n",
       "      <td>2.314233</td>\n",
       "      <td>2.376141e-05</td>\n",
       "      <td>1.245965</td>\n",
       "      <td>3.071610e-04</td>\n",
       "      <td>ICL2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>413</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0D18645g</td>\n",
       "      <td>YLR240W</td>\n",
       "      <td>1.586871</td>\n",
       "      <td>2.037084e-05</td>\n",
       "      <td>1.383917</td>\n",
       "      <td>7.911803e-04</td>\n",
       "      <td>VPS34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>757</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0B11572g</td>\n",
       "      <td>YER015W</td>\n",
       "      <td>2.595638</td>\n",
       "      <td>4.315927e-08</td>\n",
       "      <td>1.086117</td>\n",
       "      <td>3.941036e-01</td>\n",
       "      <td>FAA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>794</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0C04367g</td>\n",
       "      <td>YDL015C</td>\n",
       "      <td>1.900058</td>\n",
       "      <td>2.191820e-06</td>\n",
       "      <td>0.504446</td>\n",
       "      <td>3.139114e-01</td>\n",
       "      <td>TSC13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>930</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0C13431g</td>\n",
       "      <td>YOR100C</td>\n",
       "      <td>3.081807</td>\n",
       "      <td>9.627323e-13</td>\n",
       "      <td>4.318901</td>\n",
       "      <td>4.325369e-02</td>\n",
       "      <td>CRC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1088</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0C02739g</td>\n",
       "      <td>YDR058C</td>\n",
       "      <td>1.690946</td>\n",
       "      <td>2.639785e-06</td>\n",
       "      <td>2.218048</td>\n",
       "      <td>6.700189e-06</td>\n",
       "      <td>TGL2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1174</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0A10175g</td>\n",
       "      <td>YMR313C</td>\n",
       "      <td>2.402741</td>\n",
       "      <td>2.263349e-08</td>\n",
       "      <td>1.010312</td>\n",
       "      <td>3.231624e-03</td>\n",
       "      <td>TGL3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1450</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0A05005g</td>\n",
       "      <td>YGL010W</td>\n",
       "      <td>2.449851</td>\n",
       "      <td>3.105022e-10</td>\n",
       "      <td>2.476848</td>\n",
       "      <td>3.041543e-08</td>\n",
       "      <td>MPO1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1587</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0E13091g</td>\n",
       "      <td>YOR084W</td>\n",
       "      <td>3.015737</td>\n",
       "      <td>1.038063e-08</td>\n",
       "      <td>0.305030</td>\n",
       "      <td>5.344210e-01</td>\n",
       "      <td>LPX1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2158</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0F09526g</td>\n",
       "      <td>YGL126W</td>\n",
       "      <td>2.092161</td>\n",
       "      <td>7.571290e-08</td>\n",
       "      <td>2.019235</td>\n",
       "      <td>4.376317e-01</td>\n",
       "      <td>SCS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2425</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0C10505g</td>\n",
       "      <td>YDR294C</td>\n",
       "      <td>2.096565</td>\n",
       "      <td>1.311071e-05</td>\n",
       "      <td>0.736000</td>\n",
       "      <td>3.242495e-02</td>\n",
       "      <td>DPL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2469</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0C02035g</td>\n",
       "      <td>YLR189C</td>\n",
       "      <td>2.629811</td>\n",
       "      <td>7.331577e-12</td>\n",
       "      <td>1.521920</td>\n",
       "      <td>4.687118e-06</td>\n",
       "      <td>ATG26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2547</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0F09955g</td>\n",
       "      <td>YER026C</td>\n",
       "      <td>2.096168</td>\n",
       "      <td>2.516914e-11</td>\n",
       "      <td>0.970060</td>\n",
       "      <td>7.177456e-02</td>\n",
       "      <td>CHO1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2577</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0F16104g</td>\n",
       "      <td>YKR067W</td>\n",
       "      <td>1.950285</td>\n",
       "      <td>2.901623e-06</td>\n",
       "      <td>1.627263</td>\n",
       "      <td>1.369207e-05</td>\n",
       "      <td>GPT2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2650</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0E18613g</td>\n",
       "      <td>YHR004C</td>\n",
       "      <td>2.548557</td>\n",
       "      <td>2.084412e-08</td>\n",
       "      <td>1.745726</td>\n",
       "      <td>6.995761e-02</td>\n",
       "      <td>NEM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2668</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0E13817g</td>\n",
       "      <td>YKR009C</td>\n",
       "      <td>3.455427</td>\n",
       "      <td>3.535806e-17</td>\n",
       "      <td>2.397845</td>\n",
       "      <td>3.332038e-06</td>\n",
       "      <td>FOX2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2686</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0F20680g</td>\n",
       "      <td>YDR207C</td>\n",
       "      <td>2.761901</td>\n",
       "      <td>2.697166e-09</td>\n",
       "      <td>-0.314209</td>\n",
       "      <td>6.034268e-01</td>\n",
       "      <td>UME6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2725</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0E14213g</td>\n",
       "      <td>YPR002W</td>\n",
       "      <td>4.312088</td>\n",
       "      <td>1.377854e-05</td>\n",
       "      <td>-1.285601</td>\n",
       "      <td>3.891092e-01</td>\n",
       "      <td>PDH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2757</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0C08679g</td>\n",
       "      <td>YCR068W</td>\n",
       "      <td>2.380969</td>\n",
       "      <td>2.569959e-08</td>\n",
       "      <td>1.656607</td>\n",
       "      <td>6.587939e-01</td>\n",
       "      <td>ATG15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2834</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0D13860g</td>\n",
       "      <td>YIL002C</td>\n",
       "      <td>1.994658</td>\n",
       "      <td>1.349685e-05</td>\n",
       "      <td>1.203550</td>\n",
       "      <td>1.409712e-02</td>\n",
       "      <td>INP51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3722</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0E14235g</td>\n",
       "      <td>YPR001W</td>\n",
       "      <td>4.851200</td>\n",
       "      <td>1.081647e-08</td>\n",
       "      <td>2.257085</td>\n",
       "      <td>4.480652e-02</td>\n",
       "      <td>CIT3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4416</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0A07557g</td>\n",
       "      <td>YKR089C</td>\n",
       "      <td>2.458184</td>\n",
       "      <td>4.243260e-11</td>\n",
       "      <td>0.932734</td>\n",
       "      <td>3.385038e-03</td>\n",
       "      <td>TGL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4417</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0A07557g</td>\n",
       "      <td>YOR081C</td>\n",
       "      <td>2.458184</td>\n",
       "      <td>4.243260e-11</td>\n",
       "      <td>0.619536</td>\n",
       "      <td>2.682369e-01</td>\n",
       "      <td>TGL5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4614</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0A03443g</td>\n",
       "      <td>YAL051W</td>\n",
       "      <td>1.629952</td>\n",
       "      <td>3.213289e-06</td>\n",
       "      <td>1.123080</td>\n",
       "      <td>1.130482e-02</td>\n",
       "      <td>OAF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4615</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0A03443g</td>\n",
       "      <td>YOR363C</td>\n",
       "      <td>1.629952</td>\n",
       "      <td>3.213289e-06</td>\n",
       "      <td>1.110512</td>\n",
       "      <td>5.039133e-01</td>\n",
       "      <td>PIP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4654</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0C05940g</td>\n",
       "      <td>YMR008C</td>\n",
       "      <td>2.148433</td>\n",
       "      <td>4.820164e-09</td>\n",
       "      <td>-0.153663</td>\n",
       "      <td>7.317968e-01</td>\n",
       "      <td>PLB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4655</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0C05940g</td>\n",
       "      <td>YOL011W</td>\n",
       "      <td>2.148433</td>\n",
       "      <td>4.820164e-09</td>\n",
       "      <td>-0.743361</td>\n",
       "      <td>1.157268e-01</td>\n",
       "      <td>PLB3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      N SCer Orthologs   kl_genename sc_genename  PKA(AS)+1NMPP1_KL  \\\n",
       "51    1                 KLLA0E12651g  YHR067W     2.001996            \n",
       "336   1                 KLLA0E13905g  YPR006C     2.314233            \n",
       "413   1                 KLLA0D18645g  YLR240W     1.586871            \n",
       "757   1                 KLLA0B11572g  YER015W     2.595638            \n",
       "794   1                 KLLA0C04367g  YDL015C     1.900058            \n",
       "930   1                 KLLA0C13431g  YOR100C     3.081807            \n",
       "1088  1                 KLLA0C02739g  YDR058C     1.690946            \n",
       "1174  1                 KLLA0A10175g  YMR313C     2.402741            \n",
       "1450  1                 KLLA0A05005g  YGL010W     2.449851            \n",
       "1587  1                 KLLA0E13091g  YOR084W     3.015737            \n",
       "2158  1                 KLLA0F09526g  YGL126W     2.092161            \n",
       "2425  1                 KLLA0C10505g  YDR294C     2.096565            \n",
       "2469  1                 KLLA0C02035g  YLR189C     2.629811            \n",
       "2547  1                 KLLA0F09955g  YER026C     2.096168            \n",
       "2577  1                 KLLA0F16104g  YKR067W     1.950285            \n",
       "2650  1                 KLLA0E18613g  YHR004C     2.548557            \n",
       "2668  1                 KLLA0E13817g  YKR009C     3.455427            \n",
       "2686  1                 KLLA0F20680g  YDR207C     2.761901            \n",
       "2725  1                 KLLA0E14213g  YPR002W     4.312088            \n",
       "2757  1                 KLLA0C08679g  YCR068W     2.380969            \n",
       "2834  1                 KLLA0D13860g  YIL002C     1.994658            \n",
       "3722  1                 KLLA0E14235g  YPR001W     4.851200            \n",
       "4416  2                 KLLA0A07557g  YKR089C     2.458184            \n",
       "4417  2                 KLLA0A07557g  YOR081C     2.458184            \n",
       "4614  2                 KLLA0A03443g  YAL051W     1.629952            \n",
       "4615  2                 KLLA0A03443g  YOR363C     1.629952            \n",
       "4654  2                 KLLA0C05940g  YMR008C     2.148433            \n",
       "4655  2                 KLLA0C05940g  YOL011W     2.148433            \n",
       "\n",
       "           padj_KL  PKA(AS)+1NMPP1_SC       padj_SC SC_common_name  \n",
       "51    1.328496e-05  0.891624           4.304409e-01  HTD2           \n",
       "336   2.376141e-05  1.245965           3.071610e-04  ICL2           \n",
       "413   2.037084e-05  1.383917           7.911803e-04  VPS34          \n",
       "757   4.315927e-08  1.086117           3.941036e-01  FAA2           \n",
       "794   2.191820e-06  0.504446           3.139114e-01  TSC13          \n",
       "930   9.627323e-13  4.318901           4.325369e-02  CRC1           \n",
       "1088  2.639785e-06  2.218048           6.700189e-06  TGL2           \n",
       "1174  2.263349e-08  1.010312           3.231624e-03  TGL3           \n",
       "1450  3.105022e-10  2.476848           3.041543e-08  MPO1           \n",
       "1587  1.038063e-08  0.305030           5.344210e-01  LPX1           \n",
       "2158  7.571290e-08  2.019235           4.376317e-01  SCS3           \n",
       "2425  1.311071e-05  0.736000           3.242495e-02  DPL1           \n",
       "2469  7.331577e-12  1.521920           4.687118e-06  ATG26          \n",
       "2547  2.516914e-11  0.970060           7.177456e-02  CHO1           \n",
       "2577  2.901623e-06  1.627263           1.369207e-05  GPT2           \n",
       "2650  2.084412e-08  1.745726           6.995761e-02  NEM1           \n",
       "2668  3.535806e-17  2.397845           3.332038e-06  FOX2           \n",
       "2686  2.697166e-09 -0.314209           6.034268e-01  UME6           \n",
       "2725  1.377854e-05 -1.285601           3.891092e-01  PDH1           \n",
       "2757  2.569959e-08  1.656607           6.587939e-01  ATG15          \n",
       "2834  1.349685e-05  1.203550           1.409712e-02  INP51          \n",
       "3722  1.081647e-08  2.257085           4.480652e-02  CIT3           \n",
       "4416  4.243260e-11  0.932734           3.385038e-03  TGL4           \n",
       "4417  4.243260e-11  0.619536           2.682369e-01  TGL5           \n",
       "4614  3.213289e-06  1.123080           1.130482e-02  OAF1           \n",
       "4615  3.213289e-06  1.110512           5.039133e-01  PIP2           \n",
       "4654  4.820164e-09 -0.153663           7.317968e-01  PLB1           \n",
       "4655  4.820164e-09 -0.743361           1.157268e-01  PLB3           "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = go_term_df_cutoff.loc[\"lipid metabolic process\"]['genes common name']\n",
    "#genes = ['XBP1','RME1','COX1','COX4','COX13','COX6','GSM1','COX12','CYT1']\n",
    "#genes =  ['BAS1', 'MET32', 'GZF3','GLN3', 'MET28', 'MET4', 'GAT1'] #Purine Metabolism\n",
    "#genes = ['ADE1','ADE2','ADE4','ADE5','ADE6','ADE7','ADE8','ADE12','ADE13','ADE16']  #Purine Genes\n",
    "#genes = ['ZWF1','HIS4','SFP1']\n",
    "\n",
    "kl_sc_PKA_data[kl_sc_PKA_data['SC_common_name'].isin(genes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "term = \"lipid metabolic process\"\n",
    "term_genes = list(go_slims_aspect[go_slims_aspect['GO_term']==term]['sc_genename'])\n",
    "kl_sc_PKA_data_term = kl_sc_PKA_data[kl_sc_PKA_data['sc_genename'].isin(term_genes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "terms_of_interest = [\"lipid metabolic process\",\"biological_process\",\"conjugation\",\"response to chemical\",\n",
    "                     \"sporulation\",\"endosomal transport\",\"lipid transport\",\"response to oxidative stress\",\n",
    "                     \"cellular respiration\",\"cofactor metabolic process\",\"carbohydrate transport\",\n",
    "                     \"mitochondrial translation\", \"mitochondrion organization\",\"translational initiation\",\n",
    "                     \"cellular amino acid metabolic process\", \"cell budding\",\n",
    "                     \"nucleobase containing small molecule metabolic process\", \"snoRNA processing\", \"amino acid transport\"]\n",
    "\n",
    "term_df_dict = {}\n",
    "for term in terms_of_interest:\n",
    "    term_genes = list(go_slims_aspect[go_slims_aspect['GO_term']==term]['sc_genename'])\n",
    "    kl_sc_PKA_data_term = kl_sc_PKA_data[kl_sc_PKA_data['sc_genename'].isin(term_genes)]\n",
    "    term_df_dict[term]=kl_sc_PKA_data_term\n",
    "\n",
    "kl_sc_sig_unsig_dict = {'Unsig KL and SC' : kl_sc_PKA_data_unsig, \n",
    "                 'Sig KL only': kl_sc_PKA_data_klsig_scunsig, \n",
    "                 'Sig SC only': kl_sc_PKA_data_scsig_klunsig, \n",
    "                 'Sig KL and SC': kl_sc_PKA_data_klscsig}\n",
    "\n",
    "condition_set = {}\n",
    "for df_dict in (term_df_dict,kl_sc_sig_unsig_dict): \n",
    "    condition_set.update(df_dict)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['Unsig KL and SC', 'mitochondrion organization', 'cellular respiration', 'Sig SC only', 'translational initiation', 'conjugation', 'cell budding', 'cellular amino acid metabolic process', 'cofactor metabolic process', 'nucleobase containing small molecule metabolic process', 'endosomal transport', 'sporulation', 'Sig KL and SC', 'snoRNA processing', 'lipid transport', 'Sig KL only', 'lipid metabolic process', 'mitochondrial translation', 'response to chemical', 'amino acid transport', 'biological_process', 'carbohydrate transport', 'response to oxidative stress'])"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "condition_set.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe id=\"igraph\" scrolling=\"no\" style=\"border:none;\" seamless=\"seamless\" src=\"https://plot.ly/~heineike02/115.embed\" height=\"525px\" width=\"100%\"></iframe>"
      ],
      "text/plain": [
       "<plotly.tools.PlotlyDisplay object>"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Plot a particular set of go terms on top of my scatter plot\n",
    "\n",
    "#plot figure in plot.ly\n",
    "# condition_set = {'Unsig KL and SC' : kl_sc_PKA_data_unsig, \n",
    "#                  'Sig KL only': kl_sc_PKA_data_klsig_scunsig, \n",
    "#                  'Sig SC only': kl_sc_PKA_data_scsig_klunsig, \n",
    "#                  'Sig KL and SC': kl_sc_PKA_data_klscsig,\n",
    "#                  term: kl_sc_PKA_data_term }\n",
    "\n",
    "fig_data = [\n",
    "  \t\t{\n",
    "  \t\t\t'x': condition_set[condition]['PKA(AS)+1NMPP1_KL'], \n",
    "        \t'y': condition_set[condition]['PKA(AS)+1NMPP1_SC'], \n",
    "        \t'text': condition_set[condition]['SC_common_name'], \n",
    "        \t'mode': 'markers', \n",
    "        \t'name': condition} for condition in condition_set.keys()]\n",
    "fig = {\n",
    "    'data': fig_data,\n",
    "    'layout': {\n",
    "        'xaxis': {'title': 'KL LFC 4uM NMPP1'},\n",
    "        'yaxis': {'title': 'SC LFC 4uM NMPP1'}\n",
    "    }\n",
    "}\n",
    "\n",
    "# IPython notebook\n",
    "# py.iplot(fig, filename='pandas/multiple-scatter')\n",
    "\n",
    "py.iplot(fig, filename='sc_kl_scatter_sig_highlight_20170915_various_goterms')\n",
    "\n",
    "\n",
    "#use inline plotly. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4\n",
      "345\n",
      "104\n",
      "4864\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.32285276971249122"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#For each important set, check enrichment for go slim terms.  \n",
    "\n",
    "#Build go slim term sets (io library)\n",
    "\n",
    "#Load go slim terms\n",
    "\n",
    "#for each slim term\n",
    "#given a set of genes and a background set\n",
    "#check for go term enrichment\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {
    "collapsed": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#Generate a list of all genes in KL_SC_PKA_data and add as a list in intermine named\n",
    "#20170905_kl_sc_PKA_data_genes \n",
    "#We can then make subsets programmatically using the API\n",
    "\n",
    "#I ran this command: \n",
    "#\" \".join(kl_sc_PKA_data['sc_genename'])\n",
    "\n",
    "#Then copied and pasted the output into the yeastmine build query list. \n",
    "\n",
    "#The following items were not found in SGD: \n",
    "#Scer_YGOB_YDR134C Scer_RDN18-1 Scer_YGOB_SDC25 Scer_RDN25-1 Scer_RDN58-1 Scer_YGOB_Anc_7.495 Scer_YGOB_ADL119W Scer_RDN5-1\n",
    "\n",
    "#I wonder if I could do this just using something like this: query.add_constraint(\"Gene\", \"LOOKUP\", [\"YNL323W\",\"YIL093C\",...])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#(kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_SC']>0) & (kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_KL']>0)\n",
    "\n",
    "\n",
    "#The one gene that is significant for both, but up in SC and down in kl. \n",
    "#set(kl_sc_PKA_data_klscsig[(kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_SC']>0)]['sc_genename'])-set(kl_sc_PKA_data_klscsig[(kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_SC']>0) & (kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_KL']>0)]['sc_genename'])\n",
    "#HEF3 {'YNL014W'}\n",
    "\n",
    "a = set(kl_sc_PKA_data_klscsig[(kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_SC']<0)]['sc_genename'])\n",
    "b = set(kl_sc_PKA_data_klscsig[(kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_KL']<0)]['sc_genename'])\n",
    "c = set(kl_sc_PKA_data_klscsig[(kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_SC']<0) & (kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_KL']<0)]['sc_genename'])\n",
    "\n",
    "#YJL208C\n",
    "#NUC1 up in KL and down in SC\n",
    "#mintochondrial nuclease. \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N SCer Orthologs</th>\n",
       "      <th>kl_genename</th>\n",
       "      <th>sc_genename</th>\n",
       "      <th>PKA(AS)+1NMPP1_KL</th>\n",
       "      <th>padj_KL</th>\n",
       "      <th>PKA(AS)+1NMPP1_SC</th>\n",
       "      <th>padj_SC</th>\n",
       "      <th>SC_common_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0F18678g</td>\n",
       "      <td>YDR036C</td>\n",
       "      <td>2.075350</td>\n",
       "      <td>2.885098e-07</td>\n",
       "      <td>0.261295</td>\n",
       "      <td>5.984839e-01</td>\n",
       "      <td>EHD3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0E13905g</td>\n",
       "      <td>YPR006C</td>\n",
       "      <td>2.314233</td>\n",
       "      <td>2.376141e-05</td>\n",
       "      <td>1.245965</td>\n",
       "      <td>3.071610e-04</td>\n",
       "      <td>ICL2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>620</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0F16423g</td>\n",
       "      <td>YOR059C</td>\n",
       "      <td>1.650792</td>\n",
       "      <td>6.436999e-05</td>\n",
       "      <td>1.377572</td>\n",
       "      <td>3.806161e-05</td>\n",
       "      <td>LPL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1088</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0C02739g</td>\n",
       "      <td>YDR058C</td>\n",
       "      <td>1.690946</td>\n",
       "      <td>2.639785e-06</td>\n",
       "      <td>2.218048</td>\n",
       "      <td>6.700189e-06</td>\n",
       "      <td>TGL2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1174</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0A10175g</td>\n",
       "      <td>YMR313C</td>\n",
       "      <td>2.402741</td>\n",
       "      <td>2.263349e-08</td>\n",
       "      <td>1.010312</td>\n",
       "      <td>3.231624e-03</td>\n",
       "      <td>TGL3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1450</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0A05005g</td>\n",
       "      <td>YGL010W</td>\n",
       "      <td>2.449851</td>\n",
       "      <td>3.105022e-10</td>\n",
       "      <td>2.476848</td>\n",
       "      <td>3.041543e-08</td>\n",
       "      <td>MPO1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1587</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0E13091g</td>\n",
       "      <td>YOR084W</td>\n",
       "      <td>3.015737</td>\n",
       "      <td>1.038063e-08</td>\n",
       "      <td>0.305030</td>\n",
       "      <td>5.344210e-01</td>\n",
       "      <td>LPX1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2668</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0E13817g</td>\n",
       "      <td>YKR009C</td>\n",
       "      <td>3.455427</td>\n",
       "      <td>3.535806e-17</td>\n",
       "      <td>2.397845</td>\n",
       "      <td>3.332038e-06</td>\n",
       "      <td>FOX2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2725</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0E14213g</td>\n",
       "      <td>YPR002W</td>\n",
       "      <td>4.312088</td>\n",
       "      <td>1.377854e-05</td>\n",
       "      <td>-1.285601</td>\n",
       "      <td>3.891092e-01</td>\n",
       "      <td>PDH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2757</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0C08679g</td>\n",
       "      <td>YCR068W</td>\n",
       "      <td>2.380969</td>\n",
       "      <td>2.569959e-08</td>\n",
       "      <td>1.656607</td>\n",
       "      <td>6.587939e-01</td>\n",
       "      <td>ATG15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3722</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0E14235g</td>\n",
       "      <td>YPR001W</td>\n",
       "      <td>4.851200</td>\n",
       "      <td>1.081647e-08</td>\n",
       "      <td>2.257085</td>\n",
       "      <td>4.480652e-02</td>\n",
       "      <td>CIT3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4416</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0A07557g</td>\n",
       "      <td>YKR089C</td>\n",
       "      <td>2.458184</td>\n",
       "      <td>4.243260e-11</td>\n",
       "      <td>0.932734</td>\n",
       "      <td>3.385038e-03</td>\n",
       "      <td>TGL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4417</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0A07557g</td>\n",
       "      <td>YOR081C</td>\n",
       "      <td>2.458184</td>\n",
       "      <td>4.243260e-11</td>\n",
       "      <td>0.619536</td>\n",
       "      <td>2.682369e-01</td>\n",
       "      <td>TGL5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4614</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0A03443g</td>\n",
       "      <td>YAL051W</td>\n",
       "      <td>1.629952</td>\n",
       "      <td>3.213289e-06</td>\n",
       "      <td>1.123080</td>\n",
       "      <td>1.130482e-02</td>\n",
       "      <td>OAF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4615</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0A03443g</td>\n",
       "      <td>YOR363C</td>\n",
       "      <td>1.629952</td>\n",
       "      <td>3.213289e-06</td>\n",
       "      <td>1.110512</td>\n",
       "      <td>5.039133e-01</td>\n",
       "      <td>PIP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4654</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0C05940g</td>\n",
       "      <td>YMR008C</td>\n",
       "      <td>2.148433</td>\n",
       "      <td>4.820164e-09</td>\n",
       "      <td>-0.153663</td>\n",
       "      <td>7.317968e-01</td>\n",
       "      <td>PLB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4655</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0C05940g</td>\n",
       "      <td>YOL011W</td>\n",
       "      <td>2.148433</td>\n",
       "      <td>4.820164e-09</td>\n",
       "      <td>-0.743361</td>\n",
       "      <td>1.157268e-01</td>\n",
       "      <td>PLB3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      N SCer Orthologs   kl_genename sc_genename  PKA(AS)+1NMPP1_KL  \\\n",
       "21                   1  KLLA0F18678g     YDR036C           2.075350   \n",
       "336                  1  KLLA0E13905g     YPR006C           2.314233   \n",
       "620                  1  KLLA0F16423g     YOR059C           1.650792   \n",
       "1088                 1  KLLA0C02739g     YDR058C           1.690946   \n",
       "1174                 1  KLLA0A10175g     YMR313C           2.402741   \n",
       "1450                 1  KLLA0A05005g     YGL010W           2.449851   \n",
       "1587                 1  KLLA0E13091g     YOR084W           3.015737   \n",
       "2668                 1  KLLA0E13817g     YKR009C           3.455427   \n",
       "2725                 1  KLLA0E14213g     YPR002W           4.312088   \n",
       "2757                 1  KLLA0C08679g     YCR068W           2.380969   \n",
       "3722                 1  KLLA0E14235g     YPR001W           4.851200   \n",
       "4416                 2  KLLA0A07557g     YKR089C           2.458184   \n",
       "4417                 2  KLLA0A07557g     YOR081C           2.458184   \n",
       "4614                 2  KLLA0A03443g     YAL051W           1.629952   \n",
       "4615                 2  KLLA0A03443g     YOR363C           1.629952   \n",
       "4654                 2  KLLA0C05940g     YMR008C           2.148433   \n",
       "4655                 2  KLLA0C05940g     YOL011W           2.148433   \n",
       "\n",
       "           padj_KL  PKA(AS)+1NMPP1_SC       padj_SC SC_common_name  \n",
       "21    2.885098e-07           0.261295  5.984839e-01           EHD3  \n",
       "336   2.376141e-05           1.245965  3.071610e-04           ICL2  \n",
       "620   6.436999e-05           1.377572  3.806161e-05           LPL1  \n",
       "1088  2.639785e-06           2.218048  6.700189e-06           TGL2  \n",
       "1174  2.263349e-08           1.010312  3.231624e-03           TGL3  \n",
       "1450  3.105022e-10           2.476848  3.041543e-08           MPO1  \n",
       "1587  1.038063e-08           0.305030  5.344210e-01           LPX1  \n",
       "2668  3.535806e-17           2.397845  3.332038e-06           FOX2  \n",
       "2725  1.377854e-05          -1.285601  3.891092e-01           PDH1  \n",
       "2757  2.569959e-08           1.656607  6.587939e-01          ATG15  \n",
       "3722  1.081647e-08           2.257085  4.480652e-02           CIT3  \n",
       "4416  4.243260e-11           0.932734  3.385038e-03           TGL4  \n",
       "4417  4.243260e-11           0.619536  2.682369e-01           TGL5  \n",
       "4614  3.213289e-06           1.123080  1.130482e-02           OAF1  \n",
       "4615  3.213289e-06           1.110512  5.039133e-01           PIP2  \n",
       "4654  4.820164e-09          -0.153663  7.317968e-01           PLB1  \n",
       "4655  4.820164e-09          -0.743361  1.157268e-01           PLB3  "
      ]
     },
     "execution_count": 281,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mitophagy_list = ['YCR079W','YDL113C','YFR021W','YGR049W','YIL146C','YJL036W','YLL042C','YLR423C','YNL223W','YNR007C']\n",
    "mito_disassembly_list = ['YCR079W','YDL113C','YFR021W','YGR049W','YIL146C','YJL036W','YLL042C','YLR423C','YNL223W','YNR007C']\n",
    "lipid_catabolic_list = ['YAL051W','YCR068W','YDR036C','YDR058C','YGL010W','YKR009C','YKR089C','YMR008C','YMR313C','YOL011W','YOR059C','YOR081C','YOR084W','YOR363C','YPR001W','YPR002W','YPR006C']\n",
    "kl_sc_PKA_data[kl_sc_PKA_data['sc_genename'].isin(lipid_catabolic_list)]\n",
    "#kl_sc_PKA_data_scsig_klunsig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<ListConstraint: Gene IN 20170509_kl_sc_PKA_data_genes>"
      ]
     },
     "execution_count": 274,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Generate Query for all kl_sc_PKA_data genes\n",
    "\n",
    "# Get a new query on the class (table) you will be querying:\n",
    "query_all = service.new_query(\"Gene\")\n",
    "\n",
    "# The view specifies the output columns\n",
    "query_all.add_view(\n",
    "    \"primaryIdentifier\", \"secondaryIdentifier\", \"organism.shortName\", \"symbol\",\n",
    "    \"name\"\n",
    ")\n",
    "\n",
    "# You can edit the constraint values below\n",
    "query_all.add_constraint(\"Gene\", \"IN\", \"20170509_kl_sc_PKA_data_genes\", code = \"A\")\n",
    "\n",
    "# Uncomment and edit the code below to specify your own custom logic:\n",
    "# query.set_logic(\"A\")\n",
    "\n",
    "#gene_id_list = [str(row[\"id\"] for row in query.rows()]\n",
    "\n",
    "#type(gene_id_list[0])\n",
    "\n",
    "# query_list = list(query)\n",
    "\n",
    "# for row in query_list:     #query.rows():\n",
    "#     print(row[\"primaryIdentifier\"], row[\"secondaryIdentifier\"], row[\"organism.shortName\"], \\\n",
    "#         row[\"symbol\"], row[\"name\"], row[\"id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "klsig_act\n",
      "The following missing genes are filtered out:\n",
      "{'Scer_YGOB_SDC25'}\n",
      "{'description': 'lipid catabolic process', 'p-value': 1.4470773067871862e-06, 'matches': 16, 'identifier': 'GO:0016042', 'populationAnnotationCount': 61}\n",
      "{'description': 'lipid catabolic process', 'p-value': 1.4470773067871862e-06, 'matches': 16, 'identifier': 'GO:0016042', 'populationAnnotationCount': 61}\n",
      "{'description': 'lipid catabolic process', 'p-value': 1.4470773067871862e-06, 'matches': 16, 'identifier': 'GO:0016042', 'populationAnnotationCount': 61}\n",
      "{'description': 'cellular lipid catabolic process', 'p-value': 2.2382440433077777e-06, 'matches': 14, 'identifier': 'GO:0044242', 'populationAnnotationCount': 46}\n",
      "{'description': 'cellular lipid catabolic process', 'p-value': 2.2382440433077777e-06, 'matches': 14, 'identifier': 'GO:0044242', 'populationAnnotationCount': 46}\n",
      "{'description': 'cellular lipid catabolic process', 'p-value': 2.2382440433077777e-06, 'matches': 14, 'identifier': 'GO:0044242', 'populationAnnotationCount': 46}\n",
      "{'description': 'catabolic process', 'p-value': 0.0006057912753930454, 'matches': 55, 'identifier': 'GO:0009056', 'populationAnnotationCount': 745}\n",
      "{'description': 'catabolic process', 'p-value': 0.0006057912753930454, 'matches': 55, 'identifier': 'GO:0009056', 'populationAnnotationCount': 745}\n",
      "{'description': 'catabolic process', 'p-value': 0.0006057912753930454, 'matches': 55, 'identifier': 'GO:0009056', 'populationAnnotationCount': 745}\n",
      "{'description': 'organic substance catabolic process', 'p-value': 0.001216548343839346, 'matches': 54, 'identifier': 'GO:1901575', 'populationAnnotationCount': 740}\n",
      "{'description': 'organic substance catabolic process', 'p-value': 0.001216548343839346, 'matches': 54, 'identifier': 'GO:1901575', 'populationAnnotationCount': 740}\n",
      "{'description': 'organic substance catabolic process', 'p-value': 0.001216548343839346, 'matches': 54, 'identifier': 'GO:1901575', 'populationAnnotationCount': 740}\n",
      "{'description': 'single-organism process', 'p-value': 0.0013150219291881273, 'matches': 160, 'identifier': 'GO:0044699', 'populationAnnotationCount': 3436}\n",
      "{'description': 'single-organism process', 'p-value': 0.0013150219291881273, 'matches': 160, 'identifier': 'GO:0044699', 'populationAnnotationCount': 3436}\n",
      "{'description': 'single-organism process', 'p-value': 0.0013150219291881273, 'matches': 160, 'identifier': 'GO:0044699', 'populationAnnotationCount': 3436}\n",
      "{'description': 'cellular lipid metabolic process', 'p-value': 0.0019399188316196465, 'matches': 31, 'identifier': 'GO:0044255', 'populationAnnotationCount': 318}\n",
      "{'description': 'cellular lipid metabolic process', 'p-value': 0.0019399188316196465, 'matches': 31, 'identifier': 'GO:0044255', 'populationAnnotationCount': 318}\n",
      "{'description': 'cellular lipid metabolic process', 'p-value': 0.0019399188316196465, 'matches': 31, 'identifier': 'GO:0044255', 'populationAnnotationCount': 318}\n",
      "{'description': 'mitophagy', 'p-value': 0.002674670085493186, 'matches': 10, 'identifier': 'GO:0000422', 'populationAnnotationCount': 37}\n",
      "{'description': 'mitophagy', 'p-value': 0.002674670085493186, 'matches': 10, 'identifier': 'GO:0000422', 'populationAnnotationCount': 37}\n",
      "{'description': 'mitophagy', 'p-value': 0.002674670085493186, 'matches': 10, 'identifier': 'GO:0000422', 'populationAnnotationCount': 37}\n",
      "{'description': 'mitochondrion disassembly', 'p-value': 0.002674670085493186, 'matches': 10, 'identifier': 'GO:0061726', 'populationAnnotationCount': 37}\n",
      "{'description': 'mitochondrion disassembly', 'p-value': 0.002674670085493186, 'matches': 10, 'identifier': 'GO:0061726', 'populationAnnotationCount': 37}\n",
      "{'description': 'mitochondrion disassembly', 'p-value': 0.002674670085493186, 'matches': 10, 'identifier': 'GO:0061726', 'populationAnnotationCount': 37}\n",
      "{'description': 'cellular catabolic process', 'p-value': 0.008169879075210076, 'matches': 47, 'identifier': 'GO:0044248', 'populationAnnotationCount': 641}\n",
      "{'description': 'cellular catabolic process', 'p-value': 0.008169879075210076, 'matches': 47, 'identifier': 'GO:0044248', 'populationAnnotationCount': 641}\n",
      "{'description': 'cellular catabolic process', 'p-value': 0.008169879075210076, 'matches': 47, 'identifier': 'GO:0044248', 'populationAnnotationCount': 641}\n",
      "{'description': 'glycerolipid catabolic process', 'p-value': 0.00903333417815859, 'matches': 6, 'identifier': 'GO:0046503', 'populationAnnotationCount': 12}\n",
      "{'description': 'glycerolipid catabolic process', 'p-value': 0.00903333417815859, 'matches': 6, 'identifier': 'GO:0046503', 'populationAnnotationCount': 12}\n",
      "{'description': 'glycerolipid catabolic process', 'p-value': 0.00903333417815859, 'matches': 6, 'identifier': 'GO:0046503', 'populationAnnotationCount': 12}\n",
      "{'description': 'lipid metabolic process', 'p-value': 0.01018069081840652, 'matches': 31, 'identifier': 'GO:0006629', 'populationAnnotationCount': 343}\n",
      "{'description': 'neutral lipid catabolic process', 'p-value': 0.01683980494503397, 'matches': 5, 'identifier': 'GO:0046461', 'populationAnnotationCount': 8}\n",
      "{'description': 'single-organism metabolic process', 'p-value': 0.018074874944907327, 'matches': 84, 'identifier': 'GO:0044710', 'populationAnnotationCount': 1485}\n",
      "{'description': 'fatty acid metabolic process', 'p-value': 0.03085448833196635, 'matches': 12, 'identifier': 'GO:0006631', 'populationAnnotationCount': 69}\n",
      "klsig_rep\n",
      "The following missing genes are filtered out:\n",
      "set()\n",
      "{'description': 'organonitrogen compound biosynthetic process', 'p-value': 1.1794237425064692e-38, 'matches': 147, 'identifier': 'GO:1901566', 'populationAnnotationCount': 1159}\n",
      "{'description': 'organonitrogen compound biosynthetic process', 'p-value': 1.1794237425064692e-38, 'matches': 147, 'identifier': 'GO:1901566', 'populationAnnotationCount': 1159}\n",
      "{'description': 'organonitrogen compound biosynthetic process', 'p-value': 1.1794237425064692e-38, 'matches': 147, 'identifier': 'GO:1901566', 'populationAnnotationCount': 1159}\n",
      "{'description': 'mitochondrial translation', 'p-value': 3.480884688834449e-35, 'matches': 59, 'identifier': 'GO:0032543', 'populationAnnotationCount': 171}\n",
      "{'description': 'mitochondrial translation', 'p-value': 3.480884688834449e-35, 'matches': 59, 'identifier': 'GO:0032543', 'populationAnnotationCount': 171}\n",
      "{'description': 'mitochondrial translation', 'p-value': 3.480884688834449e-35, 'matches': 59, 'identifier': 'GO:0032543', 'populationAnnotationCount': 171}\n",
      "{'description': 'organonitrogen compound metabolic process', 'p-value': 1.677669858123977e-34, 'matches': 156, 'identifier': 'GO:1901564', 'populationAnnotationCount': 1405}\n",
      "{'description': 'organonitrogen compound metabolic process', 'p-value': 1.677669858123977e-34, 'matches': 156, 'identifier': 'GO:1901564', 'populationAnnotationCount': 1405}\n",
      "{'description': 'organonitrogen compound metabolic process', 'p-value': 1.677669858123977e-34, 'matches': 156, 'identifier': 'GO:1901564', 'populationAnnotationCount': 1405}\n",
      "{'description': 'amide biosynthetic process', 'p-value': 1.9146926650792372e-30, 'matches': 114, 'identifier': 'GO:0043604', 'populationAnnotationCount': 820}\n",
      "{'description': 'amide biosynthetic process', 'p-value': 1.9146926650792372e-30, 'matches': 114, 'identifier': 'GO:0043604', 'populationAnnotationCount': 820}\n",
      "{'description': 'amide biosynthetic process', 'p-value': 1.9146926650792372e-30, 'matches': 114, 'identifier': 'GO:0043604', 'populationAnnotationCount': 820}\n",
      "{'description': 'translation', 'p-value': 1.4107362419705543e-29, 'matches': 109, 'identifier': 'GO:0006412', 'populationAnnotationCount': 766}\n",
      "{'description': 'translation', 'p-value': 1.4107362419705543e-29, 'matches': 109, 'identifier': 'GO:0006412', 'populationAnnotationCount': 766}\n",
      "{'description': 'translation', 'p-value': 1.4107362419705543e-29, 'matches': 109, 'identifier': 'GO:0006412', 'populationAnnotationCount': 766}\n",
      "{'description': 'peptide biosynthetic process', 'p-value': 2.3180371717093495e-29, 'matches': 109, 'identifier': 'GO:0043043', 'populationAnnotationCount': 770}\n",
      "{'description': 'peptide biosynthetic process', 'p-value': 2.3180371717093495e-29, 'matches': 109, 'identifier': 'GO:0043043', 'populationAnnotationCount': 770}\n",
      "{'description': 'peptide biosynthetic process', 'p-value': 2.3180371717093495e-29, 'matches': 109, 'identifier': 'GO:0043043', 'populationAnnotationCount': 770}\n",
      "{'description': 'peptide metabolic process', 'p-value': 2.4612613479863283e-27, 'matches': 109, 'identifier': 'GO:0006518', 'populationAnnotationCount': 809}\n",
      "{'description': 'peptide metabolic process', 'p-value': 2.4612613479863283e-27, 'matches': 109, 'identifier': 'GO:0006518', 'populationAnnotationCount': 809}\n",
      "{'description': 'peptide metabolic process', 'p-value': 2.4612613479863283e-27, 'matches': 109, 'identifier': 'GO:0006518', 'populationAnnotationCount': 809}\n",
      "{'description': 'cellular amide metabolic process', 'p-value': 3.0863243964644438e-27, 'matches': 114, 'identifier': 'GO:0043603', 'populationAnnotationCount': 884}\n",
      "{'description': 'cellular amide metabolic process', 'p-value': 3.0863243964644438e-27, 'matches': 114, 'identifier': 'GO:0043603', 'populationAnnotationCount': 884}\n",
      "{'description': 'cellular amide metabolic process', 'p-value': 3.0863243964644438e-27, 'matches': 114, 'identifier': 'GO:0043603', 'populationAnnotationCount': 884}\n",
      "{'description': 'nitrogen compound metabolic process', 'p-value': 1.0493371018984419e-22, 'matches': 210, 'identifier': 'GO:0006807', 'populationAnnotationCount': 2993}\n",
      "{'description': 'nitrogen compound metabolic process', 'p-value': 1.0493371018984419e-22, 'matches': 210, 'identifier': 'GO:0006807', 'populationAnnotationCount': 2993}\n",
      "{'description': 'nitrogen compound metabolic process', 'p-value': 1.0493371018984419e-22, 'matches': 210, 'identifier': 'GO:0006807', 'populationAnnotationCount': 2993}\n",
      "{'description': 'gene expression', 'p-value': 1.8408667104081404e-16, 'matches': 163, 'identifier': 'GO:0010467', 'populationAnnotationCount': 2111}\n",
      "{'description': 'gene expression', 'p-value': 1.8408667104081404e-16, 'matches': 163, 'identifier': 'GO:0010467', 'populationAnnotationCount': 2111}\n",
      "{'description': 'gene expression', 'p-value': 1.8408667104081404e-16, 'matches': 163, 'identifier': 'GO:0010467', 'populationAnnotationCount': 2111}\n",
      "{'description': 'cellular biosynthetic process', 'p-value': 8.890527000236657e-16, 'matches': 176, 'identifier': 'GO:0044249', 'populationAnnotationCount': 2431}\n",
      "{'description': 'cellular biosynthetic process', 'p-value': 8.890527000236657e-16, 'matches': 176, 'identifier': 'GO:0044249', 'populationAnnotationCount': 2431}\n",
      "{'description': 'cellular biosynthetic process', 'p-value': 8.890527000236657e-16, 'matches': 176, 'identifier': 'GO:0044249', 'populationAnnotationCount': 2431}\n",
      "{'description': 'organic substance biosynthetic process', 'p-value': 2.2662043939456823e-15, 'matches': 176, 'identifier': 'GO:1901576', 'populationAnnotationCount': 2449}\n",
      "{'description': 'organic substance biosynthetic process', 'p-value': 2.2662043939456823e-15, 'matches': 176, 'identifier': 'GO:1901576', 'populationAnnotationCount': 2449}\n",
      "{'description': 'organic substance biosynthetic process', 'p-value': 2.2662043939456823e-15, 'matches': 176, 'identifier': 'GO:1901576', 'populationAnnotationCount': 2449}\n",
      "{'description': 'cellular nitrogen compound metabolic process', 'p-value': 4.390303470897839e-15, 'matches': 190, 'identifier': 'GO:0034641', 'populationAnnotationCount': 2798}\n",
      "{'description': 'cellular nitrogen compound metabolic process', 'p-value': 4.390303470897839e-15, 'matches': 190, 'identifier': 'GO:0034641', 'populationAnnotationCount': 2798}\n",
      "{'description': 'cellular nitrogen compound metabolic process', 'p-value': 4.390303470897839e-15, 'matches': 190, 'identifier': 'GO:0034641', 'populationAnnotationCount': 2798}\n",
      "{'description': 'cellular nitrogen compound biosynthetic process', 'p-value': 9.335390019870767e-15, 'matches': 148, 'identifier': 'GO:0044271', 'populationAnnotationCount': 1865}\n",
      "{'description': 'cellular nitrogen compound biosynthetic process', 'p-value': 9.335390019870767e-15, 'matches': 148, 'identifier': 'GO:0044271', 'populationAnnotationCount': 1865}\n",
      "{'description': 'cellular nitrogen compound biosynthetic process', 'p-value': 9.335390019870767e-15, 'matches': 148, 'identifier': 'GO:0044271', 'populationAnnotationCount': 1865}\n",
      "{'description': 'biosynthetic process', 'p-value': 1.2825055332721666e-14, 'matches': 176, 'identifier': 'GO:0009058', 'populationAnnotationCount': 2483}\n",
      "{'description': 'biosynthetic process', 'p-value': 1.2825055332721666e-14, 'matches': 176, 'identifier': 'GO:0009058', 'populationAnnotationCount': 2483}\n",
      "{'description': 'biosynthetic process', 'p-value': 1.2825055332721666e-14, 'matches': 176, 'identifier': 'GO:0009058', 'populationAnnotationCount': 2483}\n",
      "{'description': 'cellular metabolic process', 'p-value': 1.0609061549973035e-10, 'matches': 225, 'identifier': 'GO:0044237', 'populationAnnotationCount': 3993}\n",
      "{'description': 'cellular metabolic process', 'p-value': 1.0609061549973035e-10, 'matches': 225, 'identifier': 'GO:0044237', 'populationAnnotationCount': 3993}\n",
      "{'description': 'cellular metabolic process', 'p-value': 1.0609061549973035e-10, 'matches': 225, 'identifier': 'GO:0044237', 'populationAnnotationCount': 3993}\n",
      "{'description': 'primary metabolic process', 'p-value': 2.074802823938643e-10, 'matches': 221, 'identifier': 'GO:0044238', 'populationAnnotationCount': 3885}\n",
      "{'description': 'primary metabolic process', 'p-value': 2.074802823938643e-10, 'matches': 221, 'identifier': 'GO:0044238', 'populationAnnotationCount': 3885}\n",
      "{'description': 'primary metabolic process', 'p-value': 2.074802823938643e-10, 'matches': 221, 'identifier': 'GO:0044238', 'populationAnnotationCount': 3885}\n",
      "{'description': 'organic substance metabolic process', 'p-value': 4.052547453335325e-10, 'matches': 226, 'identifier': 'GO:0071704', 'populationAnnotationCount': 4056}\n",
      "{'description': 'organic substance metabolic process', 'p-value': 4.052547453335325e-10, 'matches': 226, 'identifier': 'GO:0071704', 'populationAnnotationCount': 4056}\n",
      "{'description': 'organic substance metabolic process', 'p-value': 4.052547453335325e-10, 'matches': 226, 'identifier': 'GO:0071704', 'populationAnnotationCount': 4056}\n",
      "{'description': 'metabolic process', 'p-value': 4.4119745643818933e-10, 'matches': 231, 'identifier': 'GO:0008152', 'populationAnnotationCount': 4221}\n",
      "{'description': 'metabolic process', 'p-value': 4.4119745643818933e-10, 'matches': 231, 'identifier': 'GO:0008152', 'populationAnnotationCount': 4221}\n",
      "{'description': 'metabolic process', 'p-value': 4.4119745643818933e-10, 'matches': 231, 'identifier': 'GO:0008152', 'populationAnnotationCount': 4221}\n",
      "{'description': 'cellular protein metabolic process', 'p-value': 3.750474396735555e-08, 'matches': 126, 'identifier': 'GO:0044267', 'populationAnnotationCount': 1719}\n",
      "{'description': 'cellular protein metabolic process', 'p-value': 3.750474396735555e-08, 'matches': 126, 'identifier': 'GO:0044267', 'populationAnnotationCount': 1719}\n",
      "{'description': 'cellular protein metabolic process', 'p-value': 3.750474396735555e-08, 'matches': 126, 'identifier': 'GO:0044267', 'populationAnnotationCount': 1719}\n",
      "{'description': 'ribosome biogenesis', 'p-value': 4.140751413420741e-07, 'matches': 53, 'identifier': 'GO:0042254', 'populationAnnotationCount': 475}\n",
      "{'description': 'ribosome biogenesis', 'p-value': 4.140751413420741e-07, 'matches': 53, 'identifier': 'GO:0042254', 'populationAnnotationCount': 475}\n",
      "{'description': 'ribosome biogenesis', 'p-value': 4.140751413420741e-07, 'matches': 53, 'identifier': 'GO:0042254', 'populationAnnotationCount': 475}\n",
      "{'description': 'ribonucleoprotein complex biogenesis', 'p-value': 7.576565409933483e-07, 'matches': 59, 'identifier': 'GO:0022613', 'populationAnnotationCount': 573}\n",
      "{'description': 'ribonucleoprotein complex biogenesis', 'p-value': 7.576565409933483e-07, 'matches': 59, 'identifier': 'GO:0022613', 'populationAnnotationCount': 573}\n",
      "{'description': 'ribonucleoprotein complex biogenesis', 'p-value': 7.576565409933483e-07, 'matches': 59, 'identifier': 'GO:0022613', 'populationAnnotationCount': 573}\n",
      "{'description': 'cellular macromolecule biosynthetic process', 'p-value': 9.225667726148051e-07, 'matches': 131, 'identifier': 'GO:0034645', 'populationAnnotationCount': 1897}\n",
      "{'description': 'cellular macromolecule biosynthetic process', 'p-value': 9.225667726148051e-07, 'matches': 131, 'identifier': 'GO:0034645', 'populationAnnotationCount': 1897}\n",
      "{'description': 'cellular macromolecule biosynthetic process', 'p-value': 9.225667726148051e-07, 'matches': 131, 'identifier': 'GO:0034645', 'populationAnnotationCount': 1897}\n",
      "{'description': 'protein metabolic process', 'p-value': 1.09602827500857e-06, 'matches': 128, 'identifier': 'GO:0019538', 'populationAnnotationCount': 1839}\n",
      "{'description': 'protein metabolic process', 'p-value': 1.09602827500857e-06, 'matches': 128, 'identifier': 'GO:0019538', 'populationAnnotationCount': 1839}\n",
      "{'description': 'protein metabolic process', 'p-value': 1.09602827500857e-06, 'matches': 128, 'identifier': 'GO:0019538', 'populationAnnotationCount': 1839}\n",
      "{'description': 'macromolecule biosynthetic process', 'p-value': 1.2702807756177674e-06, 'matches': 131, 'identifier': 'GO:0009059', 'populationAnnotationCount': 1905}\n",
      "{'description': 'macromolecule biosynthetic process', 'p-value': 1.2702807756177674e-06, 'matches': 131, 'identifier': 'GO:0009059', 'populationAnnotationCount': 1905}\n",
      "{'description': 'macromolecule biosynthetic process', 'p-value': 1.2702807756177674e-06, 'matches': 131, 'identifier': 'GO:0009059', 'populationAnnotationCount': 1905}\n",
      "{'description': 'cytoplasmic translation', 'p-value': 1.8803192828213717e-06, 'matches': 30, 'identifier': 'GO:0002181', 'populationAnnotationCount': 187}\n",
      "{'description': 'cytoplasmic translation', 'p-value': 1.8803192828213717e-06, 'matches': 30, 'identifier': 'GO:0002181', 'populationAnnotationCount': 187}\n",
      "{'description': 'cytoplasmic translation', 'p-value': 1.8803192828213717e-06, 'matches': 30, 'identifier': 'GO:0002181', 'populationAnnotationCount': 187}\n",
      "{'description': 'ribosome assembly', 'p-value': 3.6143736100573694e-06, 'matches': 19, 'identifier': 'GO:0042255', 'populationAnnotationCount': 79}\n",
      "{'description': 'ribosome assembly', 'p-value': 3.6143736100573694e-06, 'matches': 19, 'identifier': 'GO:0042255', 'populationAnnotationCount': 79}\n",
      "{'description': 'ribosome assembly', 'p-value': 3.6143736100573694e-06, 'matches': 19, 'identifier': 'GO:0042255', 'populationAnnotationCount': 79}\n",
      "{'description': 'translational initiation', 'p-value': 4.007650397283164e-05, 'matches': 17, 'identifier': 'GO:0006413', 'populationAnnotationCount': 72}\n",
      "{'description': 'translational initiation', 'p-value': 4.007650397283164e-05, 'matches': 17, 'identifier': 'GO:0006413', 'populationAnnotationCount': 72}\n",
      "{'description': 'translational initiation', 'p-value': 4.007650397283164e-05, 'matches': 17, 'identifier': 'GO:0006413', 'populationAnnotationCount': 72}\n",
      "{'description': 'ribonucleoprotein complex subunit organization', 'p-value': 6.022318721714574e-05, 'matches': 29, 'identifier': 'GO:0071826', 'populationAnnotationCount': 203}\n",
      "{'description': 'ribonucleoprotein complex subunit organization', 'p-value': 6.022318721714574e-05, 'matches': 29, 'identifier': 'GO:0071826', 'populationAnnotationCount': 203}\n",
      "{'description': 'ribonucleoprotein complex subunit organization', 'p-value': 6.022318721714574e-05, 'matches': 29, 'identifier': 'GO:0071826', 'populationAnnotationCount': 203}\n",
      "{'description': 'cellular macromolecule metabolic process', 'p-value': 0.00010681316547600656, 'matches': 179, 'identifier': 'GO:0044260', 'populationAnnotationCount': 3118}\n",
      "{'description': 'cellular macromolecule metabolic process', 'p-value': 0.00010681316547600656, 'matches': 179, 'identifier': 'GO:0044260', 'populationAnnotationCount': 3118}\n",
      "{'description': 'cellular macromolecule metabolic process', 'p-value': 0.00010681316547600656, 'matches': 179, 'identifier': 'GO:0044260', 'populationAnnotationCount': 3118}\n",
      "{'description': 'macromolecule metabolic process', 'p-value': 0.00018051604014907501, 'matches': 183, 'identifier': 'GO:0043170', 'populationAnnotationCount': 3233}\n",
      "{'description': 'macromolecule metabolic process', 'p-value': 0.00018051604014907501, 'matches': 183, 'identifier': 'GO:0043170', 'populationAnnotationCount': 3233}\n",
      "{'description': 'macromolecule metabolic process', 'p-value': 0.00018051604014907501, 'matches': 183, 'identifier': 'GO:0043170', 'populationAnnotationCount': 3233}\n",
      "{'description': 'ribonucleoprotein complex assembly', 'p-value': 0.0002514443653018176, 'matches': 27, 'identifier': 'GO:0022618', 'populationAnnotationCount': 191}\n",
      "{'description': 'ribonucleoprotein complex assembly', 'p-value': 0.0002514443653018176, 'matches': 27, 'identifier': 'GO:0022618', 'populationAnnotationCount': 191}\n",
      "{'description': 'ribonucleoprotein complex assembly', 'p-value': 0.0002514443653018176, 'matches': 27, 'identifier': 'GO:0022618', 'populationAnnotationCount': 191}\n",
      "{'description': 'ncRNA metabolic process', 'p-value': 0.000814086762872877, 'matches': 53, 'identifier': 'GO:0034660', 'populationAnnotationCount': 587}\n",
      "{'description': 'ncRNA metabolic process', 'p-value': 0.000814086762872877, 'matches': 53, 'identifier': 'GO:0034660', 'populationAnnotationCount': 587}\n",
      "{'description': 'ncRNA metabolic process', 'p-value': 0.000814086762872877, 'matches': 53, 'identifier': 'GO:0034660', 'populationAnnotationCount': 587}\n",
      "{'description': 'ribosomal small subunit assembly', 'p-value': 0.0011239852786390225, 'matches': 10, 'identifier': 'GO:0000028', 'populationAnnotationCount': 29}\n",
      "{'description': 'ribosomal small subunit assembly', 'p-value': 0.0011239852786390225, 'matches': 10, 'identifier': 'GO:0000028', 'populationAnnotationCount': 29}\n",
      "{'description': 'ribosomal small subunit assembly', 'p-value': 0.0011239852786390225, 'matches': 10, 'identifier': 'GO:0000028', 'populationAnnotationCount': 29}\n",
      "{'description': 'ribosomal small subunit biogenesis', 'p-value': 0.0017239602691994532, 'matches': 22, 'identifier': 'GO:0042274', 'populationAnnotationCount': 147}\n",
      "{'description': 'ribosomal small subunit biogenesis', 'p-value': 0.0017239602691994532, 'matches': 22, 'identifier': 'GO:0042274', 'populationAnnotationCount': 147}\n",
      "{'description': 'ribosomal small subunit biogenesis', 'p-value': 0.0017239602691994532, 'matches': 22, 'identifier': 'GO:0042274', 'populationAnnotationCount': 147}\n",
      "{'description': 'ncRNA processing', 'p-value': 0.001728520764575832, 'matches': 45, 'identifier': 'GO:0034470', 'populationAnnotationCount': 471}\n",
      "{'description': 'ncRNA processing', 'p-value': 0.001728520764575832, 'matches': 45, 'identifier': 'GO:0034470', 'populationAnnotationCount': 471}\n",
      "{'description': 'ncRNA processing', 'p-value': 0.001728520764575832, 'matches': 45, 'identifier': 'GO:0034470', 'populationAnnotationCount': 471}\n",
      "{'description': 'RNA processing', 'p-value': 0.005292380789208374, 'matches': 55, 'identifier': 'GO:0006396', 'populationAnnotationCount': 656}\n",
      "{'description': 'RNA processing', 'p-value': 0.005292380789208374, 'matches': 55, 'identifier': 'GO:0006396', 'populationAnnotationCount': 656}\n",
      "{'description': 'RNA processing', 'p-value': 0.005292380789208374, 'matches': 55, 'identifier': 'GO:0006396', 'populationAnnotationCount': 656}\n",
      "{'description': 'rRNA metabolic process', 'p-value': 0.006615511904754437, 'matches': 37, 'identifier': 'GO:0016072', 'populationAnnotationCount': 368}\n",
      "{'description': 'rRNA metabolic process', 'p-value': 0.006615511904754437, 'matches': 37, 'identifier': 'GO:0016072', 'populationAnnotationCount': 368}\n",
      "{'description': 'rRNA metabolic process', 'p-value': 0.006615511904754437, 'matches': 37, 'identifier': 'GO:0016072', 'populationAnnotationCount': 368}\n",
      "{'description': 'organelle assembly', 'p-value': 0.007182623028798488, 'matches': 23, 'identifier': 'GO:0070925', 'populationAnnotationCount': 172}\n",
      "{'description': 'organelle assembly', 'p-value': 0.007182623028798488, 'matches': 23, 'identifier': 'GO:0070925', 'populationAnnotationCount': 172}\n",
      "{'description': 'organelle assembly', 'p-value': 0.007182623028798488, 'matches': 23, 'identifier': 'GO:0070925', 'populationAnnotationCount': 172}\n",
      "{'description': 'rRNA processing', 'p-value': 0.016097777695048415, 'matches': 35, 'identifier': 'GO:0006364', 'populationAnnotationCount': 351}\n",
      "{'description': 'cellular amino acid biosynthetic process', 'p-value': 0.02583667477673036, 'matches': 19, 'identifier': 'GO:0008652', 'populationAnnotationCount': 134}\n",
      "{'description': 'alpha-amino acid biosynthetic process', 'p-value': 0.04151699760092002, 'matches': 18, 'identifier': 'GO:1901607', 'populationAnnotationCount': 126}\n",
      "scsig_act\n",
      "The following missing genes are filtered out:\n",
      "set()\n",
      "{'description': 'oxidation-reduction process', 'p-value': 5.230705566706802e-09, 'matches': 44, 'identifier': 'GO:0055114', 'populationAnnotationCount': 467}\n",
      "{'description': 'oxidation-reduction process', 'p-value': 5.230705566706802e-09, 'matches': 44, 'identifier': 'GO:0055114', 'populationAnnotationCount': 467}\n",
      "{'description': 'oxidation-reduction process', 'p-value': 5.230705566706802e-09, 'matches': 44, 'identifier': 'GO:0055114', 'populationAnnotationCount': 467}\n",
      "{'description': 'generation of precursor metabolites and energy', 'p-value': 2.0814838057718405e-06, 'matches': 25, 'identifier': 'GO:0006091', 'populationAnnotationCount': 198}\n",
      "{'description': 'generation of precursor metabolites and energy', 'p-value': 2.0814838057718405e-06, 'matches': 25, 'identifier': 'GO:0006091', 'populationAnnotationCount': 198}\n",
      "{'description': 'generation of precursor metabolites and energy', 'p-value': 2.0814838057718405e-06, 'matches': 25, 'identifier': 'GO:0006091', 'populationAnnotationCount': 198}\n",
      "{'description': 'single-organism metabolic process', 'p-value': 4.158541141334688e-06, 'matches': 80, 'identifier': 'GO:0044710', 'populationAnnotationCount': 1485}\n",
      "{'description': 'single-organism metabolic process', 'p-value': 4.158541141334688e-06, 'matches': 80, 'identifier': 'GO:0044710', 'populationAnnotationCount': 1485}\n",
      "{'description': 'single-organism metabolic process', 'p-value': 4.158541141334688e-06, 'matches': 80, 'identifier': 'GO:0044710', 'populationAnnotationCount': 1485}\n",
      "{'description': 'nucleotide metabolic process', 'p-value': 1.9810048434345498e-05, 'matches': 26, 'identifier': 'GO:0009117', 'populationAnnotationCount': 237}\n",
      "{'description': 'nucleotide metabolic process', 'p-value': 1.9810048434345498e-05, 'matches': 26, 'identifier': 'GO:0009117', 'populationAnnotationCount': 237}\n",
      "{'description': 'nucleotide metabolic process', 'p-value': 1.9810048434345498e-05, 'matches': 26, 'identifier': 'GO:0009117', 'populationAnnotationCount': 237}\n",
      "{'description': 'nucleoside phosphate metabolic process', 'p-value': 3.100830400286227e-05, 'matches': 26, 'identifier': 'GO:0006753', 'populationAnnotationCount': 242}\n",
      "{'description': 'nucleoside phosphate metabolic process', 'p-value': 3.100830400286227e-05, 'matches': 26, 'identifier': 'GO:0006753', 'populationAnnotationCount': 242}\n",
      "{'description': 'nucleoside phosphate metabolic process', 'p-value': 3.100830400286227e-05, 'matches': 26, 'identifier': 'GO:0006753', 'populationAnnotationCount': 242}\n",
      "{'description': 'carbohydrate metabolic process', 'p-value': 0.00014055676727744586, 'matches': 29, 'identifier': 'GO:0005975', 'populationAnnotationCount': 316}\n",
      "{'description': 'carbohydrate metabolic process', 'p-value': 0.00014055676727744586, 'matches': 29, 'identifier': 'GO:0005975', 'populationAnnotationCount': 316}\n",
      "{'description': 'carbohydrate metabolic process', 'p-value': 0.00014055676727744586, 'matches': 29, 'identifier': 'GO:0005975', 'populationAnnotationCount': 316}\n",
      "{'description': 'single-organism carbohydrate metabolic process', 'p-value': 0.00017882491237784375, 'matches': 26, 'identifier': 'GO:0044723', 'populationAnnotationCount': 263}\n",
      "{'description': 'single-organism carbohydrate metabolic process', 'p-value': 0.00017882491237784375, 'matches': 26, 'identifier': 'GO:0044723', 'populationAnnotationCount': 263}\n",
      "{'description': 'single-organism carbohydrate metabolic process', 'p-value': 0.00017882491237784375, 'matches': 26, 'identifier': 'GO:0044723', 'populationAnnotationCount': 263}\n",
      "{'description': 'nucleobase-containing small molecule metabolic process', 'p-value': 0.00019340925292711381, 'matches': 26, 'identifier': 'GO:0055086', 'populationAnnotationCount': 264}\n",
      "{'description': 'nucleobase-containing small molecule metabolic process', 'p-value': 0.00019340925292711381, 'matches': 26, 'identifier': 'GO:0055086', 'populationAnnotationCount': 264}\n",
      "{'description': 'nucleobase-containing small molecule metabolic process', 'p-value': 0.00019340925292711381, 'matches': 26, 'identifier': 'GO:0055086', 'populationAnnotationCount': 264}\n",
      "{'description': 'energy derivation by oxidation of organic compounds', 'p-value': 0.00022596939804491316, 'matches': 20, 'identifier': 'GO:0015980', 'populationAnnotationCount': 163}\n",
      "{'description': 'energy derivation by oxidation of organic compounds', 'p-value': 0.00022596939804491316, 'matches': 20, 'identifier': 'GO:0015980', 'populationAnnotationCount': 163}\n",
      "{'description': 'energy derivation by oxidation of organic compounds', 'p-value': 0.00022596939804491316, 'matches': 20, 'identifier': 'GO:0015980', 'populationAnnotationCount': 163}\n",
      "{'description': 'tricarboxylic acid cycle', 'p-value': 0.000587925144891532, 'matches': 9, 'identifier': 'GO:0006099', 'populationAnnotationCount': 30}\n",
      "{'description': 'tricarboxylic acid cycle', 'p-value': 0.000587925144891532, 'matches': 9, 'identifier': 'GO:0006099', 'populationAnnotationCount': 30}\n",
      "{'description': 'tricarboxylic acid cycle', 'p-value': 0.000587925144891532, 'matches': 9, 'identifier': 'GO:0006099', 'populationAnnotationCount': 30}\n",
      "{'description': 'citrate metabolic process', 'p-value': 0.000587925144891532, 'matches': 9, 'identifier': 'GO:0006101', 'populationAnnotationCount': 30}\n",
      "{'description': 'citrate metabolic process', 'p-value': 0.000587925144891532, 'matches': 9, 'identifier': 'GO:0006101', 'populationAnnotationCount': 30}\n",
      "{'description': 'citrate metabolic process', 'p-value': 0.000587925144891532, 'matches': 9, 'identifier': 'GO:0006101', 'populationAnnotationCount': 30}\n",
      "{'description': 'ribose phosphate metabolic process', 'p-value': 0.0007514309939815119, 'matches': 20, 'identifier': 'GO:0019693', 'populationAnnotationCount': 175}\n",
      "{'description': 'ribose phosphate metabolic process', 'p-value': 0.0007514309939815119, 'matches': 20, 'identifier': 'GO:0019693', 'populationAnnotationCount': 175}\n",
      "{'description': 'ribose phosphate metabolic process', 'p-value': 0.0007514309939815119, 'matches': 20, 'identifier': 'GO:0019693', 'populationAnnotationCount': 175}\n",
      "{'description': 'oxidoreduction coenzyme metabolic process', 'p-value': 0.0007525876857489592, 'matches': 15, 'identifier': 'GO:0006733', 'populationAnnotationCount': 99}\n",
      "{'description': 'oxidoreduction coenzyme metabolic process', 'p-value': 0.0007525876857489592, 'matches': 15, 'identifier': 'GO:0006733', 'populationAnnotationCount': 99}\n",
      "{'description': 'oxidoreduction coenzyme metabolic process', 'p-value': 0.0007525876857489592, 'matches': 15, 'identifier': 'GO:0006733', 'populationAnnotationCount': 99}\n",
      "{'description': 'tricarboxylic acid metabolic process', 'p-value': 0.0008070962668641697, 'matches': 9, 'identifier': 'GO:0072350', 'populationAnnotationCount': 31}\n",
      "{'description': 'tricarboxylic acid metabolic process', 'p-value': 0.0008070962668641697, 'matches': 9, 'identifier': 'GO:0072350', 'populationAnnotationCount': 31}\n",
      "{'description': 'tricarboxylic acid metabolic process', 'p-value': 0.0008070962668641697, 'matches': 9, 'identifier': 'GO:0072350', 'populationAnnotationCount': 31}\n",
      "{'description': 'cellular carbohydrate metabolic process', 'p-value': 0.0008161208568434072, 'matches': 22, 'identifier': 'GO:0044262', 'populationAnnotationCount': 210}\n",
      "{'description': 'cellular carbohydrate metabolic process', 'p-value': 0.0008161208568434072, 'matches': 22, 'identifier': 'GO:0044262', 'populationAnnotationCount': 210}\n",
      "{'description': 'cellular carbohydrate metabolic process', 'p-value': 0.0008161208568434072, 'matches': 22, 'identifier': 'GO:0044262', 'populationAnnotationCount': 210}\n",
      "{'description': 'purine ribonucleoside triphosphate metabolic process', 'p-value': 0.0014662432779022508, 'matches': 15, 'identifier': 'GO:0009205', 'populationAnnotationCount': 104}\n",
      "{'description': 'purine ribonucleoside triphosphate metabolic process', 'p-value': 0.0014662432779022508, 'matches': 15, 'identifier': 'GO:0009205', 'populationAnnotationCount': 104}\n",
      "{'description': 'purine ribonucleoside triphosphate metabolic process', 'p-value': 0.0014662432779022508, 'matches': 15, 'identifier': 'GO:0009205', 'populationAnnotationCount': 104}\n",
      "{'description': 'purine nucleoside triphosphate metabolic process', 'p-value': 0.0018927700394925095, 'matches': 15, 'identifier': 'GO:0009144', 'populationAnnotationCount': 106}\n",
      "{'description': 'purine nucleoside triphosphate metabolic process', 'p-value': 0.0018927700394925095, 'matches': 15, 'identifier': 'GO:0009144', 'populationAnnotationCount': 106}\n",
      "{'description': 'purine nucleoside triphosphate metabolic process', 'p-value': 0.0018927700394925095, 'matches': 15, 'identifier': 'GO:0009144', 'populationAnnotationCount': 106}\n",
      "{'description': 'mitochondrial electron transport, succinate to ubiquinone', 'p-value': 0.002182632458293142, 'matches': 5, 'identifier': 'GO:0006121', 'populationAnnotationCount': 7}\n",
      "{'description': 'mitochondrial electron transport, succinate to ubiquinone', 'p-value': 0.002182632458293142, 'matches': 5, 'identifier': 'GO:0006121', 'populationAnnotationCount': 7}\n",
      "{'description': 'mitochondrial electron transport, succinate to ubiquinone', 'p-value': 0.002182632458293142, 'matches': 5, 'identifier': 'GO:0006121', 'populationAnnotationCount': 7}\n",
      "{'description': 'ribonucleoside triphosphate metabolic process', 'p-value': 0.002427902962982402, 'matches': 15, 'identifier': 'GO:0009199', 'populationAnnotationCount': 108}\n",
      "{'description': 'ribonucleoside triphosphate metabolic process', 'p-value': 0.002427902962982402, 'matches': 15, 'identifier': 'GO:0009199', 'populationAnnotationCount': 108}\n",
      "{'description': 'ribonucleoside triphosphate metabolic process', 'p-value': 0.002427902962982402, 'matches': 15, 'identifier': 'GO:0009199', 'populationAnnotationCount': 108}\n",
      "{'description': 'pyridine nucleotide metabolic process', 'p-value': 0.004133676570612348, 'matches': 13, 'identifier': 'GO:0019362', 'populationAnnotationCount': 84}\n",
      "{'description': 'pyridine nucleotide metabolic process', 'p-value': 0.004133676570612348, 'matches': 13, 'identifier': 'GO:0019362', 'populationAnnotationCount': 84}\n",
      "{'description': 'pyridine nucleotide metabolic process', 'p-value': 0.004133676570612348, 'matches': 13, 'identifier': 'GO:0019362', 'populationAnnotationCount': 84}\n",
      "{'description': 'nicotinamide nucleotide metabolic process', 'p-value': 0.004133676570612348, 'matches': 13, 'identifier': 'GO:0046496', 'populationAnnotationCount': 84}\n",
      "{'description': 'nicotinamide nucleotide metabolic process', 'p-value': 0.004133676570612348, 'matches': 13, 'identifier': 'GO:0046496', 'populationAnnotationCount': 84}\n",
      "{'description': 'nicotinamide nucleotide metabolic process', 'p-value': 0.004133676570612348, 'matches': 13, 'identifier': 'GO:0046496', 'populationAnnotationCount': 84}\n",
      "{'description': 'ATP metabolic process', 'p-value': 0.0047709511042181635, 'matches': 14, 'identifier': 'GO:0046034', 'populationAnnotationCount': 99}\n",
      "{'description': 'ATP metabolic process', 'p-value': 0.0047709511042181635, 'matches': 14, 'identifier': 'GO:0046034', 'populationAnnotationCount': 99}\n",
      "{'description': 'ATP metabolic process', 'p-value': 0.0047709511042181635, 'matches': 14, 'identifier': 'GO:0046034', 'populationAnnotationCount': 99}\n",
      "{'description': 'carbohydrate derivative metabolic process', 'p-value': 0.00635697732837156, 'matches': 29, 'identifier': 'GO:1901135', 'populationAnnotationCount': 377}\n",
      "{'description': 'carbohydrate derivative metabolic process', 'p-value': 0.00635697732837156, 'matches': 29, 'identifier': 'GO:1901135', 'populationAnnotationCount': 377}\n",
      "{'description': 'carbohydrate derivative metabolic process', 'p-value': 0.00635697732837156, 'matches': 29, 'identifier': 'GO:1901135', 'populationAnnotationCount': 377}\n",
      "{'description': 'nucleoside triphosphate metabolic process', 'p-value': 0.006946077306957916, 'matches': 15, 'identifier': 'GO:0009141', 'populationAnnotationCount': 117}\n",
      "{'description': 'nucleoside triphosphate metabolic process', 'p-value': 0.006946077306957916, 'matches': 15, 'identifier': 'GO:0009141', 'populationAnnotationCount': 117}\n",
      "{'description': 'nucleoside triphosphate metabolic process', 'p-value': 0.006946077306957916, 'matches': 15, 'identifier': 'GO:0009141', 'populationAnnotationCount': 117}\n",
      "{'description': 'ribonucleoside metabolic process', 'p-value': 0.0070809531234694465, 'matches': 17, 'identifier': 'GO:0009119', 'populationAnnotationCount': 149}\n",
      "{'description': 'ribonucleoside metabolic process', 'p-value': 0.0070809531234694465, 'matches': 17, 'identifier': 'GO:0009119', 'populationAnnotationCount': 149}\n",
      "{'description': 'ribonucleoside metabolic process', 'p-value': 0.0070809531234694465, 'matches': 17, 'identifier': 'GO:0009119', 'populationAnnotationCount': 149}\n",
      "{'description': 'purine ribonucleoside metabolic process', 'p-value': 0.007233573926706066, 'matches': 16, 'identifier': 'GO:0046128', 'populationAnnotationCount': 133}\n",
      "{'description': 'purine ribonucleoside metabolic process', 'p-value': 0.007233573926706066, 'matches': 16, 'identifier': 'GO:0046128', 'populationAnnotationCount': 133}\n",
      "{'description': 'purine ribonucleoside metabolic process', 'p-value': 0.007233573926706066, 'matches': 16, 'identifier': 'GO:0046128', 'populationAnnotationCount': 133}\n",
      "{'description': 'purine nucleoside metabolic process', 'p-value': 0.008005235820671449, 'matches': 16, 'identifier': 'GO:0042278', 'populationAnnotationCount': 134}\n",
      "{'description': 'purine nucleoside metabolic process', 'p-value': 0.008005235820671449, 'matches': 16, 'identifier': 'GO:0042278', 'populationAnnotationCount': 134}\n",
      "{'description': 'purine nucleoside metabolic process', 'p-value': 0.008005235820671449, 'matches': 16, 'identifier': 'GO:0042278', 'populationAnnotationCount': 134}\n",
      "{'description': 'ribonucleotide metabolic process', 'p-value': 0.01233647998506758, 'matches': 17, 'identifier': 'GO:0009259', 'populationAnnotationCount': 155}\n",
      "{'description': 'organophosphate metabolic process', 'p-value': 0.012742013862656783, 'matches': 31, 'identifier': 'GO:0019637', 'populationAnnotationCount': 434}\n",
      "{'description': 'ribonucleoside monophosphate metabolic process', 'p-value': 0.014415151827929214, 'matches': 16, 'identifier': 'GO:0009161', 'populationAnnotationCount': 140}\n",
      "{'description': 'purine nucleoside monophosphate metabolic process', 'p-value': 0.014648167276291597, 'matches': 15, 'identifier': 'GO:0009126', 'populationAnnotationCount': 124}\n",
      "{'description': 'purine ribonucleoside monophosphate metabolic process', 'p-value': 0.014648167276291597, 'matches': 15, 'identifier': 'GO:0009167', 'populationAnnotationCount': 124}\n",
      "{'description': 'purine ribonucleotide metabolic process', 'p-value': 0.017400200877775794, 'matches': 16, 'identifier': 'GO:0009150', 'populationAnnotationCount': 142}\n",
      "{'description': 'nucleoside monophosphate metabolic process', 'p-value': 0.019093319798356993, 'matches': 16, 'identifier': 'GO:0009123', 'populationAnnotationCount': 143}\n",
      "{'description': 'nucleoside metabolic process', 'p-value': 0.019153316568220294, 'matches': 17, 'identifier': 'GO:0009116', 'populationAnnotationCount': 160}\n",
      "{'description': 'pyridine-containing compound metabolic process', 'p-value': 0.01955189783587873, 'matches': 13, 'identifier': 'GO:0072524', 'populationAnnotationCount': 96}\n",
      "{'description': 'purine nucleotide metabolic process', 'p-value': 0.02509597993059181, 'matches': 16, 'identifier': 'GO:0006163', 'populationAnnotationCount': 146}\n",
      "{'description': 'glycosyl compound metabolic process', 'p-value': 0.029214406704572014, 'matches': 17, 'identifier': 'GO:1901657', 'populationAnnotationCount': 165}\n",
      "{'description': 'cellular response to oxidative stress', 'p-value': 0.029571409847286127, 'matches': 14, 'identifier': 'GO:0034599', 'populationAnnotationCount': 115}\n",
      "{'description': 'cellular oxidant detoxification', 'p-value': 0.04618529750419306, 'matches': 8, 'identifier': 'GO:0098869', 'populationAnnotationCount': 37}\n",
      "scsig_rep\n",
      "The following missing genes are filtered out:\n",
      "set()\n",
      "{'description': 'ncRNA metabolic process', 'p-value': 6.415710539122057e-08, 'matches': 39, 'identifier': 'GO:0034660', 'populationAnnotationCount': 587}\n",
      "{'description': 'ncRNA metabolic process', 'p-value': 6.415710539122057e-08, 'matches': 39, 'identifier': 'GO:0034660', 'populationAnnotationCount': 587}\n",
      "{'description': 'ncRNA metabolic process', 'p-value': 6.415710539122057e-08, 'matches': 39, 'identifier': 'GO:0034660', 'populationAnnotationCount': 587}\n",
      "{'description': 'tRNA metabolic process', 'p-value': 8.590568531953101e-08, 'matches': 23, 'identifier': 'GO:0006399', 'populationAnnotationCount': 204}\n",
      "{'description': 'tRNA metabolic process', 'p-value': 8.590568531953101e-08, 'matches': 23, 'identifier': 'GO:0006399', 'populationAnnotationCount': 204}\n",
      "{'description': 'tRNA metabolic process', 'p-value': 8.590568531953101e-08, 'matches': 23, 'identifier': 'GO:0006399', 'populationAnnotationCount': 204}\n",
      "{'description': 'ncRNA processing', 'p-value': 0.0011983336102535029, 'matches': 28, 'identifier': 'GO:0034470', 'populationAnnotationCount': 471}\n",
      "{'description': 'ncRNA processing', 'p-value': 0.0011983336102535029, 'matches': 28, 'identifier': 'GO:0034470', 'populationAnnotationCount': 471}\n",
      "{'description': 'ncRNA processing', 'p-value': 0.0011983336102535029, 'matches': 28, 'identifier': 'GO:0034470', 'populationAnnotationCount': 471}\n",
      "{'description': 'cleavage involved in rRNA processing', 'p-value': 0.003758308161084862, 'matches': 11, 'identifier': 'GO:0000469', 'populationAnnotationCount': 81}\n",
      "{'description': 'cleavage involved in rRNA processing', 'p-value': 0.003758308161084862, 'matches': 11, 'identifier': 'GO:0000469', 'populationAnnotationCount': 81}\n",
      "{'description': 'cleavage involved in rRNA processing', 'p-value': 0.003758308161084862, 'matches': 11, 'identifier': 'GO:0000469', 'populationAnnotationCount': 81}\n",
      "{'description': 'maturation of 5.8S rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)', 'p-value': 0.008773176076478206, 'matches': 11, 'identifier': 'GO:0000466', 'populationAnnotationCount': 88}\n",
      "{'description': 'maturation of 5.8S rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)', 'p-value': 0.008773176076478206, 'matches': 11, 'identifier': 'GO:0000466', 'populationAnnotationCount': 88}\n",
      "{'description': 'maturation of 5.8S rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)', 'p-value': 0.008773176076478206, 'matches': 11, 'identifier': 'GO:0000466', 'populationAnnotationCount': 88}\n",
      "{'description': 'maturation of 5.8S rRNA', 'p-value': 0.009834463676021043, 'matches': 11, 'identifier': 'GO:0000460', 'populationAnnotationCount': 89}\n",
      "{'description': 'maturation of 5.8S rRNA', 'p-value': 0.009834463676021043, 'matches': 11, 'identifier': 'GO:0000460', 'populationAnnotationCount': 89}\n",
      "{'description': 'maturation of 5.8S rRNA', 'p-value': 0.009834463676021043, 'matches': 11, 'identifier': 'GO:0000460', 'populationAnnotationCount': 89}\n",
      "{'description': 'ribosome biogenesis', 'p-value': 0.017269939767013357, 'matches': 26, 'identifier': 'GO:0042254', 'populationAnnotationCount': 475}\n",
      "{'description': 'glutamine family amino acid metabolic process', 'p-value': 0.026393096582410132, 'matches': 9, 'identifier': 'GO:0009064', 'populationAnnotationCount': 63}\n",
      "{'description': 'cellular amino acid metabolic process', 'p-value': 0.02691490222273958, 'matches': 18, 'identifier': 'GO:0006520', 'populationAnnotationCount': 258}\n",
      "klscsig_act\n",
      "The following missing genes are filtered out:\n",
      "set()\n",
      "{'description': 'single-organism metabolic process', 'p-value': 2.021335666028154e-08, 'matches': 54, 'identifier': 'GO:0044710', 'populationAnnotationCount': 1485}\n",
      "{'description': 'single-organism metabolic process', 'p-value': 2.021335666028154e-08, 'matches': 54, 'identifier': 'GO:0044710', 'populationAnnotationCount': 1485}\n",
      "{'description': 'single-organism metabolic process', 'p-value': 2.021335666028154e-08, 'matches': 54, 'identifier': 'GO:0044710', 'populationAnnotationCount': 1485}\n",
      "{'description': 'oxidation-reduction process', 'p-value': 2.3980721250375784e-08, 'matches': 30, 'identifier': 'GO:0055114', 'populationAnnotationCount': 467}\n",
      "{'description': 'oxidation-reduction process', 'p-value': 2.3980721250375784e-08, 'matches': 30, 'identifier': 'GO:0055114', 'populationAnnotationCount': 467}\n",
      "{'description': 'oxidation-reduction process', 'p-value': 2.3980721250375784e-08, 'matches': 30, 'identifier': 'GO:0055114', 'populationAnnotationCount': 467}\n",
      "{'description': 'trehalose metabolic process', 'p-value': 6.875799618980739e-07, 'matches': 7, 'identifier': 'GO:0005991', 'populationAnnotationCount': 12}\n",
      "{'description': 'trehalose metabolic process', 'p-value': 6.875799618980739e-07, 'matches': 7, 'identifier': 'GO:0005991', 'populationAnnotationCount': 12}\n",
      "{'description': 'trehalose metabolic process', 'p-value': 6.875799618980739e-07, 'matches': 7, 'identifier': 'GO:0005991', 'populationAnnotationCount': 12}\n",
      "{'description': 'carbohydrate metabolic process', 'p-value': 1.2137466631579154e-06, 'matches': 23, 'identifier': 'GO:0005975', 'populationAnnotationCount': 316}\n",
      "{'description': 'carbohydrate metabolic process', 'p-value': 1.2137466631579154e-06, 'matches': 23, 'identifier': 'GO:0005975', 'populationAnnotationCount': 316}\n",
      "{'description': 'carbohydrate metabolic process', 'p-value': 1.2137466631579154e-06, 'matches': 23, 'identifier': 'GO:0005975', 'populationAnnotationCount': 316}\n",
      "{'description': 'cellular carbohydrate metabolic process', 'p-value': 1.5071557436324058e-06, 'matches': 19, 'identifier': 'GO:0044262', 'populationAnnotationCount': 210}\n",
      "{'description': 'cellular carbohydrate metabolic process', 'p-value': 1.5071557436324058e-06, 'matches': 19, 'identifier': 'GO:0044262', 'populationAnnotationCount': 210}\n",
      "{'description': 'cellular carbohydrate metabolic process', 'p-value': 1.5071557436324058e-06, 'matches': 19, 'identifier': 'GO:0044262', 'populationAnnotationCount': 210}\n",
      "{'description': 'trehalose metabolism in response to stress', 'p-value': 4.320246296185489e-06, 'matches': 5, 'identifier': 'GO:0070413', 'populationAnnotationCount': 5}\n",
      "{'description': 'trehalose metabolism in response to stress', 'p-value': 4.320246296185489e-06, 'matches': 5, 'identifier': 'GO:0070413', 'populationAnnotationCount': 5}\n",
      "{'description': 'trehalose metabolism in response to stress', 'p-value': 4.320246296185489e-06, 'matches': 5, 'identifier': 'GO:0070413', 'populationAnnotationCount': 5}\n",
      "{'description': 'single-organism carbohydrate metabolic process', 'p-value': 1.0646718455898953e-05, 'matches': 20, 'identifier': 'GO:0044723', 'populationAnnotationCount': 263}\n",
      "{'description': 'single-organism carbohydrate metabolic process', 'p-value': 1.0646718455898953e-05, 'matches': 20, 'identifier': 'GO:0044723', 'populationAnnotationCount': 263}\n",
      "{'description': 'single-organism carbohydrate metabolic process', 'p-value': 1.0646718455898953e-05, 'matches': 20, 'identifier': 'GO:0044723', 'populationAnnotationCount': 263}\n",
      "{'description': 'trehalose biosynthetic process', 'p-value': 8.848064606838846e-05, 'matches': 5, 'identifier': 'GO:0005992', 'populationAnnotationCount': 7}\n",
      "{'description': 'trehalose biosynthetic process', 'p-value': 8.848064606838846e-05, 'matches': 5, 'identifier': 'GO:0005992', 'populationAnnotationCount': 7}\n",
      "{'description': 'trehalose biosynthetic process', 'p-value': 8.848064606838846e-05, 'matches': 5, 'identifier': 'GO:0005992', 'populationAnnotationCount': 7}\n",
      "{'description': 'catabolic process', 'p-value': 0.00013468257855457852, 'matches': 32, 'identifier': 'GO:0009056', 'populationAnnotationCount': 745}\n",
      "{'description': 'catabolic process', 'p-value': 0.00013468257855457852, 'matches': 32, 'identifier': 'GO:0009056', 'populationAnnotationCount': 745}\n",
      "{'description': 'catabolic process', 'p-value': 0.00013468257855457852, 'matches': 32, 'identifier': 'GO:0009056', 'populationAnnotationCount': 745}\n",
      "{'description': 'disaccharide biosynthetic process', 'p-value': 0.0002329681214063206, 'matches': 5, 'identifier': 'GO:0046351', 'populationAnnotationCount': 8}\n",
      "{'description': 'disaccharide biosynthetic process', 'p-value': 0.0002329681214063206, 'matches': 5, 'identifier': 'GO:0046351', 'populationAnnotationCount': 8}\n",
      "{'description': 'disaccharide biosynthetic process', 'p-value': 0.0002329681214063206, 'matches': 5, 'identifier': 'GO:0046351', 'populationAnnotationCount': 8}\n",
      "{'description': 'organic substance catabolic process', 'p-value': 0.0004439353873961616, 'matches': 31, 'identifier': 'GO:1901575', 'populationAnnotationCount': 740}\n",
      "{'description': 'organic substance catabolic process', 'p-value': 0.0004439353873961616, 'matches': 31, 'identifier': 'GO:1901575', 'populationAnnotationCount': 740}\n",
      "{'description': 'organic substance catabolic process', 'p-value': 0.0004439353873961616, 'matches': 31, 'identifier': 'GO:1901575', 'populationAnnotationCount': 740}\n",
      "{'description': 'oligosaccharide biosynthetic process', 'p-value': 0.0005175586245842867, 'matches': 5, 'identifier': 'GO:0009312', 'populationAnnotationCount': 9}\n",
      "{'description': 'oligosaccharide biosynthetic process', 'p-value': 0.0005175586245842867, 'matches': 5, 'identifier': 'GO:0009312', 'populationAnnotationCount': 9}\n",
      "{'description': 'oligosaccharide biosynthetic process', 'p-value': 0.0005175586245842867, 'matches': 5, 'identifier': 'GO:0009312', 'populationAnnotationCount': 9}\n",
      "{'description': 'small molecule catabolic process', 'p-value': 0.0062178863214365976, 'matches': 12, 'identifier': 'GO:0044282', 'populationAnnotationCount': 138}\n",
      "{'description': 'small molecule catabolic process', 'p-value': 0.0062178863214365976, 'matches': 12, 'identifier': 'GO:0044282', 'populationAnnotationCount': 138}\n",
      "{'description': 'small molecule catabolic process', 'p-value': 0.0062178863214365976, 'matches': 12, 'identifier': 'GO:0044282', 'populationAnnotationCount': 138}\n",
      "{'description': 'carbohydrate biosynthetic process', 'p-value': 0.00638608549371138, 'matches': 11, 'identifier': 'GO:0016051', 'populationAnnotationCount': 114}\n",
      "{'description': 'carbohydrate biosynthetic process', 'p-value': 0.00638608549371138, 'matches': 11, 'identifier': 'GO:0016051', 'populationAnnotationCount': 114}\n",
      "{'description': 'carbohydrate biosynthetic process', 'p-value': 0.00638608549371138, 'matches': 11, 'identifier': 'GO:0016051', 'populationAnnotationCount': 114}\n",
      "{'description': 'single-organism process', 'p-value': 0.01120290404181545, 'matches': 74, 'identifier': 'GO:0044699', 'populationAnnotationCount': 3436}\n",
      "{'description': 'small molecule metabolic process', 'p-value': 0.013345722029125238, 'matches': 30, 'identifier': 'GO:0044281', 'populationAnnotationCount': 815}\n",
      "{'description': 'cellular catabolic process', 'p-value': 0.013478309047006587, 'matches': 26, 'identifier': 'GO:0044248', 'populationAnnotationCount': 641}\n",
      "klscsig_rep\n",
      "The following missing genes are filtered out:\n",
      "set()\n",
      "{'description': 'ribosome biogenesis', 'p-value': 7.330565788584461e-169, 'matches': 221, 'identifier': 'GO:0042254', 'populationAnnotationCount': 475}\n",
      "{'description': 'ribosome biogenesis', 'p-value': 7.330565788584461e-169, 'matches': 221, 'identifier': 'GO:0042254', 'populationAnnotationCount': 475}\n",
      "{'description': 'ribosome biogenesis', 'p-value': 7.330565788584461e-169, 'matches': 221, 'identifier': 'GO:0042254', 'populationAnnotationCount': 475}\n",
      "{'description': 'ribonucleoprotein complex biogenesis', 'p-value': 9.846473356716181e-151, 'matches': 224, 'identifier': 'GO:0022613', 'populationAnnotationCount': 573}\n",
      "{'description': 'ribonucleoprotein complex biogenesis', 'p-value': 9.846473356716181e-151, 'matches': 224, 'identifier': 'GO:0022613', 'populationAnnotationCount': 573}\n",
      "{'description': 'ribonucleoprotein complex biogenesis', 'p-value': 9.846473356716181e-151, 'matches': 224, 'identifier': 'GO:0022613', 'populationAnnotationCount': 573}\n",
      "{'description': 'ncRNA metabolic process', 'p-value': 6.750804816684573e-112, 'matches': 197, 'identifier': 'GO:0034660', 'populationAnnotationCount': 587}\n",
      "{'description': 'ncRNA metabolic process', 'p-value': 6.750804816684573e-112, 'matches': 197, 'identifier': 'GO:0034660', 'populationAnnotationCount': 587}\n",
      "{'description': 'ncRNA metabolic process', 'p-value': 6.750804816684573e-112, 'matches': 197, 'identifier': 'GO:0034660', 'populationAnnotationCount': 587}\n",
      "{'description': 'rRNA processing', 'p-value': 8.434904136874433e-111, 'matches': 160, 'identifier': 'GO:0006364', 'populationAnnotationCount': 351}\n",
      "{'description': 'rRNA processing', 'p-value': 8.434904136874433e-111, 'matches': 160, 'identifier': 'GO:0006364', 'populationAnnotationCount': 351}\n",
      "{'description': 'rRNA processing', 'p-value': 8.434904136874433e-111, 'matches': 160, 'identifier': 'GO:0006364', 'populationAnnotationCount': 351}\n",
      "{'description': 'rRNA metabolic process', 'p-value': 4.670762592251889e-108, 'matches': 161, 'identifier': 'GO:0016072', 'populationAnnotationCount': 368}\n",
      "{'description': 'rRNA metabolic process', 'p-value': 4.670762592251889e-108, 'matches': 161, 'identifier': 'GO:0016072', 'populationAnnotationCount': 368}\n",
      "{'description': 'rRNA metabolic process', 'p-value': 4.670762592251889e-108, 'matches': 161, 'identifier': 'GO:0016072', 'populationAnnotationCount': 368}\n",
      "{'description': 'ncRNA processing', 'p-value': 2.9684562919915473e-102, 'matches': 173, 'identifier': 'GO:0034470', 'populationAnnotationCount': 471}\n",
      "{'description': 'ncRNA processing', 'p-value': 2.9684562919915473e-102, 'matches': 173, 'identifier': 'GO:0034470', 'populationAnnotationCount': 471}\n",
      "{'description': 'ncRNA processing', 'p-value': 2.9684562919915473e-102, 'matches': 173, 'identifier': 'GO:0034470', 'populationAnnotationCount': 471}\n",
      "{'description': 'cytoplasmic translation', 'p-value': 3.3996242867028676e-92, 'matches': 113, 'identifier': 'GO:0002181', 'populationAnnotationCount': 187}\n",
      "{'description': 'cytoplasmic translation', 'p-value': 3.3996242867028676e-92, 'matches': 113, 'identifier': 'GO:0002181', 'populationAnnotationCount': 187}\n",
      "{'description': 'cytoplasmic translation', 'p-value': 3.3996242867028676e-92, 'matches': 113, 'identifier': 'GO:0002181', 'populationAnnotationCount': 187}\n",
      "{'description': 'ribosomal small subunit biogenesis', 'p-value': 6.093828082719817e-85, 'matches': 98, 'identifier': 'GO:0042274', 'populationAnnotationCount': 147}\n",
      "{'description': 'ribosomal small subunit biogenesis', 'p-value': 6.093828082719817e-85, 'matches': 98, 'identifier': 'GO:0042274', 'populationAnnotationCount': 147}\n",
      "{'description': 'ribosomal small subunit biogenesis', 'p-value': 6.093828082719817e-85, 'matches': 98, 'identifier': 'GO:0042274', 'populationAnnotationCount': 147}\n",
      "{'description': 'RNA processing', 'p-value': 4.733208080002952e-75, 'matches': 173, 'identifier': 'GO:0006396', 'populationAnnotationCount': 656}\n",
      "{'description': 'RNA processing', 'p-value': 4.733208080002952e-75, 'matches': 173, 'identifier': 'GO:0006396', 'populationAnnotationCount': 656}\n",
      "{'description': 'RNA processing', 'p-value': 4.733208080002952e-75, 'matches': 173, 'identifier': 'GO:0006396', 'populationAnnotationCount': 656}\n",
      "{'description': 'ribosomal large subunit biogenesis', 'p-value': 1.5588341262955272e-74, 'matches': 84, 'identifier': 'GO:0042273', 'populationAnnotationCount': 120}\n",
      "{'description': 'ribosomal large subunit biogenesis', 'p-value': 1.5588341262955272e-74, 'matches': 84, 'identifier': 'GO:0042273', 'populationAnnotationCount': 120}\n",
      "{'description': 'ribosomal large subunit biogenesis', 'p-value': 1.5588341262955272e-74, 'matches': 84, 'identifier': 'GO:0042273', 'populationAnnotationCount': 120}\n",
      "{'description': 'maturation of SSU-rRNA', 'p-value': 7.355947384931461e-73, 'matches': 83, 'identifier': 'GO:0030490', 'populationAnnotationCount': 120}\n",
      "{'description': 'maturation of SSU-rRNA', 'p-value': 7.355947384931461e-73, 'matches': 83, 'identifier': 'GO:0030490', 'populationAnnotationCount': 120}\n",
      "{'description': 'maturation of SSU-rRNA', 'p-value': 7.355947384931461e-73, 'matches': 83, 'identifier': 'GO:0030490', 'populationAnnotationCount': 120}\n",
      "{'description': 'gene expression', 'p-value': 1.5133585296764334e-72, 'matches': 286, 'identifier': 'GO:0010467', 'populationAnnotationCount': 2111}\n",
      "{'description': 'gene expression', 'p-value': 1.5133585296764334e-72, 'matches': 286, 'identifier': 'GO:0010467', 'populationAnnotationCount': 2111}\n",
      "{'description': 'gene expression', 'p-value': 1.5133585296764334e-72, 'matches': 286, 'identifier': 'GO:0010467', 'populationAnnotationCount': 2111}\n",
      "{'description': 'maturation of SSU-rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)', 'p-value': 3.156257750630804e-72, 'matches': 79, 'identifier': 'GO:0000462', 'populationAnnotationCount': 108}\n",
      "{'description': 'maturation of SSU-rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)', 'p-value': 3.156257750630804e-72, 'matches': 79, 'identifier': 'GO:0000462', 'populationAnnotationCount': 108}\n",
      "{'description': 'maturation of SSU-rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)', 'p-value': 3.156257750630804e-72, 'matches': 79, 'identifier': 'GO:0000462', 'populationAnnotationCount': 108}\n",
      "{'description': 'cellular component biogenesis', 'p-value': 1.2003573310346846e-69, 'matches': 228, 'identifier': 'GO:0044085', 'populationAnnotationCount': 1294}\n",
      "{'description': 'cellular component biogenesis', 'p-value': 1.2003573310346846e-69, 'matches': 228, 'identifier': 'GO:0044085', 'populationAnnotationCount': 1294}\n",
      "{'description': 'cellular component biogenesis', 'p-value': 1.2003573310346846e-69, 'matches': 228, 'identifier': 'GO:0044085', 'populationAnnotationCount': 1294}\n",
      "{'description': 'cellular nitrogen compound metabolic process', 'p-value': 4.0544306018069575e-54, 'matches': 302, 'identifier': 'GO:0034641', 'populationAnnotationCount': 2798}\n",
      "{'description': 'cellular nitrogen compound metabolic process', 'p-value': 4.0544306018069575e-54, 'matches': 302, 'identifier': 'GO:0034641', 'populationAnnotationCount': 2798}\n",
      "{'description': 'cellular nitrogen compound metabolic process', 'p-value': 4.0544306018069575e-54, 'matches': 302, 'identifier': 'GO:0034641', 'populationAnnotationCount': 2798}\n",
      "{'description': 'nitrogen compound metabolic process', 'p-value': 1.5003913192130075e-51, 'matches': 308, 'identifier': 'GO:0006807', 'populationAnnotationCount': 2993}\n",
      "{'description': 'nitrogen compound metabolic process', 'p-value': 1.5003913192130075e-51, 'matches': 308, 'identifier': 'GO:0006807', 'populationAnnotationCount': 2993}\n",
      "{'description': 'nitrogen compound metabolic process', 'p-value': 1.5003913192130075e-51, 'matches': 308, 'identifier': 'GO:0006807', 'populationAnnotationCount': 2993}\n",
      "{'description': 'maturation of 5.8S rRNA', 'p-value': 4.5631309696581573e-38, 'matches': 52, 'identifier': 'GO:0000460', 'populationAnnotationCount': 89}\n",
      "{'description': 'maturation of 5.8S rRNA', 'p-value': 4.5631309696581573e-38, 'matches': 52, 'identifier': 'GO:0000460', 'populationAnnotationCount': 89}\n",
      "{'description': 'maturation of 5.8S rRNA', 'p-value': 4.5631309696581573e-38, 'matches': 52, 'identifier': 'GO:0000460', 'populationAnnotationCount': 89}\n",
      "{'description': 'ribosome assembly', 'p-value': 1.7923467341259583e-37, 'matches': 49, 'identifier': 'GO:0042255', 'populationAnnotationCount': 79}\n",
      "{'description': 'ribosome assembly', 'p-value': 1.7923467341259583e-37, 'matches': 49, 'identifier': 'GO:0042255', 'populationAnnotationCount': 79}\n",
      "{'description': 'ribosome assembly', 'p-value': 1.7923467341259583e-37, 'matches': 49, 'identifier': 'GO:0042255', 'populationAnnotationCount': 79}\n",
      "{'description': 'maturation of 5.8S rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)', 'p-value': 5.192763469657087e-37, 'matches': 51, 'identifier': 'GO:0000466', 'populationAnnotationCount': 88}\n",
      "{'description': 'maturation of 5.8S rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)', 'p-value': 5.192763469657087e-37, 'matches': 51, 'identifier': 'GO:0000466', 'populationAnnotationCount': 88}\n",
      "{'description': 'maturation of 5.8S rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)', 'p-value': 5.192763469657087e-37, 'matches': 51, 'identifier': 'GO:0000466', 'populationAnnotationCount': 88}\n",
      "{'description': 'translation', 'p-value': 5.975403753431216e-34, 'matches': 137, 'identifier': 'GO:0006412', 'populationAnnotationCount': 766}\n",
      "{'description': 'translation', 'p-value': 5.975403753431216e-34, 'matches': 137, 'identifier': 'GO:0006412', 'populationAnnotationCount': 766}\n",
      "{'description': 'translation', 'p-value': 5.975403753431216e-34, 'matches': 137, 'identifier': 'GO:0006412', 'populationAnnotationCount': 766}\n",
      "{'description': 'RNA metabolic process', 'p-value': 1.0495177389768606e-33, 'matches': 200, 'identifier': 'GO:0016070', 'populationAnnotationCount': 1538}\n",
      "{'description': 'RNA metabolic process', 'p-value': 1.0495177389768606e-33, 'matches': 200, 'identifier': 'GO:0016070', 'populationAnnotationCount': 1538}\n",
      "{'description': 'RNA metabolic process', 'p-value': 1.0495177389768606e-33, 'matches': 200, 'identifier': 'GO:0016070', 'populationAnnotationCount': 1538}\n",
      "{'description': 'peptide biosynthetic process', 'p-value': 1.1097914776929798e-33, 'matches': 137, 'identifier': 'GO:0043043', 'populationAnnotationCount': 770}\n",
      "{'description': 'peptide biosynthetic process', 'p-value': 1.1097914776929798e-33, 'matches': 137, 'identifier': 'GO:0043043', 'populationAnnotationCount': 770}\n",
      "{'description': 'peptide biosynthetic process', 'p-value': 1.1097914776929798e-33, 'matches': 137, 'identifier': 'GO:0043043', 'populationAnnotationCount': 770}\n",
      "{'description': 'maturation of LSU-rRNA', 'p-value': 1.270303947897047e-32, 'matches': 39, 'identifier': 'GO:0000470', 'populationAnnotationCount': 55}\n",
      "{'description': 'maturation of LSU-rRNA', 'p-value': 1.270303947897047e-32, 'matches': 39, 'identifier': 'GO:0000470', 'populationAnnotationCount': 55}\n",
      "{'description': 'maturation of LSU-rRNA', 'p-value': 1.270303947897047e-32, 'matches': 39, 'identifier': 'GO:0000470', 'populationAnnotationCount': 55}\n",
      "{'description': 'cleavage involved in rRNA processing', 'p-value': 1.8089172207185676e-32, 'matches': 46, 'identifier': 'GO:0000469', 'populationAnnotationCount': 81}\n",
      "{'description': 'cleavage involved in rRNA processing', 'p-value': 1.8089172207185676e-32, 'matches': 46, 'identifier': 'GO:0000469', 'populationAnnotationCount': 81}\n",
      "{'description': 'cleavage involved in rRNA processing', 'p-value': 1.8089172207185676e-32, 'matches': 46, 'identifier': 'GO:0000469', 'populationAnnotationCount': 81}\n",
      "{'description': 'cellular macromolecule metabolic process', 'p-value': 2.243000524755414e-32, 'matches': 291, 'identifier': 'GO:0044260', 'populationAnnotationCount': 3118}\n",
      "{'description': 'cellular macromolecule metabolic process', 'p-value': 2.243000524755414e-32, 'matches': 291, 'identifier': 'GO:0044260', 'populationAnnotationCount': 3118}\n",
      "{'description': 'cellular macromolecule metabolic process', 'p-value': 2.243000524755414e-32, 'matches': 291, 'identifier': 'GO:0044260', 'populationAnnotationCount': 3118}\n",
      "{'description': 'amide biosynthetic process', 'p-value': 7.89773702559337e-32, 'matches': 139, 'identifier': 'GO:0043604', 'populationAnnotationCount': 820}\n",
      "{'description': 'amide biosynthetic process', 'p-value': 7.89773702559337e-32, 'matches': 139, 'identifier': 'GO:0043604', 'populationAnnotationCount': 820}\n",
      "{'description': 'amide biosynthetic process', 'p-value': 7.89773702559337e-32, 'matches': 139, 'identifier': 'GO:0043604', 'populationAnnotationCount': 820}\n",
      "{'description': 'endonucleolytic cleavage involved in rRNA processing', 'p-value': 3.4454490375102214e-31, 'matches': 39, 'identifier': 'GO:0000478', 'populationAnnotationCount': 58}\n",
      "{'description': 'endonucleolytic cleavage involved in rRNA processing', 'p-value': 3.4454490375102214e-31, 'matches': 39, 'identifier': 'GO:0000478', 'populationAnnotationCount': 58}\n",
      "{'description': 'endonucleolytic cleavage involved in rRNA processing', 'p-value': 3.4454490375102214e-31, 'matches': 39, 'identifier': 'GO:0000478', 'populationAnnotationCount': 58}\n",
      "{'description': 'endonucleolytic cleavage of tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)', 'p-value': 3.4454490375102214e-31, 'matches': 39, 'identifier': 'GO:0000479', 'populationAnnotationCount': 58}\n",
      "{'description': 'endonucleolytic cleavage of tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)', 'p-value': 3.4454490375102214e-31, 'matches': 39, 'identifier': 'GO:0000479', 'populationAnnotationCount': 58}\n",
      "{'description': 'endonucleolytic cleavage of tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)', 'p-value': 3.4454490375102214e-31, 'matches': 39, 'identifier': 'GO:0000479', 'populationAnnotationCount': 58}\n",
      "{'description': 'peptide metabolic process', 'p-value': 3.679960542308555e-31, 'matches': 137, 'identifier': 'GO:0006518', 'populationAnnotationCount': 809}\n",
      "{'description': 'peptide metabolic process', 'p-value': 3.679960542308555e-31, 'matches': 137, 'identifier': 'GO:0006518', 'populationAnnotationCount': 809}\n",
      "{'description': 'peptide metabolic process', 'p-value': 3.679960542308555e-31, 'matches': 137, 'identifier': 'GO:0006518', 'populationAnnotationCount': 809}\n",
      "{'description': 'macromolecule metabolic process', 'p-value': 1.1681115601676365e-30, 'matches': 294, 'identifier': 'GO:0043170', 'populationAnnotationCount': 3233}\n",
      "{'description': 'macromolecule metabolic process', 'p-value': 1.1681115601676365e-30, 'matches': 294, 'identifier': 'GO:0043170', 'populationAnnotationCount': 3233}\n",
      "{'description': 'macromolecule metabolic process', 'p-value': 1.1681115601676365e-30, 'matches': 294, 'identifier': 'GO:0043170', 'populationAnnotationCount': 3233}\n",
      "{'description': 'maturation of LSU-rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)', 'p-value': 1.7793649952127092e-28, 'matches': 33, 'identifier': 'GO:0000463', 'populationAnnotationCount': 44}\n",
      "{'description': 'maturation of LSU-rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)', 'p-value': 1.7793649952127092e-28, 'matches': 33, 'identifier': 'GO:0000463', 'populationAnnotationCount': 44}\n",
      "{'description': 'maturation of LSU-rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)', 'p-value': 1.7793649952127092e-28, 'matches': 33, 'identifier': 'GO:0000463', 'populationAnnotationCount': 44}\n",
      "{'description': 'cellular amide metabolic process', 'p-value': 4.854756592809763e-28, 'matches': 139, 'identifier': 'GO:0043603', 'populationAnnotationCount': 884}\n",
      "{'description': 'cellular amide metabolic process', 'p-value': 4.854756592809763e-28, 'matches': 139, 'identifier': 'GO:0043603', 'populationAnnotationCount': 884}\n",
      "{'description': 'cellular amide metabolic process', 'p-value': 4.854756592809763e-28, 'matches': 139, 'identifier': 'GO:0043603', 'populationAnnotationCount': 884}\n",
      "{'description': 'endonucleolytic cleavage in ITS1 to separate SSU-rRNA from 5.8S rRNA and LSU-rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)', 'p-value': 2.1187596704678572e-27, 'matches': 33, 'identifier': 'GO:0000447', 'populationAnnotationCount': 46}\n",
      "{'description': 'endonucleolytic cleavage in ITS1 to separate SSU-rRNA from 5.8S rRNA and LSU-rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)', 'p-value': 2.1187596704678572e-27, 'matches': 33, 'identifier': 'GO:0000447', 'populationAnnotationCount': 46}\n",
      "{'description': 'endonucleolytic cleavage in ITS1 to separate SSU-rRNA from 5.8S rRNA and LSU-rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)', 'p-value': 2.1187596704678572e-27, 'matches': 33, 'identifier': 'GO:0000447', 'populationAnnotationCount': 46}\n",
      "{'description': 'ribosomal large subunit assembly', 'p-value': 4.540321693197129e-25, 'matches': 30, 'identifier': 'GO:0000027', 'populationAnnotationCount': 41}\n",
      "{'description': 'ribosomal large subunit assembly', 'p-value': 4.540321693197129e-25, 'matches': 30, 'identifier': 'GO:0000027', 'populationAnnotationCount': 41}\n",
      "{'description': 'ribosomal large subunit assembly', 'p-value': 4.540321693197129e-25, 'matches': 30, 'identifier': 'GO:0000027', 'populationAnnotationCount': 41}\n",
      "{'description': 'nucleic acid metabolic process', 'p-value': 2.8786666381849832e-24, 'matches': 202, 'identifier': 'GO:0090304', 'populationAnnotationCount': 1798}\n",
      "{'description': 'nucleic acid metabolic process', 'p-value': 2.8786666381849832e-24, 'matches': 202, 'identifier': 'GO:0090304', 'populationAnnotationCount': 1798}\n",
      "{'description': 'nucleic acid metabolic process', 'p-value': 2.8786666381849832e-24, 'matches': 202, 'identifier': 'GO:0090304', 'populationAnnotationCount': 1798}\n",
      "{'description': 'organonitrogen compound biosynthetic process', 'p-value': 6.337956921366802e-24, 'matches': 155, 'identifier': 'GO:1901566', 'populationAnnotationCount': 1159}\n",
      "{'description': 'organonitrogen compound biosynthetic process', 'p-value': 6.337956921366802e-24, 'matches': 155, 'identifier': 'GO:1901566', 'populationAnnotationCount': 1159}\n",
      "{'description': 'organonitrogen compound biosynthetic process', 'p-value': 6.337956921366802e-24, 'matches': 155, 'identifier': 'GO:1901566', 'populationAnnotationCount': 1159}\n",
      "{'description': 'primary metabolic process', 'p-value': 2.4078203025248607e-23, 'matches': 312, 'identifier': 'GO:0044238', 'populationAnnotationCount': 3885}\n",
      "{'description': 'primary metabolic process', 'p-value': 2.4078203025248607e-23, 'matches': 312, 'identifier': 'GO:0044238', 'populationAnnotationCount': 3885}\n",
      "{'description': 'primary metabolic process', 'p-value': 2.4078203025248607e-23, 'matches': 312, 'identifier': 'GO:0044238', 'populationAnnotationCount': 3885}\n",
      "{'description': \"rRNA 5'-end processing\", 'p-value': 9.095930673950712e-23, 'matches': 27, 'identifier': 'GO:0000967', 'populationAnnotationCount': 36}\n",
      "{'description': \"rRNA 5'-end processing\", 'p-value': 9.095930673950712e-23, 'matches': 27, 'identifier': 'GO:0000967', 'populationAnnotationCount': 36}\n",
      "{'description': \"rRNA 5'-end processing\", 'p-value': 9.095930673950712e-23, 'matches': 27, 'identifier': 'GO:0000967', 'populationAnnotationCount': 36}\n",
      "{'description': \"ncRNA 5'-end processing\", 'p-value': 9.095930673950712e-23, 'matches': 27, 'identifier': 'GO:0034471', 'populationAnnotationCount': 36}\n",
      "{'description': \"ncRNA 5'-end processing\", 'p-value': 9.095930673950712e-23, 'matches': 27, 'identifier': 'GO:0034471', 'populationAnnotationCount': 36}\n",
      "{'description': \"ncRNA 5'-end processing\", 'p-value': 9.095930673950712e-23, 'matches': 27, 'identifier': 'GO:0034471', 'populationAnnotationCount': 36}\n",
      "{'description': 'ribonucleoprotein complex assembly', 'p-value': 1.6664240149760316e-22, 'matches': 57, 'identifier': 'GO:0022618', 'populationAnnotationCount': 191}\n",
      "{'description': 'ribonucleoprotein complex assembly', 'p-value': 1.6664240149760316e-22, 'matches': 57, 'identifier': 'GO:0022618', 'populationAnnotationCount': 191}\n",
      "{'description': 'ribonucleoprotein complex assembly', 'p-value': 1.6664240149760316e-22, 'matches': 57, 'identifier': 'GO:0022618', 'populationAnnotationCount': 191}\n",
      "{'description': \"RNA 5'-end processing\", 'p-value': 3.184900943592527e-22, 'matches': 27, 'identifier': 'GO:0000966', 'populationAnnotationCount': 37}\n",
      "{'description': \"RNA 5'-end processing\", 'p-value': 3.184900943592527e-22, 'matches': 27, 'identifier': 'GO:0000966', 'populationAnnotationCount': 37}\n",
      "{'description': \"RNA 5'-end processing\", 'p-value': 3.184900943592527e-22, 'matches': 27, 'identifier': 'GO:0000966', 'populationAnnotationCount': 37}\n",
      "{'description': 'ribonucleoprotein complex subunit organization', 'p-value': 7.257958744252648e-22, 'matches': 58, 'identifier': 'GO:0071826', 'populationAnnotationCount': 203}\n",
      "{'description': 'ribonucleoprotein complex subunit organization', 'p-value': 7.257958744252648e-22, 'matches': 58, 'identifier': 'GO:0071826', 'populationAnnotationCount': 203}\n",
      "{'description': 'ribonucleoprotein complex subunit organization', 'p-value': 7.257958744252648e-22, 'matches': 58, 'identifier': 'GO:0071826', 'populationAnnotationCount': 203}\n",
      "{'description': 'nucleobase-containing compound metabolic process', 'p-value': 1.1512279350385585e-21, 'matches': 213, 'identifier': 'GO:0006139', 'populationAnnotationCount': 2042}\n",
      "{'description': 'nucleobase-containing compound metabolic process', 'p-value': 1.1512279350385585e-21, 'matches': 213, 'identifier': 'GO:0006139', 'populationAnnotationCount': 2042}\n",
      "{'description': 'nucleobase-containing compound metabolic process', 'p-value': 1.1512279350385585e-21, 'matches': 213, 'identifier': 'GO:0006139', 'populationAnnotationCount': 2042}\n",
      "{'description': \"endonucleolytic cleavage to generate mature 5'-end of SSU-rRNA from (SSU-rRNA, 5.8S rRNA, LSU-rRNA)\", 'p-value': 1.2333493800986901e-21, 'matches': 26, 'identifier': 'GO:0000472', 'populationAnnotationCount': 35}\n",
      "{'description': \"endonucleolytic cleavage to generate mature 5'-end of SSU-rRNA from (SSU-rRNA, 5.8S rRNA, LSU-rRNA)\", 'p-value': 1.2333493800986901e-21, 'matches': 26, 'identifier': 'GO:0000472', 'populationAnnotationCount': 35}\n",
      "{'description': \"endonucleolytic cleavage to generate mature 5'-end of SSU-rRNA from (SSU-rRNA, 5.8S rRNA, LSU-rRNA)\", 'p-value': 1.2333493800986901e-21, 'matches': 26, 'identifier': 'GO:0000472', 'populationAnnotationCount': 35}\n",
      "{'description': \"endonucleolytic cleavage in 5'-ETS of tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)\", 'p-value': 4.622163016980155e-21, 'matches': 25, 'identifier': 'GO:0000480', 'populationAnnotationCount': 33}\n",
      "{'description': \"endonucleolytic cleavage in 5'-ETS of tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)\", 'p-value': 4.622163016980155e-21, 'matches': 25, 'identifier': 'GO:0000480', 'populationAnnotationCount': 33}\n",
      "{'description': \"endonucleolytic cleavage in 5'-ETS of tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)\", 'p-value': 4.622163016980155e-21, 'matches': 25, 'identifier': 'GO:0000480', 'populationAnnotationCount': 33}\n",
      "{'description': 'cellular aromatic compound metabolic process', 'p-value': 5.317386281076241e-21, 'matches': 217, 'identifier': 'GO:0006725', 'populationAnnotationCount': 2126}\n",
      "{'description': 'cellular aromatic compound metabolic process', 'p-value': 5.317386281076241e-21, 'matches': 217, 'identifier': 'GO:0006725', 'populationAnnotationCount': 2126}\n",
      "{'description': 'cellular aromatic compound metabolic process', 'p-value': 5.317386281076241e-21, 'matches': 217, 'identifier': 'GO:0006725', 'populationAnnotationCount': 2126}\n",
      "{'description': 'heterocycle metabolic process', 'p-value': 1.553368308697589e-19, 'matches': 215, 'identifier': 'GO:0046483', 'populationAnnotationCount': 2141}\n",
      "{'description': 'heterocycle metabolic process', 'p-value': 1.553368308697589e-19, 'matches': 215, 'identifier': 'GO:0046483', 'populationAnnotationCount': 2141}\n",
      "{'description': 'heterocycle metabolic process', 'p-value': 1.553368308697589e-19, 'matches': 215, 'identifier': 'GO:0046483', 'populationAnnotationCount': 2141}\n",
      "{'description': 'cellular metabolic process', 'p-value': 1.6313653024872816e-19, 'matches': 311, 'identifier': 'GO:0044237', 'populationAnnotationCount': 3993}\n",
      "{'description': 'cellular metabolic process', 'p-value': 1.6313653024872816e-19, 'matches': 311, 'identifier': 'GO:0044237', 'populationAnnotationCount': 3993}\n",
      "{'description': 'cellular metabolic process', 'p-value': 1.6313653024872816e-19, 'matches': 311, 'identifier': 'GO:0044237', 'populationAnnotationCount': 3993}\n",
      "{'description': 'organic substance metabolic process', 'p-value': 3.7880824219652074e-19, 'matches': 313, 'identifier': 'GO:0071704', 'populationAnnotationCount': 4056}\n",
      "{'description': 'organic substance metabolic process', 'p-value': 3.7880824219652074e-19, 'matches': 313, 'identifier': 'GO:0071704', 'populationAnnotationCount': 4056}\n",
      "{'description': 'organic substance metabolic process', 'p-value': 3.7880824219652074e-19, 'matches': 313, 'identifier': 'GO:0071704', 'populationAnnotationCount': 4056}\n",
      "{'description': 'organelle assembly', 'p-value': 9.755708253487189e-19, 'matches': 50, 'identifier': 'GO:0070925', 'populationAnnotationCount': 172}\n",
      "{'description': 'organelle assembly', 'p-value': 9.755708253487189e-19, 'matches': 50, 'identifier': 'GO:0070925', 'populationAnnotationCount': 172}\n",
      "{'description': 'organelle assembly', 'p-value': 9.755708253487189e-19, 'matches': 50, 'identifier': 'GO:0070925', 'populationAnnotationCount': 172}\n",
      "{'description': 'organic cyclic compound metabolic process', 'p-value': 1.0749258753002849e-18, 'matches': 217, 'identifier': 'GO:1901360', 'populationAnnotationCount': 2201}\n",
      "{'description': 'organic cyclic compound metabolic process', 'p-value': 1.0749258753002849e-18, 'matches': 217, 'identifier': 'GO:1901360', 'populationAnnotationCount': 2201}\n",
      "{'description': 'organic cyclic compound metabolic process', 'p-value': 1.0749258753002849e-18, 'matches': 217, 'identifier': 'GO:1901360', 'populationAnnotationCount': 2201}\n",
      "{'description': 'rRNA-containing ribonucleoprotein complex export from nucleus', 'p-value': 2.2937211655589648e-18, 'matches': 31, 'identifier': 'GO:0071428', 'populationAnnotationCount': 63}\n",
      "{'description': 'rRNA-containing ribonucleoprotein complex export from nucleus', 'p-value': 2.2937211655589648e-18, 'matches': 31, 'identifier': 'GO:0071428', 'populationAnnotationCount': 63}\n",
      "{'description': 'rRNA-containing ribonucleoprotein complex export from nucleus', 'p-value': 2.2937211655589648e-18, 'matches': 31, 'identifier': 'GO:0071428', 'populationAnnotationCount': 63}\n",
      "{'description': 'cellular component organization or biogenesis', 'p-value': 4.132713308527583e-18, 'matches': 235, 'identifier': 'GO:0071840', 'populationAnnotationCount': 2521}\n",
      "{'description': 'cellular component organization or biogenesis', 'p-value': 4.132713308527583e-18, 'matches': 235, 'identifier': 'GO:0071840', 'populationAnnotationCount': 2521}\n",
      "{'description': 'cellular component organization or biogenesis', 'p-value': 4.132713308527583e-18, 'matches': 235, 'identifier': 'GO:0071840', 'populationAnnotationCount': 2521}\n",
      "{'description': 'metabolic process', 'p-value': 1.001944552619371e-16, 'matches': 316, 'identifier': 'GO:0008152', 'populationAnnotationCount': 4221}\n",
      "{'description': 'metabolic process', 'p-value': 1.001944552619371e-16, 'matches': 316, 'identifier': 'GO:0008152', 'populationAnnotationCount': 4221}\n",
      "{'description': 'metabolic process', 'p-value': 1.001944552619371e-16, 'matches': 316, 'identifier': 'GO:0008152', 'populationAnnotationCount': 4221}\n",
      "{'description': 'organonitrogen compound metabolic process', 'p-value': 1.6146961030615892e-15, 'matches': 157, 'identifier': 'GO:1901564', 'populationAnnotationCount': 1405}\n",
      "{'description': 'organonitrogen compound metabolic process', 'p-value': 1.6146961030615892e-15, 'matches': 157, 'identifier': 'GO:1901564', 'populationAnnotationCount': 1405}\n",
      "{'description': 'organonitrogen compound metabolic process', 'p-value': 1.6146961030615892e-15, 'matches': 157, 'identifier': 'GO:1901564', 'populationAnnotationCount': 1405}\n",
      "{'description': 'RNA phosphodiester bond hydrolysis, endonucleolytic', 'p-value': 1.1046329136265063e-14, 'matches': 41, 'identifier': 'GO:0090502', 'populationAnnotationCount': 142}\n",
      "{'description': 'RNA phosphodiester bond hydrolysis, endonucleolytic', 'p-value': 1.1046329136265063e-14, 'matches': 41, 'identifier': 'GO:0090502', 'populationAnnotationCount': 142}\n",
      "{'description': 'RNA phosphodiester bond hydrolysis, endonucleolytic', 'p-value': 1.1046329136265063e-14, 'matches': 41, 'identifier': 'GO:0090502', 'populationAnnotationCount': 142}\n",
      "{'description': 'RNA phosphodiester bond hydrolysis', 'p-value': 9.372585996885215e-14, 'matches': 47, 'identifier': 'GO:0090501', 'populationAnnotationCount': 195}\n",
      "{'description': 'RNA phosphodiester bond hydrolysis', 'p-value': 9.372585996885215e-14, 'matches': 47, 'identifier': 'GO:0090501', 'populationAnnotationCount': 195}\n",
      "{'description': 'RNA phosphodiester bond hydrolysis', 'p-value': 9.372585996885215e-14, 'matches': 47, 'identifier': 'GO:0090501', 'populationAnnotationCount': 195}\n",
      "{'description': 'ribonucleoprotein complex export from nucleus', 'p-value': 1.7926864221285404e-13, 'matches': 38, 'identifier': 'GO:0071426', 'populationAnnotationCount': 131}\n",
      "{'description': 'ribonucleoprotein complex export from nucleus', 'p-value': 1.7926864221285404e-13, 'matches': 38, 'identifier': 'GO:0071426', 'populationAnnotationCount': 131}\n",
      "{'description': 'ribonucleoprotein complex export from nucleus', 'p-value': 1.7926864221285404e-13, 'matches': 38, 'identifier': 'GO:0071426', 'populationAnnotationCount': 131}\n",
      "{'description': 'ribonucleoprotein complex localization', 'p-value': 3.1635368364681284e-13, 'matches': 38, 'identifier': 'GO:0071166', 'populationAnnotationCount': 133}\n",
      "{'description': 'ribonucleoprotein complex localization', 'p-value': 3.1635368364681284e-13, 'matches': 38, 'identifier': 'GO:0071166', 'populationAnnotationCount': 133}\n",
      "{'description': 'ribonucleoprotein complex localization', 'p-value': 3.1635368364681284e-13, 'matches': 38, 'identifier': 'GO:0071166', 'populationAnnotationCount': 133}\n",
      "{'description': 'nuclear export', 'p-value': 1.6215008178279745e-12, 'matches': 38, 'identifier': 'GO:0051168', 'populationAnnotationCount': 139}\n",
      "{'description': 'nuclear export', 'p-value': 1.6215008178279745e-12, 'matches': 38, 'identifier': 'GO:0051168', 'populationAnnotationCount': 139}\n",
      "{'description': 'nuclear export', 'p-value': 1.6215008178279745e-12, 'matches': 38, 'identifier': 'GO:0051168', 'populationAnnotationCount': 139}\n",
      "{'description': 'nucleocytoplasmic transport', 'p-value': 1.2816314295274604e-11, 'matches': 43, 'identifier': 'GO:0006913', 'populationAnnotationCount': 186}\n",
      "{'description': 'nucleocytoplasmic transport', 'p-value': 1.2816314295274604e-11, 'matches': 43, 'identifier': 'GO:0006913', 'populationAnnotationCount': 186}\n",
      "{'description': 'nucleocytoplasmic transport', 'p-value': 1.2816314295274604e-11, 'matches': 43, 'identifier': 'GO:0006913', 'populationAnnotationCount': 186}\n",
      "{'description': 'nuclear transport', 'p-value': 1.5771428123620305e-11, 'matches': 43, 'identifier': 'GO:0051169', 'populationAnnotationCount': 187}\n",
      "{'description': 'nuclear transport', 'p-value': 1.5771428123620305e-11, 'matches': 43, 'identifier': 'GO:0051169', 'populationAnnotationCount': 187}\n",
      "{'description': 'nuclear transport', 'p-value': 1.5771428123620305e-11, 'matches': 43, 'identifier': 'GO:0051169', 'populationAnnotationCount': 187}\n",
      "{'description': 'ribosomal small subunit assembly', 'p-value': 1.4017533362469904e-10, 'matches': 17, 'identifier': 'GO:0000028', 'populationAnnotationCount': 29}\n",
      "{'description': 'ribosomal small subunit assembly', 'p-value': 1.4017533362469904e-10, 'matches': 17, 'identifier': 'GO:0000028', 'populationAnnotationCount': 29}\n",
      "{'description': 'ribosomal small subunit assembly', 'p-value': 1.4017533362469904e-10, 'matches': 17, 'identifier': 'GO:0000028', 'populationAnnotationCount': 29}\n",
      "{'description': 'regulation of translational fidelity', 'p-value': 1.4017533362469904e-10, 'matches': 17, 'identifier': 'GO:0006450', 'populationAnnotationCount': 29}\n",
      "{'description': 'regulation of translational fidelity', 'p-value': 1.4017533362469904e-10, 'matches': 17, 'identifier': 'GO:0006450', 'populationAnnotationCount': 29}\n",
      "{'description': 'regulation of translational fidelity', 'p-value': 1.4017533362469904e-10, 'matches': 17, 'identifier': 'GO:0006450', 'populationAnnotationCount': 29}\n",
      "{'description': 'cellular nitrogen compound biosynthetic process', 'p-value': 2.2964962635219464e-10, 'matches': 176, 'identifier': 'GO:0044271', 'populationAnnotationCount': 1865}\n",
      "{'description': 'cellular nitrogen compound biosynthetic process', 'p-value': 2.2964962635219464e-10, 'matches': 176, 'identifier': 'GO:0044271', 'populationAnnotationCount': 1865}\n",
      "{'description': 'cellular nitrogen compound biosynthetic process', 'p-value': 2.2964962635219464e-10, 'matches': 176, 'identifier': 'GO:0044271', 'populationAnnotationCount': 1865}\n",
      "{'description': 'regulation of translational elongation', 'p-value': 3.838595686232152e-10, 'matches': 19, 'identifier': 'GO:0006448', 'populationAnnotationCount': 39}\n",
      "{'description': 'regulation of translational elongation', 'p-value': 3.838595686232152e-10, 'matches': 19, 'identifier': 'GO:0006448', 'populationAnnotationCount': 39}\n",
      "{'description': 'regulation of translational elongation', 'p-value': 3.838595686232152e-10, 'matches': 19, 'identifier': 'GO:0006448', 'populationAnnotationCount': 39}\n",
      "{'description': 'ribosomal subunit export from nucleus', 'p-value': 3.3827175235645332e-09, 'matches': 20, 'identifier': 'GO:0000054', 'populationAnnotationCount': 48}\n",
      "{'description': 'ribosomal subunit export from nucleus', 'p-value': 3.3827175235645332e-09, 'matches': 20, 'identifier': 'GO:0000054', 'populationAnnotationCount': 48}\n",
      "{'description': 'ribosomal subunit export from nucleus', 'p-value': 3.3827175235645332e-09, 'matches': 20, 'identifier': 'GO:0000054', 'populationAnnotationCount': 48}\n",
      "{'description': 'establishment of ribosome localization', 'p-value': 3.3827175235645332e-09, 'matches': 20, 'identifier': 'GO:0033753', 'populationAnnotationCount': 48}\n",
      "{'description': 'establishment of ribosome localization', 'p-value': 3.3827175235645332e-09, 'matches': 20, 'identifier': 'GO:0033753', 'populationAnnotationCount': 48}\n",
      "{'description': 'establishment of ribosome localization', 'p-value': 3.3827175235645332e-09, 'matches': 20, 'identifier': 'GO:0033753', 'populationAnnotationCount': 48}\n",
      "{'description': 'nucleic acid phosphodiester bond hydrolysis', 'p-value': 3.7004468925030307e-09, 'matches': 50, 'identifier': 'GO:0090305', 'populationAnnotationCount': 281}\n",
      "{'description': 'nucleic acid phosphodiester bond hydrolysis', 'p-value': 3.7004468925030307e-09, 'matches': 50, 'identifier': 'GO:0090305', 'populationAnnotationCount': 281}\n",
      "{'description': 'nucleic acid phosphodiester bond hydrolysis', 'p-value': 3.7004468925030307e-09, 'matches': 50, 'identifier': 'GO:0090305', 'populationAnnotationCount': 281}\n",
      "{'description': 'ribosome localization', 'p-value': 5.406345289764284e-09, 'matches': 20, 'identifier': 'GO:0033750', 'populationAnnotationCount': 49}\n",
      "{'description': 'ribosome localization', 'p-value': 5.406345289764284e-09, 'matches': 20, 'identifier': 'GO:0033750', 'populationAnnotationCount': 49}\n",
      "{'description': 'ribosome localization', 'p-value': 5.406345289764284e-09, 'matches': 20, 'identifier': 'GO:0033750', 'populationAnnotationCount': 49}\n",
      "{'description': 'rRNA export from nucleus', 'p-value': 1.1416441095675305e-07, 'matches': 12, 'identifier': 'GO:0006407', 'populationAnnotationCount': 18}\n",
      "{'description': 'rRNA export from nucleus', 'p-value': 1.1416441095675305e-07, 'matches': 12, 'identifier': 'GO:0006407', 'populationAnnotationCount': 18}\n",
      "{'description': 'rRNA export from nucleus', 'p-value': 1.1416441095675305e-07, 'matches': 12, 'identifier': 'GO:0006407', 'populationAnnotationCount': 18}\n",
      "{'description': 'rRNA transport', 'p-value': 1.1416441095675305e-07, 'matches': 12, 'identifier': 'GO:0051029', 'populationAnnotationCount': 18}\n",
      "{'description': 'rRNA transport', 'p-value': 1.1416441095675305e-07, 'matches': 12, 'identifier': 'GO:0051029', 'populationAnnotationCount': 18}\n",
      "{'description': 'rRNA transport', 'p-value': 1.1416441095675305e-07, 'matches': 12, 'identifier': 'GO:0051029', 'populationAnnotationCount': 18}\n",
      "{'description': 'cellular macromolecule biosynthetic process', 'p-value': 7.019963923000122e-07, 'matches': 168, 'identifier': 'GO:0034645', 'populationAnnotationCount': 1897}\n",
      "{'description': 'cellular macromolecule biosynthetic process', 'p-value': 7.019963923000122e-07, 'matches': 168, 'identifier': 'GO:0034645', 'populationAnnotationCount': 1897}\n",
      "{'description': 'cellular macromolecule biosynthetic process', 'p-value': 7.019963923000122e-07, 'matches': 168, 'identifier': 'GO:0034645', 'populationAnnotationCount': 1897}\n",
      "{'description': 'macromolecule biosynthetic process', 'p-value': 1.0231722143435962e-06, 'matches': 168, 'identifier': 'GO:0009059', 'populationAnnotationCount': 1905}\n",
      "{'description': 'macromolecule biosynthetic process', 'p-value': 1.0231722143435962e-06, 'matches': 168, 'identifier': 'GO:0009059', 'populationAnnotationCount': 1905}\n",
      "{'description': 'macromolecule biosynthetic process', 'p-value': 1.0231722143435962e-06, 'matches': 168, 'identifier': 'GO:0009059', 'populationAnnotationCount': 1905}\n",
      "{'description': 'RNA secondary structure unwinding', 'p-value': 4.573975841860619e-06, 'matches': 13, 'identifier': 'GO:0010501', 'populationAnnotationCount': 27}\n",
      "{'description': 'RNA secondary structure unwinding', 'p-value': 4.573975841860619e-06, 'matches': 13, 'identifier': 'GO:0010501', 'populationAnnotationCount': 27}\n",
      "{'description': 'RNA secondary structure unwinding', 'p-value': 4.573975841860619e-06, 'matches': 13, 'identifier': 'GO:0010501', 'populationAnnotationCount': 27}\n",
      "{'description': 'assembly of large subunit precursor of preribosome', 'p-value': 3.060857629717604e-05, 'matches': 8, 'identifier': 'GO:1902626', 'populationAnnotationCount': 10}\n",
      "{'description': 'assembly of large subunit precursor of preribosome', 'p-value': 3.060857629717604e-05, 'matches': 8, 'identifier': 'GO:1902626', 'populationAnnotationCount': 10}\n",
      "{'description': 'assembly of large subunit precursor of preribosome', 'p-value': 3.060857629717604e-05, 'matches': 8, 'identifier': 'GO:1902626', 'populationAnnotationCount': 10}\n",
      "{'description': 'tRNA metabolic process', 'p-value': 3.839845814555503e-05, 'matches': 35, 'identifier': 'GO:0006399', 'populationAnnotationCount': 204}\n",
      "{'description': 'tRNA metabolic process', 'p-value': 3.839845814555503e-05, 'matches': 35, 'identifier': 'GO:0006399', 'populationAnnotationCount': 204}\n",
      "{'description': 'tRNA metabolic process', 'p-value': 3.839845814555503e-05, 'matches': 35, 'identifier': 'GO:0006399', 'populationAnnotationCount': 204}\n",
      "{'description': 'cellular protein metabolic process', 'p-value': 0.0014520196051751063, 'matches': 145, 'identifier': 'GO:0044267', 'populationAnnotationCount': 1719}\n",
      "{'description': 'cellular protein metabolic process', 'p-value': 0.0014520196051751063, 'matches': 145, 'identifier': 'GO:0044267', 'populationAnnotationCount': 1719}\n",
      "{'description': 'cellular protein metabolic process', 'p-value': 0.0014520196051751063, 'matches': 145, 'identifier': 'GO:0044267', 'populationAnnotationCount': 1719}\n",
      "{'description': 'cellular macromolecular complex assembly', 'p-value': 0.003122579225327374, 'matches': 62, 'identifier': 'GO:0034622', 'populationAnnotationCount': 560}\n",
      "{'description': 'cellular macromolecular complex assembly', 'p-value': 0.003122579225327374, 'matches': 62, 'identifier': 'GO:0034622', 'populationAnnotationCount': 560}\n",
      "{'description': 'cellular macromolecular complex assembly', 'p-value': 0.003122579225327374, 'matches': 62, 'identifier': 'GO:0034622', 'populationAnnotationCount': 560}\n",
      "{'description': 'nucleobase-containing compound transport', 'p-value': 0.0040138088332279396, 'matches': 30, 'identifier': 'GO:0015931', 'populationAnnotationCount': 192}\n",
      "{'description': 'nucleobase-containing compound transport', 'p-value': 0.0040138088332279396, 'matches': 30, 'identifier': 'GO:0015931', 'populationAnnotationCount': 192}\n",
      "{'description': 'nucleobase-containing compound transport', 'p-value': 0.0040138088332279396, 'matches': 30, 'identifier': 'GO:0015931', 'populationAnnotationCount': 192}\n",
      "{'description': 'ncRNA transcription', 'p-value': 0.005737576472936224, 'matches': 15, 'identifier': 'GO:0098781', 'populationAnnotationCount': 59}\n",
      "{'description': 'ncRNA transcription', 'p-value': 0.005737576472936224, 'matches': 15, 'identifier': 'GO:0098781', 'populationAnnotationCount': 59}\n",
      "{'description': 'ncRNA transcription', 'p-value': 0.005737576472936224, 'matches': 15, 'identifier': 'GO:0098781', 'populationAnnotationCount': 59}\n",
      "{'description': 'cellular biosynthetic process', 'p-value': 0.006669154180539723, 'matches': 187, 'identifier': 'GO:0044249', 'populationAnnotationCount': 2431}\n",
      "{'description': 'cellular biosynthetic process', 'p-value': 0.006669154180539723, 'matches': 187, 'identifier': 'GO:0044249', 'populationAnnotationCount': 2431}\n",
      "{'description': 'cellular biosynthetic process', 'p-value': 0.006669154180539723, 'matches': 187, 'identifier': 'GO:0044249', 'populationAnnotationCount': 2431}\n",
      "{'description': 'transcription from RNA polymerase I promoter', 'p-value': 0.0077382234271341725, 'matches': 16, 'identifier': 'GO:0006360', 'populationAnnotationCount': 68}\n",
      "{'description': 'transcription from RNA polymerase I promoter', 'p-value': 0.0077382234271341725, 'matches': 16, 'identifier': 'GO:0006360', 'populationAnnotationCount': 68}\n",
      "{'description': 'transcription from RNA polymerase I promoter', 'p-value': 0.0077382234271341725, 'matches': 16, 'identifier': 'GO:0006360', 'populationAnnotationCount': 68}\n",
      "{'description': 'organic substance biosynthetic process', 'p-value': 0.012047647002194887, 'matches': 187, 'identifier': 'GO:1901576', 'populationAnnotationCount': 2449}\n",
      "{'description': 'biosynthetic process', 'p-value': 0.012408891657273487, 'matches': 189, 'identifier': 'GO:0009058', 'populationAnnotationCount': 2483}\n",
      "{'description': 'ribosomal large subunit export from nucleus', 'p-value': 0.01268136272859071, 'matches': 10, 'identifier': 'GO:0000055', 'populationAnnotationCount': 28}\n",
      "{'description': 'snRNA pseudouridine synthesis', 'p-value': 0.015915834537195857, 'matches': 6, 'identifier': 'GO:0031120', 'populationAnnotationCount': 9}\n",
      "{'description': 'snRNA modification', 'p-value': 0.015915834537195857, 'matches': 6, 'identifier': 'GO:0040031', 'populationAnnotationCount': 9}\n",
      "{'description': 'macromolecular complex assembly', 'p-value': 0.031190256949736643, 'matches': 63, 'identifier': 'GO:0065003', 'populationAnnotationCount': 612}\n",
      "{'description': 'tRNA transcription from RNA polymerase III promoter', 'p-value': 0.031842550772909164, 'matches': 8, 'identifier': 'GO:0042797', 'populationAnnotationCount': 19}\n",
      "{'description': 'RNA export from nucleus', 'p-value': 0.03465344038767785, 'matches': 19, 'identifier': 'GO:0006405', 'populationAnnotationCount': 102}\n",
      "{'description': 'protein metabolic process', 'p-value': 0.03624084467492526, 'matches': 147, 'identifier': 'GO:0019538', 'populationAnnotationCount': 1839}\n",
      "{'description': 'RNA localization', 'p-value': 0.04963893130926979, 'matches': 24, 'identifier': 'GO:0006403', 'populationAnnotationCount': 153}\n"
     ]
    }
   ],
   "source": [
    "#For each set of genes\n",
    "# gene_sets = [('klsig','act'),\n",
    "#             ('scsig','act'),\n",
    "#             ('klscsig','act'),\n",
    "#             ('klsig','rep'),\n",
    "#             ('scsig','rep'),\n",
    "#             ('klscsig','rep')]\n",
    "\n",
    "#Add in the one gene that is significant for both but repressed in SC and activated in KL. \n",
    "klsig_act_genes = list(kl_sc_PKA_data_klsig_scunsig[kl_sc_PKA_data_klsig_scunsig['PKA(AS)+1NMPP1_KL']>0]['sc_genename'])\n",
    "klsig_act_genes.append('YJL208C')\n",
    "\n",
    "#Add in the one gene that is significant for both but activated in SC and repressed in KL.\n",
    "klsig_rep_genes = list(kl_sc_PKA_data_klsig_scunsig[kl_sc_PKA_data_klsig_scunsig['PKA(AS)+1NMPP1_KL']<0]['sc_genename'])\n",
    "klsig_rep_genes.append('YNL014W')\n",
    "\n",
    "#Add in the one gene that is significant for both but activated in SC and repressed in KL.\n",
    "scsig_act_genes = list(kl_sc_PKA_data_scsig_klunsig[kl_sc_PKA_data_scsig_klunsig['PKA(AS)+1NMPP1_SC']>0]['sc_genename'])\n",
    "scsig_act_genes.append('YNL014W')\n",
    "\n",
    "#Add in the one gene that is significant for both but repressed in SC and activated in KL. \n",
    "scsig_rep_genes = list(kl_sc_PKA_data_scsig_klunsig[kl_sc_PKA_data_scsig_klunsig['PKA(AS)+1NMPP1_SC']<0]['sc_genename'])\n",
    "scsig_rep_genes.append('YJL208C')\n",
    "\n",
    "gene_sets = ['klsig_act', 'klsig_rep','scsig_act','scsig_rep','klscsig_act','klscsig_rep']\n",
    "\n",
    "gene_set_dict = {'klsig_act': klsig_act_genes,\n",
    "                 'klsig_rep': klsig_rep_genes,\n",
    "                 'scsig_act': scsig_act_genes,\n",
    "                 'scsig_rep': scsig_rep_genes,\n",
    "                 'klscsig_act': list(kl_sc_PKA_data_klscsig[(kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_SC']>0) & (kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_KL']>0)]['sc_genename']),\n",
    "                 'klscsig_rep': list(kl_sc_PKA_data_klscsig[(kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_SC']<0) & (kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_KL']<0)]['sc_genename'])\n",
    "                }\n",
    "\n",
    "missing_genes = ['Scer_YGOB_YDR134C', 'Scer_RDN18-1', \n",
    "                 'Scer_YGOB_SDC25', 'Scer_RDN25-1', \n",
    "                 'Scer_RDN58-1', 'Scer_YGOB_Anc_7.495', \n",
    "                 'Scer_YGOB_ADL119W', 'Scer_RDN5-1']\n",
    "\n",
    "pval_cutoff = 0.01\n",
    "\n",
    "for gene_set in gene_sets: \n",
    "    print(gene_set)\n",
    "    gene_set_list = gene_set_dict[gene_set]\n",
    "    \n",
    "    #filter out missing genes \n",
    "    print(\"The following missing genes are filtered out:\")\n",
    "    print(set(gene_set_list) & set(missing_genes))\n",
    "    gene_set_list_filtered = list(set(gene_set_list)-set(missing_genes))\n",
    "    new_query = query_all.clone()\n",
    "    \n",
    "    #subset query based on desired gene set\n",
    "    new_query.add_constraint(\"Gene.secondaryIdentifier\", \"ONE OF\", gene_set_list_filtered)\n",
    "    \n",
    "    #Generate a list\n",
    "    new_list = service.create_list(new_query.select('Gene.id'), \"Gene\", name = \"20170905\" + gene_set)\n",
    "    \n",
    "    #Run Go enrichment on the list\n",
    "    results = new_list.calculate_enrichment('go_enrichment_for_gene')\n",
    "    results_list = list(results)    \n",
    "    for result in results_list: \n",
    "        print(result)\n",
    "        if result['p-value']<pval_cutoff:\n",
    "            print(result)\n",
    "            print(result)\n",
    "            \n",
    "    #Given a P-value cutoff\n",
    "    # Save the go term \n",
    "    # Save the contingency table\n",
    "    # Save the related genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "233\n",
      "8\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'Scer_YGOB_SDC25'}"
      ]
     },
     "execution_count": 264,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_genes = ['Scer_YGOB_YDR134C', 'Scer_RDN18-1', \n",
    "                 'Scer_YGOB_SDC25', 'Scer_RDN25-1', \n",
    "                 'Scer_RDN58-1', 'Scer_YGOB_Anc_7.495', \n",
    "                 'Scer_YGOB_ADL119W', 'Scer_RDN5-1']\n",
    "\n",
    "\n",
    "klsig_act_genes = list(kl_sc_PKA_data_klsig_scunsig[kl_sc_PKA_data_klsig_scunsig['PKA(AS)+1NMPP1_KL']>0]['sc_genename'])\n",
    "klsig_act_genes.append('YJL208C')\n",
    "print(len(set(klsig_act_genes)))\n",
    "print(len(set(missing_genes)))\n",
    "set(klsig_act_genes) & set(missing_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['YBR196C', 'YFL007W', 'YBL049W', 'YDR222W']"
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selection = [1,4,6,10]\n",
    "gene_subset = [query_list[jj].secondaryIdentifier for jj in selection]\n",
    "gene_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "list2 = list(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(list2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'primaryIdentifier'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-181-b91906e2eba9>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mrow\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mquery_list\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"primaryIdentifier\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"secondaryIdentifier\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"organism.shortName\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m         \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"symbol\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"name\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"id\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/usr/local/lib/python3.5/dist-packages/intermine-1.9.9-py3.5.egg/intermine/results.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    233\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    234\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 235\u001b[0;31m             \u001b[0mindex\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_index_for\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    236\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    237\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.5/dist-packages/intermine-1.9.9-py3.5.egg/intermine/results.py\u001b[0m in \u001b[0;36m_get_index_for\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    213\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex_map\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mheadless_view\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    214\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 215\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex_map\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    216\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    217\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__str__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'primaryIdentifier'"
     ]
    }
   ],
   "source": [
    "for row in query_list:\n",
    "    print(row[\"primaryIdentifier\"], row[\"secondaryIdentifier\"], row[\"organism.shortName\"], \\\n",
    "        row[\"symbol\"], row[\"name\"], row[\"id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'go_enrichment_for_gene': {'description': 'GO terms enriched for items in this list.',\n",
       "  'enrich': 'goAnnotation.ontologyTerm.parents.name',\n",
       "  'enrichIdentifier': 'goAnnotation.ontologyTerm.parents.identifier',\n",
       "  'filters': 'biological_process,cellular_component,molecular_function',\n",
       "  'name': 'go_enrichment_for_gene',\n",
       "  'startClass': 'Gene',\n",
       "  'startClassDisplay': 'secondaryIdentifier',\n",
       "  'targets': ['Gene'],\n",
       "  'title': 'Gene Ontology Enrichment',\n",
       "  'widgetType': 'enrichment'},\n",
       " 'interactions': {'description': 'Genes (from the list or not) that interact with genes in this list.  Counts may include the same interaction more than once if observed in multiple experiments.',\n",
       "  'filters': None,\n",
       "  'name': 'interactions',\n",
       "  'startClass': None,\n",
       "  'targets': ['Gene'],\n",
       "  'title': 'Interactions',\n",
       "  'widgetType': 'table'},\n",
       " 'orthologues': {'description': 'Shows the number of Genes in list with orthologue not the number of orthologues.',\n",
       "  'filters': None,\n",
       "  'name': 'orthologues',\n",
       "  'startClass': None,\n",
       "  'targets': ['Gene'],\n",
       "  'title': 'Orthologues',\n",
       "  'widgetType': 'table'},\n",
       " 'pathway_enrichment': {'description': 'Pathways enriched for genes in this list - data from KEGG and Reactome',\n",
       "  'enrich': 'pathways.name',\n",
       "  'enrichIdentifier': 'pathways.identifier',\n",
       "  'filters': 'All,KEGG pathways data set,Reactome data set',\n",
       "  'name': 'pathway_enrichment',\n",
       "  'startClass': 'Gene',\n",
       "  'startClassDisplay': 'secondaryIdentifier',\n",
       "  'targets': ['Gene'],\n",
       "  'title': 'Pathway Enrichment',\n",
       "  'widgetType': 'enrichment'},\n",
       " 'publication_enrichment': {'description': 'Publications enriched for genes in this list.',\n",
       "  'enrich': 'publications.title',\n",
       "  'enrichIdentifier': 'publications.pubMedId',\n",
       "  'filters': None,\n",
       "  'name': 'publication_enrichment',\n",
       "  'startClass': 'Gene',\n",
       "  'startClassDisplay': 'secondaryIdentifier',\n",
       "  'targets': ['Gene'],\n",
       "  'title': 'Publication Enrichment',\n",
       "  'widgetType': 'enrichment'}}"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##To see a list of widgets that are available\n",
    "service.widgets \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "intermine.results.EnrichmentLine"
      ]
     },
     "execution_count": 265,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "results = new_list.calculate_enrichment('go_enrichment_for_gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "results_parser = results.parser\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "res = list(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "item = res[0]\n",
    "\n",
    "item['matches']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'description': 'cellular lipid catabolic process', 'p-value': 2.0996195815425373e-06, 'matches': 14, 'identifier': 'GO:0044242', 'populationAnnotationCount': 46}\n"
     ]
    }
   ],
   "source": [
    "print(res[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Get genes from a go term.  \n",
    "\n",
    "from intermine.webservice import Service\n",
    "service = Service(\"http://yeastmine.yeastgenome.org:443/yeastmine/service\")\n",
    "query = service.new_query(\"Gene\")\n",
    "query.add_view(\n",
    "    \"primaryIdentifier\", \"secondaryIdentifier\", \"symbol\", \"featureType\",\n",
    "    \"goAnnotation.qualifier\", \"goAnnotation.ontologyTerm.identifier\",\n",
    "    \"goAnnotation.ontologyTerm.name\", \"goAnnotation.ontologyTerm.namespace\",\n",
    "    \"goAnnotation.evidence.code.code\", \"goAnnotation.evidence.code.withText\",\n",
    "    \"goAnnotation.evidence.code.annotType\",\n",
    "    \"goAnnotation.evidence.publications.pubMedId\",\n",
    "    \"goAnnotation.evidence.publications.citation\"\n",
    ")\n",
    "query.add_sort_order(\"Gene.secondaryIdentifier\", \"ASC\")\n",
    "query.add_constraint(\"goAnnotation.ontologyTerm\", \"LOOKUP\", \"GO:0000182\", code = \"A\")\n",
    "\n",
    "for row in query.rows():\n",
    "    print row[\"primaryIdentifier\"], row[\"secondaryIdentifier\"], row[\"symbol\"], row[\"featureType\"], \\\n",
    "        row[\"goAnnotation.qualifier\"], row[\"goAnnotation.ontologyTerm.identifier\"], \\\n",
    "        row[\"goAnnotation.ontologyTerm.name\"], row[\"goAnnotation.ontologyTerm.namespace\"], \\\n",
    "        row[\"goAnnotation.evidence.code.code\"], row[\"goAnnotation.evidence.code.withText\"], \\\n",
    "        row[\"goAnnotation.evidence.code.annotType\"], \\\n",
    "        row[\"goAnnotation.evidence.publications.pubMedId\"], \\\n",
    "        row[\"goAnnotation.evidence.publications.citation\"]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
