{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "what computer are you on? a = Ben's laptop, b = gpucluster, c = Ben's desktop, d = othera\n",
      "base directory is C:\\Users\\BMH_work\\github\\yeast_esr_expression_analysis\n",
      "Importing yeast_esr_exp and setting base_dir and data_processing_dir\n",
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "# %load notebook_setup.py\n",
    "#Use %load notebook_setup.py to load these commands at the top of your\n",
    "#ipython notebook.  Includes running the std_libraries.py file. \n",
    "\n",
    "%run std_libraries.py\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "# %matplotlib \n",
    "# plt.ioff()\n",
    "#%matplotlib inline\n",
    "%matplotlib notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "what computer are you on? a = Ben's laptop, b = gpucluster, c = Ben's desktop, d = othera\n",
      "base directory is C:\\Users\\BMH_work\\github\\expression_broad_data\n",
      "I am about to import a library\n",
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n",
      "are you online? Yes/No Yes\n"
     ]
    }
   ],
   "source": [
    "# # %load std_libraries.py\n",
    "# #to keep my scripts consistent, I am adding this to the top of all scripts with %load std_libraries.py\n",
    "# import sys\n",
    "# #Indicate operating environment and import core modules\n",
    "# location_input = input(\"what computer are you on? a = Ben's laptop, b = gpucluster, c = Ben's desktop, d = other\")\n",
    "# location_dict = {'a': \"C:\\\\Users\\\\BMH_work\\\\github\\\\expression_broad_data\", 'b': \"/home/heineike/github/expression_broad_data\",\n",
    "#                  'c': \"C:\\\\Users\\\\Ben\\\\Documents\\\\GitHub\\\\expression_broad_data\", 'd':'you need to add your location to the location_dict'}\n",
    "# base_dir = location_dict[location_input]\n",
    "# print(\"base directory is \" + base_dir)\n",
    "\n",
    "# if sys.path[-1] != base_dir:\n",
    "#     sys.path.append(base_dir)\n",
    "#     print(\"Added \" + base_dir + \" to path: \" )\n",
    "#     print(sys.path)\n",
    "\n",
    "# import os\n",
    "\n",
    "# print(\"I am about to import a library\")\n",
    "# from core import expression_plots \n",
    "# from core import io_library \n",
    "# %load_ext autoreload\n",
    "# %autoreload 2\n",
    "# # %matplotlib \n",
    "# # plt.ioff()\n",
    "# #%matplotlib inline\n",
    "# %matplotlib notebook\n",
    "# data_processing_dir = base_dir + os.sep + os.path.normpath(\"expression_data\") + os.sep\n",
    "\n",
    "# import pandas as pd\n",
    "# import numpy as np\n",
    "# import matplotlib.pyplot as plt\n",
    "# #import matplotlib.colormap as cm\n",
    "# #from matplotlib_venn import venn2\n",
    "# #for my windows10 laptop I had to install this package using pip rather than anaconda.  \n",
    "# #import seaborn as sns; sns.set(style=\"ticks\", color_codes=True)\n",
    "# #from sklearn import linear_model\n",
    "# import pickle\n",
    "# #import subprocess\n",
    "# #import networkx as nx\n",
    "# #import scipy.stats as stats\n",
    "# #import statsmodels.graphics.gofplots as stats_graph\n",
    "# #import scipy.cluster.hierarchy as sch\n",
    "\n",
    "# # from Bio import SeqIO\n",
    "# # from Bio import SeqFeature as sf\n",
    "# # from Bio.Alphabet import generic_dna\n",
    "# # from Bio.Seq import Seq\n",
    "\n",
    "# # import re\n",
    "\n",
    "# from collections import Counter\n",
    "# # import scipy.stats as stats\n",
    "# # from itertools import chain\n",
    "# #from itertools import product\n",
    "# #this only works if you are online\n",
    "# online_input = input(\"are you online? Yes/No \")\n",
    "# if online_input == \"Yes\": \n",
    "#     import plotly.plotly as py\n",
    "#     import plotly.graph_objs as pygo\n",
    "#     import plotly.tools as pytools\n",
    "#     py.sign_in('heineike02_student','9dMTMZgJMgUP0YX0P5mQ')\n",
    "#     #py.sign_in('heineike02', 'APPjKrtARaN2ZgUYIkqr')\n",
    "    \n",
    "# # for phylogenetic trees: \n",
    "# # from ete3 import Tree\n",
    "# #ete3 is not officially supported on windows, and so must be loaded via pip: \n",
    "# # pip install -U https://github.com/etetoolkit/ete/archive/qt5.zip\n",
    "# # ref: https://groups.google.com/forum/#!topic/etetoolkit/6NblSBPij4o\n",
    "\n",
    "# #for scraping internet data (e.g. ncbi)\n",
    "# #import requests\n",
    "# #from lxml import etree    #parses xml output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load activated and repressed genes from \n",
    "#20181031_klscpka_r1g1_m24_SC_analysis and\n",
    "#20181031_klscpka_r1g1_m24_KL_analysis\n",
    "\n",
    "target_set_files = {'SC': {'pkainh_act':'20200608_pkainh_act_SC.csv',\n",
    "                           'pkainh_rep':'20200608_pkainh_rep_SC.csv'},\n",
    "                    'KL': {'pkainh_act':'20200608_pkainh_act_KL.csv',\n",
    "                           'pkainh_rep':'20200608_pkainh_rep_KL.csv'}\n",
    "                    }\n",
    "\n",
    "target_sets = {}\n",
    "\n",
    "for spec,spec_sets in target_set_files.items():\n",
    "    target_sets_spec = {}\n",
    "    \n",
    "    for set_name, fname in spec_sets.items():\n",
    "        target_set_df = pd.read_csv(os.path.normpath(data_processing_dir + 'kl_sc_PKA_as_m24_r1g1_20181017/' + fname ), index_col=0)\n",
    "        target_set = set(target_set_df.index)\n",
    "        target_sets_spec[set_name] = target_set\n",
    "\n",
    "    target_sets[spec] = target_sets_spec\n",
    "    \n",
    "# #Older version with .pkl data\n",
    "# target_sets_dict = {}\n",
    "# for spec in ['KL', 'SC']: \n",
    "#     fname = data_processing_dir + '\\\\kl_sc_PKA_as_m24_r1g1_20181017\\\\20181203_r1g1_m24_pka_targets_' + spec + '.pkl'\n",
    "    \n",
    "#     with open(fname, 'rb') as file:\n",
    "#         target_sets_dict[spec] = pickle.load(file)\n",
    "        \n",
    "#Load Kl and SC datasets for determining background sets\n",
    "\n",
    "#Load DEseq data for SCer PKA AS -/+ NMPP1 \n",
    "#pkainh_deseq_SC = pd.read_csv(os.path.normpath(data_processing_dir + '\\\\kl_sc_PKA_as_m24_r1g1_20181017\\\\20181017_deseq_SC_AS_WT_nmpp1.csv'), index_col=0)\n",
    "pkainh_deseq_SC = pd.read_csv(os.path.normpath(data_processing_dir + '\\\\kl_sc_PKA_as_m24_r1g1_20181017\\\\20200603_deseq_SC_AS_WT_nmpp1.csv'), index_col=0)\n",
    "\n",
    "pkainh_deseq_SC['SC_common_name'] = yeast_esr_exp.SC_common_name_lookup(pkainh_deseq_SC.index)\n",
    "\n",
    "# #load Deseq data for KLac AS -/+ NMPP1\n",
    "# pkainh_deseq_KL = pd.read_csv(os.path.normpath(data_processing_dir + '\\\\kl_sc_PKA_as_m24_r1g1_20181017\\\\20181017_deseq_KL_AS_WT_nmpp1.csv'), index_col=0)\n",
    "# pkainh_deseq_KL['kl_genename'] = yeast_esr_exp.kl_genename_convert_list(list(pkainh_deseq_KL.index))\n",
    "# pkainh_deseq_KL.set_index('kl_genename', inplace=True)\n",
    "\n",
    "# #Add column for common names\n",
    "# pkainh_deseq_KL['SC_common_name'] = yeast_esr_exp.SC_common_name_lookup_KL(pkainh_deseq_KL.index)\n",
    "\n",
    "# #Load KL to SC ortholog mapping\n",
    "\n",
    "# kl_orthologs = pd.read_pickle(data_processing_dir + \"ortholog_files_YGOB/kl_orthologs.pkl\")\n",
    "kl_orthologs = yeast_esr_exp.read_orth_lookup_table('Klac', 'Scer', data_processing_dir + os.sep + 'ortholog_files_YGOB' + os.sep)\n",
    "\n",
    "#Set up go term list and database: \n",
    "GO_aspect = 'P'\n",
    "#The three GO_aspect are: \n",
    "#C = cellular_component\n",
    "#F = molecular_function\n",
    "#P = biological_process\n",
    "go_slims_aspect, go_term_list = yeast_esr_exp.load_goslim_data(GO_aspect, go_slim_fname = 'go_slim_mapping_20181204.tab')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, 2, 3, 4, 2]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = []\n",
    "b = [1,2,3]\n",
    "c = [4,2]\n",
    "a + b + c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KLLA0D08140r from pkainh_act not in kl ortholog index\n",
      "EBG00000981852 from pkainh_act not in kl ortholog index\n",
      "EBG00000982140 from pkainh_act not in kl ortholog index\n",
      "KLLA0F15059r from pkainh_act not in kl ortholog index\n",
      "EBG00000981757 from pkainh_act not in kl ortholog index\n",
      "EBG00000981576 from pkainh_act not in kl ortholog index\n",
      "EBG00000982146 from pkainh_act not in kl ortholog index\n",
      "EBG00000982313 from pkainh_act not in kl ortholog index\n",
      "KLLA0B10494r from pkainh_act not in kl ortholog index\n",
      "KLLA0E11969r from pkainh_act not in kl ortholog index\n",
      "EBG00000981954 from pkainh_act not in kl ortholog index\n",
      "EBG00000982195 from pkainh_act not in kl ortholog index\n",
      "EBG00000981603 from pkainh_act not in kl ortholog index\n",
      "KLLA0A06072r from pkainh_act not in kl ortholog index\n",
      "EBG00000982290 from pkainh_act not in kl ortholog index\n",
      "EBG00000981840 from pkainh_act not in kl ortholog index\n",
      "EBG00000982055 from pkainh_act not in kl ortholog index\n",
      "EBG00000982174 from pkainh_act not in kl ortholog index\n",
      "KLLA0D15477r from pkainh_act not in kl ortholog index\n",
      "EBG00000981753 from pkainh_act not in kl ortholog index\n",
      "KLLA0C17886r from pkainh_act not in kl ortholog index\n",
      "EBG00000982260 from pkainh_act not in kl ortholog index\n",
      "EBG00000982237 from pkainh_act not in kl ortholog index\n",
      "KLLA0D02662r from pkainh_act not in kl ortholog index\n",
      "EBG00000982282 from pkainh_act not in kl ortholog index\n",
      "KLLA0C17908r from pkainh_act not in kl ortholog index\n",
      "EBG00000981933 from pkainh_act not in kl ortholog index\n",
      "EBG00000982177 from pkainh_act not in kl ortholog index\n",
      "EBG00000982298 from pkainh_act not in kl ortholog index\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#target_sets['KL']['pkainh_act']\n",
    "\n",
    "\n",
    "\n",
    "#find its orthologs, add to a list\n",
    "\n",
    "#make the list a set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "393"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(kl_act)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'YAL027W',\n",
       " 'YAL054C',\n",
       " 'YAR035W',\n",
       " 'YBL016W',\n",
       " 'YBL049W',\n",
       " 'YBL095W',\n",
       " 'YBR026C',\n",
       " 'YBR047W',\n",
       " 'YBR057C',\n",
       " 'YBR108W',\n",
       " 'YBR203W',\n",
       " 'YBR240C',\n",
       " 'YBR273C',\n",
       " 'YCL038C',\n",
       " 'YCL048W',\n",
       " 'YCL056C',\n",
       " 'YCR010C',\n",
       " 'YCR036W',\n",
       " 'YCR068W',\n",
       " 'YCR076C',\n",
       " 'YDL078C',\n",
       " 'YDL113C',\n",
       " 'YDL115C',\n",
       " 'YDL139C',\n",
       " 'YDL146W',\n",
       " 'YDL186W',\n",
       " 'YDL234C',\n",
       " 'YDR019C',\n",
       " 'YDR020C',\n",
       " 'YDR032C',\n",
       " 'YDR051C',\n",
       " 'YDR068W',\n",
       " 'YDR078C',\n",
       " 'YDR090C',\n",
       " 'YDR104C',\n",
       " 'YDR105C',\n",
       " 'YDR106W',\n",
       " 'YDR109C',\n",
       " 'YDR144C',\n",
       " 'YDR179W-A',\n",
       " 'YDR207C',\n",
       " 'YDR222W',\n",
       " 'YDR227W',\n",
       " 'YDR248C',\n",
       " 'YDR294C',\n",
       " 'YDR326C',\n",
       " 'YDR329C',\n",
       " 'YDR384C',\n",
       " 'YDR415C',\n",
       " 'YDR425W',\n",
       " 'YDR446W',\n",
       " 'YDR480W',\n",
       " 'YDR503C',\n",
       " 'YDR522C',\n",
       " 'YDR525W-A',\n",
       " 'YEL047C',\n",
       " 'YER010C',\n",
       " 'YER015W',\n",
       " 'YER026C',\n",
       " 'YER046W',\n",
       " 'YER140W',\n",
       " 'YFL027C',\n",
       " 'YFL029C',\n",
       " 'YFL030W',\n",
       " 'YFR016C',\n",
       " 'YGL036W',\n",
       " 'YGL184C',\n",
       " 'YGL224C',\n",
       " 'YGL251C',\n",
       " 'YGR015C',\n",
       " 'YGR037C',\n",
       " 'YGR044C',\n",
       " 'YGR049W',\n",
       " 'YGR068C',\n",
       " 'YGR097W',\n",
       " 'YGR126W',\n",
       " 'YGR184C',\n",
       " 'YGR232W',\n",
       " 'YGR250C',\n",
       " 'YHL022C',\n",
       " 'YHL028W',\n",
       " 'YHR004C',\n",
       " 'YHR005C',\n",
       " 'YHR044C',\n",
       " 'YHR067W',\n",
       " 'YHR131C',\n",
       " 'YHR150W',\n",
       " 'YHR153C',\n",
       " 'YHR163W',\n",
       " 'YHR195W',\n",
       " 'YIL002C',\n",
       " 'YIL024C',\n",
       " 'YIL037C',\n",
       " 'YIL073C',\n",
       " 'YIL097W',\n",
       " 'YIL146C',\n",
       " 'YIL152W',\n",
       " 'YIR015W',\n",
       " 'YIR029W',\n",
       " 'YIR035C',\n",
       " 'YJL048C',\n",
       " 'YJL059W',\n",
       " 'YJL094C',\n",
       " 'YJL106W',\n",
       " 'YJL108C',\n",
       " 'YJL126W',\n",
       " 'YJL141C',\n",
       " 'YJL146W',\n",
       " 'YJL153C',\n",
       " 'YJR001W',\n",
       " 'YJR051W',\n",
       " 'YJR061W',\n",
       " 'YJR090C',\n",
       " 'YJR099W',\n",
       " 'YJR107W',\n",
       " 'YJR111C',\n",
       " 'YJR126C',\n",
       " 'YKL010C',\n",
       " 'YKL069W',\n",
       " 'YKL124W',\n",
       " 'YKL127W',\n",
       " 'YKL128C',\n",
       " 'YKL145W',\n",
       " 'YKL188C',\n",
       " 'YKL196C',\n",
       " 'YKL209C',\n",
       " 'YKL213C',\n",
       " 'YKR089C',\n",
       " 'YLL019C',\n",
       " 'YLL042C',\n",
       " 'YLL049W',\n",
       " 'YLR034C',\n",
       " 'YLR120C',\n",
       " 'YLR165C',\n",
       " 'YLR189C',\n",
       " 'YLR191W',\n",
       " 'YLR211C',\n",
       " 'YLR219W',\n",
       " 'YLR225C',\n",
       " 'YLR251W',\n",
       " 'YLR257W',\n",
       " 'YLR273C',\n",
       " 'YLR300W',\n",
       " 'YLR307C-A',\n",
       " 'YLR361C-A',\n",
       " 'YLR368W',\n",
       " 'YLR377C',\n",
       " 'YLR431C',\n",
       " 'YML020W',\n",
       " 'YML072C',\n",
       " 'YML076C',\n",
       " 'YMR003W',\n",
       " 'YMR004W',\n",
       " 'YMR008C',\n",
       " 'YMR052W',\n",
       " 'YMR063W',\n",
       " 'YMR083W',\n",
       " 'YMR100W',\n",
       " 'YMR119W',\n",
       " 'YMR135C',\n",
       " 'YMR140W',\n",
       " 'YMR148W',\n",
       " 'YMR160W',\n",
       " 'YMR172W',\n",
       " 'YMR181C',\n",
       " 'YMR263W',\n",
       " 'YMR274C',\n",
       " 'YMR313C',\n",
       " 'YNL008C',\n",
       " 'YNL010W',\n",
       " 'YNL042W',\n",
       " 'YNL086W',\n",
       " 'YNL117W',\n",
       " 'YNL144C',\n",
       " 'YNL165W',\n",
       " 'YNL249C',\n",
       " 'YNL275W',\n",
       " 'YNL287W',\n",
       " 'YNR049C',\n",
       " 'YOL011W',\n",
       " 'YOL016C',\n",
       " 'YOL024W',\n",
       " 'YOL054W',\n",
       " 'YOL056W',\n",
       " 'YOL065C',\n",
       " 'YOR003W',\n",
       " 'YOR028C',\n",
       " 'YOR059C',\n",
       " 'YOR081C',\n",
       " 'YOR084W',\n",
       " 'YOR132W',\n",
       " 'YOR155C',\n",
       " 'YOR190W',\n",
       " 'YOR212W',\n",
       " 'YOR245C',\n",
       " 'YOR251C',\n",
       " 'YOR356W',\n",
       " 'YOR373W',\n",
       " 'YOR386W',\n",
       " 'YPL048W',\n",
       " 'YPL049C',\n",
       " 'YPL071C',\n",
       " 'YPL096W',\n",
       " 'YPL116W',\n",
       " 'YPL121C',\n",
       " 'YPL147W',\n",
       " 'YPL192C',\n",
       " 'YPL201C',\n",
       " 'YPL206C',\n",
       " 'YPL229W',\n",
       " 'YPR002W',\n",
       " 'YPR028W',\n",
       " 'YPR049C',\n",
       " 'YPR061C',\n",
       " 'YPR066W',\n",
       " 'YPR091C',\n",
       " 'YPR095C',\n",
       " 'YPR115W',\n",
       " 'YPR122W',\n",
       " 'YPR151C',\n",
       " 'YPR157W'}"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_sets['kl_only_act']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KLLA0D08140r from pkainh_act not in kl ortholog index\n",
      "EBG00000981852 from pkainh_act not in kl ortholog index\n",
      "EBG00000982140 from pkainh_act not in kl ortholog index\n",
      "KLLA0F15059r from pkainh_act not in kl ortholog index\n",
      "EBG00000981757 from pkainh_act not in kl ortholog index\n",
      "EBG00000981576 from pkainh_act not in kl ortholog index\n",
      "EBG00000982146 from pkainh_act not in kl ortholog index\n",
      "EBG00000982313 from pkainh_act not in kl ortholog index\n",
      "KLLA0B10494r from pkainh_act not in kl ortholog index\n",
      "KLLA0E11969r from pkainh_act not in kl ortholog index\n",
      "EBG00000981954 from pkainh_act not in kl ortholog index\n",
      "EBG00000982195 from pkainh_act not in kl ortholog index\n",
      "EBG00000981603 from pkainh_act not in kl ortholog index\n",
      "KLLA0A06072r from pkainh_act not in kl ortholog index\n",
      "EBG00000982290 from pkainh_act not in kl ortholog index\n",
      "EBG00000981840 from pkainh_act not in kl ortholog index\n",
      "EBG00000982055 from pkainh_act not in kl ortholog index\n",
      "EBG00000982174 from pkainh_act not in kl ortholog index\n",
      "KLLA0D15477r from pkainh_act not in kl ortholog index\n",
      "EBG00000981753 from pkainh_act not in kl ortholog index\n",
      "KLLA0C17886r from pkainh_act not in kl ortholog index\n",
      "EBG00000982260 from pkainh_act not in kl ortholog index\n",
      "EBG00000982237 from pkainh_act not in kl ortholog index\n",
      "KLLA0D02662r from pkainh_act not in kl ortholog index\n",
      "EBG00000982282 from pkainh_act not in kl ortholog index\n",
      "KLLA0C17908r from pkainh_act not in kl ortholog index\n",
      "EBG00000981933 from pkainh_act not in kl ortholog index\n",
      "EBG00000982177 from pkainh_act not in kl ortholog index\n",
      "EBG00000982298 from pkainh_act not in kl ortholog index\n",
      "KLLA0F26620r from pkainh_rep not in kl ortholog index\n",
      "KLLA0D09823r from pkainh_rep not in kl ortholog index\n",
      "KLLA0F16687r from pkainh_rep not in kl ortholog index\n",
      "KLLA0B08349r from pkainh_rep not in kl ortholog index\n",
      "KLLA0D01364r from pkainh_rep not in kl ortholog index\n",
      "KLLA0F11385g from pkainh_rep not in kl ortholog index\n",
      "KLLA0F19877r from pkainh_rep not in kl ortholog index\n",
      "KLLA0A02042r from pkainh_rep not in kl ortholog index\n",
      "KLLA0E10759r from pkainh_rep not in kl ortholog index\n",
      "KLLA0A03784r from pkainh_rep not in kl ortholog index\n",
      "KLLA0D11214r from pkainh_rep not in kl ortholog index\n",
      "EBG00000982257 from pkainh_rep not in kl ortholog index\n",
      "KLLA0B07425r from pkainh_rep not in kl ortholog index\n",
      "KLLA0B06206r from pkainh_rep not in kl ortholog index\n",
      "EBG00000981979 from pkainh_rep not in kl ortholog index\n",
      "KLLA0E20329r from pkainh_rep not in kl ortholog index\n",
      "KLLA0E20483r from pkainh_rep not in kl ortholog index\n",
      "KLLA0F08657g from pkainh_rep not in kl ortholog index\n",
      "KLLA0C12155r from pkainh_rep not in kl ortholog index\n",
      "The following missing genes are filtered out of the background set:\n",
      "{'Scer_RDN5-1', 'Scer_RDN58-1', 'Scer_YGOB_ADL119W', 'Scer_RDN18-1', 'Scer_YGOB_YDR134C', 'Scer_YGOB_SDC25', 'Scer_RDN25-1', 'Scer_YGOB_Anc_7.495'}\n",
      "not_yet_annotated term removed: no background genes\n",
      "The following missing genes are filtered out of the background set:\n",
      "{'Scer_RDN5-1', 'Scer_RDN58-1', 'Scer_YGOB_ADL119W', 'Scer_RDN18-1', 'Scer_YGOB_YDR134C', 'Scer_YGOB_SDC25', 'Scer_RDN25-1', 'Scer_YGOB_Anc_7.495'}\n",
      "not_yet_annotated term removed: no background genes\n",
      "The following missing genes are filtered out of the background set:\n",
      "{'Scer_RDN5-1', 'Scer_RDN58-1', 'Scer_YGOB_ADL119W', 'Scer_RDN18-1', 'Scer_YGOB_YDR134C', 'Scer_YGOB_SDC25', 'Scer_RDN25-1', 'Scer_YGOB_Anc_7.495'}\n",
      "not_yet_annotated term removed: no background genes\n",
      "The following missing genes are filtered out of the background set:\n",
      "{'Scer_RDN5-1', 'Scer_RDN58-1', 'Scer_YGOB_ADL119W', 'Scer_RDN18-1', 'Scer_YGOB_YDR134C', 'Scer_YGOB_SDC25', 'Scer_RDN25-1', 'Scer_YGOB_Anc_7.495'}\n",
      "not_yet_annotated term removed: no background genes\n"
     ]
    }
   ],
   "source": [
    "#Make different gene sets to look up orthologs (using SC_genenames as a key).   For SC will have all the genes, for KL will only have\n",
    "#genes with SC orthologs. \n",
    "\n",
    "#maybe making the subsets makes more sense in the KLSC analysis file\n",
    "\n",
    "gene_set_names = ['kl_only_act', 'kl_only_rep','sc_only_act','sc_only_rep','klsc_act','klsc_rep']\n",
    "\n",
    "#For kl_act we include the names of all SC orthologs of genes activated and repressed.  \n",
    "#If there is a kl_act gene that has two orthologs in S.cer, we include both of them.  \n",
    "\n",
    "\n",
    "kl_sets = {}\n",
    "for setname in ['pkainh_act','pkainh_rep']:\n",
    "    kl_set = []\n",
    "    for kl_gene in list(target_sets['KL'][setname]): \n",
    "        if not(kl_gene in kl_orthologs.keys()):\n",
    "            print(kl_gene + ' from ' + setname + ' not in kl ortholog index')\n",
    "        else: \n",
    "            sc_orth = kl_orthologs[kl_gene]\n",
    "            if sc_orth != ['NONE']:\n",
    "                kl_set = kl_set + sc_orth\n",
    "    kl_sets[setname] = set(kl_set)\n",
    "    \n",
    "kl_act = kl_sets['pkainh_act']\n",
    "kl_rep = kl_sets['pkainh_rep']\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#kl_act = set(kl_orthologs.loc[kl_orthologs['kl_genename'].isin(target_sets['KL']['pkainh_act'],'sc_genename'])\n",
    "#kl_rep = set(kl_orthologs.loc[kl_orthologs['kl_genename'].isin(target_sets['KL']['pkainh_rep']),'sc_genename'])\n",
    "\n",
    "                              \n",
    "sc_act = target_sets['SC']['pkainh_act']\n",
    "sc_rep = target_sets['SC']['pkainh_rep']\n",
    "\n",
    "gene_sets = { 'kl_only_act' : kl_act - sc_act, \n",
    "              'kl_only_rep' : kl_rep - sc_rep, \n",
    "              'sc_only_act' : sc_act - kl_act, \n",
    "              'sc_only_rep' : sc_rep - kl_rep, \n",
    "              'klsc_act' : kl_act & sc_act, \n",
    "              'klsc_rep' : kl_rep & sc_rep}\n",
    "\n",
    "background_map = {'kl_only_act' : 'KL',\n",
    "                 'kl_only_rep' : 'KL', \n",
    "                 'sc_only_act' : 'SC',\n",
    "                 'sc_only_rep': 'SC',\n",
    "                 'klsc_act': 'KL',\n",
    "                 'klsc_rep': 'KL'}\n",
    "\n",
    "\n",
    "all_kl_orths = []\n",
    "for kl_orth in kl_orthologs.values():\n",
    "    all_kl_orths = all_kl_orths + kl_orth\n",
    "all_kl_orths_set = set(all_kl_orths)-{'NONE'}\n",
    "\n",
    "background_genes = {'SC' : set(pkainh_deseq_SC.index), \n",
    "                    'KL' : all_kl_orths_set\n",
    "                   }\n",
    "\n",
    "\n",
    "#fname = data_processing_dir + '\\\\kl_sc_PKA_as_m24_r1g1_20181017\\\\20181204_klsc_pkainh_subsets_for_enrichment.pkl'\n",
    "\n",
    "# with open(fname, 'wb') as file:\n",
    "#     pickle.dump([gene_sets, background_genes, background_map], file)\n",
    "\n",
    "#Loop through + print data\n",
    "\n",
    "go_term_enrichment = {}\n",
    "pval = 0.1\n",
    "FDR = 0.05\n",
    "\n",
    "for gene_set_name, gene_set in gene_sets.items(): \n",
    "    go_term_enrichment_raw = yeast_esr_exp.go_term_enrichment(gene_set, background_genes[background_map[gene_set_name]], go_term_list, go_slims_aspect)\n",
    "    (result,padj,_,_) = sm.stats.multipletests(go_term_enrichment_raw['pvalue'],alpha=FDR, method = 'fdr_bh')\n",
    "    go_term_enrichment_raw['detect_FDR']=result\n",
    "    go_term_enrichment_raw['padj'] = padj\n",
    "    go_term_enrichment_filtered = go_term_enrichment_raw[go_term_enrichment_raw['pvalue']<pval].sort_values('pvalue')\n",
    "    go_term_enrichment[gene_set_name] = go_term_enrichment_filtered\n",
    "\n",
    "fname = data_processing_dir + \"go_terms/go_enrichment_20181204.xls\"\n",
    "writer = pd.ExcelWriter(fname)\n",
    "for gene_set_name, go_term_enrichment_set in go_term_enrichment.items(): \n",
    "    go_term_enrichment_set.to_excel(writer, sheet_name = gene_set_name)\n",
    "writer.save()\n",
    "\n",
    "#This sets a longer column width if you want to try to look at genes here rather than in the file. \n",
    "#pd.set_option('display.max_colwidth', -1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "?pd.ExcelWriter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "fname = data_processing_dir + \"go_terms/go_enrichment_20200609.xls\"\n",
    "writer = pd.ExcelWriter(fname)\n",
    "for gene_set_name, go_term_enrichment_set in go_term_enrichment.items(): \n",
    "    go_term_enrichment_set.to_excel(writer, sheet_name = gene_set_name)\n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Table [go_enrichment]: Go term enrichment was calculated from the Go Slim Dataset [ref?] downloaded from SGD on 20180412 and p-values were calculated using fisher's exact test against a background of either all genes in S.Cerevisiae (sc_only groups) or all genes in K.Lactis that contain an SC ortholog (kl_only groups and klsc groups)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N subset genes in goterm</th>\n",
       "      <th>N genes in subset</th>\n",
       "      <th>pct goterm in subset</th>\n",
       "      <th>pct go term in background</th>\n",
       "      <th>pct of go terms genes in subset</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>genes</th>\n",
       "      <th>genes common name</th>\n",
       "      <th>oddsratio</th>\n",
       "      <th>detect_FDR</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>mitochondrial translation</th>\n",
       "      <td>75</td>\n",
       "      <td>320</td>\n",
       "      <td>0.234375</td>\n",
       "      <td>0.024131</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>1.607014e-43</td>\n",
       "      <td>[YDR347W, YNL252C, YEL050C, YJR113C, YHL004W, ...</td>\n",
       "      <td>[MRP1, MRPL17, RML2, RSM7, MRP4, MRPL20, MNP1,...</td>\n",
       "      <td>12.379592</td>\n",
       "      <td>True</td>\n",
       "      <td>1.623084e-41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mitochondrion organization</th>\n",
       "      <td>99</td>\n",
       "      <td>320</td>\n",
       "      <td>0.309375</td>\n",
       "      <td>0.072780</td>\n",
       "      <td>0.262599</td>\n",
       "      <td>1.107155e-32</td>\n",
       "      <td>[YDR347W, YNL066W, YDL045W-A, YPR166C, YCR046C...</td>\n",
       "      <td>[MRP1, SUN4, MRP10, MRP2, IMG1, CBP6, RSM19, R...</td>\n",
       "      <td>5.707083</td>\n",
       "      <td>True</td>\n",
       "      <td>5.591133e-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA modification</th>\n",
       "      <td>15</td>\n",
       "      <td>320</td>\n",
       "      <td>0.046875</td>\n",
       "      <td>0.018919</td>\n",
       "      <td>0.153061</td>\n",
       "      <td>2.116911e-03</td>\n",
       "      <td>[YHR003C, YDL104C, YKR038C, YFL001W, YGL211W, ...</td>\n",
       "      <td>[TCD1, QRI7, KAE1, DEG1, NCS6, TRM5, PAP2, MTO...</td>\n",
       "      <td>2.550351</td>\n",
       "      <td>False</td>\n",
       "      <td>7.126934e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellular amino acid metabolic process</th>\n",
       "      <td>21</td>\n",
       "      <td>320</td>\n",
       "      <td>0.065625</td>\n",
       "      <td>0.034749</td>\n",
       "      <td>0.116667</td>\n",
       "      <td>6.129912e-03</td>\n",
       "      <td>[YFL022C, YGL202W, YDR234W, YNR050C, YGR124W, ...</td>\n",
       "      <td>[FRS2, ARO8, LYS4, LYS9, ASN2, LEU3, BAT2, MET...</td>\n",
       "      <td>1.950948</td>\n",
       "      <td>False</td>\n",
       "      <td>1.415484e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tRNA processing</th>\n",
       "      <td>14</td>\n",
       "      <td>320</td>\n",
       "      <td>0.043750</td>\n",
       "      <td>0.019691</td>\n",
       "      <td>0.137255</td>\n",
       "      <td>7.007346e-03</td>\n",
       "      <td>[YHR003C, YDL104C, YKR038C, YFL001W, YGL211W, ...</td>\n",
       "      <td>[TCD1, QRI7, KAE1, DEG1, NCS6, TRM5, PAP2, MTO...</td>\n",
       "      <td>2.277714</td>\n",
       "      <td>False</td>\n",
       "      <td>1.415484e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regulation of translation</th>\n",
       "      <td>13</td>\n",
       "      <td>320</td>\n",
       "      <td>0.040625</td>\n",
       "      <td>0.019498</td>\n",
       "      <td>0.128713</td>\n",
       "      <td>1.449099e-02</td>\n",
       "      <td>[YPL215W, YGL195W, YER165W, YFR009W, YBR120C, ...</td>\n",
       "      <td>[CBP3, GCN1, PAB1, GCN20, CBP6, MBA1, PET309, ...</td>\n",
       "      <td>2.129422</td>\n",
       "      <td>False</td>\n",
       "      <td>2.439316e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>carbohydrate transport</th>\n",
       "      <td>4</td>\n",
       "      <td>320</td>\n",
       "      <td>0.012500</td>\n",
       "      <td>0.003089</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>2.596991e-02</td>\n",
       "      <td>[YDR345C, YPL026C, YDR536W, YHR096C]</td>\n",
       "      <td>[HXT3, SKS1, STL1, HXT5]</td>\n",
       "      <td>4.085443</td>\n",
       "      <td>False</td>\n",
       "      <td>3.295050e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>translational initiation</th>\n",
       "      <td>8</td>\n",
       "      <td>320</td>\n",
       "      <td>0.025000</td>\n",
       "      <td>0.010425</td>\n",
       "      <td>0.148148</td>\n",
       "      <td>2.609940e-02</td>\n",
       "      <td>[YPL119C, YER165W, YLR067C, YGL049C, YOR204W, ...</td>\n",
       "      <td>[DBP1, PAB1, PET309, TIF4632, DED1, RMD9, YGR0...</td>\n",
       "      <td>2.433998</td>\n",
       "      <td>False</td>\n",
       "      <td>3.295050e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>translational elongation</th>\n",
       "      <td>6</td>\n",
       "      <td>320</td>\n",
       "      <td>0.018750</td>\n",
       "      <td>0.007722</td>\n",
       "      <td>0.150000</td>\n",
       "      <td>4.864817e-02</td>\n",
       "      <td>[YLR069C, YNL014W, YGL195W, YFR009W, YOR133W, ...</td>\n",
       "      <td>[MEF1, HEF3, GCN1, GCN20, EFT1, TUF1]</td>\n",
       "      <td>2.455414</td>\n",
       "      <td>False</td>\n",
       "      <td>5.459406e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>amino acid transport</th>\n",
       "      <td>5</td>\n",
       "      <td>320</td>\n",
       "      <td>0.015625</td>\n",
       "      <td>0.006371</td>\n",
       "      <td>0.151515</td>\n",
       "      <td>6.712001e-02</td>\n",
       "      <td>[YGR191W, YGR055W, YOL020W, YOR348C, YNL268W]</td>\n",
       "      <td>[HIP1, MUP1, TAT2, PUT4, LYP1]</td>\n",
       "      <td>2.475709</td>\n",
       "      <td>False</td>\n",
       "      <td>6.779121e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transmembrane transport</th>\n",
       "      <td>17</td>\n",
       "      <td>320</td>\n",
       "      <td>0.053125</td>\n",
       "      <td>0.035907</td>\n",
       "      <td>0.091398</td>\n",
       "      <td>8.103843e-02</td>\n",
       "      <td>[YCR037C, YJR045C, YBR039W, YOR232W, YGR260W, ...</td>\n",
       "      <td>[PHO87, SSC1, ATP3, MGE1, TNA1, PEX21, PAM18, ...</td>\n",
       "      <td>1.506405</td>\n",
       "      <td>False</td>\n",
       "      <td>7.440801e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       N subset genes in goterm  \\\n",
       "mitochondrial translation                                    75   \n",
       "mitochondrion organization                                   99   \n",
       "RNA modification                                             15   \n",
       "cellular amino acid metabolic process                        21   \n",
       "tRNA processing                                              14   \n",
       "regulation of translation                                    13   \n",
       "carbohydrate transport                                        4   \n",
       "translational initiation                                      8   \n",
       "translational elongation                                      6   \n",
       "amino acid transport                                          5   \n",
       "transmembrane transport                                      17   \n",
       "\n",
       "                                       N genes in subset  \\\n",
       "mitochondrial translation                            320   \n",
       "mitochondrion organization                           320   \n",
       "RNA modification                                     320   \n",
       "cellular amino acid metabolic process                320   \n",
       "tRNA processing                                      320   \n",
       "regulation of translation                            320   \n",
       "carbohydrate transport                               320   \n",
       "translational initiation                             320   \n",
       "translational elongation                             320   \n",
       "amino acid transport                                 320   \n",
       "transmembrane transport                              320   \n",
       "\n",
       "                                       pct goterm in subset  \\\n",
       "mitochondrial translation                          0.234375   \n",
       "mitochondrion organization                         0.309375   \n",
       "RNA modification                                   0.046875   \n",
       "cellular amino acid metabolic process              0.065625   \n",
       "tRNA processing                                    0.043750   \n",
       "regulation of translation                          0.040625   \n",
       "carbohydrate transport                             0.012500   \n",
       "translational initiation                           0.025000   \n",
       "translational elongation                           0.018750   \n",
       "amino acid transport                               0.015625   \n",
       "transmembrane transport                            0.053125   \n",
       "\n",
       "                                       pct go term in background  \\\n",
       "mitochondrial translation                               0.024131   \n",
       "mitochondrion organization                              0.072780   \n",
       "RNA modification                                        0.018919   \n",
       "cellular amino acid metabolic process                   0.034749   \n",
       "tRNA processing                                         0.019691   \n",
       "regulation of translation                               0.019498   \n",
       "carbohydrate transport                                  0.003089   \n",
       "translational initiation                                0.010425   \n",
       "translational elongation                                0.007722   \n",
       "amino acid transport                                    0.006371   \n",
       "transmembrane transport                                 0.035907   \n",
       "\n",
       "                                       pct of go terms genes in subset  \\\n",
       "mitochondrial translation                                     0.600000   \n",
       "mitochondrion organization                                    0.262599   \n",
       "RNA modification                                              0.153061   \n",
       "cellular amino acid metabolic process                         0.116667   \n",
       "tRNA processing                                               0.137255   \n",
       "regulation of translation                                     0.128713   \n",
       "carbohydrate transport                                        0.250000   \n",
       "translational initiation                                      0.148148   \n",
       "translational elongation                                      0.150000   \n",
       "amino acid transport                                          0.151515   \n",
       "transmembrane transport                                       0.091398   \n",
       "\n",
       "                                             pvalue  \\\n",
       "mitochondrial translation              1.607014e-43   \n",
       "mitochondrion organization             1.107155e-32   \n",
       "RNA modification                       2.116911e-03   \n",
       "cellular amino acid metabolic process  6.129912e-03   \n",
       "tRNA processing                        7.007346e-03   \n",
       "regulation of translation              1.449099e-02   \n",
       "carbohydrate transport                 2.596991e-02   \n",
       "translational initiation               2.609940e-02   \n",
       "translational elongation               4.864817e-02   \n",
       "amino acid transport                   6.712001e-02   \n",
       "transmembrane transport                8.103843e-02   \n",
       "\n",
       "                                                                                   genes  \\\n",
       "mitochondrial translation              [YDR347W, YNL252C, YEL050C, YJR113C, YHL004W, ...   \n",
       "mitochondrion organization             [YDR347W, YNL066W, YDL045W-A, YPR166C, YCR046C...   \n",
       "RNA modification                       [YHR003C, YDL104C, YKR038C, YFL001W, YGL211W, ...   \n",
       "cellular amino acid metabolic process  [YFL022C, YGL202W, YDR234W, YNR050C, YGR124W, ...   \n",
       "tRNA processing                        [YHR003C, YDL104C, YKR038C, YFL001W, YGL211W, ...   \n",
       "regulation of translation              [YPL215W, YGL195W, YER165W, YFR009W, YBR120C, ...   \n",
       "carbohydrate transport                              [YDR345C, YPL026C, YDR536W, YHR096C]   \n",
       "translational initiation               [YPL119C, YER165W, YLR067C, YGL049C, YOR204W, ...   \n",
       "translational elongation               [YLR069C, YNL014W, YGL195W, YFR009W, YOR133W, ...   \n",
       "amino acid transport                       [YGR191W, YGR055W, YOL020W, YOR348C, YNL268W]   \n",
       "transmembrane transport                [YCR037C, YJR045C, YBR039W, YOR232W, YGR260W, ...   \n",
       "\n",
       "                                                                       genes common name  \\\n",
       "mitochondrial translation              [MRP1, MRPL17, RML2, RSM7, MRP4, MRPL20, MNP1,...   \n",
       "mitochondrion organization             [MRP1, SUN4, MRP10, MRP2, IMG1, CBP6, RSM19, R...   \n",
       "RNA modification                       [TCD1, QRI7, KAE1, DEG1, NCS6, TRM5, PAP2, MTO...   \n",
       "cellular amino acid metabolic process  [FRS2, ARO8, LYS4, LYS9, ASN2, LEU3, BAT2, MET...   \n",
       "tRNA processing                        [TCD1, QRI7, KAE1, DEG1, NCS6, TRM5, PAP2, MTO...   \n",
       "regulation of translation              [CBP3, GCN1, PAB1, GCN20, CBP6, MBA1, PET309, ...   \n",
       "carbohydrate transport                                          [HXT3, SKS1, STL1, HXT5]   \n",
       "translational initiation               [DBP1, PAB1, PET309, TIF4632, DED1, RMD9, YGR0...   \n",
       "translational elongation                           [MEF1, HEF3, GCN1, GCN20, EFT1, TUF1]   \n",
       "amino acid transport                                      [HIP1, MUP1, TAT2, PUT4, LYP1]   \n",
       "transmembrane transport                [PHO87, SSC1, ATP3, MGE1, TNA1, PEX21, PAM18, ...   \n",
       "\n",
       "                                       oddsratio  detect_FDR          padj  \n",
       "mitochondrial translation              12.379592        True  1.623084e-41  \n",
       "mitochondrion organization              5.707083        True  5.591133e-31  \n",
       "RNA modification                        2.550351       False  7.126934e-02  \n",
       "cellular amino acid metabolic process   1.950948       False  1.415484e-01  \n",
       "tRNA processing                         2.277714       False  1.415484e-01  \n",
       "regulation of translation               2.129422       False  2.439316e-01  \n",
       "carbohydrate transport                  4.085443       False  3.295050e-01  \n",
       "translational initiation                2.433998       False  3.295050e-01  \n",
       "translational elongation                2.455414       False  5.459406e-01  \n",
       "amino acid transport                    2.475709       False  6.779121e-01  \n",
       "transmembrane transport                 1.506405       False  7.440801e-01  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_term_enrichment['kl_only_rep']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "#various gene sets\n",
    "#genes = go_term_df_cutoff.loc[\"lipid metabolic process\"]['genes common name']\n",
    "#genes = ['XBP1','RME1','COX1','COX4','COX13','COX6','GSM1','COX12','CYT1']\n",
    "#genes =  ['BAS1', 'MET32', 'GZF3','GLN3', 'MET28', 'MET4', 'GAT1'] #Purine Metabolism\n",
    "#genes = ['ADE1','ADE2','ADE4','ADE5','ADE6','ADE7','ADE8','ADE12','ADE13','ADE16']  #Purine Genes\n",
    "#genes = ['ZWF1','HIS4','SFP1']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# terms_of_interest = [\"lipid metabolic process\",\"biological_process\",\"conjugation\",\"response to chemical\",\n",
    "#                      \"sporulation\",\"endosomal transport\",\"lipid transport\",\"response to oxidative stress\",\n",
    "#                      \"cellular respiration\",\"cofactor metabolic process\",\"carbohydrate transport\",\n",
    "#                      \"mitochondrial translation\", \"mitochondrion organization\",\"translational initiation\",\n",
    "#                      \"cellular amino acid metabolic process\", \"cell budding\",\n",
    "#                      \"nucleobase containing small molecule metabolic process\", \"snoRNA processing\", \"amino acid transport\"]\n",
    "\n",
    "# term_df_dict = {}\n",
    "# for term in terms_of_interest:\n",
    "#     term_genes = list(go_slims_aspect[go_slims_aspect['GO_term']==term]['sc_genename'])\n",
    "#     kl_sc_PKA_data_term = kl_sc_PKA_data[kl_sc_PKA_data['sc_genename'].isin(term_genes)]\n",
    "#     term_df_dict[term]=kl_sc_PKA_data_term\n",
    "\n",
    "# kl_sc_sig_unsig_dict = {'Unsig KL and SC' : kl_sc_PKA_data_unsig, \n",
    "#                  'Sig KL only': kl_sc_PKA_data_klsig_scunsig, \n",
    "#                  'Sig SC only': kl_sc_PKA_data_scsig_klunsig, \n",
    "#                  'Sig KL and SC': kl_sc_PKA_data_klscsig}\n",
    "\n",
    "# condition_set = {}\n",
    "# for df_dict in (term_df_dict,kl_sc_sig_unsig_dict): \n",
    "#     condition_set.update(df_dict)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #Plot a particular set of go terms on top of my scatter plot\n",
    "\n",
    "# #plot figure in plot.ly\n",
    "# # condition_set = {'Unsig KL and SC' : kl_sc_PKA_data_unsig, \n",
    "# #                  'Sig KL only': kl_sc_PKA_data_klsig_scunsig, \n",
    "# #                  'Sig SC only': kl_sc_PKA_data_scsig_klunsig, \n",
    "# #                  'Sig KL and SC': kl_sc_PKA_data_klscsig,\n",
    "# #                  term: kl_sc_PKA_data_term }\n",
    "\n",
    "# fig_data = [\n",
    "#   \t\t{\n",
    "#   \t\t\t'x': condition_set[condition]['PKA(AS)+1NMPP1_KL'], \n",
    "#         \t'y': condition_set[condition]['PKA(AS)+1NMPP1_SC'], \n",
    "#         \t'text': condition_set[condition]['SC_common_name'], \n",
    "#         \t'mode': 'markers', \n",
    "#         \t'name': condition} for condition in condition_set.keys()]\n",
    "# fig = {\n",
    "#     'data': fig_data,\n",
    "#     'layout': {\n",
    "#         'xaxis': {'title': 'KL LFC 4uM NMPP1'},\n",
    "#         'yaxis': {'title': 'SC LFC 4uM NMPP1'}\n",
    "#     }\n",
    "# }\n",
    "\n",
    "# # IPython notebook\n",
    "# # py.iplot(fig, filename='pandas/multiple-scatter')\n",
    "\n",
    "# py.iplot(fig, filename='sc_kl_scatter_sig_highlight_20170915_various_goterms')\n",
    "\n",
    "\n",
    "# #use inline plotly. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {
    "collapsed": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#Generate a list of all genes in KL_SC_PKA_data and add as a list in intermine named\n",
    "#20170905_kl_sc_PKA_data_genes \n",
    "#We can then make subsets programmatically using the API\n",
    "\n",
    "#I ran this command: \n",
    "#\" \".join(kl_sc_PKA_data['sc_genename'])\n",
    "\n",
    "#Then copied and pasted the output into the yeastmine build query list. \n",
    "\n",
    "#The following items were not found in SGD: \n",
    "#Scer_YGOB_YDR134C Scer_RDN18-1 Scer_YGOB_SDC25 Scer_RDN25-1 Scer_RDN58-1 Scer_YGOB_Anc_7.495 Scer_YGOB_ADL119W Scer_RDN5-1\n",
    "\n",
    "#I wonder if I could do this just using something like this: query.add_constraint(\"Gene\", \"LOOKUP\", [\"YNL323W\",\"YIL093C\",...])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mitophagy_list = ['YCR079W','YDL113C','YFR021W','YGR049W','YIL146C','YJL036W','YLL042C','YLR423C','YNL223W','YNR007C']\n",
    "# mito_disassembly_list = ['YCR079W','YDL113C','YFR021W','YGR049W','YIL146C','YJL036W','YLL042C','YLR423C','YNL223W','YNR007C']\n",
    "# lipid_catabolic_list = ['YAL051W','YCR068W','YDR036C','YDR058C','YGL010W','YKR009C','YKR089C','YMR008C','YMR313C','YOL011W','YOR059C','YOR081C','YOR084W','YOR363C','YPR001W','YPR002W','YPR006C']\n",
    "# #kl_sc_PKA_data[kl_sc_PKA_data['sc_genename'].isin(lipid_catabolic_list)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #Generate Query for all kl_sc_PKA_data genes\n",
    "\n",
    "# # Get a new query on the class (table) you will be querying:\n",
    "# query_all = service.new_query(\"Gene\")\n",
    "\n",
    "# # The view specifies the output columns\n",
    "# query_all.add_view(\n",
    "#     \"primaryIdentifier\", \"secondaryIdentifier\", \"organism.shortName\", \"symbol\",\n",
    "#     \"name\"\n",
    "# )\n",
    "\n",
    "# # You can edit the constraint values below\n",
    "# query_all.add_constraint(\"Gene\", \"IN\", \"20170509_kl_sc_PKA_data_genes\", code = \"A\")\n",
    "\n",
    "# # Uncomment and edit the code below to specify your own custom logic:\n",
    "# # query.set_logic(\"A\")\n",
    "\n",
    "# #gene_id_list = [str(row[\"id\"] for row in query.rows()]\n",
    "\n",
    "# #type(gene_id_list[0])\n",
    "\n",
    "# # query_list = list(query)\n",
    "\n",
    "# # for row in query_list:     #query.rows():\n",
    "# #     print(row[\"primaryIdentifier\"], row[\"secondaryIdentifier\"], row[\"organism.shortName\"], \\\n",
    "# #         row[\"symbol\"], row[\"name\"], row[\"id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Yeastmine queries for each geneset\n",
    "# #For each set of genes\n",
    "# # gene_sets = [('klsig','act'),\n",
    "# #             ('scsig','act'),\n",
    "# #             ('klscsig','act'),\n",
    "# #             ('klsig','rep'),\n",
    "# #             ('scsig','rep'),\n",
    "# #             ('klscsig','rep')]\n",
    "\n",
    "# #Add in the one gene that is significant for both but repressed in SC and activated in KL. \n",
    "# klsig_act_genes = list(kl_sc_PKA_data_klsig_scunsig[kl_sc_PKA_data_klsig_scunsig['PKA(AS)+1NMPP1_KL']>0]['sc_genename'])\n",
    "# klsig_act_genes.append('YJL208C')\n",
    "\n",
    "# #Add in the one gene that is significant for both but activated in SC and repressed in KL.\n",
    "# klsig_rep_genes = list(kl_sc_PKA_data_klsig_scunsig[kl_sc_PKA_data_klsig_scunsig['PKA(AS)+1NMPP1_KL']<0]['sc_genename'])\n",
    "# klsig_rep_genes.append('YNL014W')\n",
    "\n",
    "# #Add in the one gene that is significant for both but activated in SC and repressed in KL.\n",
    "# scsig_act_genes = list(kl_sc_PKA_data_scsig_klunsig[kl_sc_PKA_data_scsig_klunsig['PKA(AS)+1NMPP1_SC']>0]['sc_genename'])\n",
    "# scsig_act_genes.append('YNL014W')\n",
    "\n",
    "# #Add in the one gene that is significant for both but repressed in SC and activated in KL. \n",
    "# scsig_rep_genes = list(kl_sc_PKA_data_scsig_klunsig[kl_sc_PKA_data_scsig_klunsig['PKA(AS)+1NMPP1_SC']<0]['sc_genename'])\n",
    "# scsig_rep_genes.append('YJL208C')\n",
    "\n",
    "# gene_sets = ['klsig_act', 'klsig_rep','scsig_act','scsig_rep','klscsig_act','klscsig_rep']\n",
    "\n",
    "# gene_set_dict = {'klsig_act': klsig_act_genes,\n",
    "#                  'klsig_rep': klsig_rep_genes,\n",
    "#                  'scsig_act': scsig_act_genes,\n",
    "#                  'scsig_rep': scsig_rep_genes,\n",
    "#                  'klscsig_act': list(kl_sc_PKA_data_klscsig[(kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_SC']>0) & (kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_KL']>0)]['sc_genename']),\n",
    "#                  'klscsig_rep': list(kl_sc_PKA_data_klscsig[(kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_SC']<0) & (kl_sc_PKA_data_klscsig['PKA(AS)+1NMPP1_KL']<0)]['sc_genename'])\n",
    "#                 }\n",
    "\n",
    "# missing_genes = ['Scer_YGOB_YDR134C', 'Scer_RDN18-1', \n",
    "#                  'Scer_YGOB_SDC25', 'Scer_RDN25-1', \n",
    "#                  'Scer_RDN58-1', 'Scer_YGOB_Anc_7.495', \n",
    "#                  'Scer_YGOB_ADL119W', 'Scer_RDN5-1']\n",
    "\n",
    "# pval_cutoff = 0.01\n",
    "\n",
    "# for gene_set in gene_sets: \n",
    "#     print(gene_set)\n",
    "#     gene_set_list = gene_set_dict[gene_set]\n",
    "    \n",
    "#     #filter out missing genes \n",
    "#     print(\"The following missing genes are filtered out:\")\n",
    "#     print(set(gene_set_list) & set(missing_genes))\n",
    "#     gene_set_list_filtered = list(set(gene_set_list)-set(missing_genes))\n",
    "#     new_query = query_all.clone()\n",
    "    \n",
    "#     #subset query based on desired gene set\n",
    "#     new_query.add_constraint(\"Gene.secondaryIdentifier\", \"ONE OF\", gene_set_list_filtered)\n",
    "    \n",
    "#     #Generate a list\n",
    "#     new_list = service.create_list(new_query.select('Gene.id'), \"Gene\", name = \"20170905\" + gene_set)\n",
    "    \n",
    "#     #Run Go enrichment on the list\n",
    "#     results = new_list.calculate_enrichment('go_enrichment_for_gene')\n",
    "#     results_list = list(results)    \n",
    "#     for result in results_list: \n",
    "#         print(result)\n",
    "#         if result['p-value']<pval_cutoff:\n",
    "#             print(result)\n",
    "#             print(result)\n",
    "            \n",
    "#     #Given a P-value cutoff\n",
    "#     # Save the go term \n",
    "#     # Save the contingency table\n",
    "#     # Save the related genes\n",
    "\n",
    "# ##To see a list of widgets that are available\n",
    "# service.widgets \n",
    "\n",
    "#results = new_list.calculate_enrichment('go_enrichment_for_gene')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# #Get genes from a go term.  \n",
    "\n",
    "# from intermine.webservice import Service\n",
    "# service = Service(\"http://yeastmine.yeastgenome.org:443/yeastmine/service\")\n",
    "# query = service.new_query(\"Gene\")\n",
    "# query.add_view(\n",
    "#     \"primaryIdentifier\", \"secondaryIdentifier\", \"symbol\", \"featureType\",\n",
    "#     \"goAnnotation.qualifier\", \"goAnnotation.ontologyTerm.identifier\",\n",
    "#     \"goAnnotation.ontologyTerm.name\", \"goAnnotation.ontologyTerm.namespace\",\n",
    "#     \"goAnnotation.evidence.code.code\", \"goAnnotation.evidence.code.withText\",\n",
    "#     \"goAnnotation.evidence.code.annotType\",\n",
    "#     \"goAnnotation.evidence.publications.pubMedId\",\n",
    "#     \"goAnnotation.evidence.publications.citation\"\n",
    "# )\n",
    "# query.add_sort_order(\"Gene.secondaryIdentifier\", \"ASC\")\n",
    "# query.add_constraint(\"goAnnotation.ontologyTerm\", \"LOOKUP\", \"GO:0000182\", code = \"A\")\n",
    "\n",
    "# for row in query.rows():\n",
    "#     print row[\"primaryIdentifier\"], row[\"secondaryIdentifier\"], row[\"symbol\"], row[\"featureType\"], \\\n",
    "#         row[\"goAnnotation.qualifier\"], row[\"goAnnotation.ontologyTerm.identifier\"], \\\n",
    "#         row[\"goAnnotation.ontologyTerm.name\"], row[\"goAnnotation.ontologyTerm.namespace\"], \\\n",
    "#         row[\"goAnnotation.evidence.code.code\"], row[\"goAnnotation.evidence.code.withText\"], \\\n",
    "#         row[\"goAnnotation.evidence.code.annotType\"], \\\n",
    "#         row[\"goAnnotation.evidence.publications.pubMedId\"], \\\n",
    "#         row[\"goAnnotation.evidence.publications.citation\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# #See which genes correspond to JSO targets. \n",
    "# jso_msn2_targets = ['SSA1', 'SSA4', 'HSP104', 'CTT1', 'TPK1', 'TFS1', 'DDR2', 'TSA1', 'HSP12', 'TMA10', 'TPS2', 'RTC3', 'GPH1', 'HXK1', 'TSA2', 'CYC7', 'PNC1', 'GSY1', 'FMP48', 'SS13', 'ALD4', 'YNR014W', 'HOR2', 'YOR173W', 'RAS2', 'GLK1', 'GPD1', 'HSP26', 'SIP18', 'TKL2', 'FMP16', 'ALD3', 'PGM2']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene_set_data = kl_sc_PKA_data[kl_sc_PKA_data['SC_common_name'].isin(jso_msn2_targets)]\n",
    "# gene_set_list = gene_set_data['sc_genename']\n",
    "\n",
    "\n",
    "# go_term_data = []\n",
    "\n",
    "# go_term_index = []\n",
    "\n",
    "# for term in go_term_list: \n",
    "#     term_genes = list(go_slims_aspect[go_slims_aspect['GO_term']==term]['sc_genename'])\n",
    "#     if len(term_genes)> len(set(term_genes)):\n",
    "#         print(\"Duplicate Term: \" + term)\n",
    "#     subset_genes_in_goterm =  set(gene_set_list) & set(term_genes)\n",
    "#     N_subset_genes_in_goterm = len(subset_genes_in_goterm)\n",
    "#     N_genes_in_goterm = len(term_genes)\n",
    "#     if N_subset_genes_in_goterm >0:\n",
    "#         subset_genes_in_goterm_commonname = io_library.SC_common_name_lookup(subset_genes_in_goterm)\n",
    "#         go_term_data.append((N_subset_genes_in_goterm,\n",
    "#                              subset_genes_in_goterm,\n",
    "#                              subset_genes_in_goterm_commonname,\n",
    "#                              N_genes_in_goterm))\n",
    "#         go_term_index.append(term)\n",
    "\n",
    "# go_term_df = pd.DataFrame(go_term_data, index = go_term_index,columns = ['N subset genes in goterm',\n",
    "#                                                                             'genes',\n",
    "#                                                                             'genes common name',\n",
    "#                                                                             'N genes in goterm'])\n",
    "# go_term_df[['genes common name','N genes in goterm']]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
