{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "what computer are you on? a = Bens, b = gpucluster, c = other   a\n",
      "base directory is C:\\Users\\heine\\github\\expression_broad_data\n",
      "Added C:\\Users\\heine\\github\\expression_broad_data to path: \n",
      "['', 'C:\\\\Users\\\\heine\\\\Anaconda3\\\\python36.zip', 'C:\\\\Users\\\\heine\\\\Anaconda3\\\\DLLs', 'C:\\\\Users\\\\heine\\\\Anaconda3\\\\lib', 'C:\\\\Users\\\\heine\\\\Anaconda3', 'C:\\\\Users\\\\heine\\\\Anaconda3\\\\lib\\\\site-packages', 'C:\\\\Users\\\\heine\\\\Anaconda3\\\\lib\\\\site-packages\\\\Babel-2.5.0-py3.6.egg', 'C:\\\\Users\\\\heine\\\\Anaconda3\\\\lib\\\\site-packages\\\\win32', 'C:\\\\Users\\\\heine\\\\Anaconda3\\\\lib\\\\site-packages\\\\win32\\\\lib', 'C:\\\\Users\\\\heine\\\\Anaconda3\\\\lib\\\\site-packages\\\\Pythonwin', 'C:\\\\Users\\\\heine\\\\Anaconda3\\\\lib\\\\site-packages\\\\IPython\\\\extensions', 'C:\\\\Users\\\\heine\\\\.ipython', 'C:\\\\Users\\\\heine\\\\github\\\\expression_broad_data']\n",
      "I am about to import a library\n",
      "I am importing io_library\n",
      "what computer are you on? a = Bens, b = gpucluster, c = other   a\n",
      "base directory is C:\\Users\\heine\\github\\expression_broad_data\n",
      "data processing dir is C:\\Users\\heine\\github\\expression_broad_data\\expression_data\\\n",
      "are you online? Yes/NoYes\n"
     ]
    }
   ],
   "source": [
    "# %load std_libraries.py\n",
    "#to keep my scripts consistent, I am adding this to the top of all scripts with %load std_libraries.py\n",
    "import sys\n",
    "#Indicate operating environment and import core modules\n",
    "location_input = input(\"what computer are you on? a = Bens, b = gpucluster, c = other   \")\n",
    "location_dict = {'a': \"C:\\\\Users\\\\heine\\\\github\\\\expression_broad_data\", 'b': \"/home/heineike/github/expression_broad_data\",'c':'you need to add your location to the location_dict'}\n",
    "base_dir = location_dict[location_input]\n",
    "print(\"base directory is \" + base_dir)\n",
    "\n",
    "if sys.path[-1] != base_dir:\n",
    "    sys.path.append(base_dir)\n",
    "    print(\"Added \" + base_dir + \" to path: \" )\n",
    "    print(sys.path)\n",
    "\n",
    "import os\n",
    "\n",
    "print(\"I am about to import a library\")\n",
    "from core import expression_plots \n",
    "from core import io_library \n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "# %matplotlib \n",
    "# plt.ioff()\n",
    "#%matplotlib inline\n",
    "%matplotlib notebook\n",
    "data_processing_dir = base_dir + os.sep + os.path.normpath(\"expression_data\") + os.sep\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "#import seaborn as sns; sns.set(style=\"ticks\", color_codes=True)\n",
    "#from sklearn import linear_model\n",
    "#import pickle\n",
    "#import subprocess\n",
    "#import networkx as nx\n",
    "# import scipy.stats as stats\n",
    "import scipy.cluster.hierarchy as sch\n",
    "\n",
    "# from Bio import SeqIO\n",
    "# from Bio import SeqFeature as sf\n",
    "# from Bio.Alphabet import generic_dna\n",
    "# from Bio.Seq import Seq\n",
    "\n",
    "import re\n",
    "\n",
    "from collections import Counter\n",
    "# import scipy.stats as stats\n",
    "# from itertools import chain\n",
    "#this only works if you are online\n",
    "online_input = input(\"are you online? Yes/No\")\n",
    "if online_input == \"Yes\": \n",
    "    import plotly.plotly as py\n",
    "    import plotly.graph_objs as pygo\n",
    "    import plotly.tools as pytools\n",
    "    py.sign_in('heineike02_student','9dMTMZgJMgUP0YX0P5mQ')\n",
    "    #py.sign_in('heineike02', 'APPjKrtARaN2ZgUYIkqr')\n",
    "    \n",
    "# for phylogenetic trees: \n",
    "from ete3 import Tree\n",
    "\n",
    "#for scraping internet data (e.g. ncbi)\n",
    "import requests\n",
    "from lxml import etree    #parses xml output\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load species info table\n",
    "species_info = pd.read_csv(os.path.normpath(data_processing_dir + 'promoter_phylogenies/ygob_species_info.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "      /-Kpol_1039.57\n",
      "   /-|\n",
      "  |   \\-TPHA0M01660\n",
      "  |\n",
      "  |      /-KAFR0D02620\n",
      "  |     |\n",
      "  |   /-|      /-NCAS0I01030\n",
      "  |  |  |   /-|\n",
      "  |  |  |  |   \\-NDAI0A07980\n",
      "  |  |   \\-|\n",
      "  |  |     |   /-KNAG0B01420\n",
      "  |  |      \\-|\n",
      "  |  |        |   /-Suva_4.364\n",
      "--|  |         \\-|\n",
      "  |  |           |   /-Smik_2.257\n",
      "  |  |            \\-|\n",
      "  |  |              |   /-YBR117C\n",
      "  |  |               \\-|\n",
      "  |  |                  \\-Skud_2.244\n",
      "  |  |\n",
      "  |  |            /-Kwal_23.3955\n",
      "  |  |         /-|\n",
      "  |  |        |   \\-KLTH0F11836g\n",
      "  |  |      /-|\n",
      "  |  |     |  |   /-SAKL0B08096g\n",
      "  |  |     |   \\-|\n",
      "   \\-|     |     |   /-KLLA0B09152g\n",
      "     |     |      \\-|\n",
      "     |     |        |   /-ADL366W\n",
      "     |     |         \\-|\n",
      "     |     |            \\-Ecym_3454\n",
      "     |     |\n",
      "     |     |                  /-NDAI0A05060\n",
      "     |     |               /-|\n",
      "     |     |            /-|   \\-NCAS0A11230\n",
      "     |     |           |  |\n",
      "     |     |           |   \\-KAFR0H00840\n",
      "     |   /-|         /-|\n",
      "     |  |  |        |  |      /-Skud_16.363\n",
      "     |  |  |        |  |   /-|\n",
      "     |  |  |        |  |  |   \\-Suva_16.403\n",
      "     |  |  |        |   \\-|\n",
      "     |  |  |      /-|     |   /-YPR074C\n",
      "     |  |  |     |  |      \\-|\n",
      "     |  |  |     |  |         \\-Smik_16.318\n",
      "     |  |  |     |  |\n",
      "     |  |  |     |  |   /-TBLA0H01480\n",
      "      \\-|  |   /-|   \\-|\n",
      "        |  |  |  |     |   /-Kpol_1006.8\n",
      "        |  |  |  |      \\-|\n",
      "        |  |  |  |         \\-TPHA0A03490\n",
      "        |   \\-|  |\n",
      "        |     |  |   /-CAGL0D01298g\n",
      "        |     |   \\-|\n",
      "        |     |      \\-KNAG0A07310\n",
      "        |     |\n",
      "        |      \\-TDEL0C05140\n",
      "        |\n",
      "         \\-ZYRO0C16632g\n"
     ]
    }
   ],
   "source": [
    "#load phylogenentic tree \n",
    "sc_genes = ['GND1', 'TAL1', 'SOL3']\n",
    "sc_gene = 'TKL1'\n",
    "fname = os.path.normpath(data_processing_dir + 'promoter_phylogenies/' + sc_gene + '_tree_ygob.txt')\n",
    "with open(fname,'r') as f: \n",
    "    gene_phylogeny = f.read()\n",
    "\n",
    "tree = Tree(gene_phylogeny)\n",
    "print(tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abbreviation</th>\n",
       "      <th>ygob_gene_prefix</th>\n",
       "      <th>ncbi_period_convert</th>\n",
       "      <th>ncbi_name</th>\n",
       "      <th>in_ncbi</th>\n",
       "      <th>in_sss</th>\n",
       "      <th>in_genolevures</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Vpol</td>\n",
       "      <td>Kpol</td>\n",
       "      <td>True</td>\n",
       "      <td>Vanderwaltozyma polyspora</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tpha</td>\n",
       "      <td>TPHA0</td>\n",
       "      <td>False</td>\n",
       "      <td>Tetrapisispora phaffii</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tbla</td>\n",
       "      <td>TBLA0</td>\n",
       "      <td>False</td>\n",
       "      <td>Tetrapisispora blattae</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ndai</td>\n",
       "      <td>NDAI0</td>\n",
       "      <td>False</td>\n",
       "      <td>Naumovozyma dairenensis</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ncas</td>\n",
       "      <td>NCAS0</td>\n",
       "      <td>False</td>\n",
       "      <td>Naumovozyma castellii</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Knag</td>\n",
       "      <td>KNAG0</td>\n",
       "      <td>False</td>\n",
       "      <td>Kazachstania naganishii</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Kafr</td>\n",
       "      <td>KAFR0</td>\n",
       "      <td>False</td>\n",
       "      <td>Kazachstania africana</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Cagl</td>\n",
       "      <td>CAGL0</td>\n",
       "      <td>False</td>\n",
       "      <td>Candida glabrata</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Suva</td>\n",
       "      <td>Suva</td>\n",
       "      <td>False</td>\n",
       "      <td>Saccharomyces uvarum</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Skud</td>\n",
       "      <td>Skud</td>\n",
       "      <td>False</td>\n",
       "      <td>Saccharomyces kudriavzevii</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Smik</td>\n",
       "      <td>Smik</td>\n",
       "      <td>False</td>\n",
       "      <td>Saccharomyces mikatae</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Scer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Zrou</td>\n",
       "      <td>ZYRO0</td>\n",
       "      <td>False</td>\n",
       "      <td>Zygosaccharomyces rouxii</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Tdel</td>\n",
       "      <td>TDEL0</td>\n",
       "      <td>False</td>\n",
       "      <td>Torulaspora delbrueckii</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Klac</td>\n",
       "      <td>KLLA0</td>\n",
       "      <td>False</td>\n",
       "      <td>Kluyveromyces lactis</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Egos</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>Eremothecium gossypii</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Ecym</td>\n",
       "      <td>Ecym</td>\n",
       "      <td>False</td>\n",
       "      <td>Eremothecium cymbalariae</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Lkly</td>\n",
       "      <td>SAKL0</td>\n",
       "      <td>False</td>\n",
       "      <td>Lachancea kluyveri</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Lthe</td>\n",
       "      <td>KLTH0</td>\n",
       "      <td>False</td>\n",
       "      <td>Lachancea thermotolerans</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Lwal</td>\n",
       "      <td>Kwal</td>\n",
       "      <td>False</td>\n",
       "      <td>Lachancea waltii</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   abbreviation ygob_gene_prefix  ncbi_period_convert  \\\n",
       "0          Vpol             Kpol                 True   \n",
       "1          Tpha            TPHA0                False   \n",
       "2          Tbla            TBLA0                False   \n",
       "3          Ndai            NDAI0                False   \n",
       "4          Ncas            NCAS0                False   \n",
       "5          Knag            KNAG0                False   \n",
       "6          Kafr            KAFR0                False   \n",
       "7          Cagl            CAGL0                False   \n",
       "8          Suva             Suva                False   \n",
       "9          Skud             Skud                False   \n",
       "10         Smik             Smik                False   \n",
       "11         Scer              NaN                False   \n",
       "12         Zrou            ZYRO0                False   \n",
       "13         Tdel            TDEL0                False   \n",
       "14         Klac            KLLA0                False   \n",
       "15         Egos              NaN                False   \n",
       "16         Ecym             Ecym                False   \n",
       "17         Lkly            SAKL0                False   \n",
       "18         Lthe            KLTH0                False   \n",
       "19         Lwal             Kwal                False   \n",
       "\n",
       "                     ncbi_name  in_ncbi  in_sss  in_genolevures  \n",
       "0    Vanderwaltozyma polyspora     True   False           False  \n",
       "1       Tetrapisispora phaffii     True   False           False  \n",
       "2       Tetrapisispora blattae     True   False            True  \n",
       "3      Naumovozyma dairenensis     True   False           False  \n",
       "4        Naumovozyma castellii     True   False            True  \n",
       "5      Kazachstania naganishii     True   False            True  \n",
       "6        Kazachstania africana     True   False           False  \n",
       "7             Candida glabrata     True   False           False  \n",
       "8         Saccharomyces uvarum    False    True           False  \n",
       "9   Saccharomyces kudriavzevii    False    True           False  \n",
       "10       Saccharomyces mikatae    False    True           False  \n",
       "11    Saccharomyces cerevisiae     True    True            True  \n",
       "12    Zygosaccharomyces rouxii     True   False            True  \n",
       "13     Torulaspora delbrueckii     True   False           False  \n",
       "14        Kluyveromyces lactis     True   False            True  \n",
       "15       Eremothecium gossypii     True   False            True  \n",
       "16    Eremothecium cymbalariae     True   False           False  \n",
       "17          Lachancea kluyveri    False   False            True  \n",
       "18    Lachancea thermotolerans     True   False            True  \n",
       "19            Lachancea waltii    False   False            True  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_info = pd.read_csv(os.path.normpath(data_processing_dir + 'promoter_phylogenies/ygob_species_info.csv'))\n",
    "species_info\n",
    "#species_info[~species_info['ygob_gene_prefix'].isna()].loc[:,['ygob_gene_prefix','abbreviation']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = gene_data_xml.getchildren()[0]\n",
    "b = a.find('Entrezgene_locus')\n",
    "b != None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Entrezgene'"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a.tag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Entrezgene_locus not at position 7",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-24-44e8a79e9681>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[1;31m#check that Entrezgene_locus is still indexed at item7\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[1;32mif\u001b[0m \u001b[0mlocus_element\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtag\u001b[0m \u001b[1;33m!=\u001b[0m \u001b[1;34m'Entrezgene_locus'\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 8\u001b[1;33m     \u001b[1;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'Entrezgene_locus not at position 7'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      9\u001b[0m \u001b[0mlocus_gene_commentary_list\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mlocus_element\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfindall\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"Gene-commentary\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     10\u001b[0m \u001b[1;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlocus_gene_commentary_list\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: Entrezgene_locus not at position 7"
     ]
    }
   ],
   "source": [
    "entrez_gene_id = '5544171'\n",
    "gene_data_url = 'http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=gene&id=' + entrez_gene_id + '&retmode=xml'\n",
    "gene_data_response = requests.get(gene_data_url) \n",
    "gene_data_xml = etree.fromstring(gene_data_response.content)\n",
    "entrezgene_element = gene_data_xml.getchildren()[0]\n",
    "locus_element = entrezgene_element.find('Entrezgene_locus')\n",
    "#check that Entrezgene_locus is not none\n",
    "if locus_element == None: \n",
    "    raise ValueError('Entrezgene_locus not found ' + gene)\n",
    "locus_gene_commentary_list = locus_element.findall(\"Gene-commentary\") \n",
    "if len(locus_gene_commentary_list) == 1: \n",
    "    locus_gene_commentary = locus_gene_commentary_list[0]\n",
    "else: \n",
    "    print('NCBI has more than one gene commentary elements in the Entrezgene Locus looking for type 1 = genomic')\n",
    "    #Keep only the gene commentary that has type = 1 \n",
    "    locus_gene_commentary_shortlist = [locus_gene_commentary for locus_gene_commentary in locus_gene_commentary_list if locus_gene_commentary.find(\"Gene-commentary_type\").text=='1']\n",
    "    if len(locus_gene_commentary_shortlist) == 1: \n",
    "        locus_gene_commentary = locus_gene_commentary_shortlist[0]\n",
    "    else: \n",
    "        print('Error - more than one gene commentary listed for genomic dna.  gene = ' + gene)\n",
    "if locus_gene_commentary.find(\"Gene-commentary_type\").text != '1':\n",
    "    print('Error - gene commentary type is not 1: genomic dna.  gene = ' + gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Intergenic regions for the four sensu stricu species is here: \n",
    "#http://www.saccharomycessensustricto.org/cgi-bin/s3.cgi?data=Intergenics&version=current\n",
    "\n",
    "#genolevures has the other missing species. \n",
    "#NOt to hard to get the promoter from the web interface here - \n",
    "#wish there was an API though\n",
    "#gene_search_url = 'http://gryc.inra.fr/index.php?page=locus&seqid=SAKL0B08096g'\n",
    "#have to add it to cart, check nucleotide and then download after adding 1000bases to the 5' site."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "YHR183W\n",
      "Suva_15.384 not in NCBI - no method to get promoter yet\n",
      "Skud_8.246 not in NCBI - no method to get promoter yet\n",
      "Smik_8.271 not in NCBI - no method to get promoter yet\n",
      "YGR256W\n",
      "Skud_7.590 not in NCBI - no method to get promoter yet\n",
      "Suva_7.554 not in NCBI - no method to get promoter yet\n",
      "AFR129W\n",
      "NCBI has more than one gene commentary elements in the Entrezgene Locus looking for type 1 = genomic\n",
      "SAKL0H01848g not in NCBI - no method to get promoter yet\n",
      "Kwal_23.4361 not in NCBI - no method to get promoter yet\n",
      "ACL196W\n",
      "NCBI has more than one gene commentary elements in the Entrezgene Locus looking for type 1 = genomic\n",
      "Kwal_14.1439 not in NCBI - no method to get promoter yet\n",
      "SAKL0B04642g not in NCBI - no method to get promoter yet\n",
      "YLR354C\n",
      "Skud_12.439 not in NCBI - no method to get promoter yet\n",
      "Suva_10.455 not in NCBI - no method to get promoter yet\n",
      "Smik_12.442 not in NCBI - no method to get promoter yet\n",
      "Suva_7.328 not in NCBI - no method to get promoter yet\n",
      "Skud_7.340 not in NCBI - no method to get promoter yet\n",
      "YGR043C\n",
      "Smik_6.128 not in NCBI - no method to get promoter yet\n",
      "AEL215C\n",
      "NCBI has more than one gene commentary elements in the Entrezgene Locus looking for type 1 = genomic\n",
      "SAKL0H02376g not in NCBI - no method to get promoter yet\n",
      "Smik_8.247 not in NCBI - no method to get promoter yet\n",
      "Skud_8.227 not in NCBI - no method to get promoter yet\n",
      "YHR163W\n",
      "Suva_15.362 not in NCBI - no method to get promoter yet\n",
      "Skud_7.581 not in NCBI - no method to get promoter yet\n",
      "Suva_7.543 not in NCBI - no method to get promoter yet\n",
      "YGR248W\n",
      "Smik_16.59 not in NCBI - no method to get promoter yet\n"
     ]
    }
   ],
   "source": [
    "#load species info table\n",
    "species_info = pd.read_csv(os.path.normpath(data_processing_dir + 'promoter_phylogenies/ygob_species_info.csv'))\n",
    "sc_genes = ['GND1', 'TAL1', 'SOL3']\n",
    "\n",
    "for sc_gene in sc_genes: \n",
    "    fname = os.path.normpath(data_processing_dir + 'promoter_phylogenies/' + sc_gene + '_tree_ygob.txt')\n",
    "    with open(fname,'r') as f: \n",
    "        gene_phylogeny = f.read()\n",
    "    tree = Tree(gene_phylogeny)   \n",
    "\n",
    "    genes = tree.get_leaf_names()\n",
    "    prom_length = 1000\n",
    "    specs = []\n",
    "    promoter_outputs = {}\n",
    "    #go get its promoter sequence. \n",
    "    for gene in genes: \n",
    "        spec, promoter_output = io_library.extract_promoter_sequence(gene,prom_length)\n",
    "        specs.append(spec)\n",
    "        if isinstance(promoter_output,tuple):\n",
    "            promoter_outputs[gene] = promoter_output\n",
    "\n",
    "    promoter_data = pd.DataFrame.from_dict(promoter_outputs, orient='index')\n",
    "    promoter_data.columns =  [\"refseq\",\"loc\",\"strand\",\"prom_seq\",\"gene_seq\"]\n",
    "    promoter_data.index.name = 'genename'\n",
    "\n",
    "    gene_list = list(promoter_data.index)\n",
    "    fname = os.path.normpath(data_processing_dir + 'promoter_phylogenies/ortholog_promoters_' + sc_gene + '_' + str(prom_length) + '.fasta')\n",
    "    #Changed name from write_ame_promoter_file to write_promoter_file 10JAN18    \n",
    "    io_library.write_promoter_file(promoter_data, gene_list,fname)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "YHR183W\n",
      "YGR256W\n",
      "AFR129W\n",
      "ACL196W\n",
      "YLR354C\n",
      "YGR043C\n",
      "AEL215C\n",
      "YHR163W\n",
      "YGR248W\n",
      "YBR117C\n",
      "ADL366W\n",
      "YPR074C\n"
     ]
    }
   ],
   "source": [
    "#For a given promoter data file, do a fimo search \n",
    "\n",
    "sc_genes = ['GND1', 'TAL1', 'SOL3', 'TKL1']\n",
    "\n",
    "sc_gene_phylogenies = {}\n",
    "\n",
    "ncbi_specs = species_info[species_info['in_ncbi']]['abbreviation']\n",
    "\n",
    "\n",
    "for sc_gene in sc_genes:\n",
    "\n",
    "    #Load the fimo search data \n",
    "    fimo_fname = os.path.normpath(data_processing_dir + 'promoter_phylogenies/fimo_' + sc_gene + '_orthologs.txt')\n",
    "    fimo_hits = pd.read_table(fimo_fname)\n",
    "    \n",
    "    #Load the genes: \n",
    "    fname = os.path.normpath(data_processing_dir + 'promoter_phylogenies/' + sc_gene + '_tree_ygob.txt')\n",
    "    with open(fname,'r') as f: \n",
    "        gene_phylogeny = f.read()\n",
    "    tree = Tree(gene_phylogeny)  \n",
    "\n",
    "    #filter out the genes that didn't have promoters to search\n",
    "    genes = tree.get_leaf_names()\n",
    "    genes_ncbi = [gene for gene in genes if io_library.get_species_from_gene(gene, species_info) in set(ncbi_specs)]\n",
    "\n",
    "    motifs = list(set(fimo_hits['# motif_id']))\n",
    "\n",
    "    ortholog_motifs = pd.DataFrame(index = genes_ncbi)\n",
    "\n",
    "\n",
    "    length_filter = 700\n",
    "\n",
    "    print('this lenght filter is wrong: low numbers are farther away from the promoter - need to invert it') \n",
    "    for motif in motifs: \n",
    "        fimo_hits_motif = fimo_hits[fimo_hits['# motif_id']==motif]\n",
    "        #Filter motif to start site less than length_filter\n",
    "        fimo_hits_motif_lfilter = fimo_hits_motif[fimo_hits_motif['start']<length_filter]\n",
    "        ortholog_motifs[motif] = [Counter(fimo_hits_motif_lfilter['sequence_name'])[gene] for gene in genes_ncbi]\n",
    "\n",
    "    sc_gene_phylogenies[sc_gene] = ortholog_motifs\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "      /-Kpol_1039.57\n",
      "   /-|\n",
      "  |   \\-TPHA0M01660\n",
      "  |\n",
      "  |      /-KAFR0D02620\n",
      "  |     |\n",
      "  |   /-|      /-NCAS0I01030\n",
      "  |  |  |   /-|\n",
      "  |  |  |  |   \\-NDAI0A07980\n",
      "  |  |   \\-|\n",
      "  |  |     |   /-KNAG0B01420\n",
      "  |  |      \\-|\n",
      "  |  |        |   /-Suva_4.364\n",
      "--|  |         \\-|\n",
      "  |  |           |   /-Smik_2.257\n",
      "  |  |            \\-|\n",
      "  |  |              |   /-YBR117C\n",
      "  |  |               \\-|\n",
      "  |  |                  \\-Skud_2.244\n",
      "  |  |\n",
      "  |  |            /-Kwal_23.3955\n",
      "  |  |         /-|\n",
      "  |  |        |   \\-KLTH0F11836g\n",
      "  |  |      /-|\n",
      "  |  |     |  |   /-SAKL0B08096g\n",
      "  |  |     |   \\-|\n",
      "   \\-|     |     |   /-KLLA0B09152g\n",
      "     |     |      \\-|\n",
      "     |     |        |   /-ADL366W\n",
      "     |     |         \\-|\n",
      "     |     |            \\-Ecym_3454\n",
      "     |     |\n",
      "     |     |                  /-NDAI0A05060\n",
      "     |     |               /-|\n",
      "     |     |            /-|   \\-NCAS0A11230\n",
      "     |     |           |  |\n",
      "     |     |           |   \\-KAFR0H00840\n",
      "     |   /-|         /-|\n",
      "     |  |  |        |  |      /-Skud_16.363\n",
      "     |  |  |        |  |   /-|\n",
      "     |  |  |        |  |  |   \\-Suva_16.403\n",
      "     |  |  |        |   \\-|\n",
      "     |  |  |      /-|     |   /-YPR074C\n",
      "     |  |  |     |  |      \\-|\n",
      "     |  |  |     |  |         \\-Smik_16.318\n",
      "     |  |  |     |  |\n",
      "     |  |  |     |  |   /-TBLA0H01480\n",
      "      \\-|  |   /-|   \\-|\n",
      "        |  |  |  |     |   /-Kpol_1006.8\n",
      "        |  |  |  |      \\-|\n",
      "        |  |  |  |         \\-TPHA0A03490\n",
      "        |   \\-|  |\n",
      "        |     |  |   /-CAGL0D01298g\n",
      "        |     |   \\-|\n",
      "        |     |      \\-KNAG0A07310\n",
      "        |     |\n",
      "        |      \\-TDEL0C05140\n",
      "        |\n",
      "         \\-ZYRO0C16632g\n"
     ]
    }
   ],
   "source": [
    "#Load the genes: \n",
    "sc_gene = 'TKL1'\n",
    "fname = os.path.normpath(data_processing_dir + 'promoter_phylogenies/' + sc_gene + '_tree_ygob.txt')\n",
    "with open(fname,'r') as f: \n",
    "    gene_phylogeny = f.read()\n",
    "tree = Tree(gene_phylogeny) \n",
    "print(tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "YBR117C\n",
      "YBR117C\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'Cagl'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-75-e8d2c8a0ad0c>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     24\u001b[0m \u001b[1;31m#organize in species order\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     25\u001b[0m \u001b[1;31m#post_wgd = ['Scer','Cagl','Kafr','Knag','Ncas','Ndai','Tbla','Tpha','Vpol']\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 26\u001b[1;33m \u001b[1;33m[\u001b[0m\u001b[0mtrack1_orthologs_ncbi_spec\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mspec\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mspec\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mpost_wgd\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m<ipython-input-75-e8d2c8a0ad0c>\u001b[0m in \u001b[0;36m<listcomp>\u001b[1;34m(.0)\u001b[0m\n\u001b[0;32m     24\u001b[0m \u001b[1;31m#organize in species order\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     25\u001b[0m \u001b[1;31m#post_wgd = ['Scer','Cagl','Kafr','Knag','Ncas','Ndai','Tbla','Tpha','Vpol']\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 26\u001b[1;33m \u001b[1;33m[\u001b[0m\u001b[0mtrack1_orthologs_ncbi_spec\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mspec\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mspec\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mpost_wgd\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m: 'Cagl'"
     ]
    }
   ],
   "source": [
    "sc_genes = ['TAL1', 'SOL3', 'TKL1']\n",
    "\n",
    "phylogeny_pairs = {'YPR074C': 'KNAG0A07310', \n",
    "                   'YBR117C':'KAFR0D02620'}\n",
    "\n",
    "track1_track2 = {'TKL1':'TKL2',\n",
    "                 #'GND1':'GND2',\n",
    "                 'TAL1':'NQM1',\n",
    "                 'SOL3':'SOL4'}\n",
    "\n",
    "sc_gene = 'TKL1'\n",
    "\n",
    "fname = os.path.normpath(data_processing_dir + 'promoter_phylogenies/' + sc_gene + '_tree_ygob.txt')\n",
    "with open(fname,'r') as f: \n",
    "    gene_phylogeny = f.read()\n",
    "tree = Tree(gene_phylogeny) \n",
    "\n",
    "track1_ancestor = tree.get_common_ancestor(['YBR117C',phylogeny_pairs['YBR117C']])\n",
    "track1_orthologs = track1_ancestor.get_leaf_names()\n",
    "#remove seed_outgroup\n",
    "#track1_orthologs.remove(seed_outgroup)\n",
    "track1_orthologs_ncbi = [gene for gene in track1_orthologs if io_library.get_species_from_gene(gene, species_info) in set(ncbi_specs)]\n",
    "track1_orthologs_ncbi_spec = {io_library.get_species_from_gene(gene, species_info): gene for gene in track1_orthologs_ncbi}\n",
    "#organize in species order\n",
    "#post_wgd = ['Scer','Cagl','Kafr','Knag','Ncas','Ndai','Tbla','Tpha','Vpol']\n",
    "\n",
    "[track1_orthologs_ncbi_spec[spec] for spec in post_wgd]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "track1 = {}\n",
    "track2 = {}\n",
    "\n",
    "#T1 = TKL1, GND1, TAL1, SOL3\n",
    "#T2 = TKL2, GND2, NQM1, SOL4\n",
    "\n",
    "\n",
    "track1['GND1'] = ['YHR183W','CAGL0M13343g','KAFR0B01610','TBLA0C06670','NCAS0F02030', 'NDAI0C03490', 'Kpol_1067.2']\n",
    "track2['GND1'] = ['YGR256W','KAFR0B04490','KNAG0K00390','TPHA0A05350']\n",
    "\n",
    "track1['SOL3'] = ['YHR163W','TPHA0A05240', 'CAGL0I02200g', 'KNAG0E00640', 'Kpol_1050.78', 'NCAS0D01430', 'NDAI0C03310', 'KAFR0E02210' ]\n",
    "track2['SOL3'] = ['YGR248W', 'KAFR0B04400', 'KNAG0K00490', 'NCAS0E03860', 'NDAI0B00950', 'TBLA0C06580' ]\n",
    "\n",
    "track1['TAL1'] = ['YLR354C', 'NDAI0J01880', 'NCAS0J01140', 'KNAG0B05780', 'KAFR0A06290','CAGL0B03069g','Kpol_472.17', 'TPHA0F01080', 'TBLA0A08400']\n",
    "track2['TAL1'] = ['YGR043C', 'CAGL0K04235g', 'KNAG0D04010', 'KAFR0F03930', 'NDAI0D04080', 'NCAS0A02150', 'Kpol_530.51', 'TPHA0J00660']\n",
    "\n",
    "track1['TKL1'] = ['YPR074C','CAGL0D01298g', 'KAFR0H00840', 'KNAG0A07310', 'NCAS0A11230', 'NDAI0A05060', 'TBLA0H01480', 'TPHA0A03490','Kpol_1039.57']\n",
    "track2['TKL1'] = ['YBR117C','KAFR0D02620', 'KNAG0B01420','NCAS0I01030','NDAI0A07980', 'TPHA0M01660', 'Kpol_1006.8']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "YLR354C\n",
      "YLR354C\n",
      "YGR043C\n",
      "YGR043C\n",
      "ACL196W\n",
      "YLR354C\n",
      "YGR043C\n",
      "YHR163W\n",
      "YHR163W\n",
      "YGR248W\n",
      "YGR248W\n",
      "AEL215C\n",
      "YHR163W\n",
      "YGR248W\n",
      "YPR074C\n",
      "YPR074C\n",
      "YBR117C\n",
      "YBR117C\n",
      "YBR117C\n",
      "ADL366W\n",
      "YPR074C\n",
      "YHR183W\n",
      "YHR183W\n",
      "YGR256W\n",
      "YGR256W\n",
      "YHR183W\n",
      "YGR256W\n",
      "AFR129W\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TAL1</th>\n",
       "      <th>SOL3</th>\n",
       "      <th>TKL1</th>\n",
       "      <th>GND1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Scer_t1</th>\n",
       "      <td>YLR354C</td>\n",
       "      <td>YHR163W</td>\n",
       "      <td>YPR074C</td>\n",
       "      <td>YHR183W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cagl_t1</th>\n",
       "      <td>CAGL0B03069g</td>\n",
       "      <td>CAGL0I02200g</td>\n",
       "      <td>CAGL0D01298g</td>\n",
       "      <td>CAGL0M13343g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Kafr_t1</th>\n",
       "      <td>KAFR0A06290</td>\n",
       "      <td>KAFR0E02210</td>\n",
       "      <td>KAFR0H00840</td>\n",
       "      <td>KAFR0B01610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Knag_t1</th>\n",
       "      <td>KNAG0B05780</td>\n",
       "      <td>KNAG0E00640</td>\n",
       "      <td>KNAG0A07310</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ncas_t1</th>\n",
       "      <td>NCAS0J01140</td>\n",
       "      <td>NCAS0D01430</td>\n",
       "      <td>NCAS0A11230</td>\n",
       "      <td>NCAS0F02030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ndai_t1</th>\n",
       "      <td>NDAI0J01880</td>\n",
       "      <td>NDAI0C03310</td>\n",
       "      <td>NDAI0A05060</td>\n",
       "      <td>NDAI0C03490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tbla_t1</th>\n",
       "      <td>TBLA0A08400</td>\n",
       "      <td>None</td>\n",
       "      <td>TBLA0H01480</td>\n",
       "      <td>TBLA0C06670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tpha_t1</th>\n",
       "      <td>TPHA0F01080</td>\n",
       "      <td>TPHA0A05240</td>\n",
       "      <td>TPHA0A03490</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Vpol_t1</th>\n",
       "      <td>Kpol_472.17</td>\n",
       "      <td>Kpol_1050.78</td>\n",
       "      <td>Kpol_1039.57</td>\n",
       "      <td>Kpol_1067.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Scer_t2</th>\n",
       "      <td>YGR043C</td>\n",
       "      <td>YGR248W</td>\n",
       "      <td>YBR117C</td>\n",
       "      <td>YGR256W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cagl_t2</th>\n",
       "      <td>CAGL0K04235g</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Kafr_t2</th>\n",
       "      <td>KAFR0F03930</td>\n",
       "      <td>KAFR0B04400</td>\n",
       "      <td>KAFR0D02620</td>\n",
       "      <td>KAFR0B04490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Knag_t2</th>\n",
       "      <td>KNAG0D04010</td>\n",
       "      <td>KNAG0K00490</td>\n",
       "      <td>KNAG0B01420</td>\n",
       "      <td>KNAG0K00390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ncas_t2</th>\n",
       "      <td>NCAS0A02150</td>\n",
       "      <td>NCAS0E03860</td>\n",
       "      <td>NCAS0I01030</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ndai_t2</th>\n",
       "      <td>NDAI0D04080</td>\n",
       "      <td>NDAI0B00950</td>\n",
       "      <td>NDAI0A07980</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tbla_t2</th>\n",
       "      <td>None</td>\n",
       "      <td>TBLA0C06580</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tpha_t2</th>\n",
       "      <td>TPHA0J00660</td>\n",
       "      <td>None</td>\n",
       "      <td>TPHA0M01660</td>\n",
       "      <td>TPHA0A05350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Vpol_t2</th>\n",
       "      <td>Kpol_530.51</td>\n",
       "      <td>None</td>\n",
       "      <td>Kpol_1006.8</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Klac</th>\n",
       "      <td>KLLA0A02607g</td>\n",
       "      <td>KLLA0A05390g</td>\n",
       "      <td>KLLA0B09152g</td>\n",
       "      <td>KLLA0A09339g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lthe</th>\n",
       "      <td>KLTH0G17380g</td>\n",
       "      <td>KLTH0H00836g</td>\n",
       "      <td>KLTH0F11836g</td>\n",
       "      <td>KLTH0B08668g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Egos</th>\n",
       "      <td>ACL196W</td>\n",
       "      <td>AEL215C</td>\n",
       "      <td>ADL366W</td>\n",
       "      <td>AFR129W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ecym</th>\n",
       "      <td>Ecym_6140</td>\n",
       "      <td>Ecym_5572</td>\n",
       "      <td>Ecym_3454</td>\n",
       "      <td>Ecym_5108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tdel</th>\n",
       "      <td>TDEL0E01870</td>\n",
       "      <td>TDEL0G01020</td>\n",
       "      <td>TDEL0C05140</td>\n",
       "      <td>TDEL0G00790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zrou</th>\n",
       "      <td>ZYRO0A12914g</td>\n",
       "      <td>ZYRO0C11484g</td>\n",
       "      <td>ZYRO0C16632g</td>\n",
       "      <td>ZYRO0D07876g</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 TAL1          SOL3          TKL1          GND1\n",
       "Scer_t1       YLR354C       YHR163W       YPR074C       YHR183W\n",
       "Cagl_t1  CAGL0B03069g  CAGL0I02200g  CAGL0D01298g  CAGL0M13343g\n",
       "Kafr_t1   KAFR0A06290   KAFR0E02210   KAFR0H00840   KAFR0B01610\n",
       "Knag_t1   KNAG0B05780   KNAG0E00640   KNAG0A07310          None\n",
       "Ncas_t1   NCAS0J01140   NCAS0D01430   NCAS0A11230   NCAS0F02030\n",
       "Ndai_t1   NDAI0J01880   NDAI0C03310   NDAI0A05060   NDAI0C03490\n",
       "Tbla_t1   TBLA0A08400          None   TBLA0H01480   TBLA0C06670\n",
       "Tpha_t1   TPHA0F01080   TPHA0A05240   TPHA0A03490          None\n",
       "Vpol_t1   Kpol_472.17  Kpol_1050.78  Kpol_1039.57   Kpol_1067.2\n",
       "Scer_t2       YGR043C       YGR248W       YBR117C       YGR256W\n",
       "Cagl_t2  CAGL0K04235g          None          None          None\n",
       "Kafr_t2   KAFR0F03930   KAFR0B04400   KAFR0D02620   KAFR0B04490\n",
       "Knag_t2   KNAG0D04010   KNAG0K00490   KNAG0B01420   KNAG0K00390\n",
       "Ncas_t2   NCAS0A02150   NCAS0E03860   NCAS0I01030          None\n",
       "Ndai_t2   NDAI0D04080   NDAI0B00950   NDAI0A07980          None\n",
       "Tbla_t2          None   TBLA0C06580          None          None\n",
       "Tpha_t2   TPHA0J00660          None   TPHA0M01660   TPHA0A05350\n",
       "Vpol_t2   Kpol_530.51          None   Kpol_1006.8          None\n",
       "Klac     KLLA0A02607g  KLLA0A05390g  KLLA0B09152g  KLLA0A09339g\n",
       "Lthe     KLTH0G17380g  KLTH0H00836g  KLTH0F11836g  KLTH0B08668g\n",
       "Egos          ACL196W       AEL215C       ADL366W       AFR129W\n",
       "Ecym        Ecym_6140     Ecym_5572     Ecym_3454     Ecym_5108\n",
       "Tdel      TDEL0E01870   TDEL0G01020   TDEL0C05140   TDEL0G00790\n",
       "Zrou     ZYRO0A12914g  ZYRO0C11484g  ZYRO0C16632g  ZYRO0D07876g"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "#Make dataframe with index \n",
    "ncbi_specs = species_info[species_info['in_ncbi']]['abbreviation']\n",
    "\n",
    "pre_wgd = ['Klac','Lthe', 'Egos','Ecym','Tdel','Zrou']\n",
    "pre_wgd_ncbi = [spec for spec in pre_wgd if spec in set(ncbi_specs)]\n",
    "post_wgd = ['Scer','Cagl','Kafr','Knag','Ncas','Ndai','Tbla','Tpha','Vpol']\n",
    "post_wgd_ncbi = [spec for spec in post_wgd if spec in set(ncbi_specs)]\n",
    "\n",
    "post_wgd_t1 = [spec + '_t1' for spec in post_wgd_ncbi]\n",
    "post_wgd_t2 = [spec + '_t2' for spec in post_wgd_ncbi]\n",
    "\n",
    "spec_index = post_wgd_t1 + post_wgd_t2 + pre_wgd_ncbi\n",
    "\n",
    "sc_genes = ['TAL1', 'SOL3', 'TKL1', 'GND1']\n",
    "\n",
    "orthologs_table = pd.DataFrame(index = spec_index, columns = sc_genes)\n",
    "\n",
    "\n",
    "for sc_gene in sc_genes: \n",
    "    #make a list with a gene or None for each species that we searched in FIMO in the order of the index. \n",
    "    gene_list = []\n",
    "    for track in [track1, track2]: \n",
    "#         print(sc_gene)\n",
    "#         print(track)\n",
    "#         print(track[sc_gene])\n",
    "        track_sc_gene = track[sc_gene]\n",
    "        #keep only the genes in NCBI\n",
    "        track_ncbi = [gene for gene in track_sc_gene if io_library.get_species_from_gene(gene, species_info) in set(ncbi_specs)]\n",
    "        track_ncbi_spec = {io_library.get_species_from_gene(gene, species_info): gene for gene in track_ncbi}\n",
    "    \n",
    "        #for each species in the post_wgd list, try to slot in a gene from the track.  If none is present put in None. \n",
    "        for spec in post_wgd_ncbi:\n",
    "            if spec in track_ncbi_spec.keys(): \n",
    "                gene_list.append(track_ncbi_spec[spec])\n",
    "            else:\n",
    "                gene_list.append(None)\n",
    "    \n",
    "    #add on the species with only one track\n",
    "    \n",
    "    fname = os.path.normpath(data_processing_dir + 'promoter_phylogenies/' + sc_gene + '_tree_ygob.txt')\n",
    "    with open(fname,'r') as f: \n",
    "        gene_phylogeny = f.read()\n",
    "    tree = Tree(gene_phylogeny) \n",
    "    orthologs = tree.get_leaf_names()\n",
    "    \n",
    "    pre_wgd_orthologs = {}\n",
    "    for ortholog in orthologs: \n",
    "        spec = io_library.get_species_from_gene(ortholog,species_info)\n",
    "        if spec in set(pre_wgd_ncbi): \n",
    "            pre_wgd_orthologs[spec] = ortholog        \n",
    "            \n",
    "    for spec in pre_wgd_ncbi: \n",
    "        if spec in pre_wgd_orthologs.keys(): \n",
    "            gene_list.append(pre_wgd_orthologs[spec])\n",
    "        else:\n",
    "            gene_list.append(None)\n",
    "    \n",
    "    orthologs_table[sc_gene] = gene_list\n",
    "\n",
    "\n",
    "orthologs_table\n",
    "        \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AEL215C': 0,\n",
       " 'CAGL0I02200g': 0,\n",
       " 'Ecym_5572': 3,\n",
       " 'KAFR0B04400': 1,\n",
       " 'KAFR0E02210': 0,\n",
       " 'KLLA0A05390g': 0,\n",
       " 'KLTH0H00836g': 2,\n",
       " 'KNAG0E00640': 2,\n",
       " 'KNAG0K00490': 1,\n",
       " 'Kpol_1050.78': 0,\n",
       " 'NCAS0D01430': 0,\n",
       " 'NCAS0E03860': 1,\n",
       " 'NDAI0B00950': 0,\n",
       " 'NDAI0C03310': 4,\n",
       " 'TBLA0C06580': 1,\n",
       " 'TDEL0G01020': 0,\n",
       " 'TPHA0A05240': 0,\n",
       " 'YGR248W': 0,\n",
       " 'YHR163W': 0,\n",
       " 'ZYRO0C11484g': 1}"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "binding_site_lookup = dict(sc_gene_phylogenies['SOL3']['MSN2_Badis'])\n",
    "binding_site_lookup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 7, 0, 2, 0, 3, 1, 0, 0, 1, 0, 0, 3, 1, 1, None, 1, 1, 1, 1, 2, 4, 0, 0]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "binding_site_lookup = dict(sc_gene_phylogenies['SOL3']['MSN2_Badis'])\n",
    "binding_site_list = []\n",
    "for ortholog in orthologs_table['SOL3']: \n",
    "    if ortholog==None: \n",
    "        binding_site_list.append(None)\n",
    "    else: \n",
    "        binding_site_list.append(binding_site_lookup[ortholog])\n",
    "binding_site_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "binding_site_table = pd.DataFrame(index = spec_index, columns = sc_genes)\n",
    "\n",
    "for sc_gene in sc_genes: \n",
    "    # make a dictionary that links name to number of Msn2\n",
    "    # use that to make a new column for a new dataframe with the same index \n",
    "    binding_site_lookup = dict(sc_gene_phylogenies[sc_gene]['MSN2_Badis'])\n",
    "    binding_site_list = []\n",
    "    for ortholog in orthologs_table[sc_gene]: \n",
    "        if ortholog==None: \n",
    "            binding_site_list.append(None)\n",
    "        else: \n",
    "            binding_site_list.append(binding_site_lookup[ortholog])\n",
    "    binding_site_list\n",
    "    binding_site_table[sc_gene] = binding_site_list\n",
    "\n",
    "binding_site_table.to_csv(os.path.normpath(data_processing_dir + 'promoter_phylogenies/PPP_phylogenies_badis_700.csv'))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add additional promoters from Trevor's database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#For a given gene: \n",
    "\n",
    "#Make a promoter orthology. \n",
    "#Organize orthology to match phylogenetic tree\n",
    "\n",
    "#Search motifs for the promoter list - make a column for each search (exact matches or fimo matches (check trevor's scanning code))\n",
    "\n",
    "#save as a spreadsheet"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
