{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "what computer are you on? a = Bens, b = gpucluster, c = other   a\n",
      "base directory is C:\\Users\\heine\\github\\expression_broad_data\n",
      "I am about to import a library\n",
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n",
      "are you online? Yes/NoYes\n"
     ]
    }
   ],
   "source": [
    "# %load std_libraries.py\n",
    "#to keep my scripts consistent, I am adding this to the top of all scripts with %load std_libraries.py\n",
    "import sys\n",
    "#Indicate operating environment and import core modules\n",
    "location_input = input(\"what computer are you on? a = Bens, b = gpucluster, c = other   \")\n",
    "location_dict = {'a': \"C:\\\\Users\\\\heine\\\\github\\\\expression_broad_data\", 'b': \"/home/heineike/github/expression_broad_data\",'c':'you need to add your location to the location_dict'}\n",
    "base_dir = location_dict[location_input]\n",
    "print(\"base directory is \" + base_dir)\n",
    "\n",
    "if sys.path[-1] != base_dir:\n",
    "    sys.path.append(base_dir)\n",
    "    print(\"Added \" + base_dir + \" to path: \" )\n",
    "    print(sys.path)\n",
    "\n",
    "import os\n",
    "\n",
    "print(\"I am about to import a library\")\n",
    "from core import expression_plots \n",
    "from core import io_library \n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "# %matplotlib \n",
    "# plt.ioff()\n",
    "#%matplotlib inline\n",
    "%matplotlib notebook\n",
    "data_processing_dir = base_dir + os.sep + os.path.normpath(\"expression_data\") + os.sep\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "#import seaborn as sns; sns.set(style=\"ticks\", color_codes=True)\n",
    "#from sklearn import linear_model\n",
    "#import pickle\n",
    "#import subprocess\n",
    "#import networkx as nx\n",
    "# import scipy.stats as stats\n",
    "import scipy.cluster.hierarchy as sch\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio import SeqFeature as sf\n",
    "from Bio.Alphabet import generic_dna\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "import re\n",
    "\n",
    "from collections import Counter\n",
    "# import scipy.stats as stats\n",
    "# from itertools import chain\n",
    "#this only works if you are online\n",
    "online_input = input(\"are you online? Yes/No\")\n",
    "if online_input == \"Yes\": \n",
    "    import plotly.plotly as py\n",
    "    import plotly.graph_objs as pygo\n",
    "    import plotly.tools as pytools\n",
    "    py.sign_in('heineike02_student','9dMTMZgJMgUP0YX0P5mQ')\n",
    "    #py.sign_in('heineike02', 'APPjKrtARaN2ZgUYIkqr')\n",
    "    \n",
    "# for phylogenetic trees: \n",
    "from ete3 import Tree\n",
    "\n",
    "#for scraping internet data (e.g. ncbi)\n",
    "import requests\n",
    "from lxml import etree    #parses xml output\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get promoters for all the genes I am interested in\n",
    "\n",
    "#Calculate \n",
    "#   Presence + location of exact STRE\n",
    "#   Presence + location of core PDS\n",
    "#   Presence (using motif?) of TATA box\n",
    "\n",
    "#Display\n",
    "#   Numbers of elements for all gois\n",
    "#   Image with location on promoter for any given goi\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load species info table\n",
    "species_info = pd.read_csv(os.path.normpath(data_processing_dir + 'promoter_phylogenies/ygob_species_info.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "      /-Kpol_1039.57\n",
      "   /-|\n",
      "  |   \\-TPHA0M01660\n",
      "  |\n",
      "  |      /-KAFR0D02620\n",
      "  |     |\n",
      "  |   /-|      /-NCAS0I01030\n",
      "  |  |  |   /-|\n",
      "  |  |  |  |   \\-NDAI0A07980\n",
      "  |  |   \\-|\n",
      "  |  |     |   /-KNAG0B01420\n",
      "  |  |      \\-|\n",
      "  |  |        |   /-Suva_4.364\n",
      "--|  |         \\-|\n",
      "  |  |           |   /-Smik_2.257\n",
      "  |  |            \\-|\n",
      "  |  |              |   /-YBR117C\n",
      "  |  |               \\-|\n",
      "  |  |                  \\-Skud_2.244\n",
      "  |  |\n",
      "  |  |            /-Kwal_23.3955\n",
      "  |  |         /-|\n",
      "  |  |        |   \\-KLTH0F11836g\n",
      "  |  |      /-|\n",
      "  |  |     |  |   /-SAKL0B08096g\n",
      "  |  |     |   \\-|\n",
      "   \\-|     |     |   /-KLLA0B09152g\n",
      "     |     |      \\-|\n",
      "     |     |        |   /-ADL366W\n",
      "     |     |         \\-|\n",
      "     |     |            \\-Ecym_3454\n",
      "     |     |\n",
      "     |     |                  /-NDAI0A05060\n",
      "     |     |               /-|\n",
      "     |     |            /-|   \\-NCAS0A11230\n",
      "     |     |           |  |\n",
      "     |     |           |   \\-KAFR0H00840\n",
      "     |   /-|         /-|\n",
      "     |  |  |        |  |      /-Skud_16.363\n",
      "     |  |  |        |  |   /-|\n",
      "     |  |  |        |  |  |   \\-Suva_16.403\n",
      "     |  |  |        |   \\-|\n",
      "     |  |  |      /-|     |   /-YPR074C\n",
      "     |  |  |     |  |      \\-|\n",
      "     |  |  |     |  |         \\-Smik_16.318\n",
      "     |  |  |     |  |\n",
      "     |  |  |     |  |   /-TBLA0H01480\n",
      "      \\-|  |   /-|   \\-|\n",
      "        |  |  |  |     |   /-Kpol_1006.8\n",
      "        |  |  |  |      \\-|\n",
      "        |  |  |  |         \\-TPHA0A03490\n",
      "        |   \\-|  |\n",
      "        |     |  |   /-CAGL0D01298g\n",
      "        |     |   \\-|\n",
      "        |     |      \\-KNAG0A07310\n",
      "        |     |\n",
      "        |      \\-TDEL0C05140\n",
      "        |\n",
      "         \\-ZYRO0C16632g\n"
     ]
    }
   ],
   "source": [
    "#load phylogenentic tree \n",
    "sc_genes = ['GND1', 'TAL1', 'SOL3']\n",
    "sc_gene = 'TKL1'\n",
    "fname = os.path.normpath(data_processing_dir + 'promoter_phylogenies/' + sc_gene + '_tree_ygob.txt')\n",
    "with open(fname,'r') as f: \n",
    "    gene_phylogeny = f.read()\n",
    "\n",
    "tree = Tree(gene_phylogeny)\n",
    "print(tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abbreviation</th>\n",
       "      <th>ygob_gene_prefix</th>\n",
       "      <th>ncbi_period_convert</th>\n",
       "      <th>ncbi_name</th>\n",
       "      <th>in_ncbi</th>\n",
       "      <th>in_sss</th>\n",
       "      <th>in_genolevures</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Vpol</td>\n",
       "      <td>Kpol</td>\n",
       "      <td>True</td>\n",
       "      <td>Vanderwaltozyma polyspora</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tpha</td>\n",
       "      <td>TPHA0</td>\n",
       "      <td>False</td>\n",
       "      <td>Tetrapisispora phaffii</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tbla</td>\n",
       "      <td>TBLA0</td>\n",
       "      <td>False</td>\n",
       "      <td>Tetrapisispora blattae</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ndai</td>\n",
       "      <td>NDAI0</td>\n",
       "      <td>False</td>\n",
       "      <td>Naumovozyma dairenensis</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ncas</td>\n",
       "      <td>NCAS0</td>\n",
       "      <td>False</td>\n",
       "      <td>Naumovozyma castellii</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Knag</td>\n",
       "      <td>KNAG0</td>\n",
       "      <td>False</td>\n",
       "      <td>Kazachstania naganishii</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Kafr</td>\n",
       "      <td>KAFR0</td>\n",
       "      <td>False</td>\n",
       "      <td>Kazachstania africana</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Cagl</td>\n",
       "      <td>CAGL0</td>\n",
       "      <td>False</td>\n",
       "      <td>Candida glabrata</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Suva</td>\n",
       "      <td>Suva</td>\n",
       "      <td>False</td>\n",
       "      <td>Saccharomyces uvarum</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Skud</td>\n",
       "      <td>Skud</td>\n",
       "      <td>False</td>\n",
       "      <td>Saccharomyces kudriavzevii</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Smik</td>\n",
       "      <td>Smik</td>\n",
       "      <td>False</td>\n",
       "      <td>Saccharomyces mikatae</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Scer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>Saccharomyces cerevisiae</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Zrou</td>\n",
       "      <td>ZYRO0</td>\n",
       "      <td>False</td>\n",
       "      <td>Zygosaccharomyces rouxii</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Tdel</td>\n",
       "      <td>TDEL0</td>\n",
       "      <td>False</td>\n",
       "      <td>Torulaspora delbrueckii</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Klac</td>\n",
       "      <td>KLLA0</td>\n",
       "      <td>False</td>\n",
       "      <td>Kluyveromyces lactis</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Egos</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>Eremothecium gossypii</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Ecym</td>\n",
       "      <td>Ecym</td>\n",
       "      <td>False</td>\n",
       "      <td>Eremothecium cymbalariae</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Lkly</td>\n",
       "      <td>SAKL0</td>\n",
       "      <td>False</td>\n",
       "      <td>Lachancea kluyveri</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Lthe</td>\n",
       "      <td>KLTH0</td>\n",
       "      <td>False</td>\n",
       "      <td>Lachancea thermotolerans</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Lwal</td>\n",
       "      <td>Kwal</td>\n",
       "      <td>False</td>\n",
       "      <td>Lachancea waltii</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   abbreviation ygob_gene_prefix  ncbi_period_convert  \\\n",
       "0          Vpol             Kpol                 True   \n",
       "1          Tpha            TPHA0                False   \n",
       "2          Tbla            TBLA0                False   \n",
       "3          Ndai            NDAI0                False   \n",
       "4          Ncas            NCAS0                False   \n",
       "5          Knag            KNAG0                False   \n",
       "6          Kafr            KAFR0                False   \n",
       "7          Cagl            CAGL0                False   \n",
       "8          Suva             Suva                False   \n",
       "9          Skud             Skud                False   \n",
       "10         Smik             Smik                False   \n",
       "11         Scer              NaN                False   \n",
       "12         Zrou            ZYRO0                False   \n",
       "13         Tdel            TDEL0                False   \n",
       "14         Klac            KLLA0                False   \n",
       "15         Egos              NaN                False   \n",
       "16         Ecym             Ecym                False   \n",
       "17         Lkly            SAKL0                False   \n",
       "18         Lthe            KLTH0                False   \n",
       "19         Lwal             Kwal                False   \n",
       "\n",
       "                     ncbi_name  in_ncbi  in_sss  in_genolevures  \n",
       "0    Vanderwaltozyma polyspora     True   False           False  \n",
       "1       Tetrapisispora phaffii     True   False           False  \n",
       "2       Tetrapisispora blattae     True   False            True  \n",
       "3      Naumovozyma dairenensis     True   False           False  \n",
       "4        Naumovozyma castellii     True   False            True  \n",
       "5      Kazachstania naganishii     True   False            True  \n",
       "6        Kazachstania africana     True   False           False  \n",
       "7             Candida glabrata     True   False           False  \n",
       "8         Saccharomyces uvarum    False    True           False  \n",
       "9   Saccharomyces kudriavzevii    False    True           False  \n",
       "10       Saccharomyces mikatae    False    True           False  \n",
       "11    Saccharomyces cerevisiae     True    True            True  \n",
       "12    Zygosaccharomyces rouxii     True   False            True  \n",
       "13     Torulaspora delbrueckii     True   False           False  \n",
       "14        Kluyveromyces lactis     True   False            True  \n",
       "15       Eremothecium gossypii     True   False            True  \n",
       "16    Eremothecium cymbalariae     True   False           False  \n",
       "17          Lachancea kluyveri    False   False            True  \n",
       "18    Lachancea thermotolerans     True   False            True  \n",
       "19            Lachancea waltii    False   False            True  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_info = pd.read_csv(os.path.normpath(data_processing_dir + 'promoter_phylogenies/ygob_species_info.csv'))\n",
    "species_info\n",
    "#species_info[~species_info['ygob_gene_prefix'].isna()].loc[:,['ygob_gene_prefix','abbreviation']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = gene_data_xml.getchildren()[0]\n",
    "b = a.find('Entrezgene_locus')\n",
    "b != None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Entrezgene'"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a.tag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Entrezgene_locus not at position 7",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-24-44e8a79e9681>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[1;31m#check that Entrezgene_locus is still indexed at item7\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[1;32mif\u001b[0m \u001b[0mlocus_element\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtag\u001b[0m \u001b[1;33m!=\u001b[0m \u001b[1;34m'Entrezgene_locus'\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 8\u001b[1;33m     \u001b[1;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'Entrezgene_locus not at position 7'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      9\u001b[0m \u001b[0mlocus_gene_commentary_list\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mlocus_element\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfindall\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"Gene-commentary\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     10\u001b[0m \u001b[1;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlocus_gene_commentary_list\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: Entrezgene_locus not at position 7"
     ]
    }
   ],
   "source": [
    "entrez_gene_id = '5544171'\n",
    "gene_data_url = 'http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=gene&id=' + entrez_gene_id + '&retmode=xml'\n",
    "gene_data_response = requests.get(gene_data_url) \n",
    "gene_data_xml = etree.fromstring(gene_data_response.content)\n",
    "entrezgene_element = gene_data_xml.getchildren()[0]\n",
    "locus_element = entrezgene_element.find('Entrezgene_locus')\n",
    "#check that Entrezgene_locus is not none\n",
    "if locus_element == None: \n",
    "    raise ValueError('Entrezgene_locus not found ' + gene)\n",
    "locus_gene_commentary_list = locus_element.findall(\"Gene-commentary\") \n",
    "if len(locus_gene_commentary_list) == 1: \n",
    "    locus_gene_commentary = locus_gene_commentary_list[0]\n",
    "else: \n",
    "    print('NCBI has more than one gene commentary elements in the Entrezgene Locus looking for type 1 = genomic')\n",
    "    #Keep only the gene commentary that has type = 1 \n",
    "    locus_gene_commentary_shortlist = [locus_gene_commentary for locus_gene_commentary in locus_gene_commentary_list if locus_gene_commentary.find(\"Gene-commentary_type\").text=='1']\n",
    "    if len(locus_gene_commentary_shortlist) == 1: \n",
    "        locus_gene_commentary = locus_gene_commentary_shortlist[0]\n",
    "    else: \n",
    "        print('Error - more than one gene commentary listed for genomic dna.  gene = ' + gene)\n",
    "if locus_gene_commentary.find(\"Gene-commentary_type\").text != '1':\n",
    "    print('Error - gene commentary type is not 1: genomic dna.  gene = ' + gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Intergenic regions for the four sensu stricu species is here: \n",
    "#http://www.saccharomycessensustricto.org/cgi-bin/s3.cgi?data=Intergenics&version=current\n",
    "\n",
    "#genolevures has the other missing species. \n",
    "#NOt to hard to get the promoter from the web interface here - \n",
    "#wish there was an API though\n",
    "#gene_search_url = 'http://gryc.inra.fr/index.php?page=locus&seqid=SAKL0B08096g'\n",
    "#have to add it to cart, check nucleotide and then download after adding 1000bases to the 5' site."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "YHR183W\n",
      "Suva_15.384 not in NCBI - no method to get promoter yet\n",
      "Skud_8.246 not in NCBI - no method to get promoter yet\n",
      "Smik_8.271 not in NCBI - no method to get promoter yet\n",
      "YGR256W\n",
      "Skud_7.590 not in NCBI - no method to get promoter yet\n",
      "Suva_7.554 not in NCBI - no method to get promoter yet\n",
      "AFR129W\n",
      "NCBI has more than one gene commentary elements in the Entrezgene Locus looking for type 1 = genomic\n",
      "SAKL0H01848g not in NCBI - no method to get promoter yet\n",
      "Kwal_23.4361 not in NCBI - no method to get promoter yet\n",
      "ACL196W\n",
      "NCBI has more than one gene commentary elements in the Entrezgene Locus looking for type 1 = genomic\n",
      "Kwal_14.1439 not in NCBI - no method to get promoter yet\n",
      "SAKL0B04642g not in NCBI - no method to get promoter yet\n",
      "YLR354C\n",
      "Skud_12.439 not in NCBI - no method to get promoter yet\n",
      "Suva_10.455 not in NCBI - no method to get promoter yet\n",
      "Smik_12.442 not in NCBI - no method to get promoter yet\n",
      "Suva_7.328 not in NCBI - no method to get promoter yet\n",
      "Skud_7.340 not in NCBI - no method to get promoter yet\n",
      "YGR043C\n",
      "Smik_6.128 not in NCBI - no method to get promoter yet\n",
      "AEL215C\n",
      "NCBI has more than one gene commentary elements in the Entrezgene Locus looking for type 1 = genomic\n",
      "SAKL0H02376g not in NCBI - no method to get promoter yet\n",
      "Smik_8.247 not in NCBI - no method to get promoter yet\n",
      "Skud_8.227 not in NCBI - no method to get promoter yet\n",
      "YHR163W\n",
      "Suva_15.362 not in NCBI - no method to get promoter yet\n",
      "Skud_7.581 not in NCBI - no method to get promoter yet\n",
      "Suva_7.543 not in NCBI - no method to get promoter yet\n",
      "YGR248W\n",
      "Smik_16.59 not in NCBI - no method to get promoter yet\n"
     ]
    }
   ],
   "source": [
    "#load species info table\n",
    "species_info = pd.read_csv(os.path.normpath(data_processing_dir + 'promoter_phylogenies/ygob_species_info.csv'))\n",
    "sc_genes = ['GND1', 'TAL1', 'SOL3']\n",
    "\n",
    "for sc_gene in sc_genes: \n",
    "    fname = os.path.normpath(data_processing_dir + 'promoter_phylogenies/' + sc_gene + '_tree_ygob.txt')\n",
    "    with open(fname,'r') as f: \n",
    "        gene_phylogeny = f.read()\n",
    "    tree = Tree(gene_phylogeny)   \n",
    "\n",
    "    genes = tree.get_leaf_names()\n",
    "    prom_length = 1000\n",
    "    specs = []\n",
    "    promoter_outputs = {}\n",
    "    #go get its promoter sequence. \n",
    "    for gene in genes: \n",
    "        spec, promoter_output = io_library.extract_promoter_sequence(gene,prom_length)\n",
    "        specs.append(spec)\n",
    "        if isinstance(promoter_output,tuple):\n",
    "            promoter_outputs[gene] = promoter_output\n",
    "\n",
    "    promoter_data = pd.DataFrame.from_dict(promoter_outputs, orient='index')\n",
    "    promoter_data.columns =  [\"refseq\",\"loc\",\"strand\",\"prom_seq\",\"gene_seq\"]\n",
    "    promoter_data.index.name = 'genename'\n",
    "\n",
    "    gene_list = list(promoter_data.index)\n",
    "    fname = os.path.normpath(data_processing_dir + 'promoter_phylogenies/ortholog_promoters_' + sc_gene + '_' + str(prom_length) + '.fasta')\n",
    "    #Changed name from write_ame_promoter_file to write_promoter_file 10JAN18    \n",
    "    io_library.write_promoter_file(promoter_data, gene_list,fname)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "YHR183W\n",
      "YGR256W\n",
      "AFR129W\n",
      "ACL196W\n",
      "YLR354C\n",
      "YGR043C\n",
      "AEL215C\n",
      "YHR163W\n",
      "YGR248W\n",
      "YBR117C\n",
      "ADL366W\n",
      "YPR074C\n"
     ]
    }
   ],
   "source": [
    "#For a given promoter data file, do a fimo search \n",
    "\n",
    "sc_genes = ['GND1', 'TAL1', 'SOL3', 'TKL1']\n",
    "\n",
    "sc_gene_phylogenies = {}\n",
    "\n",
    "ncbi_specs = species_info[species_info['in_ncbi']]['abbreviation']\n",
    "\n",
    "\n",
    "for sc_gene in sc_genes:\n",
    "\n",
    "    #Load the fimo search data \n",
    "    fimo_fname = os.path.normpath(data_processing_dir + 'promoter_phylogenies/fimo_' + sc_gene + '_orthologs.txt')\n",
    "    fimo_hits = pd.read_table(fimo_fname)\n",
    "    \n",
    "    #Load the genes: \n",
    "    fname = os.path.normpath(data_processing_dir + 'promoter_phylogenies/' + sc_gene + '_tree_ygob.txt')\n",
    "    with open(fname,'r') as f: \n",
    "        gene_phylogeny = f.read()\n",
    "    tree = Tree(gene_phylogeny)  \n",
    "\n",
    "    #filter out the genes that didn't have promoters to search\n",
    "    genes = tree.get_leaf_names()\n",
    "    genes_ncbi = [gene for gene in genes if io_library.get_species_from_gene(gene, species_info) in set(ncbi_specs)]\n",
    "\n",
    "    motifs = list(set(fimo_hits['# motif_id']))\n",
    "\n",
    "    ortholog_motifs = pd.DataFrame(index = genes_ncbi)\n",
    "\n",
    "\n",
    "    length_filter = 700\n",
    "\n",
    "    print('this lenght filter is wrong: low numbers are farther away from the promoter - need to invert it') \n",
    "    for motif in motifs: \n",
    "        fimo_hits_motif = fimo_hits[fimo_hits['# motif_id']==motif]\n",
    "        #Filter motif to start site less than length_filter\n",
    "        fimo_hits_motif_lfilter = fimo_hits_motif[fimo_hits_motif['start']<length_filter]\n",
    "        ortholog_motifs[motif] = [Counter(fimo_hits_motif_lfilter['sequence_name'])[gene] for gene in genes_ncbi]\n",
    "\n",
    "    sc_gene_phylogenies[sc_gene] = ortholog_motifs\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "      /-Kpol_1039.57\n",
      "   /-|\n",
      "  |   \\-TPHA0M01660\n",
      "  |\n",
      "  |      /-KAFR0D02620\n",
      "  |     |\n",
      "  |   /-|      /-NCAS0I01030\n",
      "  |  |  |   /-|\n",
      "  |  |  |  |   \\-NDAI0A07980\n",
      "  |  |   \\-|\n",
      "  |  |     |   /-KNAG0B01420\n",
      "  |  |      \\-|\n",
      "  |  |        |   /-Suva_4.364\n",
      "--|  |         \\-|\n",
      "  |  |           |   /-Smik_2.257\n",
      "  |  |            \\-|\n",
      "  |  |              |   /-YBR117C\n",
      "  |  |               \\-|\n",
      "  |  |                  \\-Skud_2.244\n",
      "  |  |\n",
      "  |  |            /-Kwal_23.3955\n",
      "  |  |         /-|\n",
      "  |  |        |   \\-KLTH0F11836g\n",
      "  |  |      /-|\n",
      "  |  |     |  |   /-SAKL0B08096g\n",
      "  |  |     |   \\-|\n",
      "   \\-|     |     |   /-KLLA0B09152g\n",
      "     |     |      \\-|\n",
      "     |     |        |   /-ADL366W\n",
      "     |     |         \\-|\n",
      "     |     |            \\-Ecym_3454\n",
      "     |     |\n",
      "     |     |                  /-NDAI0A05060\n",
      "     |     |               /-|\n",
      "     |     |            /-|   \\-NCAS0A11230\n",
      "     |     |           |  |\n",
      "     |     |           |   \\-KAFR0H00840\n",
      "     |   /-|         /-|\n",
      "     |  |  |        |  |      /-Skud_16.363\n",
      "     |  |  |        |  |   /-|\n",
      "     |  |  |        |  |  |   \\-Suva_16.403\n",
      "     |  |  |        |   \\-|\n",
      "     |  |  |      /-|     |   /-YPR074C\n",
      "     |  |  |     |  |      \\-|\n",
      "     |  |  |     |  |         \\-Smik_16.318\n",
      "     |  |  |     |  |\n",
      "     |  |  |     |  |   /-TBLA0H01480\n",
      "      \\-|  |   /-|   \\-|\n",
      "        |  |  |  |     |   /-Kpol_1006.8\n",
      "        |  |  |  |      \\-|\n",
      "        |  |  |  |         \\-TPHA0A03490\n",
      "        |   \\-|  |\n",
      "        |     |  |   /-CAGL0D01298g\n",
      "        |     |   \\-|\n",
      "        |     |      \\-KNAG0A07310\n",
      "        |     |\n",
      "        |      \\-TDEL0C05140\n",
      "        |\n",
      "         \\-ZYRO0C16632g\n"
     ]
    }
   ],
   "source": [
    "#Load the genes: \n",
    "sc_gene = 'TKL1'\n",
    "fname = os.path.normpath(data_processing_dir + 'promoter_phylogenies/' + sc_gene + '_tree_ygob.txt')\n",
    "with open(fname,'r') as f: \n",
    "    gene_phylogeny = f.read()\n",
    "tree = Tree(gene_phylogeny) \n",
    "print(tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "YBR117C\n",
      "YBR117C\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'Cagl'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-75-e8d2c8a0ad0c>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     24\u001b[0m \u001b[1;31m#organize in species order\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     25\u001b[0m \u001b[1;31m#post_wgd = ['Scer','Cagl','Kafr','Knag','Ncas','Ndai','Tbla','Tpha','Vpol']\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 26\u001b[1;33m \u001b[1;33m[\u001b[0m\u001b[0mtrack1_orthologs_ncbi_spec\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mspec\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mspec\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mpost_wgd\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m<ipython-input-75-e8d2c8a0ad0c>\u001b[0m in \u001b[0;36m<listcomp>\u001b[1;34m(.0)\u001b[0m\n\u001b[0;32m     24\u001b[0m \u001b[1;31m#organize in species order\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     25\u001b[0m \u001b[1;31m#post_wgd = ['Scer','Cagl','Kafr','Knag','Ncas','Ndai','Tbla','Tpha','Vpol']\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 26\u001b[1;33m \u001b[1;33m[\u001b[0m\u001b[0mtrack1_orthologs_ncbi_spec\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mspec\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mspec\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mpost_wgd\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m: 'Cagl'"
     ]
    }
   ],
   "source": [
    "sc_genes = ['TAL1', 'SOL3', 'TKL1']\n",
    "\n",
    "phylogeny_pairs = {'YPR074C': 'KNAG0A07310', \n",
    "                   'YBR117C':'KAFR0D02620'}\n",
    "\n",
    "track1_track2 = {'TKL1':'TKL2',\n",
    "                 #'GND1':'GND2',\n",
    "                 'TAL1':'NQM1',\n",
    "                 'SOL3':'SOL4'}\n",
    "\n",
    "sc_gene = 'TKL1'\n",
    "\n",
    "fname = os.path.normpath(data_processing_dir + 'promoter_phylogenies/' + sc_gene + '_tree_ygob.txt')\n",
    "with open(fname,'r') as f: \n",
    "    gene_phylogeny = f.read()\n",
    "tree = Tree(gene_phylogeny) \n",
    "\n",
    "track1_ancestor = tree.get_common_ancestor(['YBR117C',phylogeny_pairs['YBR117C']])\n",
    "track1_orthologs = track1_ancestor.get_leaf_names()\n",
    "#remove seed_outgroup\n",
    "#track1_orthologs.remove(seed_outgroup)\n",
    "track1_orthologs_ncbi = [gene for gene in track1_orthologs if io_library.get_species_from_gene(gene, species_info) in set(ncbi_specs)]\n",
    "track1_orthologs_ncbi_spec = {io_library.get_species_from_gene(gene, species_info): gene for gene in track1_orthologs_ncbi}\n",
    "#organize in species order\n",
    "#post_wgd = ['Scer','Cagl','Kafr','Knag','Ncas','Ndai','Tbla','Tpha','Vpol']\n",
    "\n",
    "[track1_orthologs_ncbi_spec[spec] for spec in post_wgd]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "track1 = {}\n",
    "track2 = {}\n",
    "\n",
    "#T1 = TKL1, GND1, TAL1, SOL3\n",
    "#T2 = TKL2, GND2, NQM1, SOL4\n",
    "\n",
    "\n",
    "track1['GND1'] = ['YHR183W','CAGL0M13343g','KAFR0B01610','TBLA0C06670','NCAS0F02030', 'NDAI0C03490', 'Kpol_1067.2']\n",
    "track2['GND1'] = ['YGR256W','KAFR0B04490','KNAG0K00390','TPHA0A05350']\n",
    "\n",
    "track1['SOL3'] = ['YHR163W','TPHA0A05240', 'CAGL0I02200g', 'KNAG0E00640', 'Kpol_1050.78', 'NCAS0D01430', 'NDAI0C03310', 'KAFR0E02210' ]\n",
    "track2['SOL3'] = ['YGR248W', 'KAFR0B04400', 'KNAG0K00490', 'NCAS0E03860', 'NDAI0B00950', 'TBLA0C06580' ]\n",
    "\n",
    "track1['TAL1'] = ['YLR354C', 'NDAI0J01880', 'NCAS0J01140', 'KNAG0B05780', 'KAFR0A06290','CAGL0B03069g','Kpol_472.17', 'TPHA0F01080', 'TBLA0A08400']\n",
    "track2['TAL1'] = ['YGR043C', 'CAGL0K04235g', 'KNAG0D04010', 'KAFR0F03930', 'NDAI0D04080', 'NCAS0A02150', 'Kpol_530.51', 'TPHA0J00660']\n",
    "\n",
    "track1['TKL1'] = ['YPR074C','CAGL0D01298g', 'KAFR0H00840', 'KNAG0A07310', 'NCAS0A11230', 'NDAI0A05060', 'TBLA0H01480', 'TPHA0A03490','Kpol_1039.57']\n",
    "track2['TKL1'] = ['YBR117C','KAFR0D02620', 'KNAG0B01420','NCAS0I01030','NDAI0A07980', 'TPHA0M01660', 'Kpol_1006.8']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "YLR354C\n",
      "YLR354C\n",
      "YGR043C\n",
      "YGR043C\n",
      "ACL196W\n",
      "YLR354C\n",
      "YGR043C\n",
      "YHR163W\n",
      "YHR163W\n",
      "YGR248W\n",
      "YGR248W\n",
      "AEL215C\n",
      "YHR163W\n",
      "YGR248W\n",
      "YPR074C\n",
      "YPR074C\n",
      "YBR117C\n",
      "YBR117C\n",
      "YBR117C\n",
      "ADL366W\n",
      "YPR074C\n",
      "YHR183W\n",
      "YHR183W\n",
      "YGR256W\n",
      "YGR256W\n",
      "YHR183W\n",
      "YGR256W\n",
      "AFR129W\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TAL1</th>\n",
       "      <th>SOL3</th>\n",
       "      <th>TKL1</th>\n",
       "      <th>GND1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Scer_t1</th>\n",
       "      <td>YLR354C</td>\n",
       "      <td>YHR163W</td>\n",
       "      <td>YPR074C</td>\n",
       "      <td>YHR183W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cagl_t1</th>\n",
       "      <td>CAGL0B03069g</td>\n",
       "      <td>CAGL0I02200g</td>\n",
       "      <td>CAGL0D01298g</td>\n",
       "      <td>CAGL0M13343g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Kafr_t1</th>\n",
       "      <td>KAFR0A06290</td>\n",
       "      <td>KAFR0E02210</td>\n",
       "      <td>KAFR0H00840</td>\n",
       "      <td>KAFR0B01610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Knag_t1</th>\n",
       "      <td>KNAG0B05780</td>\n",
       "      <td>KNAG0E00640</td>\n",
       "      <td>KNAG0A07310</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ncas_t1</th>\n",
       "      <td>NCAS0J01140</td>\n",
       "      <td>NCAS0D01430</td>\n",
       "      <td>NCAS0A11230</td>\n",
       "      <td>NCAS0F02030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ndai_t1</th>\n",
       "      <td>NDAI0J01880</td>\n",
       "      <td>NDAI0C03310</td>\n",
       "      <td>NDAI0A05060</td>\n",
       "      <td>NDAI0C03490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tbla_t1</th>\n",
       "      <td>TBLA0A08400</td>\n",
       "      <td>None</td>\n",
       "      <td>TBLA0H01480</td>\n",
       "      <td>TBLA0C06670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tpha_t1</th>\n",
       "      <td>TPHA0F01080</td>\n",
       "      <td>TPHA0A05240</td>\n",
       "      <td>TPHA0A03490</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Vpol_t1</th>\n",
       "      <td>Kpol_472.17</td>\n",
       "      <td>Kpol_1050.78</td>\n",
       "      <td>Kpol_1039.57</td>\n",
       "      <td>Kpol_1067.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Scer_t2</th>\n",
       "      <td>YGR043C</td>\n",
       "      <td>YGR248W</td>\n",
       "      <td>YBR117C</td>\n",
       "      <td>YGR256W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cagl_t2</th>\n",
       "      <td>CAGL0K04235g</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Kafr_t2</th>\n",
       "      <td>KAFR0F03930</td>\n",
       "      <td>KAFR0B04400</td>\n",
       "      <td>KAFR0D02620</td>\n",
       "      <td>KAFR0B04490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Knag_t2</th>\n",
       "      <td>KNAG0D04010</td>\n",
       "      <td>KNAG0K00490</td>\n",
       "      <td>KNAG0B01420</td>\n",
       "      <td>KNAG0K00390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ncas_t2</th>\n",
       "      <td>NCAS0A02150</td>\n",
       "      <td>NCAS0E03860</td>\n",
       "      <td>NCAS0I01030</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ndai_t2</th>\n",
       "      <td>NDAI0D04080</td>\n",
       "      <td>NDAI0B00950</td>\n",
       "      <td>NDAI0A07980</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tbla_t2</th>\n",
       "      <td>None</td>\n",
       "      <td>TBLA0C06580</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tpha_t2</th>\n",
       "      <td>TPHA0J00660</td>\n",
       "      <td>None</td>\n",
       "      <td>TPHA0M01660</td>\n",
       "      <td>TPHA0A05350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Vpol_t2</th>\n",
       "      <td>Kpol_530.51</td>\n",
       "      <td>None</td>\n",
       "      <td>Kpol_1006.8</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Klac</th>\n",
       "      <td>KLLA0A02607g</td>\n",
       "      <td>KLLA0A05390g</td>\n",
       "      <td>KLLA0B09152g</td>\n",
       "      <td>KLLA0A09339g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lthe</th>\n",
       "      <td>KLTH0G17380g</td>\n",
       "      <td>KLTH0H00836g</td>\n",
       "      <td>KLTH0F11836g</td>\n",
       "      <td>KLTH0B08668g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Egos</th>\n",
       "      <td>ACL196W</td>\n",
       "      <td>AEL215C</td>\n",
       "      <td>ADL366W</td>\n",
       "      <td>AFR129W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ecym</th>\n",
       "      <td>Ecym_6140</td>\n",
       "      <td>Ecym_5572</td>\n",
       "      <td>Ecym_3454</td>\n",
       "      <td>Ecym_5108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tdel</th>\n",
       "      <td>TDEL0E01870</td>\n",
       "      <td>TDEL0G01020</td>\n",
       "      <td>TDEL0C05140</td>\n",
       "      <td>TDEL0G00790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zrou</th>\n",
       "      <td>ZYRO0A12914g</td>\n",
       "      <td>ZYRO0C11484g</td>\n",
       "      <td>ZYRO0C16632g</td>\n",
       "      <td>ZYRO0D07876g</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 TAL1          SOL3          TKL1          GND1\n",
       "Scer_t1       YLR354C       YHR163W       YPR074C       YHR183W\n",
       "Cagl_t1  CAGL0B03069g  CAGL0I02200g  CAGL0D01298g  CAGL0M13343g\n",
       "Kafr_t1   KAFR0A06290   KAFR0E02210   KAFR0H00840   KAFR0B01610\n",
       "Knag_t1   KNAG0B05780   KNAG0E00640   KNAG0A07310          None\n",
       "Ncas_t1   NCAS0J01140   NCAS0D01430   NCAS0A11230   NCAS0F02030\n",
       "Ndai_t1   NDAI0J01880   NDAI0C03310   NDAI0A05060   NDAI0C03490\n",
       "Tbla_t1   TBLA0A08400          None   TBLA0H01480   TBLA0C06670\n",
       "Tpha_t1   TPHA0F01080   TPHA0A05240   TPHA0A03490          None\n",
       "Vpol_t1   Kpol_472.17  Kpol_1050.78  Kpol_1039.57   Kpol_1067.2\n",
       "Scer_t2       YGR043C       YGR248W       YBR117C       YGR256W\n",
       "Cagl_t2  CAGL0K04235g          None          None          None\n",
       "Kafr_t2   KAFR0F03930   KAFR0B04400   KAFR0D02620   KAFR0B04490\n",
       "Knag_t2   KNAG0D04010   KNAG0K00490   KNAG0B01420   KNAG0K00390\n",
       "Ncas_t2   NCAS0A02150   NCAS0E03860   NCAS0I01030          None\n",
       "Ndai_t2   NDAI0D04080   NDAI0B00950   NDAI0A07980          None\n",
       "Tbla_t2          None   TBLA0C06580          None          None\n",
       "Tpha_t2   TPHA0J00660          None   TPHA0M01660   TPHA0A05350\n",
       "Vpol_t2   Kpol_530.51          None   Kpol_1006.8          None\n",
       "Klac     KLLA0A02607g  KLLA0A05390g  KLLA0B09152g  KLLA0A09339g\n",
       "Lthe     KLTH0G17380g  KLTH0H00836g  KLTH0F11836g  KLTH0B08668g\n",
       "Egos          ACL196W       AEL215C       ADL366W       AFR129W\n",
       "Ecym        Ecym_6140     Ecym_5572     Ecym_3454     Ecym_5108\n",
       "Tdel      TDEL0E01870   TDEL0G01020   TDEL0C05140   TDEL0G00790\n",
       "Zrou     ZYRO0A12914g  ZYRO0C11484g  ZYRO0C16632g  ZYRO0D07876g"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "#Make dataframe with index \n",
    "ncbi_specs = species_info[species_info['in_ncbi']]['abbreviation']\n",
    "\n",
    "pre_wgd = ['Klac','Lthe', 'Egos','Ecym','Tdel','Zrou']\n",
    "pre_wgd_ncbi = [spec for spec in pre_wgd if spec in set(ncbi_specs)]\n",
    "post_wgd = ['Scer','Cagl','Kafr','Knag','Ncas','Ndai','Tbla','Tpha','Vpol']\n",
    "post_wgd_ncbi = [spec for spec in post_wgd if spec in set(ncbi_specs)]\n",
    "\n",
    "post_wgd_t1 = [spec + '_t1' for spec in post_wgd_ncbi]\n",
    "post_wgd_t2 = [spec + '_t2' for spec in post_wgd_ncbi]\n",
    "\n",
    "spec_index = post_wgd_t1 + post_wgd_t2 + pre_wgd_ncbi\n",
    "\n",
    "sc_genes = ['TAL1', 'SOL3', 'TKL1', 'GND1']\n",
    "\n",
    "orthologs_table = pd.DataFrame(index = spec_index, columns = sc_genes)\n",
    "\n",
    "\n",
    "for sc_gene in sc_genes: \n",
    "    #make a list with a gene or None for each species that we searched in FIMO in the order of the index. \n",
    "    gene_list = []\n",
    "    for track in [track1, track2]: \n",
    "#         print(sc_gene)\n",
    "#         print(track)\n",
    "#         print(track[sc_gene])\n",
    "        track_sc_gene = track[sc_gene]\n",
    "        #keep only the genes in NCBI\n",
    "        track_ncbi = [gene for gene in track_sc_gene if io_library.get_species_from_gene(gene, species_info) in set(ncbi_specs)]\n",
    "        track_ncbi_spec = {io_library.get_species_from_gene(gene, species_info): gene for gene in track_ncbi}\n",
    "    \n",
    "        #for each species in the post_wgd list, try to slot in a gene from the track.  If none is present put in None. \n",
    "        for spec in post_wgd_ncbi:\n",
    "            if spec in track_ncbi_spec.keys(): \n",
    "                gene_list.append(track_ncbi_spec[spec])\n",
    "            else:\n",
    "                gene_list.append(None)\n",
    "    \n",
    "    #add on the species with only one track\n",
    "    \n",
    "    fname = os.path.normpath(data_processing_dir + 'promoter_phylogenies/' + sc_gene + '_tree_ygob.txt')\n",
    "    with open(fname,'r') as f: \n",
    "        gene_phylogeny = f.read()\n",
    "    tree = Tree(gene_phylogeny) \n",
    "    orthologs = tree.get_leaf_names()\n",
    "    \n",
    "    pre_wgd_orthologs = {}\n",
    "    for ortholog in orthologs: \n",
    "        spec = io_library.get_species_from_gene(ortholog,species_info)\n",
    "        if spec in set(pre_wgd_ncbi): \n",
    "            pre_wgd_orthologs[spec] = ortholog        \n",
    "            \n",
    "    for spec in pre_wgd_ncbi: \n",
    "        if spec in pre_wgd_orthologs.keys(): \n",
    "            gene_list.append(pre_wgd_orthologs[spec])\n",
    "        else:\n",
    "            gene_list.append(None)\n",
    "    \n",
    "    orthologs_table[sc_gene] = gene_list\n",
    "\n",
    "\n",
    "orthologs_table\n",
    "        \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AEL215C': 0,\n",
       " 'CAGL0I02200g': 0,\n",
       " 'Ecym_5572': 3,\n",
       " 'KAFR0B04400': 1,\n",
       " 'KAFR0E02210': 0,\n",
       " 'KLLA0A05390g': 0,\n",
       " 'KLTH0H00836g': 2,\n",
       " 'KNAG0E00640': 2,\n",
       " 'KNAG0K00490': 1,\n",
       " 'Kpol_1050.78': 0,\n",
       " 'NCAS0D01430': 0,\n",
       " 'NCAS0E03860': 1,\n",
       " 'NDAI0B00950': 0,\n",
       " 'NDAI0C03310': 4,\n",
       " 'TBLA0C06580': 1,\n",
       " 'TDEL0G01020': 0,\n",
       " 'TPHA0A05240': 0,\n",
       " 'YGR248W': 0,\n",
       " 'YHR163W': 0,\n",
       " 'ZYRO0C11484g': 1}"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "binding_site_lookup = dict(sc_gene_phylogenies['SOL3']['MSN2_Badis'])\n",
    "binding_site_lookup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 7, 0, 2, 0, 3, 1, 0, 0, 1, 0, 0, 3, 1, 1, None, 1, 1, 1, 1, 2, 4, 0, 0]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "binding_site_lookup = dict(sc_gene_phylogenies['SOL3']['MSN2_Badis'])\n",
    "binding_site_list = []\n",
    "for ortholog in orthologs_table['SOL3']: \n",
    "    if ortholog==None: \n",
    "        binding_site_list.append(None)\n",
    "    else: \n",
    "        binding_site_list.append(binding_site_lookup[ortholog])\n",
    "binding_site_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "binding_site_table = pd.DataFrame(index = spec_index, columns = sc_genes)\n",
    "\n",
    "for sc_gene in sc_genes: \n",
    "    # make a dictionary that links name to number of Msn2\n",
    "    # use that to make a new column for a new dataframe with the same index \n",
    "    binding_site_lookup = dict(sc_gene_phylogenies[sc_gene]['MSN2_Badis'])\n",
    "    binding_site_list = []\n",
    "    for ortholog in orthologs_table[sc_gene]: \n",
    "        if ortholog==None: \n",
    "            binding_site_list.append(None)\n",
    "        else: \n",
    "            binding_site_list.append(binding_site_lookup[ortholog])\n",
    "    binding_site_list\n",
    "    binding_site_table[sc_gene] = binding_site_list\n",
    "\n",
    "binding_site_table.to_csv(os.path.normpath(data_processing_dir + 'promoter_phylogenies/PPP_phylogenies_badis_700.csv'))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Use YGOB and Trevors intergenics to make orthologous promoters ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#For a given gene: \n",
    "\n",
    "#Given a species list from the YGOB Pillars\n",
    "\n",
    "    #get set of orthologous genes\n",
    "\n",
    "#make fasta of orthologous promoters. \n",
    "#export file. \n",
    "\n",
    "\n",
    "#count no of STREs, PDS's.  Output surrounding 2 bases and locations\n",
    "#Align and get all conserved regions\n",
    "\n",
    "#visualize list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make a dataframe based on the YGOB pillars file\n",
    "\n",
    "fname_pillars = os.path.normpath(data_processing_dir + 'ortholog_files_YGOB\\YGOB_pillars_bmh.txt')\n",
    "YGOB_pillars = pd.read_table(fname_pillars, header = None)\n",
    "\n",
    "post_wgd = ['Vpol', 'Tpha', 'Tbla', 'Ndai', 'Ncas', 'Knag', 'Kafr', 'Cgla', 'Suva', 'Skud', 'Smik','Scer']\n",
    "post_wgd_p1 = [spec + '_p1' for spec in post_wgd]\n",
    "post_wgd_p2 = [spec + '_p2' for spec in post_wgd]\n",
    "post_wgd_p2.reverse()\n",
    "pre_wgd = ['Zrou', 'Tdel','Klac','Egos','Ecym','Lklu','Lthe','Lwal']\n",
    "columns = post_wgd_p1 + ['Anc'] + pre_wgd + post_wgd_p2\n",
    "\n",
    "YGOB_pillars.columns = columns\n",
    "\n",
    "\n",
    "#    0    V. polyspora Position 1\n",
    "#    1    T. phaffii Position 1\n",
    "#    2    T. blattae Position 1\n",
    "#    3    N. dairenensis Position 1\n",
    "#    4    N. castellii Position 1\n",
    "#    5    K. naganishii Position 1\n",
    "#    6    K. africana Position 1\n",
    "#    7    C. glabrata Position 1\n",
    "#    8    S. bayanus var. uvarum Position 1\n",
    "#    9    S. kudriavzevii Position 1\n",
    "#    10   S. mikatae Position 1\n",
    "#    11   S. cerevisiae Position 1\n",
    "#    12   Ancestral Gene Order\n",
    "#    13   Z. rouxii\n",
    "#    14   T. delbrueckii\n",
    "#    15   K. lactis\n",
    "#    16   E. gossypii \n",
    "#    17   E. cymbalariae\n",
    "#    18   L. kluyveri\n",
    "#    19   L. thermotolerans\n",
    "#    20   L. waltii\n",
    "#    21   S. cerevisiae Position 2\n",
    "#    22   S. mikatae Position 2\n",
    "#    23   S. kudriavzevii Position 2\n",
    "#    24   S. bayanus var. uvarum Position 2\n",
    "#    25   C. glabrata Position 2\n",
    "#    26   K. africana Position 2\n",
    "#    27   K. naganishii Position 2\n",
    "#    28   N. castellii Position 2\n",
    "#    29   N. dairenensis Position 2\n",
    "#    30   T. blattae Position 2\n",
    "#    31   T. phaffii Position 2\n",
    "#    32   V. polyspora Position 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S.Cer orf for YJR115Wnot found\n",
      "S.Cer orf for YHR097Cnot found\n",
      "S.Cer orf for YBR241Cnot found\n",
      "S.Cer orf for YNL194Cnot found\n",
      "S.Cer orf for YMR196Wnot found\n",
      "S.Cer orf for YJL045Wnot found\n",
      "S.Cer orf for YBR139Wnot found\n"
     ]
    }
   ],
   "source": [
    "gois_gr1_m24 = pd.read_csv(data_processing_dir + '20170817_klscpka/20180327_msn24_gis1_rph1_overlap.csv')\n",
    "genes = io_library.SC_orf_lookup_by_name(gois_gr1_m24['SC_common_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HXT5\n",
      "NDE2\n",
      "PIR3\n",
      "GND2\n",
      "RGI1\n",
      "USV1\n",
      "SOL4\n",
      "TSA2\n",
      "DBP1\n",
      "YJR115W\n",
      "YHR097C\n",
      "GND1\n",
      "HSP150\n",
      "NDE1\n",
      "HXT3\n",
      "RGI2\n",
      "DED1\n",
      "PAL1\n",
      "ECM13\n",
      "SOL3\n",
      "RGM1\n",
      "TSA1\n",
      "GSY2\n",
      "YBR241C\n",
      "GLG1\n",
      "MRK1\n",
      "ATG33\n",
      "FMP45\n",
      "TSL1\n",
      "PGM2\n",
      "GSY1\n",
      "YNL194C\n",
      "VPS73\n",
      "RIM11\n",
      "TPS3\n",
      "GLG2\n",
      "SCM4\n",
      "PGM1\n",
      "HSP12\n",
      "GOR1\n",
      "YMR196W\n",
      "MBR1\n",
      "ISF1\n",
      "PEX18\n",
      "SHH4\n",
      "SSE2\n",
      "YJL045W\n",
      "SHH3\n",
      "PEX21\n",
      "SDH3\n",
      "SDH1\n",
      "SDH4\n",
      "SSE1\n",
      "YBR139W\n",
      "SDS24\n",
      "SDS23\n",
      "CYC7\n",
      "HXK1\n",
      "UBI4\n",
      "CTT1\n",
      "GSP2\n",
      "HXK2\n",
      "GSP1\n",
      "CYC1\n",
      "GIS1\n",
      "DOT6\n",
      "RPH1\n",
      "TOD6\n"
     ]
    }
   ],
   "source": [
    "#for a given list of genes:\n",
    "#Searches for exact matches of defined motifs in various sequences\n",
    "motif_dict = {'STRE':'AGGGG',\n",
    "             'Msn2_pref': 'CGGGG', \n",
    "              #'PDS1': 'TTAGGGAT',\n",
    "              #'PDS2': 'TAAGGGAT',\n",
    "              'PDS core': 'AGGGA'\n",
    "             }\n",
    "output_format = 'full'\n",
    "sequence_context = 2\n",
    "\n",
    "output_dict = {}\n",
    "gene_names = io_library.SC_common_name_lookup(genes) #list(gois_gr1_m24['SC_common_name'])\n",
    "for ind, gene in enumerate(genes): \n",
    "    gene_name = gene_names[ind]\n",
    "    print(gene_name)\n",
    "    #Lookup orthologs based on the YGOB Pillars dataframe\n",
    "    gene_pos = ''\n",
    "    for pos in ['p1','p2']: \n",
    "        Scer_pos = YGOB_pillars['Scer_' + pos]\n",
    "        Scer_pos_find = Scer_pos[Scer_pos==gene]\n",
    "        if len(Scer_pos_find)==1:\n",
    "            if gene_pos != '':\n",
    "                print('Warning: found gene in two pillars')\n",
    "            gene_pos = pos\n",
    "            gene_index = Scer_pos_find.index[0]\n",
    "        elif len(Scer_pos_find)>1: \n",
    "            print('Warning: found gene twice in same pillar')\n",
    "\n",
    "    #extract intergenic regions from trevor's file\n",
    "\n",
    "    specs = ['Kafr','Cgla','Suva', 'Skud', 'Smik','Scer']\n",
    "    ##to include pre wgd species would need to update\n",
    "    specs_pillar = [spec + '_' + gene_pos for spec in specs]\n",
    "\n",
    "    orthologs = YGOB_pillars.loc[gene_index, specs_pillar]\n",
    "\n",
    "\n",
    "    seqs = {}\n",
    "\n",
    "    for spec in specs: \n",
    "        fname_intergenics = os.path.normpath(data_processing_dir + 'promoter_phylogenies/intergenics/' + spec + '_intergenics.faa' )\n",
    "        record_iterator = SeqIO.parse(fname_intergenics, \"fasta\")\n",
    "\n",
    "        for seq in record_iterator: \n",
    "            if seq.id == orthologs[spec + '_' + gene_pos]:\n",
    "                seqs[spec] = seq\n",
    "\n",
    "    #initial dataframe has species for index, but reindex on seq id before outputting\n",
    "    seq_ids = [seq.id for seq in seqs.values()]\n",
    "    output_data_frame = pd.DataFrame(index = seq_ids)\n",
    "    output_data_frame['spec'] = seqs.keys()\n",
    "\n",
    "    for motif_name, motif in motif_dict.items():\n",
    "        L_motif = len(motif)\n",
    "        output_motif = []\n",
    "\n",
    "        for spec, seq in seqs.items(): \n",
    "\n",
    "            prom_seq = seq.seq\n",
    "            if len(prom_seq)<no_prom_thresh: \n",
    "                output_motif.append('no_promoter')\n",
    "            else: \n",
    "                output_motif_seq = io_library.exact_promoter_scan(motif, prom_seq, output_format, sequence_context)\n",
    "                output_motif.append(output_motif_seq)\n",
    "        output_data_frame[motif_name] = output_motif\n",
    "\n",
    "    output_dict[gene_name] = output_data_frame\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load my data and Msn2/4 delete data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "full_output = pd.DataFrame()\n",
    "category_dict = dict(zip(gois_gr1_m24['SC_common_name'],gois_gr1_m24['category']))\n",
    "kl_sc_PKA_data_solis = pd.read_csv(data_processing_dir + os.path.normpath('20170817_klscpka/kl_sc_PKA_data_solis.csv'))\n",
    "kl_genename_dict = dict(zip(kl_sc_PKA_data_solis['SC_common_name'],kl_sc_PKA_data_solis['kl_genename']))\n",
    "\n",
    "for gene_name in gois_gr1_m24['SC_common_name']: \n",
    "       \n",
    "    new_output = output_dict[gene_name]\n",
    "    new_output['SC_common_name'] = gene_name\n",
    "    new_output['category'] = category_dict[gene_name]\n",
    "    new_output['kl_genename'] = kl_genename_dict[gene_name]\n",
    "\n",
    "    full_output = pd.concat([full_output, new_output])\n",
    "\n",
    "full_output.to_csv(data_processing_dir + 'promoter_phylogenies/gis1_rph1_STRE_PDS_hits.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_output = pd.read_csv(data_processing_dir + 'promoter_phylogenies/gis1_rph1_STRE_PDS_hits.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>spec</th>\n",
       "      <th>STRE</th>\n",
       "      <th>Msn2_pref</th>\n",
       "      <th>PDS</th>\n",
       "      <th>SC_common_name</th>\n",
       "      <th>category</th>\n",
       "      <th>kl_genename</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAGL0A02321g</td>\n",
       "      <td>Cgla</td>\n",
       "      <td>[(884, 'rev', 'GGAGGGGGG')]</td>\n",
       "      <td>[(691, 'rev', 'CTCGGGGTG')]</td>\n",
       "      <td>[(230, 'rev', 'GGAGGGAGG'), (789, 'rev', 'AGAG...</td>\n",
       "      <td>HXT5</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0D13310g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Suva_15.291</td>\n",
       "      <td>Suva</td>\n",
       "      <td>[(635, 'fwd', 'CGAGGGGAA'), (516, 'fwd', 'TAAG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(623, 'fwd', 'GAAGGGAAA')]</td>\n",
       "      <td>HXT5</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0D13310g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Smik_8.173</td>\n",
       "      <td>Smik</td>\n",
       "      <td>[(465, 'fwd', 'TAAGGGGAC'), (295, 'fwd', 'TTAG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(668, 'fwd', 'AAAGGGAAA'), (535, 'fwd', 'CAAG...</td>\n",
       "      <td>HXT5</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0D13310g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YHR096C</td>\n",
       "      <td>Scer</td>\n",
       "      <td>[(471, 'fwd', 'TTAGGGGAC'), (303, 'fwd', 'TTAG...</td>\n",
       "      <td>[(496, 'rev', 'CACGGGGAA')]</td>\n",
       "      <td>[(541, 'fwd', 'TAAGGGATG'), (184, 'rev', 'AGAG...</td>\n",
       "      <td>HXT5</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0D13310g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KAFR0B05060</td>\n",
       "      <td>Kafr</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(69, 'rev', 'ACAGGGACG')]</td>\n",
       "      <td>NDE2</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0E21891g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CAGL0I00748g</td>\n",
       "      <td>Cgla</td>\n",
       "      <td>[(356, 'rev', 'GCAGGGGAG')]</td>\n",
       "      <td>[(136, 'rev', 'TTCGGGGGG'), (224, 'rev', 'TGCG...</td>\n",
       "      <td>[(237, 'rev', 'GTAGGGAGC'), (489, 'rev', 'GCAG...</td>\n",
       "      <td>NDE2</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0E21891g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Suva_4.160</td>\n",
       "      <td>Suva</td>\n",
       "      <td>[(137, 'rev', 'AAAGGGGAG'), (142, 'rev', 'GGAG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(181, 'fwd', 'AAAGGGACT')]</td>\n",
       "      <td>NDE2</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0E21891g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Skud_13.306</td>\n",
       "      <td>Skud</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(570, 'rev', 'GACGGGGTC'), (630, 'rev', 'GGCG...</td>\n",
       "      <td>[(520, 'fwd', 'ACAGGGATC'), (99, 'fwd', 'AGAGG...</td>\n",
       "      <td>NDE2</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0E21891g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Smik_4.151</td>\n",
       "      <td>Smik</td>\n",
       "      <td>[(157, 'rev', 'GGAGGGGCT'), (172, 'rev', 'TTAG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(153, 'rev', 'AAAGGGAGG'), (232, 'rev', 'TTAG...</td>\n",
       "      <td>NDE2</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0E21891g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>YDL085W</td>\n",
       "      <td>Scer</td>\n",
       "      <td>[(143, 'rev', 'GGAGGGGCT'), (158, 'rev', 'TTAG...</td>\n",
       "      <td>[(63, 'fwd', 'TGCGGGGAT')]</td>\n",
       "      <td>[(139, 'rev', 'AAAGGGAGG')]</td>\n",
       "      <td>NDE2</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0E21891g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CAGL0M08514g</td>\n",
       "      <td>Cgla</td>\n",
       "      <td>[(301, 'rev', 'CGAGGGGAA'), (1260, 'rev', 'GGA...</td>\n",
       "      <td>[(161, 'rev', 'GACGGGGAT'), (1252, 'rev', 'GTC...</td>\n",
       "      <td>[(1178, 'fwd', 'GTAGGGAAG'), (1173, 'fwd', 'GA...</td>\n",
       "      <td>PIR3</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0B07392g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Suva_11.61</td>\n",
       "      <td>Suva</td>\n",
       "      <td>[(826, 'fwd', 'TTAGGGGCG'), (308, 'fwd', 'AGAG...</td>\n",
       "      <td>[(855, 'rev', 'TTCGGGGCC')]</td>\n",
       "      <td>[(767, 'fwd', 'AAAGGGATA'), (197, 'rev', 'CCAG...</td>\n",
       "      <td>PIR3</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0B07392g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Skud_11.63</td>\n",
       "      <td>Skud</td>\n",
       "      <td>[(309, 'fwd', 'AGAGGGGAT'), (1229, 'rev', 'ATA...</td>\n",
       "      <td>[(381, 'rev', 'CGCGGGGCC')]</td>\n",
       "      <td>[(872, 'fwd', 'CAAGGGAGG'), (221, 'fwd', 'GAAG...</td>\n",
       "      <td>PIR3</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0B07392g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Smik_10.74</td>\n",
       "      <td>Smik</td>\n",
       "      <td>[(508, 'rev', 'GAAGGGGGG')]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(1308, 'fwd', 'CAAGGGAAC'), (442, 'fwd', 'AGA...</td>\n",
       "      <td>PIR3</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0B07392g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>YKL163W</td>\n",
       "      <td>Scer</td>\n",
       "      <td>[(1562, 'fwd', 'TTAGGGGAG'), (1037, 'fwd', 'AA...</td>\n",
       "      <td>[(373, 'rev', 'CGCGGGGCC')]</td>\n",
       "      <td>[(214, 'fwd', 'GAAGGGATA'), (192, 'rev', 'GTAG...</td>\n",
       "      <td>PIR3</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0B07392g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>KAFR0B01610</td>\n",
       "      <td>Kafr</td>\n",
       "      <td>[(633, 'rev', 'ATAGGGGCC')]</td>\n",
       "      <td>[(1076, 'fwd', 'GCCGGGGTT'), (535, 'rev', 'TTC...</td>\n",
       "      <td>[(1592, 'fwd', 'GTAGGGATC'), (1570, 'fwd', 'TG...</td>\n",
       "      <td>GND2</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0A09339g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CAGL0M13343g</td>\n",
       "      <td>Cgla</td>\n",
       "      <td>[(1525, 'fwd', 'TAAGGGGGC'), (280, 'fwd', 'ATA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(1262, 'fwd', 'GTAGGGAAT'), (189, 'fwd', 'ATA...</td>\n",
       "      <td>GND2</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0A09339g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Suva_7.554</td>\n",
       "      <td>Suva</td>\n",
       "      <td>[(161, 'rev', 'TTAGGGGGG'), (329, 'rev', 'TTAG...</td>\n",
       "      <td>[(218, 'rev', 'TTCGGGGTT')]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GND2</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0A09339g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Skud_7.590</td>\n",
       "      <td>Skud</td>\n",
       "      <td>[(150, 'rev', 'GGAGGGGGT'), (167, 'rev', 'ATAG...</td>\n",
       "      <td>[(219, 'rev', 'TTCGGGGTT')]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GND2</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0A09339g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Smik_8.271</td>\n",
       "      <td>Smik</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(334, 'rev', 'GGAGGGAAG')]</td>\n",
       "      <td>GND2</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0A09339g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>YGR256W</td>\n",
       "      <td>Scer</td>\n",
       "      <td>[(169, 'rev', 'ATAGGGGTG'), (304, 'rev', 'TTAG...</td>\n",
       "      <td>[(218, 'rev', 'TTCGGGGTT')]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GND2</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0A09339g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>KAFR0L00950</td>\n",
       "      <td>Kafr</td>\n",
       "      <td>[(285, 'fwd', 'TCAGGGGGG'), (395, 'rev', 'CCAG...</td>\n",
       "      <td>[(456, 'rev', 'TTCGGGGGG'), (596, 'rev', 'GCCG...</td>\n",
       "      <td>[(56, 'rev', 'TAAGGGACG'), (851, 'rev', 'ATAGG...</td>\n",
       "      <td>RGI1</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0C08063g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CAGL0J03080g</td>\n",
       "      <td>Cgla</td>\n",
       "      <td>[(422, 'fwd', 'TCAGGGGGG'), (380, 'fwd', 'CCAG...</td>\n",
       "      <td>[(1152, 'fwd', 'ACCGGGGTG'), (307, 'rev', 'TCC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RGI1</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0C08063g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Suva_5.179</td>\n",
       "      <td>Suva</td>\n",
       "      <td>[(1014, 'fwd', 'CCAGGGGAT'), (428, 'fwd', 'TCA...</td>\n",
       "      <td>[(444, 'rev', 'TCCGGGGTA')]</td>\n",
       "      <td>[(1187, 'fwd', 'AAAGGGAGA')]</td>\n",
       "      <td>RGI1</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0C08063g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Skud_9.112</td>\n",
       "      <td>Skud</td>\n",
       "      <td>[(241, 'fwd', 'AAAGGGGAG')]</td>\n",
       "      <td>[(222, 'fwd', 'CGCGGGGTT')]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RGI1</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0C08063g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Smik_9.132</td>\n",
       "      <td>Smik</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(299, 'fwd', 'TGCGGGGTA')]</td>\n",
       "      <td>[(319, 'fwd', 'AAAGGGATA')]</td>\n",
       "      <td>RGI1</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0C08063g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>YER067W</td>\n",
       "      <td>Scer</td>\n",
       "      <td>no_promoter</td>\n",
       "      <td>no_promoter</td>\n",
       "      <td>no_promoter</td>\n",
       "      <td>RGI1</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "      <td>KLLA0C08063g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>KAFR0F01210</td>\n",
       "      <td>Kafr</td>\n",
       "      <td>[(224, 'fwd', 'TAAGGGGAA'), (673, 'rev', 'CCAG...</td>\n",
       "      <td>[(893, 'fwd', 'CCCGGGGTA')]</td>\n",
       "      <td>[(560, 'fwd', 'AGAGGGAAA'), (182, 'rev', 'CAAG...</td>\n",
       "      <td>USV1</td>\n",
       "      <td>g1r1_m24_klsc_para</td>\n",
       "      <td>KLLA0C16005g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>CAGL0L00583g</td>\n",
       "      <td>Cgla</td>\n",
       "      <td>[(3494, 'fwd', 'CTAGGGGAA'), (3030, 'fwd', 'AG...</td>\n",
       "      <td>[(1905, 'rev', 'TGCGGGGAT'), (3012, 'rev', 'AG...</td>\n",
       "      <td>[(5217, 'fwd', 'TCAGGGATA'), (4375, 'fwd', 'AA...</td>\n",
       "      <td>USV1</td>\n",
       "      <td>g1r1_m24_klsc_para</td>\n",
       "      <td>KLLA0C16005g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Suva_16.76</td>\n",
       "      <td>Suva</td>\n",
       "      <td>[(617, 'fwd', 'GTAGGGGGC'), (549, 'fwd', 'CAAG...</td>\n",
       "      <td>[(721, 'fwd', 'TCCGGGGTA'), (163, 'rev', 'TTCG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USV1</td>\n",
       "      <td>g1r1_m24_klsc_para</td>\n",
       "      <td>KLLA0C16005g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367</th>\n",
       "      <td>Smik_15.370</td>\n",
       "      <td>Smik</td>\n",
       "      <td>[(476, 'fwd', 'CAAGGGGTC'), (447, 'fwd', 'TAAG...</td>\n",
       "      <td>[(326, 'rev', 'ATCGGGGTG')]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSP1</td>\n",
       "      <td>g1r1_m24wtdrop_para_low</td>\n",
       "      <td>KLLA0C05126g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368</th>\n",
       "      <td>YLR293C</td>\n",
       "      <td>Scer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(67, 'rev', 'TGAGGGAAA')]</td>\n",
       "      <td>GSP1</td>\n",
       "      <td>g1r1_m24wtdrop_para_low</td>\n",
       "      <td>KLLA0C05126g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369</th>\n",
       "      <td>Suva_12.132</td>\n",
       "      <td>Suva</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(421, 'rev', 'CCCGGGGAT')]</td>\n",
       "      <td>[(911, 'fwd', 'GAAGGGAAT')]</td>\n",
       "      <td>CYC1</td>\n",
       "      <td>g1r1_m24wtdrop_para_low</td>\n",
       "      <td>KLLA0F16929g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>Skud_5.51</td>\n",
       "      <td>Skud</td>\n",
       "      <td>[(293, 'fwd', 'CGAGGGGTC'), (209, 'fwd', 'CAAG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(463, 'rev', 'ACAGGGAAT')]</td>\n",
       "      <td>CYC1</td>\n",
       "      <td>g1r1_m24wtdrop_para_low</td>\n",
       "      <td>KLLA0F16929g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>Smik_5.54</td>\n",
       "      <td>Smik</td>\n",
       "      <td>[(277, 'fwd', 'CGAGGGGTC'), (193, 'fwd', 'CAAG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CYC1</td>\n",
       "      <td>g1r1_m24wtdrop_para_low</td>\n",
       "      <td>KLLA0F16929g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372</th>\n",
       "      <td>YJR048W</td>\n",
       "      <td>Scer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(336, 'fwd', 'GCCGGGGTT')]</td>\n",
       "      <td>[(886, 'fwd', 'TAAGGGAAC')]</td>\n",
       "      <td>CYC1</td>\n",
       "      <td>g1r1_m24wtdrop_para_low</td>\n",
       "      <td>KLLA0F16929g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>KAFR0A02380</td>\n",
       "      <td>Kafr</td>\n",
       "      <td>[(1050, 'rev', 'GGAGGGGAA'), (1158, 'rev', 'GG...</td>\n",
       "      <td>[(1143, 'fwd', 'ATCGGGGAA'), (671, 'rev', 'GAC...</td>\n",
       "      <td>[(1339, 'fwd', 'GGAGGGAAG'), (826, 'fwd', 'GCA...</td>\n",
       "      <td>GIS1</td>\n",
       "      <td>other</td>\n",
       "      <td>KLLA0C17710g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>CAGL0L11880g</td>\n",
       "      <td>Cgla</td>\n",
       "      <td>[(194, 'fwd', 'TAAGGGGAG'), (161, 'fwd', 'TTAG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GIS1</td>\n",
       "      <td>other</td>\n",
       "      <td>KLLA0C17710g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>Suva_5.303</td>\n",
       "      <td>Suva</td>\n",
       "      <td>[(169, 'fwd', 'TCAGGGGTC'), (129, 'fwd', 'ATAG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GIS1</td>\n",
       "      <td>other</td>\n",
       "      <td>KLLA0C17710g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>Skud_4.354</td>\n",
       "      <td>Skud</td>\n",
       "      <td>[(851, 'fwd', 'GAAGGGGCT'), (754, 'fwd', 'TGAG...</td>\n",
       "      <td>[(890, 'fwd', 'TGCGGGGTA'), (319, 'rev', 'AGCG...</td>\n",
       "      <td>[(576, 'rev', 'CAAGGGACC')]</td>\n",
       "      <td>GIS1</td>\n",
       "      <td>other</td>\n",
       "      <td>KLLA0C17710g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>Smik_4.337</td>\n",
       "      <td>Smik</td>\n",
       "      <td>[(851, 'fwd', 'AAAGGGGAG'), (758, 'fwd', 'TGAG...</td>\n",
       "      <td>[(824, 'fwd', 'GCCGGGGTT'), (822, 'rev', 'GCCG...</td>\n",
       "      <td>[(585, 'rev', 'CAAGGGACC')]</td>\n",
       "      <td>GIS1</td>\n",
       "      <td>other</td>\n",
       "      <td>KLLA0C17710g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>YDR096W</td>\n",
       "      <td>Scer</td>\n",
       "      <td>[(331, 'rev', 'GGAGGGGGG')]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(600, 'rev', 'CAAGGGACC')]</td>\n",
       "      <td>GIS1</td>\n",
       "      <td>other</td>\n",
       "      <td>KLLA0C17710g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>KAFR0L01620</td>\n",
       "      <td>Kafr</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(789, 'fwd', 'TCAGGGAGA'), (137, 'rev', 'TGAG...</td>\n",
       "      <td>DOT6</td>\n",
       "      <td>other</td>\n",
       "      <td>KLLA0E19933g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>CAGL0A04257g</td>\n",
       "      <td>Cgla</td>\n",
       "      <td>[(758, 'fwd', 'GTAGGGGAG'), (364, 'fwd', 'GGAG...</td>\n",
       "      <td>[(443, 'rev', 'CGCGGGGTA')]</td>\n",
       "      <td>[(353, 'fwd', 'AGAGGGACA')]</td>\n",
       "      <td>DOT6</td>\n",
       "      <td>other</td>\n",
       "      <td>KLLA0E19933g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>Suva_2.63</td>\n",
       "      <td>Suva</td>\n",
       "      <td>[(607, 'fwd', 'CGAGGGGCT')]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(661, 'fwd', 'ACAGGGACG'), (491, 'fwd', 'GCAG...</td>\n",
       "      <td>DOT6</td>\n",
       "      <td>other</td>\n",
       "      <td>KLLA0E19933g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>Skud_5.208</td>\n",
       "      <td>Skud</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(460, 'fwd', 'GGCGGGGGC'), (443, 'rev', 'GACG...</td>\n",
       "      <td>[(18, 'rev', 'GCAGGGAAG'), (23, 'rev', 'GAAGGG...</td>\n",
       "      <td>DOT6</td>\n",
       "      <td>other</td>\n",
       "      <td>KLLA0E19933g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383</th>\n",
       "      <td>Smik_5.228</td>\n",
       "      <td>Smik</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(476, 'rev', 'GACGGGGGC'), (507, 'rev', 'GCCG...</td>\n",
       "      <td>[(23, 'rev', 'GAAGGGAGA'), (31, 'rev', 'ACAGGG...</td>\n",
       "      <td>DOT6</td>\n",
       "      <td>other</td>\n",
       "      <td>KLLA0E19933g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384</th>\n",
       "      <td>YER088C</td>\n",
       "      <td>Scer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(500, 'rev', 'AACGGGGCC')]</td>\n",
       "      <td>[(200, 'fwd', 'ACAGGGACA'), (23, 'rev', 'GAAGG...</td>\n",
       "      <td>DOT6</td>\n",
       "      <td>other</td>\n",
       "      <td>KLLA0E19933g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>385</th>\n",
       "      <td>KAFR0B02550</td>\n",
       "      <td>Kafr</td>\n",
       "      <td>[(520, 'fwd', 'AAAGGGGAG'), (336, 'rev', 'GAAG...</td>\n",
       "      <td>[(916, 'rev', 'TACGGGGCA')]</td>\n",
       "      <td>[(595, 'fwd', 'TAAGGGAAT')]</td>\n",
       "      <td>RPH1</td>\n",
       "      <td>other_para_low</td>\n",
       "      <td>KLLA0C17710g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386</th>\n",
       "      <td>CAGL0G08107g</td>\n",
       "      <td>Cgla</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(263, 'fwd', 'AAAGGGAAT')]</td>\n",
       "      <td>RPH1</td>\n",
       "      <td>other_para_low</td>\n",
       "      <td>KLLA0C17710g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387</th>\n",
       "      <td>Suva_2.255</td>\n",
       "      <td>Suva</td>\n",
       "      <td>[(777, 'fwd', 'AGAGGGGAC'), (733, 'fwd', 'TGAG...</td>\n",
       "      <td>[(762, 'fwd', 'GCCGGGGTG')]</td>\n",
       "      <td>[(696, 'fwd', 'GAAGGGATG'), (17, 'fwd', 'ACAGG...</td>\n",
       "      <td>RPH1</td>\n",
       "      <td>other_para_low</td>\n",
       "      <td>KLLA0C17710g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388</th>\n",
       "      <td>Skud_5.308</td>\n",
       "      <td>Skud</td>\n",
       "      <td>[(126, 'fwd', 'ATAGGGGGA')]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(307, 'rev', 'TTAGGGAGT')]</td>\n",
       "      <td>RPH1</td>\n",
       "      <td>other_para_low</td>\n",
       "      <td>KLLA0C17710g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389</th>\n",
       "      <td>Smik_5.341</td>\n",
       "      <td>Smik</td>\n",
       "      <td>[(115, 'fwd', 'ATAGGGGGA'), (406, 'rev', 'TTAG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RPH1</td>\n",
       "      <td>other_para_low</td>\n",
       "      <td>KLLA0C17710g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>390</th>\n",
       "      <td>YER169W</td>\n",
       "      <td>Scer</td>\n",
       "      <td>[(125, 'fwd', 'ATAGGGGGA'), (431, 'rev', 'TTAG...</td>\n",
       "      <td>[(474, 'rev', 'GGCGGGGGA')]</td>\n",
       "      <td>[(654, 'fwd', 'CGAGGGAAG'), (52, 'fwd', 'AAAGG...</td>\n",
       "      <td>RPH1</td>\n",
       "      <td>other_para_low</td>\n",
       "      <td>KLLA0C17710g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>391</th>\n",
       "      <td>KAFR0K01720</td>\n",
       "      <td>Kafr</td>\n",
       "      <td>[(588, 'fwd', 'GAAGGGGTA')]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(1009, 'fwd', 'GCAGGGAGG'), (353, 'rev', 'AGA...</td>\n",
       "      <td>TOD6</td>\n",
       "      <td>other_para_low</td>\n",
       "      <td>KLLA0E19933g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392</th>\n",
       "      <td>CAGL0I05060g</td>\n",
       "      <td>Cgla</td>\n",
       "      <td>[(2732, 'fwd', 'AAAGGGGAT'), (547, 'fwd', 'GAA...</td>\n",
       "      <td>[(1278, 'fwd', 'TTCGGGGCA'), (1321, 'rev', 'TC...</td>\n",
       "      <td>[(5047, 'fwd', 'ATAGGGATT'), (4478, 'fwd', 'TG...</td>\n",
       "      <td>TOD6</td>\n",
       "      <td>other_para_low</td>\n",
       "      <td>KLLA0E19933g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>393</th>\n",
       "      <td>Suva_5.202</td>\n",
       "      <td>Suva</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOD6</td>\n",
       "      <td>other_para_low</td>\n",
       "      <td>KLLA0E19933g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394</th>\n",
       "      <td>Skud_2.50</td>\n",
       "      <td>Skud</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(595, 'fwd', 'GACGGGGTC')]</td>\n",
       "      <td>[(480, 'fwd', 'GCAGGGATT'), (478, 'rev', 'CCAG...</td>\n",
       "      <td>TOD6</td>\n",
       "      <td>other_para_low</td>\n",
       "      <td>KLLA0E19933g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>Smik_2.62</td>\n",
       "      <td>Smik</td>\n",
       "      <td>[(576, 'fwd', 'TGAGGGGCT'), (609, 'rev', 'TAAG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TOD6</td>\n",
       "      <td>other_para_low</td>\n",
       "      <td>KLLA0E19933g</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>396</th>\n",
       "      <td>YBL054W</td>\n",
       "      <td>Scer</td>\n",
       "      <td>[(600, 'fwd', 'GTAGGGGCT')]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[(14, 'rev', 'CGAGGGAAC')]</td>\n",
       "      <td>TOD6</td>\n",
       "      <td>other_para_low</td>\n",
       "      <td>KLLA0E19933g</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>397 rows  8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0  spec                                               STRE  \\\n",
       "0    CAGL0A02321g  Cgla                        [(884, 'rev', 'GGAGGGGGG')]   \n",
       "1     Suva_15.291  Suva  [(635, 'fwd', 'CGAGGGGAA'), (516, 'fwd', 'TAAG...   \n",
       "2      Smik_8.173  Smik  [(465, 'fwd', 'TAAGGGGAC'), (295, 'fwd', 'TTAG...   \n",
       "3         YHR096C  Scer  [(471, 'fwd', 'TTAGGGGAC'), (303, 'fwd', 'TTAG...   \n",
       "4     KAFR0B05060  Kafr                                                NaN   \n",
       "5    CAGL0I00748g  Cgla                        [(356, 'rev', 'GCAGGGGAG')]   \n",
       "6      Suva_4.160  Suva  [(137, 'rev', 'AAAGGGGAG'), (142, 'rev', 'GGAG...   \n",
       "7     Skud_13.306  Skud                                                NaN   \n",
       "8      Smik_4.151  Smik  [(157, 'rev', 'GGAGGGGCT'), (172, 'rev', 'TTAG...   \n",
       "9         YDL085W  Scer  [(143, 'rev', 'GGAGGGGCT'), (158, 'rev', 'TTAG...   \n",
       "10   CAGL0M08514g  Cgla  [(301, 'rev', 'CGAGGGGAA'), (1260, 'rev', 'GGA...   \n",
       "11     Suva_11.61  Suva  [(826, 'fwd', 'TTAGGGGCG'), (308, 'fwd', 'AGAG...   \n",
       "12     Skud_11.63  Skud  [(309, 'fwd', 'AGAGGGGAT'), (1229, 'rev', 'ATA...   \n",
       "13     Smik_10.74  Smik                        [(508, 'rev', 'GAAGGGGGG')]   \n",
       "14        YKL163W  Scer  [(1562, 'fwd', 'TTAGGGGAG'), (1037, 'fwd', 'AA...   \n",
       "15    KAFR0B01610  Kafr                        [(633, 'rev', 'ATAGGGGCC')]   \n",
       "16   CAGL0M13343g  Cgla  [(1525, 'fwd', 'TAAGGGGGC'), (280, 'fwd', 'ATA...   \n",
       "17     Suva_7.554  Suva  [(161, 'rev', 'TTAGGGGGG'), (329, 'rev', 'TTAG...   \n",
       "18     Skud_7.590  Skud  [(150, 'rev', 'GGAGGGGGT'), (167, 'rev', 'ATAG...   \n",
       "19     Smik_8.271  Smik                                                NaN   \n",
       "20        YGR256W  Scer  [(169, 'rev', 'ATAGGGGTG'), (304, 'rev', 'TTAG...   \n",
       "21    KAFR0L00950  Kafr  [(285, 'fwd', 'TCAGGGGGG'), (395, 'rev', 'CCAG...   \n",
       "22   CAGL0J03080g  Cgla  [(422, 'fwd', 'TCAGGGGGG'), (380, 'fwd', 'CCAG...   \n",
       "23     Suva_5.179  Suva  [(1014, 'fwd', 'CCAGGGGAT'), (428, 'fwd', 'TCA...   \n",
       "24     Skud_9.112  Skud                        [(241, 'fwd', 'AAAGGGGAG')]   \n",
       "25     Smik_9.132  Smik                                                NaN   \n",
       "26        YER067W  Scer                                        no_promoter   \n",
       "27    KAFR0F01210  Kafr  [(224, 'fwd', 'TAAGGGGAA'), (673, 'rev', 'CCAG...   \n",
       "28   CAGL0L00583g  Cgla  [(3494, 'fwd', 'CTAGGGGAA'), (3030, 'fwd', 'AG...   \n",
       "29     Suva_16.76  Suva  [(617, 'fwd', 'GTAGGGGGC'), (549, 'fwd', 'CAAG...   \n",
       "..            ...   ...                                                ...   \n",
       "367   Smik_15.370  Smik  [(476, 'fwd', 'CAAGGGGTC'), (447, 'fwd', 'TAAG...   \n",
       "368       YLR293C  Scer                                                NaN   \n",
       "369   Suva_12.132  Suva                                                NaN   \n",
       "370     Skud_5.51  Skud  [(293, 'fwd', 'CGAGGGGTC'), (209, 'fwd', 'CAAG...   \n",
       "371     Smik_5.54  Smik  [(277, 'fwd', 'CGAGGGGTC'), (193, 'fwd', 'CAAG...   \n",
       "372       YJR048W  Scer                                                NaN   \n",
       "373   KAFR0A02380  Kafr  [(1050, 'rev', 'GGAGGGGAA'), (1158, 'rev', 'GG...   \n",
       "374  CAGL0L11880g  Cgla  [(194, 'fwd', 'TAAGGGGAG'), (161, 'fwd', 'TTAG...   \n",
       "375    Suva_5.303  Suva  [(169, 'fwd', 'TCAGGGGTC'), (129, 'fwd', 'ATAG...   \n",
       "376    Skud_4.354  Skud  [(851, 'fwd', 'GAAGGGGCT'), (754, 'fwd', 'TGAG...   \n",
       "377    Smik_4.337  Smik  [(851, 'fwd', 'AAAGGGGAG'), (758, 'fwd', 'TGAG...   \n",
       "378       YDR096W  Scer                        [(331, 'rev', 'GGAGGGGGG')]   \n",
       "379   KAFR0L01620  Kafr                                                NaN   \n",
       "380  CAGL0A04257g  Cgla  [(758, 'fwd', 'GTAGGGGAG'), (364, 'fwd', 'GGAG...   \n",
       "381     Suva_2.63  Suva                        [(607, 'fwd', 'CGAGGGGCT')]   \n",
       "382    Skud_5.208  Skud                                                NaN   \n",
       "383    Smik_5.228  Smik                                                NaN   \n",
       "384       YER088C  Scer                                                NaN   \n",
       "385   KAFR0B02550  Kafr  [(520, 'fwd', 'AAAGGGGAG'), (336, 'rev', 'GAAG...   \n",
       "386  CAGL0G08107g  Cgla                                                NaN   \n",
       "387    Suva_2.255  Suva  [(777, 'fwd', 'AGAGGGGAC'), (733, 'fwd', 'TGAG...   \n",
       "388    Skud_5.308  Skud                        [(126, 'fwd', 'ATAGGGGGA')]   \n",
       "389    Smik_5.341  Smik  [(115, 'fwd', 'ATAGGGGGA'), (406, 'rev', 'TTAG...   \n",
       "390       YER169W  Scer  [(125, 'fwd', 'ATAGGGGGA'), (431, 'rev', 'TTAG...   \n",
       "391   KAFR0K01720  Kafr                        [(588, 'fwd', 'GAAGGGGTA')]   \n",
       "392  CAGL0I05060g  Cgla  [(2732, 'fwd', 'AAAGGGGAT'), (547, 'fwd', 'GAA...   \n",
       "393    Suva_5.202  Suva                                                NaN   \n",
       "394     Skud_2.50  Skud                                                NaN   \n",
       "395     Smik_2.62  Smik  [(576, 'fwd', 'TGAGGGGCT'), (609, 'rev', 'TAAG...   \n",
       "396       YBL054W  Scer                        [(600, 'fwd', 'GTAGGGGCT')]   \n",
       "\n",
       "                                             Msn2_pref  \\\n",
       "0                          [(691, 'rev', 'CTCGGGGTG')]   \n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                          [(496, 'rev', 'CACGGGGAA')]   \n",
       "4                                                  NaN   \n",
       "5    [(136, 'rev', 'TTCGGGGGG'), (224, 'rev', 'TGCG...   \n",
       "6                                                  NaN   \n",
       "7    [(570, 'rev', 'GACGGGGTC'), (630, 'rev', 'GGCG...   \n",
       "8                                                  NaN   \n",
       "9                           [(63, 'fwd', 'TGCGGGGAT')]   \n",
       "10   [(161, 'rev', 'GACGGGGAT'), (1252, 'rev', 'GTC...   \n",
       "11                         [(855, 'rev', 'TTCGGGGCC')]   \n",
       "12                         [(381, 'rev', 'CGCGGGGCC')]   \n",
       "13                                                 NaN   \n",
       "14                         [(373, 'rev', 'CGCGGGGCC')]   \n",
       "15   [(1076, 'fwd', 'GCCGGGGTT'), (535, 'rev', 'TTC...   \n",
       "16                                                 NaN   \n",
       "17                         [(218, 'rev', 'TTCGGGGTT')]   \n",
       "18                         [(219, 'rev', 'TTCGGGGTT')]   \n",
       "19                                                 NaN   \n",
       "20                         [(218, 'rev', 'TTCGGGGTT')]   \n",
       "21   [(456, 'rev', 'TTCGGGGGG'), (596, 'rev', 'GCCG...   \n",
       "22   [(1152, 'fwd', 'ACCGGGGTG'), (307, 'rev', 'TCC...   \n",
       "23                         [(444, 'rev', 'TCCGGGGTA')]   \n",
       "24                         [(222, 'fwd', 'CGCGGGGTT')]   \n",
       "25                         [(299, 'fwd', 'TGCGGGGTA')]   \n",
       "26                                         no_promoter   \n",
       "27                         [(893, 'fwd', 'CCCGGGGTA')]   \n",
       "28   [(1905, 'rev', 'TGCGGGGAT'), (3012, 'rev', 'AG...   \n",
       "29   [(721, 'fwd', 'TCCGGGGTA'), (163, 'rev', 'TTCG...   \n",
       "..                                                 ...   \n",
       "367                        [(326, 'rev', 'ATCGGGGTG')]   \n",
       "368                                                NaN   \n",
       "369                        [(421, 'rev', 'CCCGGGGAT')]   \n",
       "370                                                NaN   \n",
       "371                                                NaN   \n",
       "372                        [(336, 'fwd', 'GCCGGGGTT')]   \n",
       "373  [(1143, 'fwd', 'ATCGGGGAA'), (671, 'rev', 'GAC...   \n",
       "374                                                NaN   \n",
       "375                                                NaN   \n",
       "376  [(890, 'fwd', 'TGCGGGGTA'), (319, 'rev', 'AGCG...   \n",
       "377  [(824, 'fwd', 'GCCGGGGTT'), (822, 'rev', 'GCCG...   \n",
       "378                                                NaN   \n",
       "379                                                NaN   \n",
       "380                        [(443, 'rev', 'CGCGGGGTA')]   \n",
       "381                                                NaN   \n",
       "382  [(460, 'fwd', 'GGCGGGGGC'), (443, 'rev', 'GACG...   \n",
       "383  [(476, 'rev', 'GACGGGGGC'), (507, 'rev', 'GCCG...   \n",
       "384                        [(500, 'rev', 'AACGGGGCC')]   \n",
       "385                        [(916, 'rev', 'TACGGGGCA')]   \n",
       "386                                                NaN   \n",
       "387                        [(762, 'fwd', 'GCCGGGGTG')]   \n",
       "388                                                NaN   \n",
       "389                                                NaN   \n",
       "390                        [(474, 'rev', 'GGCGGGGGA')]   \n",
       "391                                                NaN   \n",
       "392  [(1278, 'fwd', 'TTCGGGGCA'), (1321, 'rev', 'TC...   \n",
       "393                                                NaN   \n",
       "394                        [(595, 'fwd', 'GACGGGGTC')]   \n",
       "395                                                NaN   \n",
       "396                                                NaN   \n",
       "\n",
       "                                                   PDS SC_common_name  \\\n",
       "0    [(230, 'rev', 'GGAGGGAGG'), (789, 'rev', 'AGAG...           HXT5   \n",
       "1                          [(623, 'fwd', 'GAAGGGAAA')]           HXT5   \n",
       "2    [(668, 'fwd', 'AAAGGGAAA'), (535, 'fwd', 'CAAG...           HXT5   \n",
       "3    [(541, 'fwd', 'TAAGGGATG'), (184, 'rev', 'AGAG...           HXT5   \n",
       "4                           [(69, 'rev', 'ACAGGGACG')]           NDE2   \n",
       "5    [(237, 'rev', 'GTAGGGAGC'), (489, 'rev', 'GCAG...           NDE2   \n",
       "6                          [(181, 'fwd', 'AAAGGGACT')]           NDE2   \n",
       "7    [(520, 'fwd', 'ACAGGGATC'), (99, 'fwd', 'AGAGG...           NDE2   \n",
       "8    [(153, 'rev', 'AAAGGGAGG'), (232, 'rev', 'TTAG...           NDE2   \n",
       "9                          [(139, 'rev', 'AAAGGGAGG')]           NDE2   \n",
       "10   [(1178, 'fwd', 'GTAGGGAAG'), (1173, 'fwd', 'GA...           PIR3   \n",
       "11   [(767, 'fwd', 'AAAGGGATA'), (197, 'rev', 'CCAG...           PIR3   \n",
       "12   [(872, 'fwd', 'CAAGGGAGG'), (221, 'fwd', 'GAAG...           PIR3   \n",
       "13   [(1308, 'fwd', 'CAAGGGAAC'), (442, 'fwd', 'AGA...           PIR3   \n",
       "14   [(214, 'fwd', 'GAAGGGATA'), (192, 'rev', 'GTAG...           PIR3   \n",
       "15   [(1592, 'fwd', 'GTAGGGATC'), (1570, 'fwd', 'TG...           GND2   \n",
       "16   [(1262, 'fwd', 'GTAGGGAAT'), (189, 'fwd', 'ATA...           GND2   \n",
       "17                                                 NaN           GND2   \n",
       "18                                                 NaN           GND2   \n",
       "19                         [(334, 'rev', 'GGAGGGAAG')]           GND2   \n",
       "20                                                 NaN           GND2   \n",
       "21   [(56, 'rev', 'TAAGGGACG'), (851, 'rev', 'ATAGG...           RGI1   \n",
       "22                                                 NaN           RGI1   \n",
       "23                        [(1187, 'fwd', 'AAAGGGAGA')]           RGI1   \n",
       "24                                                 NaN           RGI1   \n",
       "25                         [(319, 'fwd', 'AAAGGGATA')]           RGI1   \n",
       "26                                         no_promoter           RGI1   \n",
       "27   [(560, 'fwd', 'AGAGGGAAA'), (182, 'rev', 'CAAG...           USV1   \n",
       "28   [(5217, 'fwd', 'TCAGGGATA'), (4375, 'fwd', 'AA...           USV1   \n",
       "29                                                 NaN           USV1   \n",
       "..                                                 ...            ...   \n",
       "367                                                NaN           GSP1   \n",
       "368                         [(67, 'rev', 'TGAGGGAAA')]           GSP1   \n",
       "369                        [(911, 'fwd', 'GAAGGGAAT')]           CYC1   \n",
       "370                        [(463, 'rev', 'ACAGGGAAT')]           CYC1   \n",
       "371                                                NaN           CYC1   \n",
       "372                        [(886, 'fwd', 'TAAGGGAAC')]           CYC1   \n",
       "373  [(1339, 'fwd', 'GGAGGGAAG'), (826, 'fwd', 'GCA...           GIS1   \n",
       "374                                                NaN           GIS1   \n",
       "375                                                NaN           GIS1   \n",
       "376                        [(576, 'rev', 'CAAGGGACC')]           GIS1   \n",
       "377                        [(585, 'rev', 'CAAGGGACC')]           GIS1   \n",
       "378                        [(600, 'rev', 'CAAGGGACC')]           GIS1   \n",
       "379  [(789, 'fwd', 'TCAGGGAGA'), (137, 'rev', 'TGAG...           DOT6   \n",
       "380                        [(353, 'fwd', 'AGAGGGACA')]           DOT6   \n",
       "381  [(661, 'fwd', 'ACAGGGACG'), (491, 'fwd', 'GCAG...           DOT6   \n",
       "382  [(18, 'rev', 'GCAGGGAAG'), (23, 'rev', 'GAAGGG...           DOT6   \n",
       "383  [(23, 'rev', 'GAAGGGAGA'), (31, 'rev', 'ACAGGG...           DOT6   \n",
       "384  [(200, 'fwd', 'ACAGGGACA'), (23, 'rev', 'GAAGG...           DOT6   \n",
       "385                        [(595, 'fwd', 'TAAGGGAAT')]           RPH1   \n",
       "386                        [(263, 'fwd', 'AAAGGGAAT')]           RPH1   \n",
       "387  [(696, 'fwd', 'GAAGGGATG'), (17, 'fwd', 'ACAGG...           RPH1   \n",
       "388                        [(307, 'rev', 'TTAGGGAGT')]           RPH1   \n",
       "389                                                NaN           RPH1   \n",
       "390  [(654, 'fwd', 'CGAGGGAAG'), (52, 'fwd', 'AAAGG...           RPH1   \n",
       "391  [(1009, 'fwd', 'GCAGGGAGG'), (353, 'rev', 'AGA...           TOD6   \n",
       "392  [(5047, 'fwd', 'ATAGGGATT'), (4478, 'fwd', 'TG...           TOD6   \n",
       "393                                                NaN           TOD6   \n",
       "394  [(480, 'fwd', 'GCAGGGATT'), (478, 'rev', 'CCAG...           TOD6   \n",
       "395                                                NaN           TOD6   \n",
       "396                         [(14, 'rev', 'CGAGGGAAC')]           TOD6   \n",
       "\n",
       "                    category   kl_genename  \n",
       "0           g1r1_m24_sc_para  KLLA0D13310g  \n",
       "1           g1r1_m24_sc_para  KLLA0D13310g  \n",
       "2           g1r1_m24_sc_para  KLLA0D13310g  \n",
       "3           g1r1_m24_sc_para  KLLA0D13310g  \n",
       "4           g1r1_m24_sc_para  KLLA0E21891g  \n",
       "5           g1r1_m24_sc_para  KLLA0E21891g  \n",
       "6           g1r1_m24_sc_para  KLLA0E21891g  \n",
       "7           g1r1_m24_sc_para  KLLA0E21891g  \n",
       "8           g1r1_m24_sc_para  KLLA0E21891g  \n",
       "9           g1r1_m24_sc_para  KLLA0E21891g  \n",
       "10          g1r1_m24_sc_para  KLLA0B07392g  \n",
       "11          g1r1_m24_sc_para  KLLA0B07392g  \n",
       "12          g1r1_m24_sc_para  KLLA0B07392g  \n",
       "13          g1r1_m24_sc_para  KLLA0B07392g  \n",
       "14          g1r1_m24_sc_para  KLLA0B07392g  \n",
       "15          g1r1_m24_sc_para  KLLA0A09339g  \n",
       "16          g1r1_m24_sc_para  KLLA0A09339g  \n",
       "17          g1r1_m24_sc_para  KLLA0A09339g  \n",
       "18          g1r1_m24_sc_para  KLLA0A09339g  \n",
       "19          g1r1_m24_sc_para  KLLA0A09339g  \n",
       "20          g1r1_m24_sc_para  KLLA0A09339g  \n",
       "21          g1r1_m24_sc_para  KLLA0C08063g  \n",
       "22          g1r1_m24_sc_para  KLLA0C08063g  \n",
       "23          g1r1_m24_sc_para  KLLA0C08063g  \n",
       "24          g1r1_m24_sc_para  KLLA0C08063g  \n",
       "25          g1r1_m24_sc_para  KLLA0C08063g  \n",
       "26          g1r1_m24_sc_para  KLLA0C08063g  \n",
       "27        g1r1_m24_klsc_para  KLLA0C16005g  \n",
       "28        g1r1_m24_klsc_para  KLLA0C16005g  \n",
       "29        g1r1_m24_klsc_para  KLLA0C16005g  \n",
       "..                       ...           ...  \n",
       "367  g1r1_m24wtdrop_para_low  KLLA0C05126g  \n",
       "368  g1r1_m24wtdrop_para_low  KLLA0C05126g  \n",
       "369  g1r1_m24wtdrop_para_low  KLLA0F16929g  \n",
       "370  g1r1_m24wtdrop_para_low  KLLA0F16929g  \n",
       "371  g1r1_m24wtdrop_para_low  KLLA0F16929g  \n",
       "372  g1r1_m24wtdrop_para_low  KLLA0F16929g  \n",
       "373                    other  KLLA0C17710g  \n",
       "374                    other  KLLA0C17710g  \n",
       "375                    other  KLLA0C17710g  \n",
       "376                    other  KLLA0C17710g  \n",
       "377                    other  KLLA0C17710g  \n",
       "378                    other  KLLA0C17710g  \n",
       "379                    other  KLLA0E19933g  \n",
       "380                    other  KLLA0E19933g  \n",
       "381                    other  KLLA0E19933g  \n",
       "382                    other  KLLA0E19933g  \n",
       "383                    other  KLLA0E19933g  \n",
       "384                    other  KLLA0E19933g  \n",
       "385           other_para_low  KLLA0C17710g  \n",
       "386           other_para_low  KLLA0C17710g  \n",
       "387           other_para_low  KLLA0C17710g  \n",
       "388           other_para_low  KLLA0C17710g  \n",
       "389           other_para_low  KLLA0C17710g  \n",
       "390           other_para_low  KLLA0C17710g  \n",
       "391           other_para_low  KLLA0E19933g  \n",
       "392           other_para_low  KLLA0E19933g  \n",
       "393           other_para_low  KLLA0E19933g  \n",
       "394           other_para_low  KLLA0E19933g  \n",
       "395           other_para_low  KLLA0E19933g  \n",
       "396           other_para_low  KLLA0E19933g  \n",
       "\n",
       "[397 rows x 8 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spec</th>\n",
       "      <th>STRE</th>\n",
       "      <th>Msn2_pref</th>\n",
       "      <th>PDS</th>\n",
       "      <th>SC_common_name</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KAFR0K01720</th>\n",
       "      <td>Kafr</td>\n",
       "      <td>[(588, fwd, GAAGGGGTA)]</td>\n",
       "      <td>None</td>\n",
       "      <td>[(1009, fwd, GCAGGGAGG), (353, rev, AGAGGGAAA)]</td>\n",
       "      <td>TOD6</td>\n",
       "      <td>other_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAGL0I05060g</th>\n",
       "      <td>Cgla</td>\n",
       "      <td>[(2732, fwd, AAAGGGGAT), (547, fwd, GAAGGGGGT)...</td>\n",
       "      <td>[(1278, fwd, TTCGGGGCA), (1321, rev, TCCGGGGTT...</td>\n",
       "      <td>[(5047, fwd, ATAGGGATT), (4478, fwd, TGAGGGACA...</td>\n",
       "      <td>TOD6</td>\n",
       "      <td>other_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Suva_5.202</th>\n",
       "      <td>Suva</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TOD6</td>\n",
       "      <td>other_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Skud_2.50</th>\n",
       "      <td>Skud</td>\n",
       "      <td>None</td>\n",
       "      <td>[(595, fwd, GACGGGGTC)]</td>\n",
       "      <td>[(480, fwd, GCAGGGATT), (478, rev, CCAGGGAAA)]</td>\n",
       "      <td>TOD6</td>\n",
       "      <td>other_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Smik_2.62</th>\n",
       "      <td>Smik</td>\n",
       "      <td>[(576, fwd, TGAGGGGCT), (609, rev, TAAGGGGAT)]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TOD6</td>\n",
       "      <td>other_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YBL054W</th>\n",
       "      <td>Scer</td>\n",
       "      <td>[(600, fwd, GTAGGGGCT)]</td>\n",
       "      <td>None</td>\n",
       "      <td>[(14, rev, CGAGGGAAC)]</td>\n",
       "      <td>TOD6</td>\n",
       "      <td>other_para_low</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              spec                                               STRE  \\\n",
       "KAFR0K01720   Kafr                            [(588, fwd, GAAGGGGTA)]   \n",
       "CAGL0I05060g  Cgla  [(2732, fwd, AAAGGGGAT), (547, fwd, GAAGGGGGT)...   \n",
       "Suva_5.202    Suva                                               None   \n",
       "Skud_2.50     Skud                                               None   \n",
       "Smik_2.62     Smik     [(576, fwd, TGAGGGGCT), (609, rev, TAAGGGGAT)]   \n",
       "YBL054W       Scer                            [(600, fwd, GTAGGGGCT)]   \n",
       "\n",
       "                                                      Msn2_pref  \\\n",
       "KAFR0K01720                                                None   \n",
       "CAGL0I05060g  [(1278, fwd, TTCGGGGCA), (1321, rev, TCCGGGGTT...   \n",
       "Suva_5.202                                                 None   \n",
       "Skud_2.50                               [(595, fwd, GACGGGGTC)]   \n",
       "Smik_2.62                                                  None   \n",
       "YBL054W                                                    None   \n",
       "\n",
       "                                                            PDS  \\\n",
       "KAFR0K01720     [(1009, fwd, GCAGGGAGG), (353, rev, AGAGGGAAA)]   \n",
       "CAGL0I05060g  [(5047, fwd, ATAGGGATT), (4478, fwd, TGAGGGACA...   \n",
       "Suva_5.202                                                 None   \n",
       "Skud_2.50        [(480, fwd, GCAGGGATT), (478, rev, CCAGGGAAA)]   \n",
       "Smik_2.62                                                  None   \n",
       "YBL054W                                  [(14, rev, CGAGGGAAC)]   \n",
       "\n",
       "             SC_common_name        category  \n",
       "KAFR0K01720            TOD6  other_para_low  \n",
       "CAGL0I05060g           TOD6  other_para_low  \n",
       "Suva_5.202             TOD6  other_para_low  \n",
       "Skud_2.50              TOD6  other_para_low  \n",
       "Smik_2.62              TOD6  other_para_low  \n",
       "YBL054W                TOD6  other_para_low  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_output\n",
    "full_output = pd.DataFrame()\n",
    "pd.concat([full_output, new_output])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>SC_common_name</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>HXT5</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>NDE2</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>PIR3</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>GND2</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>RGI1</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>USV1</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>SOL4</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>TSA2</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>DBP1</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>YJR115W</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>YHR097C</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>MBR1</td>\n",
       "      <td>g1r1_m24_sc_para</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>GND1</td>\n",
       "      <td>g1r1_m24_sc_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>HSP150</td>\n",
       "      <td>g1r1_m24_sc_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>NDE1</td>\n",
       "      <td>g1r1_m24_sc_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>HXT3</td>\n",
       "      <td>g1r1_m24_sc_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>RGI2</td>\n",
       "      <td>g1r1_m24_sc_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>DED1</td>\n",
       "      <td>g1r1_m24_sc_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>PAL1</td>\n",
       "      <td>g1r1_m24_sc_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>ECM13</td>\n",
       "      <td>g1r1_m24_sc_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>SOL3</td>\n",
       "      <td>g1r1_m24_sc_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>21</td>\n",
       "      <td>RGM1</td>\n",
       "      <td>g1r1_m24_sc_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>22</td>\n",
       "      <td>TSA1</td>\n",
       "      <td>g1r1_m24_sc_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>23</td>\n",
       "      <td>ISF1</td>\n",
       "      <td>g1r1_m24_sc_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>24</td>\n",
       "      <td>GSY2</td>\n",
       "      <td>g1r1_m24_klsc_para</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>25</td>\n",
       "      <td>YBR241C</td>\n",
       "      <td>g1r1_m24_klsc_para</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26</td>\n",
       "      <td>USV1</td>\n",
       "      <td>g1r1_m24_klsc_para</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>27</td>\n",
       "      <td>GLG1</td>\n",
       "      <td>g1r1_m24_klsc_para</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>28</td>\n",
       "      <td>ISF1</td>\n",
       "      <td>g1r1_m24_klsc_para</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>29</td>\n",
       "      <td>MRK1</td>\n",
       "      <td>g1r1_m24_klsc_para</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>42</td>\n",
       "      <td>PGM1</td>\n",
       "      <td>g1r1_m24_klsc_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>43</td>\n",
       "      <td>MBR1</td>\n",
       "      <td>g1r1_m24_klsc_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>44</td>\n",
       "      <td>HSP12</td>\n",
       "      <td>g1r1_m24_nonpara</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>45</td>\n",
       "      <td>GOR1</td>\n",
       "      <td>g1r1_m24_nonpara</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>46</td>\n",
       "      <td>YMR196W</td>\n",
       "      <td>g1r1_m24_nonpara</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>47</td>\n",
       "      <td>PEX18</td>\n",
       "      <td>m24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>48</td>\n",
       "      <td>SHH4</td>\n",
       "      <td>m24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>49</td>\n",
       "      <td>SSE2</td>\n",
       "      <td>m24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>50</td>\n",
       "      <td>YJL045W</td>\n",
       "      <td>m24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>51</td>\n",
       "      <td>SHH3</td>\n",
       "      <td>m24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>52</td>\n",
       "      <td>PEX21</td>\n",
       "      <td>m24_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>53</td>\n",
       "      <td>SDH3</td>\n",
       "      <td>m24_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>54</td>\n",
       "      <td>SDH1</td>\n",
       "      <td>m24_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>55</td>\n",
       "      <td>SDH4</td>\n",
       "      <td>m24_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>56</td>\n",
       "      <td>SSE1</td>\n",
       "      <td>m24_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>57</td>\n",
       "      <td>YBR139W</td>\n",
       "      <td>g1r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>58</td>\n",
       "      <td>SDS24</td>\n",
       "      <td>g1r1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>59</td>\n",
       "      <td>SDS23</td>\n",
       "      <td>g1r1_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>60</td>\n",
       "      <td>CYC7</td>\n",
       "      <td>g1r1_m24wtdrop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>61</td>\n",
       "      <td>HXK1</td>\n",
       "      <td>g1r1_m24wtdrop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>62</td>\n",
       "      <td>UBI4</td>\n",
       "      <td>g1r1_m24wtdrop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>63</td>\n",
       "      <td>CTT1</td>\n",
       "      <td>g1r1_m24wtdrop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>64</td>\n",
       "      <td>GSP2</td>\n",
       "      <td>g1r1_m24wtdrop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>65</td>\n",
       "      <td>HXK2</td>\n",
       "      <td>g1r1_m24wtdrop_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>66</td>\n",
       "      <td>GSP1</td>\n",
       "      <td>g1r1_m24wtdrop_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>67</td>\n",
       "      <td>CYC1</td>\n",
       "      <td>g1r1_m24wtdrop_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>68</td>\n",
       "      <td>GIS1</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>69</td>\n",
       "      <td>DOT6</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>70</td>\n",
       "      <td>RPH1</td>\n",
       "      <td>other_para_low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>71</td>\n",
       "      <td>TOD6</td>\n",
       "      <td>other_para_low</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0 SC_common_name                 category\n",
       "0            0           HXT5         g1r1_m24_sc_para\n",
       "1            1           NDE2         g1r1_m24_sc_para\n",
       "2            2           PIR3         g1r1_m24_sc_para\n",
       "3            3           GND2         g1r1_m24_sc_para\n",
       "4            4           RGI1         g1r1_m24_sc_para\n",
       "5            5           USV1         g1r1_m24_sc_para\n",
       "6            6           SOL4         g1r1_m24_sc_para\n",
       "7            7           TSA2         g1r1_m24_sc_para\n",
       "8            8           DBP1         g1r1_m24_sc_para\n",
       "9            9        YJR115W         g1r1_m24_sc_para\n",
       "10          10        YHR097C         g1r1_m24_sc_para\n",
       "11          11           MBR1         g1r1_m24_sc_para\n",
       "12          12           GND1     g1r1_m24_sc_para_low\n",
       "13          13         HSP150     g1r1_m24_sc_para_low\n",
       "14          14           NDE1     g1r1_m24_sc_para_low\n",
       "15          15           HXT3     g1r1_m24_sc_para_low\n",
       "16          16           RGI2     g1r1_m24_sc_para_low\n",
       "17          17           DED1     g1r1_m24_sc_para_low\n",
       "18          18           PAL1     g1r1_m24_sc_para_low\n",
       "19          19          ECM13     g1r1_m24_sc_para_low\n",
       "20          20           SOL3     g1r1_m24_sc_para_low\n",
       "21          21           RGM1     g1r1_m24_sc_para_low\n",
       "22          22           TSA1     g1r1_m24_sc_para_low\n",
       "23          23           ISF1     g1r1_m24_sc_para_low\n",
       "24          24           GSY2       g1r1_m24_klsc_para\n",
       "25          25        YBR241C       g1r1_m24_klsc_para\n",
       "26          26           USV1       g1r1_m24_klsc_para\n",
       "27          27           GLG1       g1r1_m24_klsc_para\n",
       "28          28           ISF1       g1r1_m24_klsc_para\n",
       "29          29           MRK1       g1r1_m24_klsc_para\n",
       "..         ...            ...                      ...\n",
       "42          42           PGM1   g1r1_m24_klsc_para_low\n",
       "43          43           MBR1   g1r1_m24_klsc_para_low\n",
       "44          44          HSP12         g1r1_m24_nonpara\n",
       "45          45           GOR1         g1r1_m24_nonpara\n",
       "46          46        YMR196W         g1r1_m24_nonpara\n",
       "47          47          PEX18                      m24\n",
       "48          48           SHH4                      m24\n",
       "49          49           SSE2                      m24\n",
       "50          50        YJL045W                      m24\n",
       "51          51           SHH3                      m24\n",
       "52          52          PEX21             m24_para_low\n",
       "53          53           SDH3             m24_para_low\n",
       "54          54           SDH1             m24_para_low\n",
       "55          55           SDH4             m24_para_low\n",
       "56          56           SSE1             m24_para_low\n",
       "57          57        YBR139W                     g1r1\n",
       "58          58          SDS24                     g1r1\n",
       "59          59          SDS23            g1r1_para_low\n",
       "60          60           CYC7           g1r1_m24wtdrop\n",
       "61          61           HXK1           g1r1_m24wtdrop\n",
       "62          62           UBI4           g1r1_m24wtdrop\n",
       "63          63           CTT1           g1r1_m24wtdrop\n",
       "64          64           GSP2           g1r1_m24wtdrop\n",
       "65          65           HXK2  g1r1_m24wtdrop_para_low\n",
       "66          66           GSP1  g1r1_m24wtdrop_para_low\n",
       "67          67           CYC1  g1r1_m24wtdrop_para_low\n",
       "68          68           GIS1                    other\n",
       "69          69           DOT6                    other\n",
       "70          70           RPH1           other_para_low\n",
       "71          71           TOD6           other_para_low\n",
       "\n",
       "[72 rows x 3 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#print data to file\n",
    "for gene in SC_common_name: \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'YGOB_pillars' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-2-03c10d2466b5>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     15\u001b[0m \u001b[0mgene_pos\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;34m''\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     16\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mpos\u001b[0m \u001b[1;32min\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;34m'p1'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m'p2'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 17\u001b[1;33m     \u001b[0mScer_pos\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mYGOB_pillars\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'Scer_'\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mpos\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     18\u001b[0m     \u001b[0mScer_pos_find\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mScer_pos\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mScer_pos\u001b[0m\u001b[1;33m==\u001b[0m\u001b[0mgene\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     19\u001b[0m     \u001b[1;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mScer_pos_find\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m==\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'YGOB_pillars' is not defined"
     ]
    }
   ],
   "source": [
    "motif_dict = {'STRE':'AGGGG',\n",
    "             'Msn2_pref': 'CGGGG', \n",
    "              'PDS1': 'TTAGGGAT',\n",
    "              'PDS2': 'TAAGGGAT',\n",
    "              'PDS core': 'AGGGA'\n",
    "             }\n",
    "output_format = 'full'\n",
    "sequence_context = 2\n",
    "no_prom_thresh = 20 #if length of intergenic region is less than this, call no promoter\n",
    "\n",
    "gene_name = 'SOL4'\n",
    "gene = 'YGR248W'\n",
    "\n",
    "#Lookup orthologs based on the YGOB Pillars dataframe\n",
    "gene_pos = ''\n",
    "for pos in ['p1','p2']: \n",
    "    Scer_pos = YGOB_pillars['Scer_' + pos]\n",
    "    Scer_pos_find = Scer_pos[Scer_pos==gene]\n",
    "    if len(Scer_pos_find)==1:\n",
    "        if gene_pos != '':\n",
    "            print('Warning: found gene in two pillars')\n",
    "        gene_pos = pos\n",
    "        gene_index = Scer_pos_find.index[0]\n",
    "    elif len(Scer_pos_find)>1: \n",
    "        print('Warning: found gene twice in same pillar')\n",
    "\n",
    "#extract intergenic regions from trevor's file\n",
    "\n",
    "specs = ['Kafr','Cgla','Suva', 'Skud', 'Smik','Scer']\n",
    "##to include pre wgd species would need to update\n",
    "specs_pillar = [spec + '_' + gene_pos for spec in specs]\n",
    "\n",
    "orthologs = YGOB_pillars.loc[gene_index, specs_pillar]\n",
    "\n",
    "\n",
    "seqs = {}\n",
    "\n",
    "for spec in specs: \n",
    "    fname_intergenics = os.path.normpath(data_processing_dir + 'promoter_phylogenies/intergenics/' + spec + '_intergenics.faa' )\n",
    "    record_iterator = SeqIO.parse(fname_intergenics, \"fasta\")\n",
    "\n",
    "    for seq in record_iterator: \n",
    "        if seq.id == orthologs[spec + '_' + gene_pos]:\n",
    "            seqs[spec] = seq\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#promoter_counts = io_library.exact_promoter_scan(gene_list, motif_dict, promoter_database, output_format, sequence_context)\n",
    "\n",
    "#initial dataframe has species for index, but reindex on seq id before outputting\n",
    "seq_ids = [seq.id for seq in seqs.values()]\n",
    "output_data_frame = pd.DataFrame(index = seq_ids)\n",
    "output_data_frame['spec'] = seqs.keys()\n",
    "\n",
    "for motif_name, motif in motif_dict.items():\n",
    "    L_motif = len(motif)\n",
    "    output_motif = []\n",
    "\n",
    "    for spec, seq in seqs.items(): \n",
    "\n",
    "        prom_seq = seq.seq\n",
    "        if len(prom_seq)<no_prom_thresh: \n",
    "            output_motif.append('no_promoter')\n",
    "        else: \n",
    "            output_motif_seq = io_library.exact_promoter_scan(motif, prom_seq, output_format, sequence_context)\n",
    "            output_motif.append(output_motif_seq)\n",
    "    output_data_frame[motif_name] = output_motif\n",
    "\n",
    "output_data_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Cgla': SeqRecord(seq=Seq('AGAAGATTGTAAAACCGGCAACTAGGAGAACAACATTTCAATAATCTCTATTTA...GTG', SingleLetterAlphabet()), id='CAGL0I02200g', name='CAGL0I02200g', description='CAGL0I02200g', dbxrefs=[]),\n",
       " 'Kafr': SeqRecord(seq=Seq('AGTTTTCAAGCAACAATATTCGTATCTATATAGTCGGAAGTTTCAAAGTGATTT...CTA', SingleLetterAlphabet()), id='KAFR0B04400', name='KAFR0B04400', description='KAFR0B04400', dbxrefs=[]),\n",
       " 'Scer': SeqRecord(seq=Seq('AGCATACATTTATGTGCACTGTAAGTAATTATATAGAATTTTTTTTTTTTTTTT...AAA', SingleLetterAlphabet()), id='YGR248W', name='YGR248W', description='YGR248W', dbxrefs=[]),\n",
       " 'Skud': SeqRecord(seq=Seq('AGGGCACGTGTATGGATTGTGTAGGTAAATATATAGCAGGTTTTACAAGCTTGT...GAG', SingleLetterAlphabet()), id='Skud_7.581', name='Skud_7.581', description='Skud_7.581', dbxrefs=[]),\n",
       " 'Smik': SeqRecord(seq=Seq('GTACACATGTATACATATTTTAAGTAATTTATGGCAAGGTTTTTCATATCTGTC...ACT', SingleLetterAlphabet()), id='Smik_16.59', name='Smik_16.59', description='Smik_16.59', dbxrefs=[]),\n",
       " 'Suva': SeqRecord(seq=Seq('GCAGGACAGCAAGAGAAAAGCGCAAATATATAACGTATATATACTGTAAGTAAT...ATC', SingleLetterAlphabet()), id='Suva_7.543', name='Suva_7.543', description='Suva_7.543', dbxrefs=[])}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_name = 'SOL3'\n",
    "gene = 'YHR163W'\n",
    "\n",
    "## Flag if no gene comes up\n",
    "#Lookup orthologs based on the YGOB Pillars dataframe\n",
    "gene_pos = ''\n",
    "for pos in ['p1','p2']: \n",
    "    Scer_pos = YGOB_pillars['Scer_' + pos]\n",
    "    Scer_pos_find = Scer_pos[Scer_pos==gene]\n",
    "    if len(Scer_pos_find)==1:\n",
    "        if gene_pos != '':\n",
    "            print('Warning: found gene in two pillars')\n",
    "        gene_pos = pos\n",
    "        gene_index = Scer_pos_find.index[0]\n",
    "    elif len(Scer_pos_find)>1: \n",
    "        print('Warning: found gene twice in same pillar')\n",
    "\n",
    "#extract intergenic regions from trevor's file\n",
    "\n",
    "specs = ['Kafr','Cgla','Suva', 'Skud', 'Smik','Scer']\n",
    "##to include pre wgd species would need to update\n",
    "specs_pillar = [spec + '_' + gene_pos for spec in specs]\n",
    "\n",
    "orthologs = YGOB_pillars.loc[gene_index, specs_pillar]\n",
    "\n",
    "\n",
    "seqs = {}\n",
    "\n",
    "for spec in specs: \n",
    "    fname_intergenics = os.path.normpath(data_processing_dir + 'promoter_phylogenies/intergenics/' + spec + '_intergenics.faa' )\n",
    "    record_iterator = SeqIO.parse(fname_intergenics, \"fasta\")\n",
    "\n",
    "    for seq in record_iterator: \n",
    "        if seq.id == orthologs[spec + '_' + gene_pos]:\n",
    "            seqs[spec] = seq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq = seqs['Scer']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "motif = 'AGGGG'\n",
    "sequence_context = 2\n",
    "output_format = 'full'\n",
    "prom_seq = seq.seq\n",
    "\n",
    "#exact_promoter_scan(motif, prom_seq, output_format='count', sequence_context=0):\n",
    "#this function is for a single sequence - same name used to be for one that searched a list of genes \n",
    "# now called exact_promoter_scan_genelist\n",
    "# \n",
    "# scans a sequence (biopython object) for a motif in both forward and backward directions \n",
    "# If output format is 'count', then just lists the number of hits. \n",
    "# If output format is 'full', gives the location, direction (relative to the gene), and\n",
    "# sequence context (number of bases on either side of the hit)output_format\n",
    "\n",
    "\n",
    "L_motif = len(motif)\n",
    "\n",
    "#fwd search\n",
    "prom_seq_fwd = str(prom_seq)\n",
    "L_prom = len(prom_seq_fwd)\n",
    "motif_sites_fwd = [m.start() for m in re.finditer(motif, prom_seq_fwd)]\n",
    "#to find overlapping motifs\n",
    "#[m.start() fr m in re.finditer('(?=' + motif + ')', prom_seq_rev)]\n",
    "\n",
    "#reverse search\n",
    "#rev search\n",
    "prom_seq_rev = str(prom_seq.reverse_complement())\n",
    "#location is in reference to the fwd seq\n",
    "motif_sites_rev = [len(prom_seq_rev) - m.start() - L_motif for m in re.finditer(motif, prom_seq_rev)]\n",
    "\n",
    "if output_format == 'count':  \n",
    "    n_motifs_gene = len(motif_sites_fwd) + len(motif_sites_rev)\n",
    "    output = n_motifs_gene\n",
    "elif output_format == 'full': \n",
    "    loc_hitseq_gene = []\n",
    "    for motif_site in motif_sites_fwd: \n",
    "        loc = L_prom-1-motif_site\n",
    "        hitseq = prom_seq_fwd[motif_site-sequence_context:motif_site+L_motif+sequence_context]  \n",
    "        hitdir = 'fwd'\n",
    "        loc_hitseq_gene.append((loc,hitdir,hitseq))\n",
    "    for motif_site in motif_sites_rev: \n",
    "        loc = L_prom - motif_site+L_motif\n",
    "        hitseq = prom_seq_rev[(L_prom-motif_site-L_motif-sequence_context):(L_prom-motif_site+sequence_context)]\n",
    "        hitdir = 'rev'\n",
    "        loc_hitseq_gene.append((loc,hitdir,hitseq))\n",
    "\n",
    "    if len(loc_hitseq_gene)==0: \n",
    "        output = None\n",
    "    else:\n",
    "        output = loc_hitseq_gene\n",
    "else: \n",
    "    print(\"choose output_format: 'count' or 'full' \")\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(23, 'rev', 'CCAGGGGTA')]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AACGCCGAATTTACAACTCGAAACTTTGCTTTGTCAAAGATATGTACGTAAGCATATATATCCGGATTGTATAATATTCATAAAATAGTAAAGTATATTTTCTTTGCGCGGTGTTATAATTTGCTCTTTTGCATTTTTCCGTTTTTTTCTTACCTTTTTATCCCCGACAATGGTAATAGTGGTATATTATATAAAGAGGAAAGGCCTCGAGGATAATAGAAGGCAATGCACCATCAATTGCTTTACCCCTGGTCCGCGACCAA'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prom_seq_fwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TTGGTCGCGGACCAGGGGTAAAGCAATTGATGGTGCATTGCCTTCTATTATCCTCGAGGCCTTTCCTCTTTATATAATATACCACTATTACCATTGTCGGGGATAAAAAGGTAAGAAAAAAACGGAAAAATGCAAAAGAGCAAATTATAACACCGCGCAAAGAAAATATACTTTACTATTTTATGAATATTATACAATCCGGATATATATGCTTACGTACATATCTTTGACAAAGCAAAGTTTCGAGTTGTAAATTCGGCGTT'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prom_seq_rev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AGGGG'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[245]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[len(prom_seq_rev) - m.start() - L_motif for m in re.finditer(motif, prom_seq_rev)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['HXT5',\n",
       " 'NDE2',\n",
       " 'PIR3',\n",
       " 'GND2',\n",
       " 'RGI1',\n",
       " 'USV1',\n",
       " 'SOL4',\n",
       " 'TSA2',\n",
       " 'DBP1',\n",
       " 'YJR115W',\n",
       " 'YHR097C',\n",
       " 'MBR1',\n",
       " 'GND1',\n",
       " 'HSP150',\n",
       " 'NDE1',\n",
       " 'HXT3',\n",
       " 'RGI2',\n",
       " 'DED1',\n",
       " 'PAL1',\n",
       " 'ECM13',\n",
       " 'SOL3',\n",
       " 'RGM1',\n",
       " 'TSA1',\n",
       " 'ISF1',\n",
       " 'GSY2',\n",
       " 'YBR241C',\n",
       " 'USV1',\n",
       " 'GLG1',\n",
       " 'ISF1',\n",
       " 'MRK1',\n",
       " 'ATG33',\n",
       " 'FMP45',\n",
       " 'TSL1',\n",
       " 'PGM2',\n",
       " 'GSY1',\n",
       " 'YNL194C',\n",
       " 'RGM1',\n",
       " 'VPS73',\n",
       " 'RIM11',\n",
       " 'TPS3',\n",
       " 'GLG2',\n",
       " 'SCM4',\n",
       " 'PGM1',\n",
       " 'MBR1',\n",
       " 'HSP12',\n",
       " 'GOR1',\n",
       " 'YMR196W',\n",
       " 'PEX18',\n",
       " 'SHH4',\n",
       " 'SSE2',\n",
       " 'YJL045W',\n",
       " 'SHH3',\n",
       " 'PEX21',\n",
       " 'SDH3',\n",
       " 'SDH1',\n",
       " 'SDH4',\n",
       " 'SSE1',\n",
       " 'YBR139W',\n",
       " 'SDS24',\n",
       " 'SDS23',\n",
       " 'CYC7',\n",
       " 'HXK1',\n",
       " 'UBI4',\n",
       " 'CTT1',\n",
       " 'GSP2',\n",
       " 'HXK2',\n",
       " 'GSP1',\n",
       " 'CYC1',\n",
       " 'GIS1',\n",
       " 'DOT6',\n",
       " 'RPH1',\n",
       " 'TOD6']"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#suspect Smik gnd2 is actually Smik gnd1 - check intergenics\n",
    "#RGI1 overlaps with dubious orf.  \n",
    "#Smik 4.730 is missing from intergenics.  I think that intergenics is not as good as 700bp upstream  \n",
    "#YER067W overlaps with dubious orf so isn't in trevor's file.  If you take the promoter as the 1kb upstream then it there are several STREs and one PDS. \n",
    "\n",
    "\n",
    "gene_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spec</th>\n",
       "      <th>STRE</th>\n",
       "      <th>Msn2_pref</th>\n",
       "      <th>PDS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Suva_10.389</th>\n",
       "      <td>Suva</td>\n",
       "      <td>[(229, rev, TTAGGGGAG)]</td>\n",
       "      <td>None</td>\n",
       "      <td>[(234, rev, GGAGGGACA)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Skud_15.352</th>\n",
       "      <td>Skud</td>\n",
       "      <td>[(487, fwd, CAAGGGGCA), (458, fwd, CAAGGGGAG),...</td>\n",
       "      <td>None</td>\n",
       "      <td>[(439, fwd, TAAGGGAGG), (711, rev, AAAGGGAAT)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Smik_12.371</th>\n",
       "      <td>Smik</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[(65, rev, TGAGGGAAA), (271, rev, AAAGGGATC)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YOR185C</th>\n",
       "      <td>Scer</td>\n",
       "      <td>[(482, fwd, TAAGGGGCC), (150, rev, TAAGGGGTG),...</td>\n",
       "      <td>None</td>\n",
       "      <td>[(525, fwd, CCAGGGAAT), (473, fwd, ACAGGGAAA),...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             spec                                               STRE  \\\n",
       "Suva_10.389  Suva                            [(229, rev, TTAGGGGAG)]   \n",
       "Skud_15.352  Skud  [(487, fwd, CAAGGGGCA), (458, fwd, CAAGGGGAG),...   \n",
       "Smik_12.371  Smik                                               None   \n",
       "YOR185C      Scer  [(482, fwd, TAAGGGGCC), (150, rev, TAAGGGGTG),...   \n",
       "\n",
       "            Msn2_pref                                                PDS  \n",
       "Suva_10.389      None                            [(234, rev, GGAGGGACA)]  \n",
       "Skud_15.352      None     [(439, fwd, TAAGGGAGG), (711, rev, AAAGGGAAT)]  \n",
       "Smik_12.371      None      [(65, rev, TGAGGGAAA), (271, rev, AAAGGGATC)]  \n",
       "YOR185C          None  [(525, fwd, CCAGGGAAT), (473, fwd, ACAGGGAAA),...  "
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_dict['GSP2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'YDL085W'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-88-a370cf57d660>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0moutput_dict\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mio_library\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mSC_orf_lookup_by_name\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'NDE2'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m: 'YDL085W'"
     ]
    }
   ],
   "source": [
    "output_dict[io_library.SC_orf_lookup_by_name(['NDE2'])[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "orthologs = YGOB_pillars.loc[gene_index, specs]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = next(record_iterator)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Cgla': SeqRecord(seq=Seq('TATGTGTTTCCACCAACTAAAAGAAAAAACTTGCCAAGAAGATACTTTCAACTA...GCA', SingleLetterAlphabet()), id='CAGL0L11880g', name='CAGL0L11880g', description='CAGL0L11880g', dbxrefs=[]),\n",
       " 'Kafr': SeqRecord(seq=Seq('GTATGGGGGAGGGAAGTGGGTTTTGGAGTGTTATTGCGTCGAGTAAATATATAT...AAG', SingleLetterAlphabet()), id='KAFR0A02380', name='KAFR0A02380', description='KAFR0A02380', dbxrefs=[]),\n",
       " 'Scer': SeqRecord(seq=Seq('TTCTCTTGTCTTACAGGTCCCTTGAATTGGCATTTACGTTTACATCCGATCCGG...ACA', SingleLetterAlphabet()), id='YDR096W', name='YDR096W', description='YDR096W', dbxrefs=[]),\n",
       " 'Skud': SeqRecord(seq=Seq('ACGATAACATTTTGTTCTTTTGTACCATGTATAAGCAATGTCATATGATATAGA...ATT', SingleLetterAlphabet()), id='Skud_4.354', name='Skud_4.354', description='Skud_4.354', dbxrefs=[]),\n",
       " 'Smik': SeqRecord(seq=Seq('ATGATAACATTTTTTGTTCTTTTGTACTATGTATAACAAATGTCATATGATATA...ATA', SingleLetterAlphabet()), id='Smik_4.337', name='Smik_4.337', description='Smik_4.337', dbxrefs=[]),\n",
       " 'Suva': SeqRecord(seq=Seq('TACACACACGAGCAACACTATCAACCACACTGTCGCTCTCCAATATCGAAGACA...ATT', SingleLetterAlphabet()), id='Suva_5.303', name='Suva_5.303', description='Suva_5.303', dbxrefs=[])}"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#find and extract all instances of STREs and PDSs\n",
    "\n",
    "\n",
    "seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Cgla': SeqRecord(seq=Seq('TATGTGTTTCCACCAACTAAAAGAAAAAACTTGCCAAGAAGATACTTTCAACTA...GCA', SingleLetterAlphabet()), id='CAGL0L11880g', name='CAGL0L11880g', description='CAGL0L11880g', dbxrefs=[]),\n",
       " 'Kafr': SeqRecord(seq=Seq('GTATGGGGGAGGGAAGTGGGTTTTGGAGTGTTATTGCGTCGAGTAAATATATAT...AAG', SingleLetterAlphabet()), id='KAFR0A02380', name='KAFR0A02380', description='KAFR0A02380', dbxrefs=[]),\n",
       " 'Scer': SeqRecord(seq=Seq('TTCTCTTGTCTTACAGGTCCCTTGAATTGGCATTTACGTTTACATCCGATCCGG...ACA', SingleLetterAlphabet()), id='YDR096W', name='YDR096W', description='YDR096W', dbxrefs=[]),\n",
       " 'Skud': SeqRecord(seq=Seq('ACGATAACATTTTGTTCTTTTGTACCATGTATAAGCAATGTCATATGATATAGA...ATT', SingleLetterAlphabet()), id='Skud_4.354', name='Skud_4.354', description='Skud_4.354', dbxrefs=[]),\n",
       " 'Smik': SeqRecord(seq=Seq('ATGATAACATTTTTTGTTCTTTTGTACTATGTATAACAAATGTCATATGATATA...ATA', SingleLetterAlphabet()), id='Smik_4.337', name='Smik_4.337', description='Smik_4.337', dbxrefs=[]),\n",
       " 'Suva': SeqRecord(seq=Seq('TACACACACGAGCAACACTATCAACCACACTGTCGCTCTCCAATATCGAAGACA...ATT', SingleLetterAlphabet()), id='Suva_5.303', name='Suva_5.303', description='Suva_5.303', dbxrefs=[])}"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spec</th>\n",
       "      <th>STRE</th>\n",
       "      <th>Msn2_pref</th>\n",
       "      <th>PDS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KAFR0B04400</th>\n",
       "      <td>Kafr</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAGL0I02200g</th>\n",
       "      <td>Cgla</td>\n",
       "      <td>[(151, rev, GCAGGGGAA), (190, rev, AAAGGGGTG)]</td>\n",
       "      <td>None</td>\n",
       "      <td>[(353, rev, CAAGGGATT)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Suva_7.543</th>\n",
       "      <td>Suva</td>\n",
       "      <td>[(148, rev, TTAGGGGGA)]</td>\n",
       "      <td>[(21, rev, CTCGGGGTG)]</td>\n",
       "      <td>[(182, fwd, TAAGGGATT), (175, rev, CAAGGGATA)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Skud_7.581</th>\n",
       "      <td>Skud</td>\n",
       "      <td>[(47, rev, TTAGGGGGG)]</td>\n",
       "      <td>None</td>\n",
       "      <td>[(70, fwd, CAAGGGATA), (63, rev, CAAGGGATT)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Smik_16.59</th>\n",
       "      <td>Smik</td>\n",
       "      <td>[(135, rev, TTAGGGGGT)]</td>\n",
       "      <td>None</td>\n",
       "      <td>[(160, fwd, CAAGGGATA), (154, rev, CAAGGGATT)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YGR248W</th>\n",
       "      <td>Scer</td>\n",
       "      <td>[(131, rev, TTAGGGGGG)]</td>\n",
       "      <td>None</td>\n",
       "      <td>[(160, fwd, CAAGGGATG), (155, rev, CAAGGGATT)]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              spec                                            STRE  \\\n",
       "KAFR0B04400   Kafr                                            None   \n",
       "CAGL0I02200g  Cgla  [(151, rev, GCAGGGGAA), (190, rev, AAAGGGGTG)]   \n",
       "Suva_7.543    Suva                         [(148, rev, TTAGGGGGA)]   \n",
       "Skud_7.581    Skud                          [(47, rev, TTAGGGGGG)]   \n",
       "Smik_16.59    Smik                         [(135, rev, TTAGGGGGT)]   \n",
       "YGR248W       Scer                         [(131, rev, TTAGGGGGG)]   \n",
       "\n",
       "                           Msn2_pref  \\\n",
       "KAFR0B04400                     None   \n",
       "CAGL0I02200g                    None   \n",
       "Suva_7.543    [(21, rev, CTCGGGGTG)]   \n",
       "Skud_7.581                      None   \n",
       "Smik_16.59                      None   \n",
       "YGR248W                         None   \n",
       "\n",
       "                                                         PDS  \n",
       "KAFR0B04400                                             None  \n",
       "CAGL0I02200g                         [(353, rev, CAAGGGATT)]  \n",
       "Suva_7.543    [(182, fwd, TAAGGGATT), (175, rev, CAAGGGATA)]  \n",
       "Skud_7.581      [(70, fwd, CAAGGGATA), (63, rev, CAAGGGATT)]  \n",
       "Smik_16.59    [(160, fwd, CAAGGGATA), (154, rev, CAAGGGATT)]  \n",
       "YGR248W       [(160, fwd, CAAGGGATG), (155, rev, CAAGGGATT)]  "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_YGOB_orth_lookup_table(species1, species2, base_dir, all_ortholog_file):\n",
    "    #for each position in species 1\n",
    "    #Assign orthologs or 'NONE' to each column from Position 2 and 3\n",
    "    fname = os.path.normpath(base_dir + all_ortholog_file)\n",
    "    orth_positions = {'Kluyveromyces lactis': [15], 'Saccharomyces cerevisiae' : [11,21]}\n",
    "    #YGOB_Pillars.txt order of species: \n",
    "    #    0    V. polyspora Position 1\n",
    "    #    1    T. phaffii Position 1\n",
    "    #    2    T. blattae Position 1\n",
    "    #    3    N. dairenensis Position 1\n",
    "    #    4    N. castellii Position 1\n",
    "    #    5    K. naganishii Position 1\n",
    "    #    6    K. africana Position 1\n",
    "    #    7    C. glabrata Position 1\n",
    "    #    8    S. bayanus var. uvarum Position 1\n",
    "    #    9    S. kudriavzevii Position 1\n",
    "    #    10   S. mikatae Position 1\n",
    "    #    11   S. cerevisiae Position 1\n",
    "    #    12   Ancestral Gene Order\n",
    "    #    13   Z. rouxii\n",
    "    #    14   T. delbrueckii\n",
    "    #    15   K. lactis\n",
    "    #    16   E. gossypii \n",
    "    #    17   E. cymbalariae\n",
    "    #    18   L. kluyveri\n",
    "    #    19   L. thermotolerans\n",
    "    #    20   L. waltii\n",
    "    #    21   S. cerevisiae Position 2\n",
    "    #    22   S. mikatae Position 2\n",
    "    #    23   S. kudriavzevii Position 2\n",
    "    #    24   S. bayanus var. uvarum Position 2\n",
    "    #    25   C. glabrata Position 2\n",
    "    #    26   K. africana Position 2\n",
    "    #    27   K. naganishii Position 2\n",
    "    #    28   N. castellii Position 2\n",
    "    #    29   N. dairenensis Position 2\n",
    "    #    30   T. blattae Position 2\n",
    "    #    31   T. phaffii Position 2\n",
    "    #    32   V. polyspora Position 2\n",
    "    \n",
    "    species1_columns = orth_positions[species1]\n",
    "    species2_columns = orth_positions[species2]\n",
    "    \n",
    "    with open(fname) as f:\n",
    "        orth_lookup = []\n",
    "        for line in f:\n",
    "            linesp = line.split()\n",
    "            for column1 in species1_columns: \n",
    "                if linesp[column1]!= '---':\n",
    "                    species1_gene = [linesp[column1]]\n",
    "                    species2_genes = []\n",
    "                    for column2 in species2_columns: \n",
    "                        if linesp[column2]!='---': \n",
    "                            species2_genes.append(linesp[column2])\n",
    "                    if len(species2_genes) == 0:\n",
    "                        species2_genes.append('NONE')\n",
    "                    orth_lookup.append(species1_gene + species2_genes)\n",
    "                    \n",
    "    orth_file_abbrev = {'Kluyveromyces lactis': 'Klac', 'Saccharomyces cerevisiae': 'Scer', 'Candida glabrata':'Cgla', 'Saccharomyces castellii' : 'Scas', 'Saccharomyces bayanus' : 'Sbay'}\n",
    "    orth_lookup_outputfname = os.path.normpath(base_dir + '\\expression_data\\ortholog_files_YGOB\\\\' + orth_file_abbrev[species1] + \"-\" + orth_file_abbrev[species2] + \"-orthologs.txt\"  )\n",
    "    orth_lookup_outputfile = open(orth_lookup_outputfname, 'w')\n",
    "    \n",
    "    for gene in orth_lookup:\n",
    "        line = '\\t'.join(gene)+'\\n'\n",
    "        orth_lookup_outputfile.write(line)\n",
    "    orth_lookup_outputfile.close()\n",
    "    \n",
    "    return orth_lookup \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Motif phylogenies for Nov 2018 CTTE meeting\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "LIst of genes to look at: \n",
    "\n",
    "\tM24\n",
    "\t\tParalogs: GND2, SOL4, TKL2 (PPP) , TSA1, SHH3, PIR3, YCL048W-A, YDR034W-B, SHH4, RTN2\n",
    "\t\tOthers: CYB2, PUT4\n",
    "\tRph1\n",
    "\t\tParalogs: CYC7, YSC84, GSP2, SSE2, GPD1, SHC1, TPK1\n",
    "\t\tOthers: ZWF1 (1st step in PPP), YTP1, TPK2\n",
    "\tBoth\n",
    "\t\tParalogs: TSL1, RGI1\n",
    "\t\tOthers: CTT1, EGO4, RTC3, COX26\n",
    "\n",
    "\n",
    "GPA1/PRM3 (mating related) either gained control by Msn2/4 in KL or lost it in SC \n",
    "Similarly Fus3 (mating related) re Rph1\n",
    "\tPhenotype: KL require Msn2/4 for mating phenotype (is this already known?)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "SIP1 and GRE1  \n",
    " under Rph1/Gis1 in KL - in fact Rph1/Gis1 deletion causes even bigger response in KL under PKA inhibition (so PKA inhibition doesn't fully inhibit Rph1 in KL)  msn2/4 deletion does not change response to drug. \n",
    " under Msn2/4 control in S.Cer.  Rph1/Gis1 deletion may cause a very small incresase in SIP18.  Fairly big increase in GRE1 under Rph1/Gis1\n",
    "\n",
    "Similar: GPH1, SPG4, SPI1 (which has paralog), also check YNL194C\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "List of motifs to follow: \n",
    "    \n",
    "\n",
    "    motif_dict = {'STRE':'AGGGG',\n",
    "             'Msn2_pref': 'CGGGG', \n",
    "              'PDS1': 'TTAGGGAT',\n",
    "              'PDS2': 'TAAGGGAT',\n",
    "              'PDS core': 'AGGGA'\n",
    "             }\n",
    "    \n",
    "    TATA: \n",
    "        KL\n",
    "        \n",
    "MOTIF TATAHATA DREME-4\n",
    "#             Word    RC Word        Pos        Neg    P-value    E-value\n",
    "# BEST    TATAHATA   TATDTATA        175       1522   4.5e-009   1.9e-004\n",
    "#         TATAAATA   TATTTATA         72        605   2.9e-004   1.2e+001\n",
    "#         TATATATA   TATATATA         89        797   3.7e-004   1.6e+001\n",
    "#         TATACATA   TATGTATA         51        415   1.3e-003   5.5e+001\n",
    "letter-probability matrix: alength= 4 w= 8 nsites= 250 E= 1.9e-004\n",
    "0.000000 0.000000 0.000000 1.000000\n",
    "1.000000 0.000000 0.000000 0.000000\n",
    "0.000000 0.000000 0.000000 1.000000\n",
    "1.000000 0.000000 0.000000 0.000000\n",
    "0.312000 0.220000 0.000000 0.468000\n",
    "1.000000 0.000000 0.000000 0.000000\n",
    "0.000000 0.000000 0.000000 1.000000\n",
    "1.000000 0.000000 0.000000 0.000000\n",
    "\n",
    "SC\n",
    "#             Word    RC Word        Pos        Neg    P-value    E-value\n",
    "# BEST    CWTATATA   TATATAWG        159       1368   3.0e-008   1.3e-003\n",
    "#         CTTATATA   TATATAAG         71        550   2.4e-005   1.0e+000\n",
    "#         CATATATA   TATATATG         96        946   3.0e-003   1.3e+002\n",
    "letter-probability matrix: alength= 4 w= 8 nsites= 188 E= 1.3e-003\n",
    "0.000000 1.000000 0.000000 0.000000\n",
    "0.638298 0.000000 0.000000 0.361702\n",
    "0.000000 0.000000 0.000000 1.000000\n",
    "1.000000 0.000000 0.000000 0.000000\n",
    "0.000000 0.000000 0.000000 1.000000\n",
    "1.000000 0.000000 0.000000 0.000000\n",
    "0.000000 0.000000 0.000000 1.000000\n",
    "1.000000 0.000000 0.000000 0.000000\n",
    "\n",
    "\n",
    "KL mystery motif\n",
    "MOTIF TTTGCTW DREME-2\n",
    "#             Word    RC Word        Pos        Neg    P-value    E-value\n",
    "# BEST     TTTGCTW    WAGCAAA        176       1403   3.2e-012   1.3e-007\n",
    "#          TTTGCTT    AAGCAAA        123        954   1.0e-008   4.4e-004\n",
    "#          TTTGCTA    TAGCAAA         78        563   1.0e-006   4.3e-002\n",
    "letter-probability matrix: alength= 4 w= 7 nsites= 235 E= 1.3e-007\n",
    "0.000000 0.000000 0.000000 1.000000\n",
    "0.000000 0.000000 0.000000 1.000000\n",
    "0.000000 0.000000 0.000000 1.000000\n",
    "0.000000 0.000000 1.000000 0.000000\n",
    "0.000000 1.000000 0.000000 0.000000\n",
    "0.000000 0.000000 0.000000 1.000000\n",
    "0.382979 0.000000 0.000000 0.617021\n",
    "\n",
    "Jaspar NHP6B / 6A\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
