{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "what computer are you on? a = Bens, b = gpucluster, c = other   a\n",
      "base directory is C:\\Users\\heine\\github\\expression_broad_data\n",
      "I am about to import a library\n",
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n",
      "are you online? Yes/NoNo\n"
     ]
    }
   ],
   "source": [
    "# %load std_libraries.py\n",
    "#to keep my scripts consistent, I am adding this to the top of all scripts with %load std_libraries.py\n",
    "import sys\n",
    "#Indicate operating environment and import core modules\n",
    "location_input = input(\"what computer are you on? a = Bens, b = gpucluster, c = other   \")\n",
    "location_dict = {'a': \"C:\\\\Users\\\\heine\\\\github\\\\expression_broad_data\", 'b': \"/home/heineike/github/expression_broad_data\",'c':'you need to add your location to the location_dict'}\n",
    "base_dir = location_dict[location_input]\n",
    "print(\"base directory is \" + base_dir)\n",
    "\n",
    "if sys.path[-1] != base_dir:\n",
    "    sys.path.append(base_dir)\n",
    "    print(\"Added \" + base_dir + \" to path: \" )\n",
    "    print(sys.path)\n",
    "\n",
    "import os\n",
    "\n",
    "print(\"I am about to import a library\")\n",
    "from core import expression_plots \n",
    "from core import io_library \n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "# %matplotlib \n",
    "# plt.ioff()\n",
    "#%matplotlib inline\n",
    "%matplotlib notebook\n",
    "data_processing_dir = base_dir + os.sep + os.path.normpath(\"expression_data\") + os.sep\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.cm as cm\n",
    "#from matplotlib_venn import venn2\n",
    "#import seaborn as sns; sns.set(style=\"ticks\", color_codes=True)\n",
    "#from sklearn import linear_model\n",
    "import pickle\n",
    "#import subprocess\n",
    "#import networkx as nx\n",
    "#import scipy.stats as stats\n",
    "#import statsmodels.graphics.gofplots as stats_graph\n",
    "#import scipy.cluster.hierarchy as sch\n",
    "\n",
    "# from Bio import SeqIO\n",
    "# from Bio import SeqFeature as sf\n",
    "# from Bio.Alphabet import generic_dna\n",
    "# from Bio.Seq import Seq\n",
    "\n",
    "# import re\n",
    "\n",
    "from collections import Counter\n",
    "# import scipy.stats as stats\n",
    "# from itertools import chain\n",
    "#from itertools import product\n",
    "#this only works if you are online\n",
    "online_input = input(\"are you online? Yes/No\")\n",
    "if online_input == \"Yes\": \n",
    "    import plotly.plotly as py\n",
    "    import plotly.graph_objs as pygo\n",
    "    import plotly.tools as pytools\n",
    "    py.sign_in('heineike02_student','9dMTMZgJMgUP0YX0P5mQ')\n",
    "    #py.sign_in('heineike02', 'APPjKrtARaN2ZgUYIkqr')\n",
    "    \n",
    "# for phylogenetic trees: \n",
    "# from ete3 import Tree\n",
    "\n",
    "#for scraping internet data (e.g. ncbi)\n",
    "#import requests\n",
    "#from lxml import etree    #parses xml output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import KL/SC data\n",
    "fname_cross_spec_comparison = data_processing_dir + os.path.normpath(\"20170817_klscpka/cross_species_compare_ASminus_ASplus_YPD.csv\")\n",
    "kl_sc_PKA_data = pd.read_csv(fname_cross_spec_comparison, index_col=0)\n",
    "\n",
    "#Load group lists from promoter analysis\n",
    "fname = os.path.normpath(data_processing_dir + '20170817_klscpka/20180410_act_sets_P1.pkl')\n",
    "with open(fname, mode = 'rb') as f: \n",
    "    gene_list_dict = pickle.load(f)\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['sc_act', 'kl_act', 'sc_only_act', 'kl_only_act', 'sc_sckl_act', 'kl_sckl_act', 'sc_klonly_act', 'kl_sconly_act'])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_list_dict.keys()\n",
    " total = len(gene_list_dict[gene_set_name])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sc_act\n",
      "184\n",
      "kl_act\n",
      "71\n",
      "sc_only_act\n",
      "156\n",
      "kl_only_act\n",
      "47\n",
      "sc_sckl_act\n",
      "28\n",
      "kl_sckl_act\n",
      "24\n",
      "sc_klonly_act\n",
      "27\n",
      "kl_sconly_act\n",
      "112\n"
     ]
    }
   ],
   "source": [
    "for key, value in gene_list_dict.items():\n",
    "    print(key)\n",
    "    print(len(value))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'sc_klonly_act': {'no_paralog': ['YJL057C',\n",
       "   'YHR044C',\n",
       "   'YBL016W',\n",
       "   'YPR026W',\n",
       "   'YGL036W',\n",
       "   'YBR057C',\n",
       "   'YPL054W',\n",
       "   'YBR149W',\n",
       "   'YLL019C',\n",
       "   'YML070W',\n",
       "   'YIL024C',\n",
       "   'YMR063W',\n",
       "   'YGL081W',\n",
       "   'YDR446W',\n",
       "   'YPR002W',\n",
       "   'YOR386W',\n",
       "   'YOL082W',\n",
       "   'YNL010W',\n",
       "   'YHR005C',\n",
       "   'YIR017C',\n",
       "   'YAL008W',\n",
       "   'YDR227W',\n",
       "   'YDR051C'],\n",
       "  'paralogs_diff_ind_act': [],\n",
       "  'paralogs_diff_ind_low': [],\n",
       "  'paralogs_similar_ind': ['YEL060C', 'YOR003W', 'YDR259C', 'YOR028C']},\n",
       " 'sc_only_act': {'no_paralog': ['YHR087W',\n",
       "   'YML128C',\n",
       "   'YPL222W',\n",
       "   'YIL155C',\n",
       "   'YHL021C',\n",
       "   'YML004C',\n",
       "   'YOR161C',\n",
       "   'YLR149C',\n",
       "   'YKL086W',\n",
       "   'YKL065W-A',\n",
       "   'YBR126C',\n",
       "   'YOR374W',\n",
       "   'YLR251W',\n",
       "   'YMR196W',\n",
       "   'YOR348C',\n",
       "   'YOL084W',\n",
       "   'YFL014W',\n",
       "   'YBL064C',\n",
       "   'YGL010W',\n",
       "   'YKL151C',\n",
       "   'YPR151C',\n",
       "   'YHR140W',\n",
       "   'YDR119W-A',\n",
       "   'YKL103C',\n",
       "   'YHR104W',\n",
       "   'YHL024W',\n",
       "   'YNL274C',\n",
       "   'YBR230C',\n",
       "   'YBL078C',\n",
       "   'YNL200C',\n",
       "   'YGR088W',\n",
       "   'YLL039C',\n",
       "   'YDR171W',\n",
       "   'YPL123C',\n",
       "   'YBR072W',\n",
       "   'YBR006W',\n",
       "   'YML054C',\n",
       "   'YAL061W',\n",
       "   'YGR236C',\n",
       "   'YMR170C',\n",
       "   'YGR174C',\n",
       "   'YNL237W',\n",
       "   'YDR070C',\n",
       "   'YJL163C',\n",
       "   'YDL110C',\n",
       "   'YIL101C',\n",
       "   'YGR194C',\n",
       "   'YIL087C',\n",
       "   'YHR138C',\n",
       "   'YJL141C',\n",
       "   'YJR096W',\n",
       "   'YDR074W',\n",
       "   'YPL186C',\n",
       "   'YGR281W',\n",
       "   'YPL247C'],\n",
       "  'paralogs_diff_ind_act': ['YDR018C',\n",
       "   'YGR256W',\n",
       "   'YER103W',\n",
       "   'YHR160C',\n",
       "   'YHR016C',\n",
       "   'YDL204W',\n",
       "   'YJR005C-A',\n",
       "   'YMR118C',\n",
       "   'YKL163W',\n",
       "   'YPR030W',\n",
       "   'YDL085W',\n",
       "   'YHR096C',\n",
       "   'YMR271C',\n",
       "   'YNR002C',\n",
       "   'YER054C',\n",
       "   'YER067W',\n",
       "   'YBR147W',\n",
       "   'YHR097C',\n",
       "   'YIL111W',\n",
       "   'YGR248W',\n",
       "   'YFR053C',\n",
       "   'YDR034W-B',\n",
       "   'YOR178C',\n",
       "   'YMR206W',\n",
       "   'YNL093W',\n",
       "   'YFR017C',\n",
       "   'YLR080W',\n",
       "   'YPL230W',\n",
       "   'YEL041W',\n",
       "   'YOR185C',\n",
       "   'YDL130W-A',\n",
       "   'YGR043C',\n",
       "   'YIL113W',\n",
       "   'YBR241C',\n",
       "   'YCL048W-A',\n",
       "   'YOL131W',\n",
       "   'YKL187C',\n",
       "   'YML100W',\n",
       "   'YJL045W',\n",
       "   'YBR214W',\n",
       "   'YKL091C',\n",
       "   'YLR356W',\n",
       "   'YLR164W',\n",
       "   'YDL022W',\n",
       "   'YNR034W-A',\n",
       "   'YER150W',\n",
       "   'YDR453C',\n",
       "   'YBR117C',\n",
       "   'YMR105C',\n",
       "   'YBR169C',\n",
       "   'YDL021W',\n",
       "   'YKL093W',\n",
       "   'YEL039C'],\n",
       "  'paralogs_diff_ind_low': ['YBR042C',\n",
       "   'YHR183W',\n",
       "   'YBL075C',\n",
       "   'YGR239C',\n",
       "   'YFR024C-A',\n",
       "   'YDR233C',\n",
       "   'YGR169C-A',\n",
       "   'YKL141W',\n",
       "   'YJL159W',\n",
       "   'YBL101C',\n",
       "   'YMR145C',\n",
       "   'YDR345C',\n",
       "   'YML106W',\n",
       "   'YCR010C',\n",
       "   'YIL045W',\n",
       "   'YIL057C',\n",
       "   'YOL092W',\n",
       "   'YDR348C',\n",
       "   'YNL052W',\n",
       "   'YHR163W',\n",
       "   'YGL253W',\n",
       "   'YBR056W-A',\n",
       "   'YLR273C',\n",
       "   'YNR014W',\n",
       "   'YOR089C',\n",
       "   'YOL024W',\n",
       "   'YFL048C',\n",
       "   'YMR182C',\n",
       "   'YJR049C',\n",
       "   'YLR293C',\n",
       "   'YDL181W',\n",
       "   'YLR354C',\n",
       "   'YNL053W',\n",
       "   'YGL104C',\n",
       "   'YDR524C-B',\n",
       "   'YNL309W',\n",
       "   'YLR413W',\n",
       "   'YMR261C',\n",
       "   'YKL148C',\n",
       "   'YGL056C',\n",
       "   'YMR079W',\n",
       "   'YGR049W',\n",
       "   'YDR178W',\n",
       "   'YOL059W',\n",
       "   'YCR075W-A',\n",
       "   'YDR077W',\n",
       "   'YML028W',\n",
       "   'YPR074C',\n",
       "   'YKL127W',\n",
       "   'YPL106C',\n",
       "   'YOL056W',\n",
       "   'YMR081C',\n",
       "   'YJR048W'],\n",
       "  'paralogs_similar_ind': ['YFR015C', 'YLR258W', 'YLR270W', 'YOR173W']},\n",
       " 'sc_sckl_act': {'no_paralog': ['YKR076W',\n",
       "   'YNL015W',\n",
       "   'YPR160W',\n",
       "   'YJL161W',\n",
       "   'YKR049C',\n",
       "   'YLR178C',\n",
       "   'YMR169C',\n",
       "   'YBR285W',\n",
       "   'YKL107W',\n",
       "   'YIL160C',\n",
       "   'YMR107W',\n",
       "   'YIL136W',\n",
       "   'YBL049W',\n",
       "   'YGL037C',\n",
       "   'YOR289W',\n",
       "   'YEL011W',\n",
       "   'YDL169C',\n",
       "   'YMR250W',\n",
       "   'YJR008W',\n",
       "   'YGR052W'],\n",
       "  'paralogs_diff_ind_act': [],\n",
       "  'paralogs_diff_ind_low': [],\n",
       "  'paralogs_similar_ind': ['YDL222C',\n",
       "   'YNL194C',\n",
       "   'YMR175W',\n",
       "   'YPL223C',\n",
       "   'YGR008C',\n",
       "   'YLR327C',\n",
       "   'YNL195C',\n",
       "   'YDL223C']}}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Find paralogs and add paralog groups\n",
    "pka_subsets = ['sc_only_act', 'sc_sckl_act', 'sc_klonly_act']\n",
    "\n",
    "gene_sets = {}\n",
    "for pka_subset in pka_subsets: \n",
    "\n",
    "    gene_sets_subset = {}\n",
    "\n",
    "    kl_sc_PKA_data_subset_paralogs = kl_sc_PKA_data[kl_sc_PKA_data['sc_genename'].isin(gene_list_dict[pka_subset]) & (kl_sc_PKA_data['N SCer Orthologs']==2)]\n",
    "    kl_genename_counts = Counter(kl_sc_PKA_data_subset_paralogs['kl_genename'])\n",
    "\n",
    "    paralogs_diff_ind_act_kl_genename = [kl_genename for kl_genename, NN in kl_genename_counts.items() if NN==1]\n",
    "    #get the active paralog of the differentially induced paralogs\n",
    "    paralogs_diff_ind_act = list(kl_sc_PKA_data_subset_paralogs[kl_sc_PKA_data_subset_paralogs['kl_genename'].isin(paralogs_diff_ind_act_kl_genename)]['sc_genename'])\n",
    "    gene_sets_subset['paralogs_diff_ind_act']= paralogs_diff_ind_act\n",
    "\n",
    "    paralogs_diff_ind_low = list(io_library.get_other_paralogs_from_dataframe(io_library.SC_common_name_lookup(paralogs_diff_ind_act), kl_sc_PKA_data)['sc_genename'])\n",
    "    gene_sets_subset['paralogs_diff_ind_low'] = paralogs_diff_ind_low\n",
    "\n",
    "    paralogs_similar_ind_kl_genename = [kl_genename for kl_genename, NN in kl_genename_counts.items() if NN==2]\n",
    "    #all of these should be in the subset already\n",
    "    paralogs_similar_ind = list(kl_sc_PKA_data_subset_paralogs[kl_sc_PKA_data_subset_paralogs['kl_genename'].isin(paralogs_similar_ind_kl_genename)]['sc_genename'])\n",
    "    gene_sets_subset['paralogs_similar_ind'] = paralogs_similar_ind                            \n",
    "\n",
    "    kl_sc_PKA_data_subset_no_paralogs = kl_sc_PKA_data[kl_sc_PKA_data['sc_genename'].isin(gene_list_dict[pka_subset]) & (kl_sc_PKA_data['N SCer Orthologs']==1)]\n",
    "    no_paralog = list(kl_sc_PKA_data_subset_no_paralogs['sc_genename'])\n",
    "    gene_sets_subset['no_paralog'] = no_paralog\n",
    "\n",
    "    gene_sets[pka_subset] = gene_sets_subset\n",
    "\n",
    "gene_sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sc_only_act Total: 156\n",
      "Total with ortholog: 112\n",
      "paralogs_diff_ind_act: 53\n",
      "paralogs_diff_ind_low: 53\n",
      "paralogs_similar_ind: 4\n",
      "no_paralog: 55\n",
      "sc_sckl_act Total: 28\n",
      "Total with ortholog: 28\n",
      "paralogs_diff_ind_act: 0\n",
      "paralogs_diff_ind_low: 0\n",
      "paralogs_similar_ind: 8\n",
      "no_paralog: 20\n",
      "sc_klonly_act Total: 27\n",
      "Total with ortholog: 27\n",
      "paralogs_diff_ind_act: 0\n",
      "paralogs_diff_ind_low: 0\n",
      "paralogs_similar_ind: 4\n",
      "no_paralog: 23\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SCer Only Activated</th>\n",
       "      <th>SCer and KLac Activated</th>\n",
       "      <th>KLac Only Activated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>156</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total w/ Ortholog</th>\n",
       "      <td>112</td>\n",
       "      <td>28</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Diff Induced Paralogs (high induction)</th>\n",
       "      <td>53</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Diff Induced Paralogs (low induction)</th>\n",
       "      <td>53</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Similarly Induced Paralogs</th>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>No Paralogs</th>\n",
       "      <td>55</td>\n",
       "      <td>20</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        SCer Only Activated  \\\n",
       "Total                                                   156   \n",
       "Total w/ Ortholog                                       112   \n",
       "Diff Induced Paralogs (high induction)                   53   \n",
       "Diff Induced Paralogs (low induction)                    53   \n",
       "Similarly Induced Paralogs                                4   \n",
       "No Paralogs                                              55   \n",
       "\n",
       "                                        SCer and KLac Activated  \\\n",
       "Total                                                        28   \n",
       "Total w/ Ortholog                                            28   \n",
       "Diff Induced Paralogs (high induction)                        0   \n",
       "Diff Induced Paralogs (low induction)                         0   \n",
       "Similarly Induced Paralogs                                    8   \n",
       "No Paralogs                                                  20   \n",
       "\n",
       "                                        KLac Only Activated  \n",
       "Total                                                    27  \n",
       "Total w/ Ortholog                                        27  \n",
       "Diff Induced Paralogs (high induction)                    0  \n",
       "Diff Induced Paralogs (low induction)                     0  \n",
       "Similarly Induced Paralogs                                4  \n",
       "No Paralogs                                              23  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_sets_table = pd.DataFrame(index = ['Total', 'Total w/ Ortholog', \n",
    "                                        'Diff Induced Paralogs (high induction)', 'Diff Induced Paralogs (low induction)',\n",
    "                                        'Similarly Induced Paralogs', 'No Paralogs'])\n",
    "for gene_set_name, gene_set in gene_sets.items():\n",
    "    table_data = []\n",
    "    total = len(gene_list_dict[gene_set_name])\n",
    "    table_data.append(total)\n",
    "    print(gene_set_name + \" Total: \" + str(total))\n",
    "    \n",
    "    total_orth = len(kl_sc_PKA_data[kl_sc_PKA_data['sc_genename'].isin(gene_list_dict[gene_set_name])])\n",
    "    table_data.append(total_orth)\n",
    "    print(\"Total with ortholog: \" + str(total_orth))\n",
    "    for gene_subset_name, gene_subset in gene_set.items():\n",
    "        table_data.append(len(gene_subset))\n",
    "        print(gene_subset_name + \": \" + str(len(gene_subset)))\n",
    "    gene_sets_table[gene_set_name]=table_data\n",
    "gene_sets_table.rename(columns={'sc_only_act': 'SCer Only Activated', 'sc_sckl_act': 'SCer and KLac Activated', 'sc_klonly_act': 'KLac Only Activated'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'rgba(1.0, 0.49803921568627452, 0.054901960784313725, 1.0)'"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cmap = cm.get_cmap('tab10')\n",
    "color = cmap(1/10)\n",
    "'rgba'+str(color)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe id=\"igraph\" scrolling=\"no\" style=\"border:none;\" seamless=\"seamless\" src=\"https://plot.ly/~heineike02_student/102.embed\" height=\"525px\" width=\"100%\"></iframe>"
      ],
      "text/plain": [
       "<plotly.tools.PlotlyDisplay object>"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Plot raw expression for various gene sets\n",
    "\n",
    "hover_text = []\n",
    "for ind, row in kl_sc_PKA_data.iterrows():\n",
    "    hover_text.append(('{SC_common_name}<br>kl cont: {klmin:.2f}<br>kl nmpp1: {klnmpp1:.2f}').format(\n",
    "                              SC_common_name=row['SC_common_name'],\n",
    "                              klmin = row['rlog_AS_noNMPP1_YPD_t40_mean_KL'],\n",
    "                              klnmpp1 = row['rlog_AS_yesNMPP1_YPD_t40_mean_KL']\n",
    "                              ))\n",
    "\n",
    "kl_sc_PKA_data['hover'] = hover_text\n",
    "\n",
    "cmap = cm.get_cmap('tab10')\n",
    "\n",
    "data = []\n",
    "NN = 0\n",
    "for gene_set_name, gene_set in gene_sets.items():\n",
    "     for gene_subset_name, gene_subset in gene_set.items():\n",
    "        kl_sc_PKA_data_subset = kl_sc_PKA_data[kl_sc_PKA_data['sc_genename'].isin(gene_subset)]\n",
    "\n",
    "        trace = pygo.Scatter(\n",
    "            x = kl_sc_PKA_data_subset['rlog_AS_noNMPP1_YPD_t40_mean_SC'], \n",
    "            y =  kl_sc_PKA_data_subset['rlog_AS_yesNMPP1_YPD_t40_mean_SC'],\n",
    "            text = kl_sc_PKA_data_subset['hover'],\n",
    "            mode = 'markers',\n",
    "            marker = {'opacity': 0.5, \n",
    "                      'color': 'rgba'+str(cmap(NN/10))}, \n",
    "            name = gene_set_name + ': ' + gene_subset_name\n",
    "        )\n",
    "        \n",
    "        data.append(trace)\n",
    "        NN = NN+1\n",
    "\n",
    "layout = pygo.Layout(\n",
    "    xaxis= {\n",
    "        #\"range\":[-2, 20],\n",
    "        \"title\":'SC rlog control'\n",
    "    },\n",
    "    yaxis= {\n",
    "        #\"range\":[-2, 20],\n",
    "        \"title\":'SC rlog + NMPP1'\n",
    "    }\n",
    ")\n",
    "\n",
    "\n",
    "fig = pygo.Figure(data=data, layout = layout)\n",
    "py.iplot(fig, filename='raw_expression_PKA_activated_genes')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
