{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import os\n",
    "base_dir = 'C:\\Users\\Ben\\Documents\\GitHub\\expression_broad_data'\n",
    "os.chdir(base_dir) \n",
    "from core import expression_plots \n",
    "from core import io_library \n",
    "# from IPython.core.debugger import Tracer\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns\n",
    "import csv\n",
    "import scipy.cluster.hierarchy as sch\n",
    "from collections import Counter, defaultdict\n",
    "import math\n",
    "import plotly.plotly as py\n",
    "py.sign_in('heineike02', 'APPjKrtARaN2ZgUYIkqr')\n",
    "# Learn about API authentication here: https://plot.ly/pandas/getting-started\n",
    "# Find your api_key here: https://plot.ly/settings/api\n",
    "\n",
    "#from sklearn.cluster import AgglomerativeClustering\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#This file should be main one for comparing SC and KL PKA inhibition directly to one another. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Load NMPP1 Data for KL\n",
    "kl_nmpp1_lfc = pd.read_pickle(base_dir + \"\\expression_data\\kl_PKA_as_20160824\\LFC_KLac_ASmin_ASplus.pkl\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Load NMPP1 Data for SC\n",
    "\n",
    "#oshea PKA inhibition data\n",
    "sc_nmpp1_oshea = io_library.load_oshea_NMPP1_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Load KL to SC ortholog mapping\n",
    "\n",
    "kl_orthologs = pd.read_pickle(base_dir + \"\\expression_data\\ortholog_files_YGOB\\kl_orthologs.pkl\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Map KL PKA inhibition data onto related SC orthologs\n",
    "#First change column name for kl_nmpp1_lfc data to match ortholog table for merge. \n",
    "kl_nmpp1_lfc_renamed = kl_nmpp1_lfc.copy()\n",
    "kl_nmpp1_lfc_renamed.columns = ['kl_genename'] + list(kl_nmpp1_lfc.columns[1:])\n",
    "kl_PKA_data_sc_orthologs = pd.merge(kl_orthologs, kl_nmpp1_lfc_renamed, on = 'kl_genename', how = 'inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Merge SC and KL Data\n",
    "#KL data is from DEseq2, and NMPP1 is from averaged O'shea Microarrays. \n",
    "#kl_sc_PKA_data = pd.merge(kl_PKA_data_sc_orthologs,mace_SC_PKA_data, on = 'sc_genename', how = 'inner')\n",
    "kl_sc_PKA_data = kl_PKA_data_sc_orthologs.join(sc_nmpp1_oshea, on = 'sc_genename')\n",
    "\n",
    "#would be good to output here how many genes don't merge (missing orthologs in either direction)\n",
    "#Much of this is to be done in kl_sc_orthologs_pblast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#save kl_sc_PKA_data\n",
    "kl_sc_PKA_data.to_pickle(base_dir + \"\\expression_data\\kl_PKA_as_20160824\\kl_sc_PKA_data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#load kl_sc_PKA_data\n",
    "kl_sc_PKA_data = pd.read_pickle(base_dir + \"\\expression_data\\kl_PKA_as_20160824\\kl_sc_PKA_data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N SCer Orthologs</th>\n",
       "      <th>kl_genename</th>\n",
       "      <th>sc_genename</th>\n",
       "      <th>KL_PKA(AS)+1NMPP1</th>\n",
       "      <th>padj</th>\n",
       "      <th>SC_PKA(AS)+1NMPP1</th>\n",
       "      <th>SC_common_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3850</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0D04224g</td>\n",
       "      <td>YLL024C</td>\n",
       "      <td>-3.827877</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.568383</td>\n",
       "      <td>SSA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0F03333g</td>\n",
       "      <td>YNL064C</td>\n",
       "      <td>-2.271297</td>\n",
       "      <td>2.052095e-220</td>\n",
       "      <td>-1.124353</td>\n",
       "      <td>YDJ1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1171</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0C12221g</td>\n",
       "      <td>YNL281W</td>\n",
       "      <td>-1.102161</td>\n",
       "      <td>8.344354e-52</td>\n",
       "      <td>-0.640484</td>\n",
       "      <td>HCH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4435</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0D12958g</td>\n",
       "      <td>YPL240C</td>\n",
       "      <td>-0.671964</td>\n",
       "      <td>1.908416e-34</td>\n",
       "      <td>2.186125</td>\n",
       "      <td>HSP82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4434</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0D12958g</td>\n",
       "      <td>YMR186W</td>\n",
       "      <td>-0.671964</td>\n",
       "      <td>1.908416e-34</td>\n",
       "      <td>0.443065</td>\n",
       "      <td>HSC82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>633</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0D07656g</td>\n",
       "      <td>YFL016C</td>\n",
       "      <td>-0.340166</td>\n",
       "      <td>5.580102e-05</td>\n",
       "      <td>0.976233</td>\n",
       "      <td>MDJ1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3169</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0F25014g</td>\n",
       "      <td>YOR298C-A</td>\n",
       "      <td>-0.161868</td>\n",
       "      <td>4.969806e-03</td>\n",
       "      <td>0.571472</td>\n",
       "      <td>MBF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>723</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0D04818g</td>\n",
       "      <td>YNL007C</td>\n",
       "      <td>0.107579</td>\n",
       "      <td>1.774874e-01</td>\n",
       "      <td>0.536960</td>\n",
       "      <td>SIS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3331</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0A07062g</td>\n",
       "      <td>YOR027W</td>\n",
       "      <td>0.190130</td>\n",
       "      <td>1.057140e-03</td>\n",
       "      <td>0.656043</td>\n",
       "      <td>STI1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1086</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0E11375g</td>\n",
       "      <td>YBR101C</td>\n",
       "      <td>0.248760</td>\n",
       "      <td>2.182012e-04</td>\n",
       "      <td>0.057950</td>\n",
       "      <td>FES1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>585</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0A04147g</td>\n",
       "      <td>YDR214W</td>\n",
       "      <td>0.310917</td>\n",
       "      <td>7.737007e-06</td>\n",
       "      <td>0.055784</td>\n",
       "      <td>AHA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3677</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0F05093g</td>\n",
       "      <td>YLR216C</td>\n",
       "      <td>0.321984</td>\n",
       "      <td>4.569103e-06</td>\n",
       "      <td>0.651441</td>\n",
       "      <td>CPR6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4575</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0E03763g</td>\n",
       "      <td>YPR158W</td>\n",
       "      <td>1.187937</td>\n",
       "      <td>4.382969e-45</td>\n",
       "      <td>0.135912</td>\n",
       "      <td>CUR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4574</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0E03763g</td>\n",
       "      <td>YGR142W</td>\n",
       "      <td>1.187937</td>\n",
       "      <td>4.382969e-45</td>\n",
       "      <td>1.056388</td>\n",
       "      <td>BTN2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3635</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0F18194g</td>\n",
       "      <td>YLL026W</td>\n",
       "      <td>1.874518</td>\n",
       "      <td>2.518469e-29</td>\n",
       "      <td>3.405946</td>\n",
       "      <td>HSP104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3260</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0E16897g</td>\n",
       "      <td>YDR258C</td>\n",
       "      <td>2.409128</td>\n",
       "      <td>9.864244e-151</td>\n",
       "      <td>1.508280</td>\n",
       "      <td>HSP78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2394</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0D17446g</td>\n",
       "      <td>YDR171W</td>\n",
       "      <td>2.908165</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.173787</td>\n",
       "      <td>HSP42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      N SCer Orthologs   kl_genename sc_genename  KL_PKA(AS)+1NMPP1  \\\n",
       "3850                 1  KLLA0D04224g     YLL024C          -3.827877   \n",
       "564                  1  KLLA0F03333g     YNL064C          -2.271297   \n",
       "1171                 1  KLLA0C12221g     YNL281W          -1.102161   \n",
       "4435                 2  KLLA0D12958g     YPL240C          -0.671964   \n",
       "4434                 2  KLLA0D12958g     YMR186W          -0.671964   \n",
       "633                  1  KLLA0D07656g     YFL016C          -0.340166   \n",
       "3169                 1  KLLA0F25014g   YOR298C-A          -0.161868   \n",
       "723                  1  KLLA0D04818g     YNL007C           0.107579   \n",
       "3331                 1  KLLA0A07062g     YOR027W           0.190130   \n",
       "1086                 1  KLLA0E11375g     YBR101C           0.248760   \n",
       "585                  1  KLLA0A04147g     YDR214W           0.310917   \n",
       "3677                 1  KLLA0F05093g     YLR216C           0.321984   \n",
       "4575                 2  KLLA0E03763g     YPR158W           1.187937   \n",
       "4574                 2  KLLA0E03763g     YGR142W           1.187937   \n",
       "3635                 1  KLLA0F18194g     YLL026W           1.874518   \n",
       "3260                 1  KLLA0E16897g     YDR258C           2.409128   \n",
       "2394                 1  KLLA0D17446g     YDR171W           2.908165   \n",
       "\n",
       "               padj  SC_PKA(AS)+1NMPP1 SC_common_name  \n",
       "3850   0.000000e+00           0.568383           SSA2  \n",
       "564   2.052095e-220          -1.124353           YDJ1  \n",
       "1171   8.344354e-52          -0.640484           HCH1  \n",
       "4435   1.908416e-34           2.186125          HSP82  \n",
       "4434   1.908416e-34           0.443065          HSC82  \n",
       "633    5.580102e-05           0.976233           MDJ1  \n",
       "3169   4.969806e-03           0.571472           MBF1  \n",
       "723    1.774874e-01           0.536960           SIS1  \n",
       "3331   1.057140e-03           0.656043           STI1  \n",
       "1086   2.182012e-04           0.057950           FES1  \n",
       "585    7.737007e-06           0.055784           AHA1  \n",
       "3677   4.569103e-06           0.651441           CPR6  \n",
       "4575   4.382969e-45           0.135912           CUR1  \n",
       "4574   4.382969e-45           1.056388           BTN2  \n",
       "3635   2.518469e-29           3.405946         HSP104  \n",
       "3260  9.864244e-151           1.508280          HSP78  \n",
       "2394   0.000000e+00           4.173787          HSP42  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Lookup by SC common name\n",
    "jso_msn2_targets = ['SSA1', 'SSA4', 'HSP104', 'CTT1', 'TPK1', 'TFS1', 'DDR2', 'TSA1', 'HSP12', 'TMA10', 'TPS2', 'RTC3', 'GPH1', 'HXK1', 'TSA2', 'CYC7', 'PNC1', 'GSY1', 'FMP48', 'SS13', 'ALD4', 'YNR014W', 'HOR2', 'YOR173W', 'RAS2', 'GLK1', 'GPD1', 'HSP26', 'SIP18', 'TKL2', 'FMP16', 'ALD3', 'PGM2']\n",
    "common_name_query = ['SSA1', 'SSA2','AHA1','BTN2', 'CPR6','CUR1','FES1','HCH1','HSC82','HSP42','HSP104','HSP78','HSP82','MBF1','MDJ1','SIS1','STI1','YDJ1']\n",
    "kl_sc_PKA_data[kl_sc_PKA_data['SC_common_name'].isin(common_name_query)].sort_values(by='KL_PKA(AS)+1NMPP1')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>N SCer Orthologs</th>\n",
       "      <th>kl_genename</th>\n",
       "      <th>sc_genename</th>\n",
       "      <th>KL_PKA(AS)+1NMPP1</th>\n",
       "      <th>padj</th>\n",
       "      <th>SC_PKA(AS)+1NMPP1</th>\n",
       "      <th>SC_common_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4855</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0C04004g</td>\n",
       "      <td>YOL059W</td>\n",
       "      <td>-1.419206</td>\n",
       "      <td>1.077434e-144</td>\n",
       "      <td>-1.046553</td>\n",
       "      <td>GPD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4854</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0C04004g</td>\n",
       "      <td>YDL022W</td>\n",
       "      <td>-1.419206</td>\n",
       "      <td>1.077434e-144</td>\n",
       "      <td>2.991343</td>\n",
       "      <td>GPD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3635</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0F18194g</td>\n",
       "      <td>YLL026W</td>\n",
       "      <td>1.874518</td>\n",
       "      <td>2.518469e-29</td>\n",
       "      <td>3.405946</td>\n",
       "      <td>HSP104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0F10065g</td>\n",
       "      <td>YPR160W</td>\n",
       "      <td>2.011643</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.375787</td>\n",
       "      <td>GPH1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2503</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0F18909g</td>\n",
       "      <td>YBR072W</td>\n",
       "      <td>2.044388</td>\n",
       "      <td>5.699660e-52</td>\n",
       "      <td>3.584261</td>\n",
       "      <td>HSP26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4073</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0F23133g</td>\n",
       "      <td>YLR258W</td>\n",
       "      <td>2.588160</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.842569</td>\n",
       "      <td>GSY2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4072</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0F23133g</td>\n",
       "      <td>YFR015C</td>\n",
       "      <td>2.588160</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.083677</td>\n",
       "      <td>GSY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4908</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0B12694g</td>\n",
       "      <td>YKL127W</td>\n",
       "      <td>2.897526</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-0.375948</td>\n",
       "      <td>PGM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4909</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0B12694g</td>\n",
       "      <td>YMR105C</td>\n",
       "      <td>2.897526</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.988579</td>\n",
       "      <td>PGM2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3913</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0A02717g</td>\n",
       "      <td>YGR052W</td>\n",
       "      <td>2.998799</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.409699</td>\n",
       "      <td>FMP48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0C01738g</td>\n",
       "      <td>YLR178C</td>\n",
       "      <td>3.268890</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.335641</td>\n",
       "      <td>TFS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1984</th>\n",
       "      <td>1</td>\n",
       "      <td>KLLA0F22242g</td>\n",
       "      <td>YGL037C</td>\n",
       "      <td>3.437444</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>3.685762</td>\n",
       "      <td>PNC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4069</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0E20527g</td>\n",
       "      <td>YER103W</td>\n",
       "      <td>3.581376</td>\n",
       "      <td>9.596069e-122</td>\n",
       "      <td>2.775551</td>\n",
       "      <td>SSA4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4068</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0E20527g</td>\n",
       "      <td>YBL075C</td>\n",
       "      <td>3.581376</td>\n",
       "      <td>9.596069e-122</td>\n",
       "      <td>1.296421</td>\n",
       "      <td>SSA3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4518</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0C18117g</td>\n",
       "      <td>YGR008C</td>\n",
       "      <td>5.914304</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.809141</td>\n",
       "      <td>STF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4519</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0C18117g</td>\n",
       "      <td>YLR327C</td>\n",
       "      <td>5.914304</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.868421</td>\n",
       "      <td>TMA10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4500</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0C16225g</td>\n",
       "      <td>YMR175W</td>\n",
       "      <td>6.911274</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.263623</td>\n",
       "      <td>SIP18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4501</th>\n",
       "      <td>2</td>\n",
       "      <td>KLLA0C16225g</td>\n",
       "      <td>YPL223C</td>\n",
       "      <td>6.911274</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.564510</td>\n",
       "      <td>GRE1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      N SCer Orthologs   kl_genename sc_genename  KL_PKA(AS)+1NMPP1  \\\n",
       "4855                 2  KLLA0C04004g     YOL059W          -1.419206   \n",
       "4854                 2  KLLA0C04004g     YDL022W          -1.419206   \n",
       "3635                 1  KLLA0F18194g     YLL026W           1.874518   \n",
       "67                   1  KLLA0F10065g     YPR160W           2.011643   \n",
       "2503                 1  KLLA0F18909g     YBR072W           2.044388   \n",
       "4073                 2  KLLA0F23133g     YLR258W           2.588160   \n",
       "4072                 2  KLLA0F23133g     YFR015C           2.588160   \n",
       "4908                 2  KLLA0B12694g     YKL127W           2.897526   \n",
       "4909                 2  KLLA0B12694g     YMR105C           2.897526   \n",
       "3913                 1  KLLA0A02717g     YGR052W           2.998799   \n",
       "170                  1  KLLA0C01738g     YLR178C           3.268890   \n",
       "1984                 1  KLLA0F22242g     YGL037C           3.437444   \n",
       "4069                 2  KLLA0E20527g     YER103W           3.581376   \n",
       "4068                 2  KLLA0E20527g     YBL075C           3.581376   \n",
       "4518                 2  KLLA0C18117g     YGR008C           5.914304   \n",
       "4519                 2  KLLA0C18117g     YLR327C           5.914304   \n",
       "4500                 2  KLLA0C16225g     YMR175W           6.911274   \n",
       "4501                 2  KLLA0C16225g     YPL223C           6.911274   \n",
       "\n",
       "               padj  SC_PKA(AS)+1NMPP1 SC_common_name  \n",
       "4855  1.077434e-144          -1.046553           GPD2  \n",
       "4854  1.077434e-144           2.991343           GPD1  \n",
       "3635   2.518469e-29           3.405946         HSP104  \n",
       "67     0.000000e+00           5.375787           GPH1  \n",
       "2503   5.699660e-52           3.584261          HSP26  \n",
       "4073   0.000000e+00           3.842569           GSY2  \n",
       "4072   0.000000e+00           3.083677           GSY1  \n",
       "4908   0.000000e+00          -0.375948           PGM1  \n",
       "4909   0.000000e+00           4.988579           PGM2  \n",
       "3913   0.000000e+00           3.409699          FMP48  \n",
       "170    0.000000e+00           4.335641           TFS1  \n",
       "1984   0.000000e+00           3.685762           PNC1  \n",
       "4069  9.596069e-122           2.775551           SSA4  \n",
       "4068  9.596069e-122           1.296421           SSA3  \n",
       "4518   0.000000e+00           4.809141           STF2  \n",
       "4519   0.000000e+00           4.868421          TMA10  \n",
       "4500   0.000000e+00           4.263623          SIP18  \n",
       "4501   0.000000e+00           1.564510           GRE1  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Lookup by kl_genename common name\n",
    "stre_eq1 = [kl_gene for kl_gene, n in jso_prom_hits_kl_hitcount.items() if n==1]\n",
    "stre_gt1 = [kl_gene for kl_gene, n in jso_prom_hits_kl_hitcount.items() if n>1]\n",
    "common_name_query = []\n",
    "kl_sc_PKA_data[kl_sc_PKA_data['kl_genename'].isin(all_stre)].sort_values(by='KL_PKA(AS)+1NMPP1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-80-948478efb9aa>, line 17)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-80-948478efb9aa>\"\u001b[1;36m, line \u001b[1;32m17\u001b[0m\n\u001b[1;33m    {\u001b[0m\n\u001b[1;37m    ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "#Highlight hits that are statistically significant for K.Lactis \n",
    "\n",
    "#Add column which highlights only pvalues below a threshold\n",
    "pthreshold = 10e-25\n",
    "\n",
    "kl_sc_PKA_data_sig = kl_sc_PKA_data[kl_sc_PKA_data['padj'] < pthreshold]\n",
    "kl_sc_PKA_data_unsig = kl_sc_PKA_data[kl_sc_PKA_data['padj'] > pthreshold]\n",
    "\n",
    "fig = {\n",
    "    'data': [\n",
    "  \t\t{\n",
    "  \t\t\t'x': kl_sc_PKA_data_sig['KL_PKA(AS)+1NMPP1'], \n",
    "        \t'y': kl_sc_PKA_data_sig['SC_PKA(AS)+1NMPP1'], \n",
    "        \t'text': kl_sc_PKA_data_sig['SC_common_name'], \n",
    "        \t'mode': 'markers', \n",
    "        \t'name': 'kl padj < ' + str(pthreshold)},\n",
    "        {\n",
    "        \t'x': kl_sc_PKA_data_unsig['KL_PKA(AS)+1NMPP1'], \n",
    "        \t'y': kl_sc_PKA_data_unsig['SC_PKA(AS)+1NMPP1'], \n",
    "        \t'text': kl_sc_PKA_data_unsig['SC_common_name'], \n",
    "        \t'mode': 'markers', \n",
    "        \t'name': 'kl padj > ' + str(pthreshold)}\n",
    "    ],\n",
    "    'layout': {\n",
    "        'xaxis': {'title': 'KL LFC 4uM NMPP1'},\n",
    "        'yaxis': {'title': \"SC averaged Fold Change\"}\n",
    "    }\n",
    "}\n",
    "\n",
    "# IPython notebook\n",
    "# py.iplot(fig, filename='pandas/multiple-scatter')\n",
    "\n",
    "url = py.plot(fig, filename='sc_kl_scatter_sig_highlight')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From Stewart-Ornstein et al 2013  http://www.sciencedirect.com/science/article/pii/S0960982213011901 there were 32 MSN2 target promoters that had \"measureable basal expression and greater than 4-fold induction in one of the perturbations and lost at least 50% of this induction in a Δmsn2/4 background\".  This is out of 40 that were identified from Gasch et al 2000, Huebert et al 2014, and Capaldi et al 2008. Those genes were: \n",
    "\n",
    "['SSA1', 'SSA4', 'HSP104', 'CTT1', 'TPK1', 'TFS1', 'DDR2', 'TSA1', 'HSP12', 'TMA10', 'TPS2', 'RTC3', 'GPH1', 'HXK1', 'TSA2', 'CYC7', 'PNC1', 'GSY1', 'FMP48', 'SS13', 'ALD4', 'YNR014W', 'HOR2', 'YOR173W', 'RAS2', 'GLK1', 'GPD1', 'HSP26', 'SIP18', 'TKL2', 'FMP16', 'ALD3']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Highlight hits that are MSN2 Hits for K.Lactis \n",
    "\n",
    "kl_sc_PKA_data_msn2 = kl_sc_PKA_data[kl_sc_PKA_data['SC_common_name'].isin(jso_msn2_targets)]\n",
    "kl_sc_PKA_data_notmsn2 = kl_sc_PKA_data[~kl_sc_PKA_data['SC_common_name'].isin(jso_msn2_targets)]\n",
    "\n",
    "fig = {\n",
    "    'data': [\n",
    "  \t\t{\n",
    "        \t'x': kl_sc_PKA_data_notmsn2['KL_PKA(AS)+1NMPP1'], \n",
    "        \t'y': kl_sc_PKA_data_notmsn2['SC_PKA(AS)+1NMPP1'], \n",
    "        \t'text': kl_sc_PKA_data_notmsn2['SC_common_name'], \n",
    "        \t'mode': 'markers', \n",
    "        \t'name': 'rest of genes'},\n",
    "  \t\t{\n",
    "  \t\t\t'x': kl_sc_PKA_data_msn2['KL_PKA(AS)+1NMPP1'], \n",
    "        \t'y': kl_sc_PKA_data_msn2['SC_PKA(AS)+1NMPP1'], \n",
    "        \t'text': kl_sc_PKA_data_msn2['SC_common_name'], \n",
    "        \t'mode': 'markers', \n",
    "        \t'name': 'JSO Msn2 Targets'}\n",
    "            ],\n",
    "    'layout': {\n",
    "        'xaxis': {'title': 'KL LFC 4uM NMPP1'},\n",
    "        'yaxis': {'title': \"SC averaged Fold Change\"}\n",
    "    }\n",
    "}\n",
    "\n",
    "# IPython notebook\n",
    "# py.iplot(fig, filename='pandas/multiple-scatter')\n",
    "\n",
    "url = py.plot(fig, filename='sc_kl_scatter_jso_msn2')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Highlight hits that are MSN2 Hits for K.Lactis \n",
    "kl_sc_PKA_data_msn2 = kl_sc_PKA_data[kl_sc_PKA_data['SC_common_name'].isin(jso_msn2_targets)]\n",
    "kl_sc_PKA_data_msn2_stre_eq1 = kl_sc_PKA_data_msn2[kl_sc_PKA_data_msn2['kl_genename'].isin(stre_eq1)]\n",
    "kl_sc_PKA_data_msn2_stre_gt1 = kl_sc_PKA_data_msn2[kl_sc_PKA_data_msn2['kl_genename'].isin(stre_gt1)]\n",
    "kl_sc_PKA_data_msn2_stre_no = kl_sc_PKA_data_msn2[~kl_sc_PKA_data_msn2['kl_genename'].isin(stre_eq1+stre_gt1)]\n",
    "kl_sc_PKA_data_notmsn2 = kl_sc_PKA_data[~kl_sc_PKA_data['SC_common_name'].isin(jso_msn2_targets)]\n",
    "\n",
    "fig = {\n",
    "    'data': [\n",
    "  \t\t{\n",
    "        \t'x': kl_sc_PKA_data_notmsn2['KL_PKA(AS)+1NMPP1'], \n",
    "        \t'y': kl_sc_PKA_data_notmsn2['SC_PKA(AS)+1NMPP1'], \n",
    "        \t'text': kl_sc_PKA_data_notmsn2['SC_common_name'], \n",
    "        \t'mode': 'markers', \n",
    "        \t'name': 'rest of genes'},\n",
    "\t\t{\n",
    "        \t'x': kl_sc_PKA_data_msn2_stre_no['KL_PKA(AS)+1NMPP1'], \n",
    "        \t'y': kl_sc_PKA_data_msn2_stre_no['SC_PKA(AS)+1NMPP1'], \n",
    "        \t'text': kl_sc_PKA_data_msn2_stre_no['SC_common_name'], \n",
    "        \t'mode': 'markers', \n",
    "        \t'name': 'JSO Msn2 Targets no stre'},\n",
    "\t\t{\n",
    "        \t'x': kl_sc_PKA_data_msn2_stre_eq1['KL_PKA(AS)+1NMPP1'], \n",
    "        \t'y': kl_sc_PKA_data_msn2_stre_eq1['SC_PKA(AS)+1NMPP1'], \n",
    "        \t'text': kl_sc_PKA_data_msn2_stre_eq1['SC_common_name'], \n",
    "        \t'mode': 'markers', \n",
    "        \t'name': 'JSO Msn2 Targets 1 stre'},\n",
    "  \t\t{\n",
    "  \t\t\t'x': kl_sc_PKA_data_msn2_stre_gt1['KL_PKA(AS)+1NMPP1'], \n",
    "        \t'y': kl_sc_PKA_data_msn2_stre_gt1['SC_PKA(AS)+1NMPP1'], \n",
    "        \t'text': kl_sc_PKA_data_msn2_stre_gt1['SC_common_name'], \n",
    "        \t'mode': 'markers', \n",
    "        \t'name': 'JSO Msn2 Targets >1 stre'}\n",
    "            ],\n",
    "    'layout': {\n",
    "        'xaxis': {'title': 'KL LFC 4uM NMPP1'},\n",
    "        'yaxis': {'title': \"SC averaged Fold Change\"}\n",
    "    }\n",
    "}\n",
    "\n",
    "# IPython notebook\n",
    "# py.iplot(fig, filename='pandas/multiple-scatter')\n",
    "\n",
    "url = py.plot(fig, filename='sc_kl_scatter_jso_msn2_stre')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Plot SC \"ESR activated\" and \"ESR repressed\" genes onto SC/VS KL plot\n",
    "# #For activation and repression\n",
    "ESR_types = ['activated','repressed']\n",
    "ESR_genes = {}\n",
    "for ESR_type in ESR_types:\n",
    "     ESR_genes[ESR_type] = io_library.get_gasch_ESR_list(ESR_type)\n",
    "\n",
    "#241 activated genes\n",
    "#539 repressed genes\n",
    "#4196 remaining genes\n",
    "kl_sc_PKA_data_ESR_activated = kl_sc_PKA_data[kl_sc_PKA_data['sc_genename'].isin(ESR_genes['activated'])]\n",
    "kl_sc_PKA_data_ESR_repressed = kl_sc_PKA_data[kl_sc_PKA_data['sc_genename'].isin(ESR_genes['repressed'])]\n",
    "kl_sc_PKA_data_notESR = kl_sc_PKA_data[~kl_sc_PKA_data['sc_genename'].isin(ESR_genes['repressed']+ESR_genes['activated'])]\n",
    "\n",
    "fig = {\n",
    "    'data': [\n",
    "  \t\t{\n",
    "        \t'x': kl_sc_PKA_data_notESR['KL_PKA(AS)+1NMPP1'], \n",
    "        \t'y': kl_sc_PKA_data_notESR['SC_PKA(AS)+1NMPP1'], \n",
    "        \t'text': kl_sc_PKA_data_notESR['SC_common_name'], \n",
    "        \t'mode': 'markers', \n",
    "        \t'name': 'non-ESR genes'},\n",
    "  \t\t{\n",
    "        \t'x': kl_sc_PKA_data_ESR_activated['KL_PKA(AS)+1NMPP1'], \n",
    "        \t'y': kl_sc_PKA_data_ESR_activated['SC_PKA(AS)+1NMPP1'], \n",
    "        \t'text': kl_sc_PKA_data_ESR_activated['SC_common_name'], \n",
    "        \t'mode': 'markers', \n",
    "        \t'name': 'ESR Activated genes'},\n",
    "  \t\t{\n",
    "  \t\t\t'x': kl_sc_PKA_data_ESR_repressed['KL_PKA(AS)+1NMPP1'], \n",
    "        \t'y': kl_sc_PKA_data_ESR_repressed['SC_PKA(AS)+1NMPP1'], \n",
    "        \t'text': kl_sc_PKA_data_ESR_repressed['SC_common_name'], \n",
    "        \t'mode': 'markers', \n",
    "        \t'name': 'ESR Repressed genes'}\n",
    "            ],\n",
    "    'layout': {\n",
    "        'xaxis': {'title': 'KL LFC 4uM NMPP1'},\n",
    "        'yaxis': {'title': \"SC averaged Fold Change\"}\n",
    "    }\n",
    "}\n",
    "\n",
    "url = py.plot(fig, filename='ESR Activated and Repressed genes from Gasch Data')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index([u'#pattern name', u'sequence name', u'start', u'stop', u'strand',\n",
       "       u'score', u'p-value', u'q-value', u'matched sequence'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "jso_prom_hits_kl_orthologs = pd.read_table(base_dir + \"\\expression_data\\kl_PKA_as_20160824\\\\fimo_MSN2_KL_jso_targets.txt\")\n",
    "jso_prom_hits_kl_orthologs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['KLLA0F18909g',\n",
       " 'KLLA0B12694g',\n",
       " 'KLLA0C18117g',\n",
       " 'KLLA0C16225g',\n",
       " 'KLLA0F10065g',\n",
       " 'KLLA0E20527g']"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "jso_prom_hits_kl_orthologs_exact = jso_prom_hits_kl_orthologs[jso_prom_hits_kl_orthologs['matched sequence'].isin(['AGGGG'])]\n",
    "jso_prom_hits_kl_hitcount = Counter(jso_prom_hits_kl_orthologs_exact['sequence name'])# counts the elements' frequency\n",
    "jso_prom_hits_kl_hitcount.values() \n",
    "[kl_gene for kl_gene, n in jso_prom_hits_kl_hitcount.items() if n>1]\n",
    "#{k: v for k, v in points.items() if v[0] < 5}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#pattern name</th>\n",
       "      <th>sequence name</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>p-value</th>\n",
       "      <th>q-value</th>\n",
       "      <th>matched sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MA0341.1</td>\n",
       "      <td>KLLA0C04004g</td>\n",
       "      <td>86</td>\n",
       "      <td>90</td>\n",
       "      <td>+</td>\n",
       "      <td>9.72368</td>\n",
       "      <td>0.000709</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MA0341.1</td>\n",
       "      <td>KLLA0C04004g</td>\n",
       "      <td>87</td>\n",
       "      <td>91</td>\n",
       "      <td>+</td>\n",
       "      <td>9.09211</td>\n",
       "      <td>0.001290</td>\n",
       "      <td>1</td>\n",
       "      <td>GGGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>MA0341.1</td>\n",
       "      <td>KLLA0C04004g</td>\n",
       "      <td>352</td>\n",
       "      <td>356</td>\n",
       "      <td>+</td>\n",
       "      <td>6.90789</td>\n",
       "      <td>0.001870</td>\n",
       "      <td>1</td>\n",
       "      <td>CGGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>MA0341.1</td>\n",
       "      <td>KLLA0C04004g</td>\n",
       "      <td>116</td>\n",
       "      <td>120</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.42105</td>\n",
       "      <td>0.002580</td>\n",
       "      <td>1</td>\n",
       "      <td>TGGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>MA0341.1</td>\n",
       "      <td>KLLA0C04004g</td>\n",
       "      <td>201</td>\n",
       "      <td>205</td>\n",
       "      <td>-</td>\n",
       "      <td>-1.42105</td>\n",
       "      <td>0.002580</td>\n",
       "      <td>1</td>\n",
       "      <td>TGGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>MA0341.1</td>\n",
       "      <td>KLLA0C04004g</td>\n",
       "      <td>175</td>\n",
       "      <td>179</td>\n",
       "      <td>-</td>\n",
       "      <td>-2.35526</td>\n",
       "      <td>0.006040</td>\n",
       "      <td>1</td>\n",
       "      <td>AGGGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>MA0341.1</td>\n",
       "      <td>KLLA0C04004g</td>\n",
       "      <td>176</td>\n",
       "      <td>180</td>\n",
       "      <td>-</td>\n",
       "      <td>-2.35526</td>\n",
       "      <td>0.006040</td>\n",
       "      <td>1</td>\n",
       "      <td>AAGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>MA0341.1</td>\n",
       "      <td>KLLA0C04004g</td>\n",
       "      <td>664</td>\n",
       "      <td>668</td>\n",
       "      <td>+</td>\n",
       "      <td>-2.35526</td>\n",
       "      <td>0.006040</td>\n",
       "      <td>1</td>\n",
       "      <td>AGAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>MA0341.1</td>\n",
       "      <td>KLLA0C04004g</td>\n",
       "      <td>303</td>\n",
       "      <td>307</td>\n",
       "      <td>-</td>\n",
       "      <td>-2.40789</td>\n",
       "      <td>0.008880</td>\n",
       "      <td>1</td>\n",
       "      <td>ACGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>MA0341.1</td>\n",
       "      <td>KLLA0C04004g</td>\n",
       "      <td>368</td>\n",
       "      <td>372</td>\n",
       "      <td>-</td>\n",
       "      <td>-2.40789</td>\n",
       "      <td>0.008880</td>\n",
       "      <td>1</td>\n",
       "      <td>AGCGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    #pattern name sequence name  start  stop strand    score   p-value  \\\n",
       "1        MA0341.1  KLLA0C04004g     86    90      +  9.72368  0.000709   \n",
       "20       MA0341.1  KLLA0C04004g     87    91      +  9.09211  0.001290   \n",
       "41       MA0341.1  KLLA0C04004g    352   356      +  6.90789  0.001870   \n",
       "49       MA0341.1  KLLA0C04004g    116   120      - -1.42105  0.002580   \n",
       "55       MA0341.1  KLLA0C04004g    201   205      - -1.42105  0.002580   \n",
       "86       MA0341.1  KLLA0C04004g    175   179      - -2.35526  0.006040   \n",
       "87       MA0341.1  KLLA0C04004g    176   180      - -2.35526  0.006040   \n",
       "214      MA0341.1  KLLA0C04004g    664   668      + -2.35526  0.006040   \n",
       "229      MA0341.1  KLLA0C04004g    303   307      - -2.40789  0.008880   \n",
       "234      MA0341.1  KLLA0C04004g    368   372      - -2.40789  0.008880   \n",
       "\n",
       "     q-value matched sequence  \n",
       "1          1            AGGGG  \n",
       "20         1            GGGGG  \n",
       "41         1            CGGGG  \n",
       "49         1            TGGGG  \n",
       "55         1            TGGGG  \n",
       "86         1            AGGGA  \n",
       "87         1            AAGGG  \n",
       "214        1            AGAGG  \n",
       "229        1            ACGGG  \n",
       "234        1            AGCGG  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "jso_prom_hits_kl_orthologs[jso_prom_hits_kl_orthologs['sequence name'].isin(['KLLA0F00440g','KLLA0F06138g', 'KLLA0D09999g'])]\n",
    "jso_prom_hits_kl_orthologs[jso_prom_hits_kl_orthologs['sequence name'].isin(['KLLA0C01155g','KLLA0F21736g', 'KLLA0E10253g'])]\n",
    "jso_prom_hits_kl_orthologs[jso_prom_hits_kl_orthologs['sequence name'].isin(['KLLA0C04004g'])]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Plot Msn2 hits on top of correlation plot between SCer and KLac\n",
    "# fname = \"gasch_fig7_all_conditions.txt\"\n",
    "# gasch_Msn24_top_hits = io_library.read_gasch_data([conditions_Msn24[0]],fname) #only need indices so using dummy variable. \n",
    "# orth_expression_df['orfs']=orth_expression_df['SCer_gene']\n",
    "# top_hit_check = lambda orf: orf in set(gasch_Msn24_top_hits.index)\n",
    "# orth_expression_df['Msn24_hit'] = orth_expression_df['orfs'].apply(top_hit_check)\n",
    "\n",
    "# sns.pairplot(orth_expression_df, x_vars = ['KLac_exp'], y_vars = 'SCer_exp', hue = 'Msn24_hit')\n",
    "\n",
    "\n",
    "\n",
    "#Highlight Msn2 Targets\n",
    "\n",
    "# #Plot Msn2 hits on top of correlation plot between SCer and KLac\n",
    "# fname = \"gasch_fig7_all_conditions.txt\"\n",
    "# gasch_Msn24_top_hits = io_library.read_gasch_data([conditions_Msn24[0]],fname) #only need indices so using dummy variable. \n",
    "# orth_expression_df['orfs']=orth_expression_df['SCer_gene']\n",
    "# top_hit_check = lambda orf: orf in set(gasch_Msn24_top_hits.index)\n",
    "# orth_expression_df['Msn24_hit'] = orth_expression_df['orfs'].apply(top_hit_check)\n",
    "\n",
    "# sns.pairplot(orth_expression_df, x_vars = ['KLac_exp'], y_vars = 'SCer_exp', hue = 'Msn24_hit')\n",
    "\n",
    "\n",
    "# fname = \"gasch_complete_dataset.txt\"\n",
    "# conditions_Msn24 = ['DBY7286 37degree heat - 20 min', 'DBYmsn2-4- 37degree heat - 20 min',  'DBYmsn2/4 (real strain) + 37degrees (20 min)', 'DBY7286 + 0.3 mM H2O2 (20 min)','DBYmsn2msn4 (good strain) + 0.32 mM H2O2','DBYmsn2/4 (real strain) + 0.32 mM H2O2 (20 min)','Msn2 overexpression (repeat)','Msn4 overexpression']\n",
    "# gasch_Msn24_data = io_library.read_gasch_data(conditions_Msn24,fname)\n",
    "# gasch_Msn24_data.columns = ['WT HS','msn2/4 HS 1','msn2/4 HS2','WT Ox','msn2/4 Ox 1', 'msn2/4 Ox 2', 'msn2 Overexp', 'msn4 Overexp']\n",
    "# #fig, axarr = expression_plots.multi_scatter_plot(gasch_Msn24_data, conditions) # xlim = xlim, ylim = ylim)\n",
    "\n",
    "# #Get top Msn24 hits\n",
    "# fname = \"gasch_fig7_all_conditions.txt\"\n",
    "# gasch_Msn24_top_hits = io_library.read_gasch_data([conditions_Msn24[0]],fname) #only need indices so using dummy variable. \n",
    "# gasch_Msn24_data['orfs']=gasch_Msn24_data.index\n",
    "# top_hit_check = lambda orf: orf in set(gasch_Msn24_top_hits.index)\n",
    "# gasch_Msn24_data['Msn24_hit'] = gasch_Msn24_data['orfs'].apply(top_hit_check)\n",
    "\n",
    "# conditions = ['WT HS','msn2/4 HS 1','msn2/4 HS2','WT Ox','msn2/4 Ox 1', 'msn2/4 Ox 2', 'msn2 Overexp', 'msn4 Overexp']\n",
    "# sns.pairplot(gasch_Msn24_data,x_vars=[conditions[3]],y_vars = [conditions[5]], hue=\"Msn24_hit\")\n",
    "# #sns.pairplot(gasch_Msn24_data,x_vars=[conditions[0],conditions[3]],y_vars = [conditions[2],conditions[5],conditions[6]], hue=\"Msn24_hit\")\n",
    "# #sns.lmplot(x=conditions[1], y=conditions[3], data=gasch_Msn24_data, hue=\"Msn24_hit\")\n",
    "#            #col_wrap=2, ci=None, palette=\"muted\", size=4,\n",
    "#            #scatter_kws={\"s\": 50, \"alpha\": 1})\n",
    "        \n",
    "        \n",
    "#Hightlight Msn2 Targets from Jacob's Paper. \n",
    "\n",
    "#Highlight Statistically significant items.  \n",
    "\n",
    "#Look for STRE in sets of Msn2 targets (all SC targets, all KL targets, Targets with shared or different PKA induction)\n",
    "#Unbiased STRE search in KLac genes\n",
    "#unbiased motif search in activated / repressed KLac genes\n",
    "\n",
    "#Dot6 Targets\n",
    "#Glucose Gropout Targets\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Import SC Count data \n",
    "# #Load Kieran's PKA inhibition data\n",
    "# fname_mace_SC_PKA_data = os.path.normpath(base_dir + '\\expression_data\\SCer_NMPP1_RNA_Seq\\SC_1NMPP1_From_Kieran.tsv')\n",
    "# mace_SC_PKA_data = pd.read_table(fname_mace_SC_PKA_data, sep = ' ')\n",
    "# mace_SC_PKA_data.columns = ['sc_genename','SC_WT_minus','SC_AS_plus','SC_WT_plus']\n",
    "\n",
    "\n",
    "# #Makes a new column that combines sc_genename and kl_genename into a single index, and stores a subset of the data with just the counts\n",
    "# #as a .csv in order to import to R. \n",
    "# kl_sc_PKA_conditions = ['AS1_minus', 'AS1_plus','AS2_minus', 'AS2_plus', 'WT_minus', 'WT_plus',\n",
    "#        'SC_AS_plus', 'SC_WT_minus', 'SC_WT_plus']\n",
    "# kl_sc_PKA_data['sc_kl_genename'] = kl_sc_PKA_data['kl_genename'] + '_' + kl_sc_PKA_data['sc_genename']\n",
    "# kl_sc_PKA_data_subset = kl_sc_PKA_data.loc[:,kl_sc_PKA_conditions + ['sc_kl_genename']]\n",
    "# kl_sc_PKA_data_subset.set_index('sc_kl_genename', inplace = True)\n",
    "# kl_sc_PKA_data_subset.to_csv(base_dir + '\\expression_data\\kl_PKA_as_20160824\\kl_sc_NMPP1_count_data.csv')\n",
    "\n",
    "#will eventually want to include SC Orf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "# #for each row in the SC PKA inhibition data add a tuple with (SC Gene name, SC Expression, KL Gene Name, KL Expression)\n",
    "# orth_expression = []\n",
    "# SC_data = SC_PKA_data_norm['Pincus']\n",
    "# for line in SC_data.iteritems(): \n",
    "#     SCer_gene = line[0]\n",
    "#     SCer_exp = line[1]\n",
    "#     try:\n",
    "#         kl_orths = orth_lookup_SCer_KLac_dict[SCer_gene]\n",
    "#         if kl_orths[0] == 'NONE':\n",
    "#             KLac_gene = None\n",
    "#             KLac_exp = None\n",
    "#         else:\n",
    "#             for KLac_gene in kl_orths:\n",
    "#                 try: \n",
    "#                     #KLac_exp = kl_PKA_data_norm['AS_log_fold_change'][KLac_gene]\n",
    "#                     KLac_exp = kl_PKA_data_norm['log_AS_fold_change/WT_fold_change'][KLac_gene]\n",
    "                    \n",
    "#                 except KeyError: \n",
    "#                     KLac_exp = None\n",
    "#                 orth_expression.append((SCer_gene, SCer_exp, KLac_gene, KLac_exp))\n",
    "                \n",
    "#     except KeyError:\n",
    "#         KLac_gene = None\n",
    "#         KLac_exp = None\n",
    "#         orth_expression.append((SCer_gene, SCer_exp, KLac_gene, KLac_exp))\n",
    "\n",
    "# orth_expression_df = pd.DataFrame(orth_expression, columns = ['SCer_gene','SCer_exp','KLac_gene','KLac_exp'])\n",
    "# #output is a list of ression values for each ortholog\n",
    "\n",
    "# #After generating a dataframe with orthologs and expression values for each SC Gene, make plotting vectors: \n",
    "\n",
    "# # Iterate through each SC Gene\n",
    "# # If there is an orthololog and there is an expression value\n",
    "# # If there's just one value, put that in the x column\n",
    "# # If there's more than one value put that in he \n",
    "    \n",
    "# #Convert KL_orth to joined string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Old code for comparing orthologs \n",
    "\n",
    "# #Load SC data from Pincus and KLac NMPP1 gene expression data in order to compare expression of orthologs\n",
    "\n",
    "# # To Do: Apples to Apples Comparison first and then adjust items that go up in just WT + drug\n",
    "\n",
    "# species2 = 'Kluyveromyces lactis'\n",
    "# species1 = 'Saccharomyces cerevisiae' \n",
    "# orth_lookup_SCer_KLac_dict = io_library.read_orth_lookup_table(species1, species2) #Should have directory name\n",
    "\n",
    "# #for each row in the SC PKA inhibition data add a tuple with (SC Gene name, SC Expression, KL Gene Name, KL Expression)\n",
    "# orth_expression = []\n",
    "# SC_data = SC_PKA_data_norm['Pincus']\n",
    "# for line in SC_data.iteritems(): \n",
    "#     SCer_gene = line[0]\n",
    "#     SCer_exp = line[1]\n",
    "#     try:\n",
    "#         kl_orths = orth_lookup_SCer_KLac_dict[SCer_gene]\n",
    "#         if kl_orths[0] == 'NONE':\n",
    "#             KLac_gene = None\n",
    "#             KLac_exp = None\n",
    "#         else:\n",
    "#             for KLac_gene in kl_orths:\n",
    "#                 try: \n",
    "#                     #KLac_exp = kl_PKA_data_norm['AS_log_fold_change'][KLac_gene]\n",
    "#                     KLac_exp = kl_PKA_data_norm['log_AS_fold_change/WT_fold_change'][KLac_gene]\n",
    "                    \n",
    "#                 except KeyError: \n",
    "#                     KLac_exp = None\n",
    "#                 orth_expression.append((SCer_gene, SCer_exp, KLac_gene, KLac_exp))\n",
    "                \n",
    "#     except KeyError:\n",
    "#         KLac_gene = None\n",
    "#         KLac_exp = None\n",
    "#         orth_expression.append((SCer_gene, SCer_exp, KLac_gene, KLac_exp))\n",
    "\n",
    "# orth_expression_df = pd.DataFrame(orth_expression, columns = ['SCer_gene','SCer_exp','KLac_gene','KLac_exp'])\n",
    "# #output is a list of ression values for each ortholog\n",
    "\n",
    "# #After generating a dataframe with orthologs and expression values for each SC Gene, make plotting vectors: \n",
    "\n",
    "# # Iterate through each SC Gene\n",
    "# # If there is an orthololog and there is an expression value\n",
    "# # If there's just one value, put that in the x column\n",
    "# # If there's more than one value put that in he \n",
    "    \n",
    "# #Convert KL_orth to joined string\n",
    "\n",
    "# #Plot Correlation between SCer and KLac orthologs\n",
    "# sns.pairplot(orth_expression_df, x_vars = ['KLac_exp'], y_vars = 'SCer_exp')\n",
    "\n",
    "# #Plot Msn2 hits on top of correlation plot between SCer and KLac\n",
    "# fname = \"gasch_fig7_all_conditions.txt\"\n",
    "# gasch_Msn24_top_hits = io_library.read_gasch_data([conditions_Msn24[0]],fname) #only need indices so using dummy variable. \n",
    "# orth_expression_df['orfs']=orth_expression_df['SCer_gene']\n",
    "# top_hit_check = lambda orf: orf in set(gasch_Msn24_top_hits.index)\n",
    "# orth_expression_df['Msn24_hit'] = orth_expression_df['orfs'].apply(top_hit_check)\n",
    "\n",
    "# sns.pairplot(orth_expression_df, x_vars = ['KLac_exp'], y_vars = 'SCer_exp', hue = 'Msn24_hit')\n",
    "\n",
    "# # fig, ax = plt.subplots()\n",
    "# # xx = orth_expression_df['KLac_exp'][orth_expression_df['KLac_exp'].notnull()]\n",
    "# # yy = orth_expression_df['SCer_exp'][orth_expression_df['KLac_exp'].notnull()]\n",
    "# # ax.scatter(xx,yy)\n",
    "# # ax.set_xlabel('KLac Log Fold Change w/ 1NMPP1')\n",
    "# # ax.set_ylabel('SCer Log Fold Change w/ 1NMPP1 (Pincus)')\n",
    "\n",
    "\n",
    "\n",
    "# #Gene subsets\n",
    "\n",
    "# gene_subset = (set(xx[yy>-2.25].index) & set(xx[xx<-4.00].index))\n",
    "# fig, ax = plt.subplots()\n",
    "# xx = orth_expression_df['KLac_exp'][gene_subset]\n",
    "# yy = orth_expression_df['SCer_exp'][gene_subset]\n",
    "# ax.scatter(xx,yy)\n",
    "# ax.set_xlabel('KLac Log Fold Change w/ 1NMPP1')\n",
    "# ax.set_ylabel('SCer Log Fold Change w/ 1NMPP1 (OShea Lab)')\n",
    "\n",
    "# #Make list of SC Gene, KL Gene and their respective annotations. \n",
    "# orth_expression_subset = orth_expression_df.loc[gene_subset,:]\n",
    "# orth_expression_subset.sort_values(by = 'KLac_exp', inplace = True)\n",
    "\n",
    "# SC_orfs_lookup, SC_genename_lookup, SC_features_lookup = io_library.read_SGD_features()\n",
    "\n",
    "# annotation_KL =[]\n",
    "# annotation_SC =[]\n",
    "# SCer_gene_name = []\n",
    "\n",
    "# for gene_exp in orth_expression_subset.iterrows():\n",
    "#     SCer_gene = gene_exp[1]['SCer_gene']\n",
    "#     KLac_gene = gene_exp[1]['KLac_gene']\n",
    "#     SCer_gene_name.append(SC_genename_lookup[SCer_gene])\n",
    "#     annotation_SC.append(SC_features_lookup[SCer_gene])\n",
    "#     annotation_KL.append(annotation_lookup[KLac_gene])\n",
    "\n",
    "\n",
    "# orth_expression_subset['SCer name'] = SCer_gene_name\n",
    "# orth_expression_subset['SCer annotation'] = annotation_SC\n",
    "# orth_expression_subset['KLac annotation'] = annotation_KL\n",
    "\n",
    "# orth_expression_subset\n",
    "\n",
    "\n",
    "\n",
    "# fname = \"gasch_complete_dataset.txt\"\n",
    "# conditions_Msn24 = ['DBY7286 37degree heat - 20 min', 'DBYmsn2-4- 37degree heat - 20 min',  'DBYmsn2/4 (real strain) + 37degrees (20 min)', 'DBY7286 + 0.3 mM H2O2 (20 min)','DBYmsn2msn4 (good strain) + 0.32 mM H2O2','DBYmsn2/4 (real strain) + 0.32 mM H2O2 (20 min)','Msn2 overexpression (repeat)','Msn4 overexpression']\n",
    "# gasch_Msn24_data = io_library.read_gasch_data(conditions_Msn24,fname)\n",
    "# gasch_Msn24_data.columns = ['WT HS','msn2/4 HS 1','msn2/4 HS2','WT Ox','msn2/4 Ox 1', 'msn2/4 Ox 2', 'msn2 Overexp', 'msn4 Overexp']\n",
    "# #fig, axarr = expression_plots.multi_scatter_plot(gasch_Msn24_data, conditions) # xlim = xlim, ylim = ylim)\n",
    "\n",
    "# #Get top Msn24 hits\n",
    "# fname = \"gasch_fig7_all_conditions.txt\"\n",
    "# gasch_Msn24_top_hits = io_library.read_gasch_data([conditions_Msn24[0]],fname) #only need indices so using dummy variable. \n",
    "# gasch_Msn24_data['orfs']=gasch_Msn24_data.index\n",
    "# top_hit_check = lambda orf: orf in set(gasch_Msn24_top_hits.index)\n",
    "# gasch_Msn24_data['Msn24_hit'] = gasch_Msn24_data['orfs'].apply(top_hit_check)\n",
    "\n",
    "# conditions = ['WT HS','msn2/4 HS 1','msn2/4 HS2','WT Ox','msn2/4 Ox 1', 'msn2/4 Ox 2', 'msn2 Overexp', 'msn4 Overexp']\n",
    "# sns.pairplot(gasch_Msn24_data,x_vars=[conditions[3]],y_vars = [conditions[5]], hue=\"Msn24_hit\")\n",
    "# #sns.pairplot(gasch_Msn24_data,x_vars=[conditions[0],conditions[3]],y_vars = [conditions[2],conditions[5],conditions[6]], hue=\"Msn24_hit\")\n",
    "# #sns.lmplot(x=conditions[1], y=conditions[3], data=gasch_Msn24_data, hue=\"Msn24_hit\")\n",
    "#            #col_wrap=2, ci=None, palette=\"muted\", size=4,\n",
    "#            #scatter_kws={\"s\": 50, \"alpha\": 1})\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
